// @mastra/core/dist/auth/ee/index.cjs
'use strict';

var chunkLC7HRBCC_cjs = require('../../chunk-LC7HRBCC.cjs');
var chunkETRMQ4H5_cjs = require('../../chunk-ETRMQ4H5.cjs');



Object.defineProperty(exports, "ACTIONS", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.ACTIONS; }
});
Object.defineProperty(exports, "DEFAULT_ROLES", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.DEFAULT_ROLES; }
});
Object.defineProperty(exports, "MastraFGAPermissions", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.MastraFGAPermissions; }
});
Object.defineProperty(exports, "PERMISSIONS", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.PERMISSIONS; }
});
Object.defineProperty(exports, "PERMISSION_PATTERNS", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.PERMISSION_PATTERNS; }
});
Object.defineProperty(exports, "RESOURCES", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.RESOURCES; }
});
Object.defineProperty(exports, "StaticRBACProvider", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.StaticRBACProvider; }
});
Object.defineProperty(exports, "buildCapabilities", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.buildCapabilities; }
});
Object.defineProperty(exports, "getDefaultRole", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.getDefaultRole; }
});
Object.defineProperty(exports, "hasPermission", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.hasPermission; }
});
Object.defineProperty(exports, "isAuthenticated", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.isAuthenticated; }
});
Object.defineProperty(exports, "isValidPermissionPattern", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.isValidPermissionPattern; }
});
Object.defineProperty(exports, "matchesPermission", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.matchesPermission; }
});
Object.defineProperty(exports, "resolvePermissions", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.resolvePermissions; }
});
Object.defineProperty(exports, "resolvePermissionsFromMapping", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.resolvePermissionsFromMapping; }
});
Object.defineProperty(exports, "validatePermissions", {
  enumerable: true,
  get: function () { return chunkLC7HRBCC_cjs.validatePermissions; }
});
Object.defineProperty(exports, "FGADeniedError", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.FGADeniedError; }
});
Object.defineProperty(exports, "checkFGA", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.checkFGA; }
});
Object.defineProperty(exports, "clearLicenseCache", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.clearLicenseCache; }
});
Object.defineProperty(exports, "getAgentFGAResourceId", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.getAgentFGAResourceId; }
});
Object.defineProperty(exports, "getAgentToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.getAgentToolFGAResourceId; }
});
Object.defineProperty(exports, "getMCPToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.getMCPToolFGAResourceId; }
});
Object.defineProperty(exports, "getStandaloneToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.getStandaloneToolFGAResourceId; }
});
Object.defineProperty(exports, "getWorkflowFGAResourceId", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.getWorkflowFGAResourceId; }
});
Object.defineProperty(exports, "isDevEnvironment", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.isDevEnvironment; }
});
Object.defineProperty(exports, "isEEEnabled", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.isEEEnabled; }
});
Object.defineProperty(exports, "isEELicenseValid", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.isEELicenseValid; }
});
Object.defineProperty(exports, "isFeatureEnabled", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.isFeatureEnabled; }
});
Object.defineProperty(exports, "isLicenseValid", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.isLicenseValid; }
});
Object.defineProperty(exports, "requireFGA", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.requireFGA; }
});
Object.defineProperty(exports, "validateLicense", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.validateLicense; }
});
Object.defineProperty(exports, "warnIfDevEENeedsLicense", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.warnIfDevEENeedsLicense; }
});
