// @mastra/core/dist/chunk-AOQKFEGW.cjs
'use strict';

var chunk7DI6OU5T_cjs = require('./chunk-7DI6OU5T.cjs');
var v4 = require('zod/v4');

var TASKS_REQUEST_CONTEXT_KEY = "mastra:tasks";
var TASKS_STATE_ID = "tasks";
var TASK_STATE_TYPE = "task";
var NO_MEMORY_MESSAGE = "Task tools require agent memory (a memory-backed thread). No task was recorded. Configure the agent with Memory to use the task list.";
var taskIdSchema = v4.z.string().min(1).describe("Stable task identifier (for example, 'task_investigate_tests'). Keep this unchanged across updates.");
var taskItemInputSchema = v4.z.object({
  id: taskIdSchema.optional(),
  content: v4.z.string().min(1).describe("Task description in imperative form (e.g., 'Fix authentication bug')"),
  status: v4.z.enum(["pending", "in_progress", "completed"]).describe("Current task status"),
  activeForm: v4.z.string().min(1).describe("Present continuous form shown during execution (e.g., 'Fixing authentication bug')")
});
var taskItemSchema = taskItemInputSchema.extend({
  id: taskIdSchema
});
var taskToolResultSchema = v4.z.object({
  content: v4.z.string(),
  tasks: v4.z.array(taskItemSchema),
  isError: v4.z.boolean()
});
var taskCheckSummarySchema = v4.z.object({
  total: v4.z.number().int().nonnegative(),
  completed: v4.z.number().int().nonnegative(),
  inProgress: v4.z.number().int().nonnegative(),
  pending: v4.z.number().int().nonnegative(),
  incomplete: v4.z.number().int().nonnegative(),
  hasTasks: v4.z.boolean(),
  allCompleted: v4.z.boolean()
});
var taskCheckResultSchema = taskToolResultSchema.extend({
  summary: taskCheckSummarySchema,
  incompleteTasks: v4.z.array(taskItemSchema)
});
var TASK_ID_SLUG_MAX_LENGTH = 48;
function slugifyTaskContent(content) {
  let slug = "";
  let pendingSeparator = false;
  for (const char of content.toLowerCase()) {
    const code = char.charCodeAt(0);
    const isAsciiLetter = code >= 97 && code <= 122;
    const isDigit = code >= 48 && code <= 57;
    if (isAsciiLetter || isDigit) {
      if (pendingSeparator && slug.length > 0 && slug.length < TASK_ID_SLUG_MAX_LENGTH) {
        slug += "_";
      }
      if (slug.length >= TASK_ID_SLUG_MAX_LENGTH) break;
      slug += char;
      pendingSeparator = false;
      continue;
    }
    pendingSeparator = slug.length > 0;
  }
  return slug;
}
function createDeterministicTaskId(task, occurrence) {
  const slug = slugifyTaskContent(task.content);
  const suffix = occurrence > 1 ? `_${occurrence}` : "";
  return `task_${slug || "item"}${suffix}`;
}
function makeUniqueTaskId(id, usedIds, reservedIds = /* @__PURE__ */ new Set()) {
  if (!usedIds.has(id) && !reservedIds.has(id)) return id;
  let suffix = 2;
  let nextId = `${id}_${suffix}`;
  while (usedIds.has(nextId) || reservedIds.has(nextId)) {
    suffix += 1;
    nextId = `${id}_${suffix}`;
  }
  return nextId;
}
function assignTaskIds(tasks, previousTasks = []) {
  const usedIds = /* @__PURE__ */ new Set();
  const contentOccurrences = /* @__PURE__ */ new Map();
  const omittedContentCounts = /* @__PURE__ */ new Map();
  const explicitTaskIds = new Set(tasks.map((task) => task.id).filter((id) => Boolean(id)));
  const reusablePreviousIds = /* @__PURE__ */ new Map();
  for (const task of tasks) {
    if (!task.id) {
      omittedContentCounts.set(task.content, (omittedContentCounts.get(task.content) ?? 0) + 1);
    }
  }
  tasks.forEach((task, index) => {
    if (task.id || omittedContentCounts.get(task.content) !== 1) return;
    const previousMatches = previousTasks.filter(
      (previous) => previous.content === task.content && !explicitTaskIds.has(previous.id)
    );
    if (previousMatches.length === 1) {
      reusablePreviousIds.set(index, previousMatches[0].id);
    }
  });
  const reservedIds = /* @__PURE__ */ new Set([...explicitTaskIds, ...reusablePreviousIds.values()]);
  return tasks.map((task, index) => {
    const contentOccurrence = (contentOccurrences.get(task.content) ?? 0) + 1;
    contentOccurrences.set(task.content, contentOccurrence);
    const fallbackId = createDeterministicTaskId(task, contentOccurrence);
    const reusablePreviousId = reusablePreviousIds.get(index);
    const requestedId = task.id && !usedIds.has(task.id) ? task.id : void 0;
    const id = requestedId ?? (reusablePreviousId && !usedIds.has(reusablePreviousId) ? reusablePreviousId : makeUniqueTaskId(fallbackId, usedIds, reservedIds));
    usedIds.add(id);
    return {
      id,
      content: task.content,
      status: task.status,
      activeForm: task.activeForm
    };
  });
}
function formatTaskListResult(tasks) {
  const completed = tasks.filter((t) => t.status === "completed").length;
  const inProgress = tasks.find((t) => t.status === "in_progress");
  const total = tasks.length;
  let summary = `Tasks updated: [${completed}/${total} completed]`;
  if (inProgress) {
    summary += `
Currently: ${inProgress.activeForm} (${inProgress.id})`;
  }
  if (tasks.length > 0) {
    summary += `
Task IDs:
${tasks.map((t) => `- ${t.id}: ${t.content} (${t.status})`).join("\n")}`;
  }
  return summary;
}
function summarizeTaskCheck(tasks) {
  const completedTasks = tasks.filter((task) => task.status === "completed");
  const inProgressTasks = tasks.filter((task) => task.status === "in_progress");
  const pendingTasks = tasks.filter((task) => task.status === "pending");
  const incompleteTasks = [...inProgressTasks, ...pendingTasks];
  return {
    summary: {
      total: tasks.length,
      completed: completedTasks.length,
      inProgress: inProgressTasks.length,
      pending: pendingTasks.length,
      incomplete: incompleteTasks.length,
      hasTasks: tasks.length > 0,
      allCompleted: tasks.length > 0 && incompleteTasks.length === 0
    },
    inProgressTasks,
    pendingTasks,
    incompleteTasks
  };
}
function formatTaskCheckResult(taskCheck) {
  const { summary, inProgressTasks, pendingTasks } = taskCheck;
  if (!summary.hasTasks) {
    return "No tasks found. Consider using task_write to create a task list for complex work.";
  }
  let response = `Task Status: [${summary.completed}/${summary.total} completed]
`;
  response += `- Completed: ${summary.completed}
`;
  response += `- In Progress: ${summary.inProgress}
`;
  response += `- Pending: ${summary.pending}
`;
  response += `
All tasks completed: ${summary.allCompleted ? "YES" : "NO"}`;
  if (!summary.allCompleted) {
    response += "\n\nIncomplete tasks:";
    if (inProgressTasks.length > 0) {
      response += "\n\nIn Progress:";
      inProgressTasks.forEach((t) => {
        response += `
- ${t.id}: ${t.content}`;
      });
    }
    if (pendingTasks.length > 0) {
      response += "\n\nPending:";
      pendingTasks.forEach((t) => {
        response += `
- ${t.id}: ${t.content}`;
      });
    }
    response += "\n\nContinue working on these tasks before ending.";
  }
  return response;
}
function hasMultipleInProgress(tasks) {
  return tasks.filter((task) => task.status === "in_progress").length > 1;
}
function multipleInProgressError(tasks) {
  return {
    content: "Only one task can be in_progress at a time.",
    tasks,
    isError: true
  };
}
function demoteExtraInProgress(tasks, preferredIndex) {
  const inProgressIndices = tasks.reduce((acc, t, i) => {
    if (t.status === "in_progress") acc.push(i);
    return acc;
  }, []);
  if (inProgressIndices.length <= 1) return tasks;
  const keepIndex = preferredIndex !== void 0 && inProgressIndices.includes(preferredIndex) ? preferredIndex : inProgressIndices[inProgressIndices.length - 1];
  return tasks.map(
    (t, i) => t.status === "in_progress" && i !== keepIndex ? { ...t, status: "pending" } : t
  );
}
function formatAvailableTaskIds(tasks) {
  if (tasks.length === 0) return "No tasks are currently tracked.";
  return `Available task IDs:
${tasks.map((t) => `- ${t.id}: ${t.content} (${t.status})`).join("\n")}`;
}
function isMemoryBacked(agent) {
  return Boolean(agent?.threadId && agent?.resourceId);
}
function isThreadStateStore(value) {
  return !!value && typeof value.getState === "function" && typeof value.setState === "function";
}
async function resolveTaskStore(context) {
  const store = await context.mastra?.getStorage?.()?.getStore("threadState");
  return isThreadStateStore(store) ? store : void 0;
}
function emitTaskDisplayUpdate(requestContext, tasks) {
  const harnessCtx = requestContext?.get("harness");
  harnessCtx?.emitEvent?.({ type: "task_updated", tasks });
}
function noMemoryResult() {
  return { content: NO_MEMORY_MESSAGE, tasks: [], isError: true };
}
function noMemoryCheckResult() {
  const emptyCheck = summarizeTaskCheck([]);
  return {
    content: NO_MEMORY_MESSAGE,
    tasks: [],
    summary: emptyCheck.summary,
    incompleteTasks: emptyCheck.incompleteTasks,
    isError: true
  };
}
async function readTaskStore(context) {
  const store = await resolveTaskStore(context);
  const threadId = context.agent?.threadId;
  if (!store || !threadId) return [];
  const tasks = await store.getState({ threadId, type: TASK_STATE_TYPE });
  return Array.isArray(tasks) ? tasks : [];
}
async function applyTaskMutation(context, mutation) {
  const store = await resolveTaskStore(context);
  const threadId = context.agent?.threadId;
  if (!store || !threadId) return noMemoryResult();
  const currentTasks = await readTaskStore(context);
  const result = mutation(currentTasks);
  if (!result.isError) {
    await store.setState({ threadId, type: TASK_STATE_TYPE, value: result.tasks });
    context.requestContext?.set(TASKS_REQUEST_CONTEXT_KEY, result.tasks);
    emitTaskDisplayUpdate(context.requestContext, result.tasks);
  }
  return result;
}
var taskWriteTool = chunk7DI6OU5T_cjs.createTool({
  id: "task_write",
  description: `Create and manage a structured task list for your current coding session. This helps you track progress, organize complex tasks, and demonstrate thoroughness to the user.

Usage:
- Use this to create the initial task list or replace the whole list after replanning
- Pass the FULL task list each time this tool is called (replaces the previous list)
- Each task has: id (stable identifier), content (imperative), status (pending, in_progress, or completed), activeForm (present continuous)
- IDs must be unique. If duplicate explicit IDs are provided, the duplicate task is returned with a generated fallback ID
- Keep task IDs stable across updates. If omitted, IDs are generated and returned in the tool result
- When an ID is omitted while rewriting an existing list, one unambiguous matching task may reuse an existing ID
- Prefer single-task update tools when they are available
- Mark tasks in_progress BEFORE starting work (only ONE at a time)
- Mark tasks completed IMMEDIATELY after finishing
- Use this for multi-step tasks requiring 3+ distinct actions

States:
- pending: Not yet started
- in_progress: Currently working on (limit to ONE)
- completed: Finished successfully`,
  inputSchema: v4.z.object({
    tasks: v4.z.array(taskItemInputSchema).describe("The complete updated task list")
  }),
  outputSchema: taskToolResultSchema,
  execute: async ({ tasks }, context) => {
    try {
      if (!isMemoryBacked(context?.agent)) return noMemoryResult();
      return applyTaskMutation(context, (currentTasks) => {
        const normalizedTasks = assignTaskIds(tasks, currentTasks);
        if (hasMultipleInProgress(normalizedTasks)) {
          return multipleInProgressError(currentTasks);
        }
        return {
          content: formatTaskListResult(normalizedTasks),
          tasks: normalizedTasks,
          isError: false
        };
      });
    } catch (error) {
      const msg = error instanceof Error ? error.message : "Unknown error";
      return {
        content: `Failed to update tasks: ${msg}`,
        tasks: [],
        isError: true
      };
    }
  }
});
var taskUpdateTool = chunk7DI6OU5T_cjs.createTool({
  id: "task_update",
  description: `Update one task in the current task list by stable ID. Use this for targeted changes to one existing task.

Usage:
- Provide the task ID returned by the task-list tools
- Include only the fields that changed
- Use status to move a task between pending, in_progress, and completed
- Use task_complete when only marking a task completed
- If the ID is unknown, the tool returns an error with available task IDs`,
  inputSchema: v4.z.object({
    id: taskIdSchema,
    content: v4.z.string().min(1).optional().describe("New task description in imperative form"),
    status: v4.z.enum(["pending", "in_progress", "completed"]).optional().describe("New task status"),
    activeForm: v4.z.string().min(1).optional().describe("New present continuous form shown during execution")
  }).refine((input) => input.content !== void 0 || input.status !== void 0 || input.activeForm !== void 0, {
    message: "Provide at least one field to update."
  }),
  outputSchema: taskToolResultSchema,
  execute: async ({ id, content, status, activeForm }, context) => {
    try {
      if (!isMemoryBacked(context?.agent)) return noMemoryResult();
      return applyTaskMutation(context, (tasks) => {
        const taskIndex = tasks.findIndex((task) => task.id === id);
        if (taskIndex === -1) {
          return {
            content: `Task not found: ${id}

${formatAvailableTaskIds(tasks)}`,
            tasks,
            isError: true
          };
        }
        const updatedTasks = demoteExtraInProgress(
          tasks.map(
            (task, index) => index === taskIndex ? {
              ...task,
              ...content !== void 0 ? { content } : {},
              ...status !== void 0 ? { status } : {},
              ...activeForm !== void 0 ? { activeForm } : {}
            } : task
          ),
          taskIndex
        );
        return {
          content: formatTaskListResult(updatedTasks),
          tasks: updatedTasks,
          isError: false
        };
      });
    } catch (error) {
      const msg = error instanceof Error ? error.message : "Unknown error";
      return {
        content: `Failed to update task: ${msg}`,
        tasks: [],
        isError: true
      };
    }
  }
});
var taskCompleteTool = chunk7DI6OU5T_cjs.createTool({
  id: "task_complete",
  description: `Mark one task completed by stable ID. Use this when one tracked task is finished.

Usage:
- Provide the task ID returned by the task-list tools
- If the ID is unknown, the tool returns an error with available task IDs`,
  inputSchema: v4.z.object({
    id: taskIdSchema
  }),
  outputSchema: taskToolResultSchema,
  execute: async ({ id }, context) => {
    try {
      if (!isMemoryBacked(context?.agent)) return noMemoryResult();
      return applyTaskMutation(context, (tasks) => {
        const taskIndex = tasks.findIndex((task) => task.id === id);
        if (taskIndex === -1) {
          return {
            content: `Task not found: ${id}

${formatAvailableTaskIds(tasks)}`,
            tasks,
            isError: true
          };
        }
        const updatedTasks = tasks.map(
          (task, index) => index === taskIndex ? {
            ...task,
            status: "completed"
          } : task
        );
        return {
          content: formatTaskListResult(updatedTasks),
          tasks: updatedTasks,
          isError: false
        };
      });
    } catch (error) {
      const msg = error instanceof Error ? error.message : "Unknown error";
      return {
        content: `Failed to complete task: ${msg}`,
        tasks: [],
        isError: true
      };
    }
  }
});
var taskCheckTool = chunk7DI6OU5T_cjs.createTool({
  id: "task_check",
  description: `Check the completion status of your current task list. Use this before finishing tracked work to ensure all tasks are completed.

Returns:
- Human-readable content summary with task counts and incomplete task IDs
- Structured task list snapshot with stable IDs
- summary object with total, completed, inProgress, pending, incomplete, hasTasks, and allCompleted
- incompleteTasks array for tasks that still need work

summary.allCompleted is true only when at least one tracked task exists and every tracked task is completed. If no tasks exist, summary.hasTasks is false and summary.allCompleted is false.`,
  inputSchema: v4.z.object({}),
  // No input needed
  outputSchema: taskCheckResultSchema,
  execute: async ({}, context) => {
    try {
      if (!isMemoryBacked(context?.agent)) return noMemoryCheckResult();
      const tasks = await readTaskStore(context);
      const taskCheck = summarizeTaskCheck(tasks);
      return {
        content: formatTaskCheckResult(taskCheck),
        tasks,
        summary: taskCheck.summary,
        incompleteTasks: taskCheck.incompleteTasks,
        isError: false
      };
    } catch (error) {
      const msg = error instanceof Error ? error.message : "Unknown error";
      const emptyCheck = summarizeTaskCheck([]);
      return {
        content: `Failed to check tasks: ${msg}`,
        tasks: [],
        summary: emptyCheck.summary,
        incompleteTasks: emptyCheck.incompleteTasks,
        isError: true
      };
    }
  }
});
function isTaskItemArray(value) {
  return Array.isArray(value) && value.every(
    (item) => item && typeof item === "object" && typeof item.id === "string" && typeof item.content === "string" && typeof item.status === "string" && typeof item.activeForm === "string"
  );
}
function getTasksFromRequestContext(requestContext) {
  const carried = requestContext?.get(TASKS_REQUEST_CONTEXT_KEY);
  return isTaskItemArray(carried) ? carried : void 0;
}

// src/tools/builtin/task-state-processor.ts
function isThreadStateStore2(value) {
  return !!value && typeof value.getState === "function";
}
function renderTaskList(tasks) {
  if (tasks.length === 0) return "";
  const lines = tasks.map((task) => {
    const icon = task.status === "completed" ? "\u2713" : task.status === "in_progress" ? "\u25B8" : "\u25CB";
    return `  ${icon} [${task.status}] {id: ${task.id}} ${task.content}`;
  });
  return `
${lines.join("\n")}
`;
}
var DELTA_SNAPSHOT_CAP = 10;
function getTasksFromSnapshot(snapshot) {
  const value = snapshot?.metadata?.value;
  const tasks = value?.tasks;
  if (Array.isArray(tasks)) return tasks;
  return [];
}
function getOpsFromDelta(signal) {
  const delta = signal?.metadata?.delta;
  const ops = delta?.ops;
  return Array.isArray(ops) ? ops : [];
}
function applyOps(tasks, ops) {
  const next = tasks.slice();
  for (const op of ops) {
    if (op.op === "remove") {
      const idx2 = next.findIndex((t) => t.id === op.id);
      if (idx2 >= 0) next.splice(idx2, 1);
      continue;
    }
    const task = op.task;
    const idx = next.findIndex((t) => t.id === task.id);
    if (idx >= 0) next[idx] = task;
    else next.push(task);
  }
  return next;
}
function effectivePriorTasks(args) {
  let tasks = getTasksFromSnapshot(args.lastSnapshot);
  for (const delta of args.deltasSinceSnapshot ?? []) {
    tasks = applyOps(tasks, getOpsFromDelta(delta));
  }
  return tasks;
}
function lp(value) {
  return `${value.length}:${value}`;
}
function taskFingerprint(t) {
  return `${lp(t.id)}${lp(t.status)}${lp(t.content)}${lp(t.activeForm)}`;
}
function stableTasksCacheKey(tasks) {
  return `tasks:${tasks.map(taskFingerprint).join("|")}`;
}
function diffTasks(prior, current) {
  const ops = [];
  const priorById = new Map(prior.map((t) => [t.id, t]));
  const currentIds = new Set(current.map((t) => t.id));
  for (const task of current) {
    const before = priorById.get(task.id);
    if (!before) ops.push({ op: "add", task });
    else if (taskFingerprint(before) !== taskFingerprint(task)) ops.push({ op: "update", task });
  }
  for (const task of prior) {
    if (!currentIds.has(task.id)) ops.push({ op: "remove", id: task.id });
  }
  return ops;
}
function renderDelta(ops) {
  const lines = ops.map((op) => {
    if (op.op === "remove") return `  \u2212 removed {id: ${op.id}}`;
    const { task } = op;
    const icon = task.status === "completed" ? "\u2713" : task.status === "in_progress" ? "\u25B8" : "\u25CB";
    const verb = op.op === "add" ? "+" : icon;
    return `  ${verb} {id: ${task.id}} [${task.status}] ${task.content}`;
  });
  return `
${lines.join("\n")}
`;
}
var TaskStateProcessor = class {
  id = "task-state";
  stateId = TASKS_STATE_ID;
  /**
   * The Mastra instance this processor is registered with, used to resolve the
   * thread-scoped task store. Set by the agent/Mastra runtime via
   * `__registerMastra`.
   *
   * We implement this hook inline rather than extending `BaseProcessor`: a
   * *value* import of `BaseProcessor` from `processors/index` pulls that module's
   * runtime graph, which forms an initialization cycle through this tools module.
   * At the test entry point that surfaces as `TypeError: Class extends value
   * undefined` (BaseProcessor is not yet initialized when this class evaluates).
   * Implementing the (structurally trivial) hook here keeps all imports from
   * `processors/index` type-only, so there is no runtime edge and no cycle.
   */
  mastra;
  __registerMastra(mastra) {
    this.mastra = mastra;
  }
  async resolveTaskStore() {
    const store = await this.mastra?.getStorage?.()?.getStore("threadState");
    return isThreadStateStore2(store) ? store : void 0;
  }
  async computeStateSignal(args) {
    const priorTasks = effectivePriorTasks(args);
    const carried = getTasksFromRequestContext(args.requestContext);
    let currentTasks;
    if (carried !== void 0) {
      currentTasks = carried;
    } else {
      const store = await this.resolveTaskStore();
      const stored = store ? await store.getState({ threadId: args.threadId, type: TASK_STATE_TYPE }) : void 0;
      currentTasks = Array.isArray(stored) ? stored : priorTasks;
    }
    if (currentTasks.length === 0 && priorTasks.length === 0) return;
    const hasBase = Boolean(args.lastSnapshot) && args.contextWindow.hasSnapshot;
    const deltaCount = args.deltasSinceSnapshot?.length ?? 0;
    const ops = diffTasks(priorTasks, currentTasks);
    if (ops.length === 0 && hasBase) return;
    const mustSnapshot = !hasBase || deltaCount >= DELTA_SNAPSHOT_CAP;
    if (mustSnapshot) {
      return {
        id: TASKS_STATE_ID,
        cacheKey: stableTasksCacheKey(currentTasks),
        mode: "snapshot",
        // `current-task-list` is the signal's own tag. The framework wraps and
        // escapes `contents` inside it, so `renderTaskList` returns only the
        // inner lines — no inline tag here, or the model would see
        // double-wrapped, XML-escaped markup.
        tagName: "current-task-list",
        contents: renderTaskList(currentTasks),
        value: { tasks: currentTasks },
        attributes: { count: currentTasks.length },
        metadata: { value: { tasks: currentTasks } }
      };
    }
    return {
      id: TASKS_STATE_ID,
      cacheKey: stableTasksCacheKey(currentTasks),
      mode: "delta",
      tagName: "task-list-update",
      contents: renderDelta(ops),
      value: { tasks: currentTasks },
      delta: { ops },
      attributes: { changes: ops.length },
      metadata: { value: { tasks: currentTasks }, delta: { ops } }
    };
  }
};

exports.TASKS_REQUEST_CONTEXT_KEY = TASKS_REQUEST_CONTEXT_KEY;
exports.TASKS_STATE_ID = TASKS_STATE_ID;
exports.TASK_STATE_TYPE = TASK_STATE_TYPE;
exports.TaskStateProcessor = TaskStateProcessor;
exports.assignTaskIds = assignTaskIds;
exports.demoteExtraInProgress = demoteExtraInProgress;
exports.formatTaskListResult = formatTaskListResult;
exports.getTasksFromRequestContext = getTasksFromRequestContext;
exports.hasMultipleInProgress = hasMultipleInProgress;
exports.summarizeTaskCheck = summarizeTaskCheck;
exports.taskCheckTool = taskCheckTool;
exports.taskCompleteTool = taskCompleteTool;
exports.taskUpdateTool = taskUpdateTool;
exports.taskWriteTool = taskWriteTool;
