// @mastra/core/dist/chunk-AXT7KAIO.js
import { createHash } from 'crypto';
import os from 'os';
import { PostHog } from 'posthog-node';

// src/telemetry/posthog.ts
var POSTHOG_API_KEY = "phc_SBLpZVAB6jmHOct9CABq3PF0Yn5FU3G2FgT4xUr2XrT";
var POSTHOG_HOST = "https://us.posthog.com";
var TRUTHY_DISABLED_VALUES = /* @__PURE__ */ new Set(["1", "true", "yes"]);
var client = null;
function isEETelemetryEnabled() {
  const value = process.env["MASTRA_TELEMETRY_DISABLED"];
  if (!value) {
    return true;
  }
  return !TRUTHY_DISABLED_VALUES.has(value.trim().toLowerCase());
}
function hashTelemetryValue(value) {
  return createHash("sha256").update(value).digest("hex");
}
function getHashedHostname() {
  return hashTelemetryValue(os.hostname() || "unknown-host").slice(0, 16);
}
function getEETelemetryFallbackDistinctId() {
  return `mastra-${getHashedHostname()}`;
}
function getClient() {
  if (!isEETelemetryEnabled()) {
    return null;
  }
  if (!client) {
    client = new PostHog(POSTHOG_API_KEY, {
      host: POSTHOG_HOST,
      flushAt: 1,
      flushInterval: 0,
      disableGeoip: false
    });
  }
  return client;
}
function getSystemProperties() {
  return {
    os: process.platform,
    os_version: os.release(),
    node_version: process.version,
    platform: process.arch,
    machine_id: getHashedHostname(),
    mastra_version: process.env["npm_package_version"] || "unknown"
  };
}
function captureTelemetryEvent(event, distinctId, properties) {
  try {
    const posthog = getClient();
    if (!posthog) {
      return;
    }
    posthog.capture({
      distinctId: distinctId || getEETelemetryFallbackDistinctId(),
      event,
      properties: {
        ...getSystemProperties(),
        ...properties
      }
    });
  } catch {
  }
}
function captureEEEvent(event, distinctId, properties) {
  captureTelemetryEvent(event, distinctId, properties);
}

export { captureEEEvent, captureTelemetryEvent, getEETelemetryFallbackDistinctId, hashTelemetryValue, isEETelemetryEnabled };
