// @mastra/core/dist/chunk-Y7R26VZI.cjs
'use strict';

// src/signals/signal-provider.ts
var SignalProvider = class {
  name;
  /**
   * The Mastra instance this provider is registered with.
   * Set by the framework when the agent is registered with Mastra.
   */
  mastra;
  /**
   * @internal Called when the provider's agent is registered with a Mastra instance.
   */
  __registerMastra(mastra) {
    this.mastra = mastra;
  }
  /**
   * The agent this provider is connected to.
   * Set automatically when passed to `Agent({ signals: [...] })`.
   */
  #connectedAgent;
  /**
   * In-memory subscription registry.
   * Key: `${resourceId}:${threadId}:${externalResourceId}`
   */
  #subscriptions = /* @__PURE__ */ new Map();
  /**
   * Index: externalResourceId → set of subscription keys
   */
  #subscriptionsByResource = /* @__PURE__ */ new Map();
  /**
   * Index: `${resourceId}:${threadId}` → set of subscription keys
   */
  #subscriptionsByThread = /* @__PURE__ */ new Map();
  /** Active polling timer, if any */
  #pollTimer;
  /** Guard to prevent overlapping poll cycles */
  #isPollRunning = false;
  // ── Connection ──────────────────────────────────────────────────────
  /**
   * Called by the Agent constructor to establish the bidirectional link.
   * Override to perform additional setup (always call `super.connect(agent)`).
   */
  connect(agent) {
    this.#connectedAgent = agent;
  }
  /**
   * Whether this provider is already connected to an agent.
   * Used to skip re-wiring when an Agent is forked via `__fork()`.
   */
  get isConnected() {
    return this.#connectedAgent !== void 0;
  }
  /**
   * The connected agent. Available after `connect()` has been called.
   * Use this to send signals and notification signals back into agent threads.
   */
  get agent() {
    return this.#connectedAgent;
  }
  // ── Subscription tracking ──────────────────────────────────────────
  /**
   * Subscribe a thread to an external resource.
   *
   * @param target - The thread to receive signals
   * @param externalResourceId - Provider-specific resource identifier
   *   (e.g., `"github:mastra-ai/mastra#123"`, `"slack:C0B01RW7A4T"`)
   * @param metadata - Optional provider-specific metadata for the subscription
   */
  subscribe(target, externalResourceId, metadata = {}) {
    const key = this.#subscriptionKey(target, externalResourceId);
    const existing = this.#subscriptions.get(key);
    if (existing) {
      existing.metadata = { ...existing.metadata, ...metadata };
      return existing;
    }
    const subscription = {
      id: crypto.randomUUID(),
      providerId: this.id,
      threadId: target.threadId,
      resourceId: target.resourceId,
      externalResourceId,
      subscribedAt: /* @__PURE__ */ new Date(),
      metadata
    };
    this.#subscriptions.set(key, subscription);
    let resourceSet = this.#subscriptionsByResource.get(externalResourceId);
    if (!resourceSet) {
      resourceSet = /* @__PURE__ */ new Set();
      this.#subscriptionsByResource.set(externalResourceId, resourceSet);
    }
    resourceSet.add(key);
    const threadKey = this.#threadKey(target);
    let threadSet = this.#subscriptionsByThread.get(threadKey);
    if (!threadSet) {
      threadSet = /* @__PURE__ */ new Set();
      this.#subscriptionsByThread.set(threadKey, threadSet);
    }
    threadSet.add(key);
    return subscription;
  }
  /**
   * Unsubscribe a thread from an external resource.
   *
   * @returns `true` if a subscription was removed, `false` if none existed
   */
  unsubscribe(target, externalResourceId) {
    const key = this.#subscriptionKey(target, externalResourceId);
    const subscription = this.#subscriptions.get(key);
    if (!subscription) return false;
    this.#subscriptions.delete(key);
    const resourceSet = this.#subscriptionsByResource.get(externalResourceId);
    if (resourceSet) {
      resourceSet.delete(key);
      if (resourceSet.size === 0) this.#subscriptionsByResource.delete(externalResourceId);
    }
    const threadKey = this.#threadKey(target);
    const threadSet = this.#subscriptionsByThread.get(threadKey);
    if (threadSet) {
      threadSet.delete(key);
      if (threadSet.size === 0) this.#subscriptionsByThread.delete(threadKey);
    }
    return true;
  }
  /**
   * Get all active subscriptions for this provider.
   */
  getSubscriptions() {
    return [...this.#subscriptions.values()];
  }
  /**
   * Get all subscriptions for a specific external resource.
   *
   * @example
   * ```ts
   * const subs = this.getSubscriptionsForResource('github:mastra-ai/mastra#123');
   * for (const sub of subs) {
   *   await this.notify({ ... }, { resourceId: sub.resourceId, threadId: sub.threadId });
   * }
   * ```
   */
  getSubscriptionsForResource(externalResourceId) {
    const keys = this.#subscriptionsByResource.get(externalResourceId);
    if (!keys) return [];
    return [...keys].map((key) => this.#subscriptions.get(key)).filter(Boolean);
  }
  /**
   * Get all subscriptions for a specific thread.
   */
  getSubscriptionsForThread(target) {
    const threadKey = this.#threadKey(target);
    const keys = this.#subscriptionsByThread.get(threadKey);
    if (!keys) return [];
    return [...keys].map((key) => this.#subscriptions.get(key)).filter(Boolean);
  }
  /**
   * Check if a thread is subscribed to a specific external resource.
   */
  hasSubscription(target, externalResourceId) {
    return this.#subscriptions.has(this.#subscriptionKey(target, externalResourceId));
  }
  /**
   * Remove all subscriptions for a thread.
   */
  unsubscribeAll(target) {
    const threadSubscriptions = this.getSubscriptionsForThread(target);
    let removed = 0;
    for (const sub of threadSubscriptions) {
      if (this.unsubscribe(target, sub.externalResourceId)) removed++;
    }
    return removed;
  }
  /**
   * Total number of active subscriptions.
   */
  get subscriptionCount() {
    return this.#subscriptions.size;
  }
  // ── Polling ────────────────────────────────────────────────────────
  /**
   * Optional poll interval in milliseconds.
   * When defined, the framework calls `poll()` on this interval
   * with all active subscriptions.
   *
   * Set to `undefined` or `0` for webhook-only providers that don't poll.
   */
  pollInterval;
  /**
   * Start the polling timer. Called automatically by the Agent after `connect()`.
   * Can also be called manually to restart polling after `stopPolling()`.
   */
  startPolling() {
    if (this.#pollTimer) return;
    const interval = this.pollInterval;
    if (!interval || interval <= 0 || typeof this.poll !== "function") return;
    this.#pollTimer = setInterval(() => {
      if (this.#isPollRunning) return;
      const subscriptions = this.getSubscriptions();
      if (subscriptions.length === 0) return;
      this.#isPollRunning = true;
      void Promise.resolve(this.poll(subscriptions)).catch((error) => {
        console.warn(`[${this.id}] poll failed:`, error);
      }).finally(() => {
        this.#isPollRunning = false;
      });
    }, interval);
    this.#pollTimer.unref?.();
  }
  /**
   * Stop the polling timer.
   */
  stopPolling() {
    if (this.#pollTimer) {
      clearInterval(this.#pollTimer);
      this.#pollTimer = void 0;
    }
  }
  /**
   * Called on shutdown. Override to clean up resources.
   * Default implementation stops polling and clears all subscriptions.
   */
  stop() {
    this.stopPolling();
    this.#subscriptions.clear();
    this.#subscriptionsByResource.clear();
    this.#subscriptionsByThread.clear();
  }
  // ── Convenience ────────────────────────────────────────────────────
  /**
   * Send a notification signal to the connected agent.
   * Convenience wrapper around `this.agent.sendNotificationSignal()`.
   *
   * @throws If no agent is connected
   */
  async notify(notification, target) {
    const agent = this.#connectedAgent;
    if (!agent) {
      throw new Error(
        `[${this.id}] Cannot send notification: no agent connected. Was this provider passed to Agent({ signals: [...] })?`
      );
    }
    await agent.sendNotificationSignal(notification, {
      resourceId: target.resourceId,
      threadId: target.threadId
    });
  }
  // ── Internal ───────────────────────────────────────────────────────
  #subscriptionKey(target, externalResourceId) {
    return `${target.resourceId}:${target.threadId}:${externalResourceId}`;
  }
  #threadKey(target) {
    return `${target.resourceId}:${target.threadId}`;
  }
};
function isSignalProvider(obj) {
  return obj instanceof SignalProvider;
}

// src/signals/webhook-signal-provider.ts
var WebhookSignalProvider = class extends SignalProvider {
  id;
  name;
  #options;
  constructor(options = {}) {
    super();
    this.id = options.id ?? "webhook-signals";
    this.name = options.name ?? "Webhook Signals";
    this.#options = options;
  }
  // ── Static signal factories ────────────────────────────────────────
  /**
   * Create signal inputs for subscribing/unsubscribing threads via signals.
   */
  static signals = {
    subscribe(resource) {
      return {
        type: "reactive",
        tagName: "webhook-subscribe",
        contents: `Subscribe to webhook resource: ${resource}`,
        attributes: { resource }
      };
    },
    unsubscribe(resource) {
      return {
        type: "reactive",
        tagName: "webhook-unsubscribe",
        contents: `Unsubscribe from webhook resource: ${resource}`,
        attributes: { resource }
      };
    }
  };
  // ── Public API ─────────────────────────────────────────────────────
  /**
   * Programmatically subscribe a thread to an external resource.
   */
  subscribeThread(target, externalResourceId, metadata) {
    return this.subscribe(target, externalResourceId, metadata);
  }
  /**
   * Programmatically unsubscribe a thread from an external resource.
   */
  unsubscribeThread(target, externalResourceId) {
    return this.unsubscribe(target, externalResourceId);
  }
  // ── Webhook handling ───────────────────────────────────────────────
  /**
   * Handle an incoming webhook. Matches the payload against subscriptions
   * and emits notification signals to matching threads.
   */
  async handleWebhook(request) {
    const payload = request.body;
    const resourceIds = [...new Set(this.#extractResourceIds(payload))];
    if (resourceIds.length === 0) {
      return { status: 200, body: { matched: 0 } };
    }
    let matched = 0;
    for (const resourceId of resourceIds) {
      const subscriptions = this.getSubscriptionsForResource(resourceId);
      for (const subscription of subscriptions) {
        const notification = this.#buildNotification(payload, subscription);
        try {
          await this.notify(notification, {
            threadId: subscription.threadId,
            resourceId: subscription.resourceId
          });
          matched++;
        } catch (error) {
          console.warn(`[${this.id}] Failed to notify thread ${subscription.threadId}:`, error);
        }
      }
    }
    return { status: 200, body: { matched } };
  }
  // ── Internal ───────────────────────────────────────────────────────
  #extractResourceIds(payload) {
    if (this.#options.extractResourceId) {
      const result = this.#options.extractResourceId(payload);
      if (!result) return [];
      return Array.isArray(result) ? result : [result];
    }
    if (payload && typeof payload === "object") {
      const obj = payload;
      if (typeof obj.resource === "string") return [obj.resource];
      if (typeof obj.externalResourceId === "string") return [obj.externalResourceId];
    }
    return [];
  }
  #buildNotification(payload, subscription) {
    if (this.#options.buildNotification) {
      return this.#options.buildNotification(payload, subscription);
    }
    return {
      source: this.id,
      kind: "webhook-event",
      priority: "medium",
      summary: `Webhook event for ${subscription.externalResourceId}`,
      payload,
      dedupeKey: `${this.id}:${subscription.externalResourceId}:${Date.now()}`,
      coalesceKey: `${this.id}:${subscription.externalResourceId}`
    };
  }
};

exports.SignalProvider = SignalProvider;
exports.WebhookSignalProvider = WebhookSignalProvider;
exports.isSignalProvider = isSignalProvider;
