// @mastra/core/dist/chunk-64V3I723.js
import { Tool, MASTRA_TOOL_MARKER } from './chunk-3SZ22TE5.js';

// src/tools/toolchecks.ts
function isMastraTool(tool) {
  return tool instanceof Tool || typeof tool === "object" && tool !== null && MASTRA_TOOL_MARKER in tool;
}
function isVercelTool(tool) {
  return !!(tool && !isMastraTool(tool) && ("parameters" in tool || "execute" in tool && typeof tool.execute === "function" && "inputSchema" in tool));
}
function isProviderDefinedTool(tool) {
  if (typeof tool !== "object" || tool === null) return false;
  const t = tool;
  const isProviderType = t.type === "provider-defined" || t.type === "provider";
  return isProviderType && typeof t.id === "string";
}
var isProviderTool = isProviderDefinedTool;
function getProviderToolName(providerId) {
  return providerId.split(".").slice(1).join(".");
}
function getNeedsApprovalFn(tool) {
  if (typeof tool !== "object" || tool === null) return void 0;
  const fn = tool.needsApprovalFn;
  return typeof fn === "function" ? fn : void 0;
}

export { getNeedsApprovalFn, getProviderToolName, isMastraTool, isProviderDefinedTool, isProviderTool, isVercelTool };
