// @mastra/core/dist/chunk-5IG64QT5.js
import { StorageDomain } from './chunk-NFQ5V6OA.js';
import { parseMemoryRequestContext, applyStateSignal } from './chunk-4WYUSA24.js';
import { EventEmitterPubSub } from './chunk-SFA7SIBL.js';
import { createMessageSignal, createSignal, resolveDeliveryAttributes } from './chunk-TQ3ECMVI.js';
import { MASTRA_THREAD_ID_KEY, MASTRA_RESOURCE_ID_KEY } from './chunk-RI37F2KY.js';
import { randomUUID } from 'crypto';

var AGENT_THREAD_KEY_SEPARATOR = "\0";
var AGENT_THREAD_STREAM_TOPIC_PREFIX = "agent.thread-stream";
var defaultAgentThreadPubSub = new EventEmitterPubSub();
function withThreadMemory(memory, resourceId, threadId) {
  return {
    ...memory && typeof memory === "object" ? memory : {},
    resource: memory?.resource ?? resourceId,
    thread: memory?.thread ?? threadId
  };
}
function createRuntimeState() {
  return {
    threadRunsById: /* @__PURE__ */ new Map(),
    threadKeysByRunId: /* @__PURE__ */ new Map(),
    activeThreadRunIds: /* @__PURE__ */ new Map(),
    approvalSuspendedRunIds: /* @__PURE__ */ new Set(),
    pendingSignalsByThread: /* @__PURE__ */ new Map(),
    preRunSignalsByThread: /* @__PURE__ */ new Map(),
    pendingIdleSignalsByThread: /* @__PURE__ */ new Map(),
    pendingContinuationsByThread: /* @__PURE__ */ new Map(),
    watchedThreadRunIds: /* @__PURE__ */ new Set(),
    preparedRunsById: /* @__PURE__ */ new Map(),
    abortedRunIds: /* @__PURE__ */ new Set()
  };
}
var AgentThreadStreamRuntime = class {
  #id;
  #statesByPubSub = /* @__PURE__ */ new WeakMap();
  #getPubSub(pubsub) {
    return pubsub ?? defaultAgentThreadPubSub;
  }
  #getSourceId() {
    this.#id ??= randomUUID();
    return this.#id;
  }
  #getState(pubsub) {
    const resolvedPubSub = this.#getPubSub(pubsub);
    let state = this.#statesByPubSub.get(resolvedPubSub);
    if (!state) {
      state = createRuntimeState();
      this.#statesByPubSub.set(resolvedPubSub, state);
    }
    return state;
  }
  #threadKey(resourceId, threadId) {
    return [resourceId ?? "", threadId].join(AGENT_THREAD_KEY_SEPARATOR);
  }
  #threadTopic(key) {
    return `${AGENT_THREAD_STREAM_TOPIC_PREFIX}.${encodeURIComponent(key)}`;
  }
  #isApprovalSuspendedRun(state, runId) {
    return state.approvalSuspendedRunIds.has(runId);
  }
  #isThreadBlockingRun(state, record) {
    return record.output.status === "running" || this.#isApprovalSuspendedRun(state, record.runId);
  }
  #serializeSignal(signal) {
    return signal;
  }
  getThreadState(options, pubsub) {
    const state = this.#getState(pubsub);
    const key = this.#threadKey(options.resourceId, options.threadId);
    const activeRunId = state.activeThreadRunIds.get(key);
    if (!activeRunId) return "idle";
    const activeRecord = state.threadRunsById.get(activeRunId);
    if (activeRecord && !this.#isThreadBlockingRun(state, activeRecord)) {
      state.activeThreadRunIds.delete(key);
      return "idle";
    }
    return "active";
  }
  #publish(pubsub, key, event) {
    void this.#publishAndWait(pubsub, key, event).catch(() => {
    });
  }
  async #publishAndWait(pubsub, key, event) {
    await this.#getPubSub(pubsub).publish(this.#threadTopic(key), {
      type: event.type,
      runId: event.runId,
      data: event
    });
  }
  #withBroadcastStream(output, pubsub, key) {
    const runtime = this;
    const parts = [];
    const waiters = /* @__PURE__ */ new Set();
    let started = false;
    let done = false;
    let error;
    const wake = () => {
      const pending = [...waiters];
      waiters.clear();
      for (const waiter of pending) waiter();
    };
    const emitPart = async (part) => {
      if (part && typeof part === "object" && "type" in part && part.type === "tool-call-approval") {
        runtime.#getState(pubsub).approvalSuspendedRunIds.add(output.runId);
      }
      parts.push(part);
      await runtime.#publishAndWait(pubsub, key, {
        type: "stream-part",
        runId: output.runId,
        part,
        sourceId: runtime.#getSourceId()
      });
      wake();
    };
    const start = () => {
      if (started) return;
      started = true;
      void (async () => {
        try {
          const source = output.fullStream;
          if (!source) return;
          if (typeof source.getReader === "function") {
            const reader = source.getReader();
            try {
              while (true) {
                const { value: part, done: streamDone } = await reader.read();
                if (streamDone) break;
                await emitPart(part);
              }
            } finally {
              reader.releaseLock();
            }
          } else {
            for await (const part of source) {
              await emitPart(part);
            }
          }
        } catch (caught) {
          error = caught;
        } finally {
          done = true;
          wake();
        }
      })();
    };
    const createStream = () => {
      let index = 0;
      let closed = false;
      let waiter;
      return new ReadableStream({
        async pull(controller) {
          start();
          while (!closed) {
            if (index < parts.length) {
              controller.enqueue(parts[index++]);
              return;
            }
            if (error) {
              controller.error(error);
              return;
            }
            if (done) {
              controller.close();
              return;
            }
            await new Promise((resolve) => {
              waiter = resolve;
              waiters.add(resolve);
            });
            if (waiter) {
              waiters.delete(waiter);
              waiter = void 0;
            }
          }
        },
        cancel() {
          closed = true;
          if (waiter) {
            waiters.delete(waiter);
            waiter();
            waiter = void 0;
          }
        }
      });
    };
    return { output, createSubscriberStream: createStream, startBroadcast: start };
  }
  #getThreadTarget(options) {
    const thread = options?.memory?.thread;
    const threadId = options?.requestContext?.get(MASTRA_THREAD_ID_KEY) || (typeof thread === "string" ? thread : thread?.id);
    const resourceId = options?.requestContext?.get(MASTRA_RESOURCE_ID_KEY) || options?.memory?.resource;
    return { threadId, resourceId };
  }
  prepareRunOptions(options, pubsub) {
    const { threadId } = this.#getThreadTarget(options);
    if (!threadId || !options.runId) return options;
    const state = this.#getState(pubsub);
    const abortController = new AbortController();
    const upstreamAbortSignal = options.abortSignal;
    const abort = () => abortController.abort();
    if (upstreamAbortSignal?.aborted) {
      abort();
    } else {
      upstreamAbortSignal?.addEventListener("abort", abort, { once: true });
    }
    state.preparedRunsById.set(options.runId, {
      abortController,
      cleanup: () => upstreamAbortSignal?.removeEventListener("abort", abort)
    });
    if (state.abortedRunIds.has(options.runId)) {
      abort();
    }
    return {
      ...options,
      abortSignal: abortController.signal
    };
  }
  abortRun(runId, pubsub) {
    const state = this.#getState(pubsub);
    const preparedRun = state.preparedRunsById.get(runId);
    if (!preparedRun) {
      state.abortedRunIds.add(runId);
      return false;
    }
    preparedRun.abortController.abort();
    state.abortedRunIds.add(runId);
    const key = state.threadKeysByRunId.get(runId);
    if (key) {
      this.#publish(pubsub, key, { type: "run-aborted", runId });
    }
    return true;
  }
  getActiveThreadRunId(options, pubsub) {
    const state = this.#getState(pubsub);
    const key = this.#threadKey(options.resourceId, options.threadId);
    const activeRunId = state.activeThreadRunIds.get(key);
    if (!activeRunId) return void 0;
    const record = state.threadRunsById.get(activeRunId);
    if (record && !this.#isThreadBlockingRun(state, record)) return void 0;
    return activeRunId;
  }
  abortThread(options, pubsub) {
    const activeRunId = this.getActiveThreadRunId(options, pubsub);
    if (!activeRunId) return false;
    return this.abortRun(activeRunId, pubsub);
  }
  /** @internal */
  resetForTests() {
    for (const pubsub of [defaultAgentThreadPubSub]) {
      this.#resetState(pubsub);
      void pubsub.close?.();
    }
    defaultAgentThreadPubSub = new EventEmitterPubSub();
  }
  #resetState(pubsub) {
    const state = this.#statesByPubSub.get(pubsub);
    if (!state) return;
    state.preparedRunsById.forEach((preparedRun) => {
      preparedRun.abortController.abort();
      preparedRun.cleanup();
    });
    state.threadRunsById.clear();
    state.threadKeysByRunId.clear();
    state.activeThreadRunIds.clear();
    state.approvalSuspendedRunIds.clear();
    state.pendingSignalsByThread.clear();
    state.preRunSignalsByThread.clear();
    state.pendingIdleSignalsByThread.clear();
    state.pendingContinuationsByThread.clear();
    state.watchedThreadRunIds.clear();
    state.preparedRunsById.clear();
    state.abortedRunIds.clear();
  }
  #cleanupPreparedRun(state, runId) {
    state.preparedRunsById.get(runId)?.cleanup();
    state.preparedRunsById.delete(runId);
    state.abortedRunIds.delete(runId);
  }
  async #persistSignal(agent, signal, resourceId, threadId, requestContext) {
    const memory = await agent.getMemory({ requestContext });
    if (!memory) return;
    await memory.saveMessages({
      messages: [signal.toDBMessage({ resourceId, threadId })]
    });
  }
  #broadcastPersistedSignal(state, pubsub, key, runId, signal, resourceId, threadId) {
    let finish;
    const finished = new Promise((resolve) => {
      finish = resolve;
    });
    const parts = [
      { type: "start", runId },
      { ...signal.toDataPart(), runId },
      {
        type: "finish",
        runId,
        payload: {
          stepResult: { reason: "stop" },
          usage: { inputTokens: 0, outputTokens: 0, totalTokens: 0 }
        }
      }
    ];
    const output = {
      runId,
      status: "running",
      fullStream: new ReadableStream({
        start(controller) {
          for (const part of parts) controller.enqueue(part);
          controller.close();
          finish();
        }
      }),
      _waitUntilFinished: () => finished
    };
    const {
      output: outputForSubscribers,
      createSubscriberStream,
      startBroadcast
    } = this.#withBroadcastStream(output, pubsub, key);
    const record = {
      agent: { id: `persisted-signal:${signal.id}` },
      output: outputForSubscribers,
      runId,
      threadId,
      resourceId,
      streamOptions: {},
      createSubscriberStream
    };
    state.threadRunsById.set(runId, record);
    state.threadKeysByRunId.set(runId, key);
    const registered = this.#publishAndWait(pubsub, key, { type: "run-registered", runId });
    void registered.then(startBroadcast, startBroadcast);
    void outputForSubscribers._waitUntilFinished().finally(() => {
      setTimeout(() => {
        state.threadRunsById.delete(runId);
        state.threadKeysByRunId.delete(runId);
        if (state.activeThreadRunIds.get(key) === runId) {
          state.activeThreadRunIds.delete(key);
        }
        this.#publish(pubsub, key, { type: "run-completed", runId });
      }, 0);
    });
  }
  async #persistAndBroadcastIdleSignal(state, pubsub, key, runId, agent, signal, resourceId, threadId, requestContext) {
    await this.#persistSignal(agent, signal, resourceId, threadId, requestContext);
    this.#broadcastPersistedSignal(state, pubsub, key, runId, signal, resourceId, threadId);
  }
  registerRun(agent, output, streamOptions, pubsub) {
    const { threadId, resourceId } = this.#getThreadTarget(streamOptions);
    if (!threadId) return;
    const state = this.#getState(pubsub);
    const key = this.#threadKey(resourceId, threadId);
    const {
      output: outputForSubscribers,
      createSubscriberStream,
      startBroadcast
    } = this.#withBroadcastStream(output, pubsub, key);
    const record = {
      agent,
      output: outputForSubscribers,
      runId: output.runId,
      threadId,
      resourceId,
      streamOptions,
      createSubscriberStream
    };
    state.threadRunsById.set(output.runId, record);
    state.threadKeysByRunId.set(output.runId, key);
    state.activeThreadRunIds.set(key, output.runId);
    const registered = this.#publishAndWait(pubsub, key, { type: "run-registered", runId: output.runId });
    void registered.then(startBroadcast, startBroadcast);
    this.#watchThreadRunCompletion(state, pubsub, key, record);
  }
  #watchThreadRunCompletion(state, pubsub, key, record) {
    if (state.watchedThreadRunIds.has(record.runId)) return;
    state.watchedThreadRunIds.add(record.runId);
    void record.output._waitUntilFinished().finally(() => {
      state.watchedThreadRunIds.delete(record.runId);
      this.#cleanupPreparedRun(state, record.runId);
      if (record.output.status === "suspended" && this.#isApprovalSuspendedRun(state, record.runId)) {
        this.#publish(pubsub, key, { type: "run-suspended", runId: record.runId });
        return;
      }
      state.approvalSuspendedRunIds.delete(record.runId);
      state.threadRunsById.delete(record.runId);
      state.threadKeysByRunId.delete(record.runId);
      if (state.activeThreadRunIds.get(key) === record.runId) {
        state.activeThreadRunIds.delete(key);
      }
      this.#publish(pubsub, key, { type: "run-completed", runId: record.runId });
      void this.#drainPendingSignals(state, pubsub, key, record);
    });
  }
  async #drainPendingSignals(state, pubsub, key, previousRun) {
    if (state.activeThreadRunIds.has(key)) {
      return;
    }
    const preRunLeftover = state.preRunSignalsByThread.get(key);
    if (preRunLeftover?.length) {
      state.preRunSignalsByThread.delete(key);
      state.pendingSignalsByThread.set(key, [...preRunLeftover, ...state.pendingSignalsByThread.get(key) ?? []]);
    }
    const queue = state.pendingSignalsByThread.get(key);
    const signal = queue?.shift();
    if (signal && queue) {
      if (queue.length === 0) {
        state.pendingSignalsByThread.delete(key);
      }
      const output = await previousRun.agent.stream(signal, {
        ...previousRun.streamOptions,
        runId: randomUUID(),
        memory: withThreadMemory(
          previousRun.streamOptions.memory,
          previousRun.resourceId ?? "",
          previousRun.threadId ?? ""
        )
      });
      if (queue.length > 0) {
        const nextRecord = state.threadRunsById.get(output.runId);
        if (nextRecord) {
          this.#watchThreadRunCompletion(state, pubsub, key, nextRecord);
        }
      }
      return;
    }
    if (await this.#drainPendingContinuations(state, pubsub, key)) {
      return;
    }
    await this.#drainPendingIdleSignals(state, pubsub, key);
  }
  async #drainPendingContinuations(state, pubsub, key) {
    if (state.activeThreadRunIds.has(key)) {
      return false;
    }
    const queue = state.pendingContinuationsByThread.get(key);
    const pending = queue?.shift();
    if (!pending || !queue) {
      return false;
    }
    if (queue.length === 0) {
      state.pendingContinuationsByThread.delete(key);
    }
    this.#startContinuation(state, pubsub, key, pending);
    return true;
  }
  #startContinuation(state, pubsub, key, pending) {
    state.activeThreadRunIds.set(key, pending.runId);
    state.threadKeysByRunId.set(pending.runId, key);
    void pending.agent.stream(pending.messages, {
      ...pending.streamOptions,
      runId: pending.runId,
      memory: withThreadMemory(pending.streamOptions?.memory, pending.resourceId, pending.threadId)
    }).then((output) => {
      if ((state.pendingContinuationsByThread.get(key)?.length ?? 0) > 0) {
        const nextRecord = state.threadRunsById.get(output.runId);
        if (nextRecord) {
          this.#watchThreadRunCompletion(state, pubsub, key, nextRecord);
        }
      }
    }).catch((err) => {
      state.threadKeysByRunId.delete(pending.runId);
      this.#cleanupPreparedRun(state, pending.runId);
      if (state.activeThreadRunIds.get(key) === pending.runId) {
        state.activeThreadRunIds.delete(key);
      }
      this.#publish(pubsub, key, {
        type: "run-failed",
        runId: pending.runId,
        error: err instanceof Error ? err.message : String(err)
      });
      void this.#drainPendingContinuations(state, pubsub, key).then((started) => {
        if (!started) {
          void this.#drainPendingIdleSignals(state, pubsub, key);
        }
      });
    });
  }
  continueWithMessages(agent, messages, target, pubsub) {
    const state = this.#getState(pubsub);
    const key = this.#threadKey(target.resourceId, target.threadId);
    const runId = target.runId ?? randomUUID();
    const pending = {
      agent,
      messages,
      runId,
      resourceId: target.resourceId,
      threadId: target.threadId,
      streamOptions: target.streamOptions
    };
    const activeRunId = state.activeThreadRunIds.get(key);
    const activeRecord = activeRunId ? state.threadRunsById.get(activeRunId) : void 0;
    if (state.activeThreadRunIds.has(key)) {
      const queue = state.pendingContinuationsByThread.get(key) ?? [];
      queue.push(pending);
      state.pendingContinuationsByThread.set(key, queue);
      if (activeRecord) {
        this.#watchThreadRunCompletion(state, pubsub, key, activeRecord);
      }
      return { accepted: true, runId };
    }
    this.#startContinuation(state, pubsub, key, pending);
    return { accepted: true, runId };
  }
  async #drainPendingIdleSignals(state, pubsub, key) {
    if (state.activeThreadRunIds.has(key)) {
      return;
    }
    const idleQueue = state.pendingIdleSignalsByThread.get(key);
    const pendingIdle = idleQueue?.shift();
    if (!pendingIdle || !idleQueue) {
      return;
    }
    if (idleQueue.length === 0) {
      state.pendingIdleSignalsByThread.delete(key);
    }
    state.activeThreadRunIds.set(key, pendingIdle.runId);
    state.threadKeysByRunId.set(pendingIdle.runId, key);
    try {
      const output = await pendingIdle.agent.stream(pendingIdle.signal, {
        ...pendingIdle.streamOptions,
        runId: pendingIdle.runId,
        memory: withThreadMemory(pendingIdle.streamOptions?.memory, pendingIdle.resourceId, pendingIdle.threadId)
      });
      if ((idleQueue?.length ?? 0) > 0) {
        const nextRecord = state.threadRunsById.get(output.runId);
        if (nextRecord) {
          this.#watchThreadRunCompletion(state, pubsub, key, nextRecord);
        }
      }
    } catch (err) {
      state.threadKeysByRunId.delete(pendingIdle.runId);
      this.#cleanupPreparedRun(state, pendingIdle.runId);
      if (state.activeThreadRunIds.get(key) === pendingIdle.runId) {
        state.activeThreadRunIds.delete(key);
      }
      this.#publish(pubsub, key, {
        type: "run-failed",
        runId: pendingIdle.runId,
        error: err instanceof Error ? err.message : String(err)
      });
      void this.#drainPendingIdleSignals(state, pubsub, key);
    }
  }
  /**
   * Drains queued signals for a run.
   *
   * - `scope: 'pending'` (default) returns active-run follow-up signals — each
   *   becomes its own model turn via `signalDrainStep`.
   * - `scope: 'pre-run'` returns signals queued before the run's first model
   *   request — the first LLM step folds these into that request.
   */
  drainPendingSignals(runId, pubsub, scope = "pending") {
    const state = this.#getState(pubsub);
    const record = state.threadRunsById.get(runId);
    const key = record ? this.#threadKey(record.resourceId, record.threadId) : state.threadKeysByRunId.get(runId);
    if (!key) return [];
    const signalsByThread = scope === "pre-run" ? state.preRunSignalsByThread : state.pendingSignalsByThread;
    const queue = signalsByThread.get(key);
    if (!queue || queue.length === 0) {
      return [];
    }
    signalsByThread.delete(key);
    return queue;
  }
  async waitForCrossAgentThreadRun(agent, options, pubsub) {
    const { threadId, resourceId } = this.#getThreadTarget(options);
    if (!threadId) return;
    const state = this.#getState(pubsub);
    const key = this.#threadKey(resourceId, threadId);
    while (true) {
      const activeRunId = state.activeThreadRunIds.get(key);
      if (!activeRunId) return;
      const activeRecord = state.threadRunsById.get(activeRunId);
      if (activeRecord) {
        if (activeRecord.agent.id === agent.id || !this.#isThreadBlockingRun(state, activeRecord)) {
          return;
        }
        await activeRecord.output._waitUntilFinished().catch(() => {
        });
        continue;
      }
      if (state.threadKeysByRunId.get(activeRunId) === key) return;
      await this.#waitForRemoteRunToFinish(pubsub, key, activeRunId);
    }
  }
  async #waitForRemoteRunToFinish(pubsub, key, runId) {
    const resolvedPubSub = this.#getPubSub(pubsub);
    const topic = this.#threadTopic(key);
    await new Promise((resolve) => {
      const onEvent = (event) => {
        const data = event.data;
        if ((data?.type === "run-completed" || data?.type === "run-aborted" || data?.type === "run-failed") && data.runId === runId) {
          void resolvedPubSub.unsubscribe(topic, onEvent).catch(() => {
          });
          resolve();
        }
      };
      void resolvedPubSub.subscribe(topic, onEvent).catch(() => resolve());
    });
  }
  async subscribeToThread(agent, options, pubsub) {
    const resolvedPubSub = this.#getPubSub(pubsub);
    const state = this.#getState(resolvedPubSub);
    const key = this.#threadKey(options.resourceId, options.threadId);
    const topic = this.#threadTopic(key);
    const seenRunIds = /* @__PURE__ */ new Set();
    const pendingRuns = [];
    const waiters = [];
    const remoteRuns = /* @__PURE__ */ new Map();
    let done = false;
    const wake = () => {
      while (waiters.length) waiters.shift()?.();
    };
    const activeRunId = () => {
      const runId = state.activeThreadRunIds.get(key);
      if (!runId) return null;
      const record = state.threadRunsById.get(runId);
      if (!record) return runId;
      return this.#isThreadBlockingRun(state, record) ? runId : null;
    };
    const enqueueRun = (record) => {
      if (done || seenRunIds.has(record.runId)) return;
      seenRunIds.add(record.runId);
      pendingRuns.push(record);
      wake();
    };
    const createRemoteRun = (runId) => {
      const remoteRun = {
        parts: [],
        waiters: [],
        finishWaiters: [],
        done: false,
        stream: void 0,
        closed: false
      };
      remoteRun.stream = new ReadableStream({
        pull(controller) {
          const drain = () => {
            if (remoteRun.closed) return;
            while (remoteRun.parts.length > 0) {
              controller.enqueue(remoteRun.parts.shift());
            }
            if (remoteRun.done) {
              remoteRun.closed = true;
              controller.close();
            }
          };
          drain();
          if (!remoteRun.done && !remoteRun.closed) {
            remoteRun.waiters.push(drain);
          }
        },
        cancel() {
          remoteRun.done = true;
          remoteRun.closed = true;
          remoteRun.waiters.length = 0;
          while (remoteRun.finishWaiters.length) remoteRun.finishWaiters.shift()?.();
        }
      });
      remoteRuns.set(runId, remoteRun);
      return {
        agent,
        output: {
          runId,
          status: "running",
          fullStream: remoteRun.stream,
          _waitUntilFinished: async () => {
            if (remoteRun.done) return;
            await new Promise((resolve) => remoteRun.finishWaiters.push(resolve));
          }
        },
        runId,
        threadId: options.threadId,
        resourceId: options.resourceId,
        streamOptions: {}
      };
    };
    const onEvent = (event) => {
      const data = event.data;
      if (!data) return;
      if (data.type === "run-registered") {
        state.activeThreadRunIds.set(key, data.runId);
        const record = state.threadRunsById.get(data.runId) ?? createRemoteRun(data.runId);
        enqueueRun(record);
        wake();
        return;
      }
      if (data.type === "stream-part") {
        if (data.sourceId === this.#id) return;
        let remoteRun = remoteRuns.get(data.runId);
        if (!remoteRun) {
          state.activeThreadRunIds.set(key, data.runId);
          enqueueRun(createRemoteRun(data.runId));
          remoteRun = remoteRuns.get(data.runId);
          if (!remoteRun) return;
        }
        remoteRun.parts.push(data.part);
        while (remoteRun.waiters.length) remoteRun.waiters.shift()?.();
        return;
      }
      if (data.type === "signal-enqueued") {
        if (data.sourceId === this.#id) return;
        const signalsByThread = data.preRun ? state.preRunSignalsByThread : state.pendingSignalsByThread;
        const queue = signalsByThread.get(key) ?? [];
        queue.push(createSignal(data.signal));
        signalsByThread.set(key, queue);
        return;
      }
      if (data.type === "run-failed") {
        if (state.activeThreadRunIds.get(key) === data.runId) {
          state.activeThreadRunIds.delete(key);
        }
        const errorRun = createRemoteRun(data.runId);
        const remoteRun = remoteRuns.get(data.runId);
        if (remoteRun) {
          remoteRun.parts.push({ type: "error", payload: { error: new Error(data.error) } });
          remoteRun.done = true;
          while (remoteRun.waiters.length) remoteRun.waiters.shift()?.();
          while (remoteRun.finishWaiters.length) remoteRun.finishWaiters.shift()?.();
          remoteRuns.delete(data.runId);
        }
        enqueueRun(errorRun);
        seenRunIds.delete(data.runId);
        void this.#drainPendingIdleSignals(state, resolvedPubSub, key);
        wake();
        return;
      }
      if (data.type === "run-completed" || data.type === "run-aborted" || data.type === "run-suspended") {
        if ((data.type !== "run-suspended" || !state.approvalSuspendedRunIds.has(data.runId)) && state.activeThreadRunIds.get(key) === data.runId) {
          state.activeThreadRunIds.delete(key);
        }
        if (data.type !== "run-suspended") {
          state.approvalSuspendedRunIds.delete(data.runId);
        }
        const remoteRun = remoteRuns.get(data.runId);
        if (remoteRun) {
          remoteRun.done = true;
          while (remoteRun.waiters.length) remoteRun.waiters.shift()?.();
          while (remoteRun.finishWaiters.length) remoteRun.finishWaiters.shift()?.();
          remoteRuns.delete(data.runId);
        }
        if (data.type === "run-aborted" && activeReaderRunId === data.runId && currentReader) {
          cancelledByAbort = true;
          try {
            void currentReader.cancel();
          } catch {
          }
        }
        seenRunIds.delete(data.runId);
        if (data.type !== "run-suspended") {
          void this.#drainPendingIdleSignals(state, resolvedPubSub, key);
        }
        wake();
      }
    };
    await resolvedPubSub.subscribe(topic, onEvent);
    const currentRunId = activeRunId();
    const currentRecord = currentRunId ? state.threadRunsById.get(currentRunId) : void 0;
    if (currentRecord) {
      enqueueRun(currentRecord);
    }
    let currentReader = null;
    let activeReaderRunId = null;
    let cancelledByAbort = false;
    const unsubscribe = () => {
      if (done) return;
      done = true;
      void resolvedPubSub.unsubscribe(topic, onEvent).catch(() => {
      });
      if (currentReader) {
        try {
          void currentReader.cancel();
        } catch {
        }
      }
      wake();
    };
    return {
      activeRunId,
      abort: () => this.abortThread(options, resolvedPubSub),
      unsubscribe,
      stream: (async function* () {
        try {
          while (!done || pendingRuns.length > 0) {
            if (pendingRuns.length === 0) {
              await new Promise((resolve) => waiters.push(resolve));
              continue;
            }
            const run = pendingRuns.shift();
            const subscriberStream = run.createSubscriberStream?.() ?? run.output.fullStream;
            const reader = subscriberStream.getReader();
            currentReader = reader;
            activeReaderRunId = run.runId;
            let readerReleased = false;
            try {
              while (true) {
                const { value: part, done: streamDone } = await reader.read();
                if (streamDone) break;
                const typedPart = part;
                yield typedPart;
                if (done) break;
                if (typedPart.type === "finish" || typedPart.type === "error" || typedPart.type === "abort" || typedPart.type === "tool-call-suspended") {
                  readerReleased = true;
                  void (async () => {
                    try {
                      while (true) {
                        const { done: d } = await reader.read();
                        if (d) break;
                      }
                    } catch {
                    }
                    reader.releaseLock();
                  })();
                  break;
                }
              }
              if (!readerReleased && !done && cancelledByAbort) {
                yield { type: "abort", runId: run.runId };
                cancelledByAbort = false;
              }
            } finally {
              currentReader = null;
              activeReaderRunId = null;
              if (!readerReleased) {
                reader.releaseLock();
              }
            }
          }
        } finally {
          unsubscribe();
        }
      })()
    };
  }
  sendMessage(agent, message, target, pubsub) {
    return this.sendSignal(agent, createMessageSignal(message, { acceptedAt: /* @__PURE__ */ new Date() }), target, pubsub);
  }
  queueMessage(agent, message, target, pubsub) {
    const state = this.#getState(pubsub);
    const signal = createMessageSignal(message, { acceptedAt: /* @__PURE__ */ new Date() });
    let key;
    let runId = target.runId;
    let activeRecord;
    if (target.resourceId && target.threadId) {
      key = this.#threadKey(target.resourceId, target.threadId);
      const activeRunId = state.activeThreadRunIds.get(key);
      activeRecord = activeRunId ? state.threadRunsById.get(activeRunId) : void 0;
      if (activeRecord && !this.#isThreadBlockingRun(state, activeRecord)) {
        state.activeThreadRunIds.delete(key);
        activeRecord = void 0;
      }
      runId ??= activeRunId;
    }
    if (runId) {
      activeRecord ??= state.threadRunsById.get(runId);
      if (activeRecord) {
        key ??= this.#threadKey(activeRecord.resourceId, activeRecord.threadId);
      }
    }
    const resourceId = target.resourceId ?? activeRecord?.resourceId;
    const threadId = target.threadId ?? activeRecord?.threadId;
    if (!resourceId || !threadId) {
      throw new Error("resourceId and threadId are required to queue a message");
    }
    key ??= this.#threadKey(resourceId, threadId);
    const queuedRunId = randomUUID();
    const queuedStreamOptions = target.ifIdle?.streamOptions ?? activeRecord?.streamOptions;
    if (activeRecord) {
      const idleQueue = state.pendingIdleSignalsByThread.get(key) ?? [];
      idleQueue.push({ agent, signal, runId: queuedRunId, resourceId, threadId, streamOptions: queuedStreamOptions });
      state.pendingIdleSignalsByThread.set(key, idleQueue);
      this.#watchThreadRunCompletion(state, pubsub, key, activeRecord);
      return { accepted: true, runId: queuedRunId, signal };
    }
    return this.sendSignal(
      agent,
      signal,
      { ...target, runId, resourceId, threadId, ifIdle: { ...target.ifIdle, behavior: "wake" } },
      pubsub
    );
  }
  async sendStateSignal(agent, stateInput, target, pubsub) {
    if (!target.resourceId || !target.threadId) {
      throw new Error("resourceId and threadId are required to send a state signal");
    }
    const resourceId = target.resourceId;
    const threadId = target.threadId;
    const requestContext = target.ifIdle?.streamOptions?.requestContext;
    const memoryContext = parseMemoryRequestContext(requestContext);
    const memory = await agent.getMemory({ requestContext });
    if (!memory) {
      throw new Error("sendStateSignal requires Mastra memory");
    }
    const loadedThread = await memory.getThreadById({ threadId }) ?? memoryContext?.thread;
    if (!loadedThread) {
      throw new Error(`sendStateSignal could not load thread ${threadId}`);
    }
    const thread = {
      ...loadedThread,
      id: threadId,
      resourceId: loadedThread.resourceId ?? resourceId,
      createdAt: loadedThread.createdAt ?? /* @__PURE__ */ new Date(),
      updatedAt: loadedThread.updatedAt ?? /* @__PURE__ */ new Date(),
      metadata: loadedThread.metadata
    };
    const applied = await applyStateSignal({
      input: stateInput,
      memory,
      thread,
      resourceId,
      threadId,
      memoryConfig: memoryContext?.memoryConfig,
      acceptedAt: /* @__PURE__ */ new Date()
    });
    if (applied.skipped) {
      return { accepted: true, skipped: true, reason: "unchanged" };
    }
    return this.sendSignal(agent, applied.signal, target, pubsub);
  }
  /**
   * Routes a signal to an agent thread.
   *
   * Signals can land in three places:
   * - an active same-agent run, where they are queued for the execution loop to drain;
   * - a reserved thread run that has not registered its stream record yet;
   * - a new idle-started run, when the caller opts into `ifIdle`.
   *
   * Cross-agent active runs are intentionally not interrupted here. They either finish first
   * through `waitForCrossAgentThreadRun()` on the stream path, or this method falls through to
   * the idle-start path when the caller provided a resource/thread target and `ifIdle` options.
   */
  sendSignal(agent, signalInput, target, pubsub) {
    const state = this.#getState(pubsub);
    let signal = createSignal({ ...signalInput, acceptedAt: /* @__PURE__ */ new Date() });
    let key;
    let runId = target.runId;
    const activeBehavior = target.ifActive?.behavior ?? "deliver";
    const idleBehavior = target.ifIdle?.behavior ?? "wake";
    let activeRecord;
    if (target.resourceId && target.threadId) {
      key = this.#threadKey(target.resourceId, target.threadId);
      const activeRunId = state.activeThreadRunIds.get(key);
      activeRecord = activeRunId ? state.threadRunsById.get(activeRunId) : void 0;
      if (activeRecord && !this.#isThreadBlockingRun(state, activeRecord)) {
        state.activeThreadRunIds.delete(key);
        activeRecord = void 0;
      }
      if (activeRecord && activeRecord.agent.id === agent.id) {
        runId = activeRecord.runId;
      } else if (activeRunId && !activeRecord) {
        runId = activeRunId;
      }
    }
    const isActiveTarget = Boolean(
      runId && (activeRecord?.output.status === "running" || key && state.activeThreadRunIds.get(key) === runId)
    );
    const resourceId = target.resourceId ?? activeRecord?.resourceId;
    const threadId = target.threadId ?? activeRecord?.threadId;
    signal = resolveDeliveryAttributes(
      signal,
      isActiveTarget ? target.ifActive?.attributes : target.ifIdle?.attributes
    );
    if (isActiveTarget && activeBehavior !== "deliver") {
      if (activeBehavior === "persist") {
        if (!resourceId || !threadId) {
          throw new Error("resourceId and threadId are required to persist an active signal");
        }
        const persisted = this.#persistSignal(
          agent,
          signal,
          resourceId,
          threadId,
          target.ifIdle?.streamOptions?.requestContext
        );
        void persisted.catch(() => {
        });
        return { accepted: true, runId, signal, persisted };
      }
      return { accepted: true, runId, signal };
    }
    if (runId) {
      activeRecord ??= state.threadRunsById.get(runId);
      if (activeRecord && this.#isThreadBlockingRun(state, activeRecord)) {
        key ??= this.#threadKey(activeRecord.resourceId, activeRecord.threadId);
        if (activeRecord.agent.id === agent.id) {
          const queue = state.pendingSignalsByThread.get(key) ?? [];
          queue.push(signal);
          state.pendingSignalsByThread.set(key, queue);
          this.#publish(pubsub, key, {
            type: "signal-enqueued",
            runId,
            signal: this.#serializeSignal(signal),
            sourceId: this.#getSourceId()
          });
          this.#watchThreadRunCompletion(state, pubsub, key, activeRecord);
          return { accepted: true, runId, signal };
        }
      }
      if (key && state.activeThreadRunIds.get(key) === runId) {
        const isLocalReservedRun = state.threadKeysByRunId.get(runId) === key;
        if (isLocalReservedRun) {
          const queue = state.preRunSignalsByThread.get(key) ?? [];
          queue.push(signal);
          state.preRunSignalsByThread.set(key, queue);
        }
        this.#publish(pubsub, key, {
          type: "signal-enqueued",
          runId,
          signal: this.#serializeSignal(signal),
          sourceId: this.#getSourceId(),
          preRun: isLocalReservedRun
        });
        return { accepted: true, runId, signal };
      }
    }
    if (!resourceId || !threadId) {
      throw new Error("No active agent run found for signal target");
    }
    runId = randomUUID();
    key ??= this.#threadKey(resourceId, threadId);
    if (idleBehavior === "persist") {
      const persisted = this.#persistAndBroadcastIdleSignal(
        state,
        pubsub,
        key,
        runId,
        agent,
        signal,
        resourceId,
        threadId,
        target.ifIdle?.streamOptions?.requestContext
      );
      void persisted.catch(() => {
      });
      return { accepted: true, runId, signal, persisted };
    }
    if (idleBehavior !== "wake") {
      return { accepted: true, runId, signal };
    }
    if (state.activeThreadRunIds.has(key)) {
      const idleQueue = state.pendingIdleSignalsByThread.get(key) ?? [];
      idleQueue.push({ agent, signal, runId, resourceId, threadId, streamOptions: target.ifIdle?.streamOptions });
      state.pendingIdleSignalsByThread.set(key, idleQueue);
      if (activeRecord) {
        this.#watchThreadRunCompletion(state, pubsub, key, activeRecord);
      }
      return { accepted: true, runId, signal };
    }
    state.activeThreadRunIds.set(key, runId);
    state.threadKeysByRunId.set(runId, key);
    void agent.stream(signal, {
      ...target.ifIdle?.streamOptions,
      runId,
      memory: withThreadMemory(target.ifIdle?.streamOptions?.memory, resourceId, threadId)
    }).catch((err) => {
      state.threadKeysByRunId.delete(runId);
      this.#cleanupPreparedRun(state, runId);
      if (state.activeThreadRunIds.get(key) === runId) {
        state.activeThreadRunIds.delete(key);
      }
      this.#publish(pubsub, key, {
        type: "run-failed",
        runId,
        error: err instanceof Error ? err.message : String(err)
      });
      void this.#drainPendingIdleSignals(state, pubsub, key);
    });
    return { accepted: true, runId, signal };
  }
};
var agentThreadStreamRuntime = new AgentThreadStreamRuntime();

// src/notifications/signals.ts
function notificationSignalAttributes(notification) {
  return {
    ...notification.attributes,
    id: notification.id,
    source: notification.source,
    type: notification.kind,
    kind: notification.kind,
    priority: notification.priority,
    status: notification.status,
    ...notification.coalescedCount && notification.coalescedCount > 1 ? { coalescedCount: notification.coalescedCount } : {}
  };
}
function notificationSummaryContents(summary) {
  const sources = Object.entries(summary.bySource).sort(([a], [b]) => a.localeCompare(b)).map(([source, count]) => `${source}: ${count}`).join(", ");
  return sources || "No pending notifications";
}
var priorityOrder = ["low", "medium", "high", "urgent"];
function highestPriority(byPriority) {
  for (let i = priorityOrder.length - 1; i >= 0; i -= 1) {
    const priority = priorityOrder[i];
    if (priority && (byPriority[priority] ?? 0) > 0) return priority;
  }
  return void 0;
}
function notificationSignalMetadata(notification) {
  return {
    signal: "notification",
    recordId: notification.id,
    source: notification.source,
    kind: notification.kind,
    priority: notification.priority,
    status: notification.status,
    ...notification.coalescedCount && notification.coalescedCount > 1 ? { coalescedCount: notification.coalescedCount } : {},
    ...notification.deliveredAt ? { deliveredAt: notification.deliveredAt.toISOString() } : {},
    ...notification.seenAt ? { seenAt: notification.seenAt.toISOString() } : {}
  };
}
function notificationSummarySignalMetadata(summary) {
  const priority = highestPriority(summary.byPriority);
  return {
    signal: "summary",
    pending: summary.pending,
    groups: Object.entries(summary.bySource).sort(([a], [b]) => a.localeCompare(b)).map(([source, count]) => ({ source, count })),
    byPriority: summary.byPriority,
    notificationIds: summary.notificationIds,
    ...priority ? { priority } : {}
  };
}
function createNotificationSignal(notification) {
  return createSignal({
    type: "notification",
    tagName: "notification",
    contents: notification.summary,
    attributes: notificationSignalAttributes(notification),
    metadata: { ...notification.metadata, notification: notificationSignalMetadata(notification) }
  });
}
function createNotificationSummarySignal(summary) {
  const notification = notificationSummarySignalMetadata(summary);
  return createSignal({
    type: "notification",
    tagName: "notification-summary",
    contents: notificationSummaryContents(summary),
    attributes: {
      pending: summary.pending,
      ...notification.priority ? { priority: notification.priority } : {}
    },
    metadata: { notification, notificationSummary: summary, notificationIds: summary.notificationIds }
  });
}
function summarizeNotifications(notifications) {
  const pendingNotifications = notifications.filter((notification) => notification.status === "pending");
  const first = pendingNotifications[0] ?? notifications[0];
  return pendingNotifications.reduce(
    (summary, notification) => {
      summary.pending += 1;
      summary.bySource[notification.source] = (summary.bySource[notification.source] ?? 0) + 1;
      summary.byPriority[notification.priority] = (summary.byPriority[notification.priority] ?? 0) + 1;
      summary.notificationIds.push(notification.id);
      return summary;
    },
    {
      threadId: first?.threadId ?? "",
      resourceId: first?.resourceId,
      agentId: first?.agentId,
      pending: 0,
      bySource: {},
      byPriority: {},
      notificationIds: []
    }
  );
}

// src/notifications/dispatcher.ts
var errorMessage = (error) => error instanceof Error ? error.message : String(error);
var isSummaryDue = (record, now) => Boolean(record.summaryAt && record.summaryAt.getTime() <= now.getTime());
var deliveryPriority = {
  urgent: 0,
  high: 1,
  medium: 2,
  low: 3
};
var compareDueDispatchItems = (a, b) => deliveryPriority[a.priority] - deliveryPriority[b.priority] || a.createdAt.getTime() - b.createdAt.getTime();
var getHighestPriority = (records) => records.reduce(
  (highest, record) => deliveryPriority[record.priority] < deliveryPriority[highest] ? record.priority : highest,
  "low"
);
var getEarliestCreatedAt = (records) => records.reduce(
  (earliest, record) => record.createdAt.getTime() < earliest.getTime() ? record.createdAt : earliest,
  records[0].createdAt
);
var groupKey = (record) => {
  if (!record.agentId || !record.resourceId || !record.threadId) return void 0;
  return [record.agentId, record.resourceId, record.threadId].join("\0");
};
async function recordDeliveryFailure({
  storage,
  record,
  now,
  error
}) {
  await storage.updateNotification({
    id: record.id,
    threadId: record.threadId,
    deliveryAttempts: (record.deliveryAttempts ?? 0) + 1,
    lastDeliveryAttemptAt: now,
    lastDeliveryError: errorMessage(error)
  });
}
async function sendNotificationRecord({
  mastra,
  storage,
  record,
  now,
  batchThreadState
}) {
  const current = await storage.getNotification({ threadId: record.threadId, id: record.id });
  if (!current || current.status !== "pending" || current.deliveredSignalId) return null;
  if (!current.agentId) throw new Error(`Notification ${current.id} is missing agentId`);
  if (!current.resourceId) throw new Error(`Notification ${current.id} is missing resourceId`);
  const agent = await mastra.getAgentById(current.agentId);
  if (current.priority === "high" && current.summarySignalId) {
    const threadState = batchThreadState ?? agentThreadStreamRuntime.getThreadState(
      { resourceId: current.resourceId, threadId: current.threadId },
      agent.getPubSub?.()
    );
    if (threadState === "active") return null;
  }
  const signal = createNotificationSignal({
    ...current,
    status: "delivered",
    deliveredAt: now});
  const target = { resourceId: current.resourceId, threadId: current.threadId };
  const result = agent.sendSignal(signal, target);
  await result.persisted;
  if (!result.accepted) {
    throw new Error(`Notification ${current.id} signal was rejected`);
  }
  const updated = await storage.updateNotification({
    id: current.id,
    threadId: current.threadId,
    status: "delivered",
    deliveredSignalId: result.signal.id,
    lastDeliveryAttemptAt: now
  });
  return { record: updated, signal: result.signal };
}
async function sendNotificationSummary({
  mastra,
  storage,
  records,
  now
}) {
  const first = records[0];
  if (!first?.agentId) throw new Error("Notification summary is missing agentId");
  if (!first.resourceId) throw new Error("Notification summary is missing resourceId");
  const agent = await mastra.getAgentById(first.agentId);
  const summary = summarizeNotifications(records);
  const signal = createNotificationSummarySignal(summary);
  const target = records.every((record) => record.priority === "low") ? { resourceId: first.resourceId, threadId: first.threadId, ifIdle: { behavior: "persist" } } : { resourceId: first.resourceId, threadId: first.threadId };
  const result = agent.sendSignal(signal, target);
  await result.persisted;
  if (!result.accepted) {
    throw new Error(`Notification summary for thread ${first.threadId} was rejected`);
  }
  const updatedRecords = [];
  for (const record of records) {
    updatedRecords.push(
      await storage.updateNotification({
        id: record.id,
        threadId: record.threadId,
        summaryAt: null,
        summarySignalId: result.signal.id,
        lastDeliveryAttemptAt: now
      })
    );
  }
  return { records: updatedRecords, signal: result.signal };
}
async function getBatchThreadState({
  mastra,
  group
}) {
  const agent = await mastra.getAgentById(group.agentId);
  return agentThreadStreamRuntime.getThreadState(
    { resourceId: group.resourceId, threadId: group.threadId },
    agent.getPubSub?.()
  );
}
async function dispatchDueNotifications({
  mastra,
  storage,
  now = /* @__PURE__ */ new Date(),
  limit = 100
}) {
  const due = await storage.listDueNotifications({ now, limit });
  const delivered = [];
  const failed = [];
  const signals = [];
  const groups = /* @__PURE__ */ new Map();
  const ungroupedIndividual = [];
  for (const record of due) {
    const key = groupKey(record);
    if (!key) {
      if (isSummaryDue(record, now)) {
        const error = new Error(
          `Notification ${record.id} cannot be summarized without agentId, resourceId, and threadId`
        );
        await recordDeliveryFailure({ storage, record, now, error });
        failed.push({ record, error: error.message });
      } else {
        ungroupedIndividual.push(record);
      }
      continue;
    }
    const group = groups.get(key) ?? {
      key,
      agentId: record.agentId,
      resourceId: record.resourceId,
      threadId: record.threadId,
      summaryRecords: [],
      individualRecords: []
    };
    if (isSummaryDue(record, now)) {
      group.summaryRecords.push(record);
    } else {
      group.individualRecords.push(record);
    }
    groups.set(key, group);
  }
  for (const group of groups.values()) {
    const records = [...group.summaryRecords, ...group.individualRecords];
    let batchThreadState;
    try {
      batchThreadState = await getBatchThreadState({ mastra, group });
    } catch (error) {
      for (const record of records) {
        await recordDeliveryFailure({ storage, record, now, error });
        failed.push({ record, error: errorMessage(error) });
      }
      continue;
    }
    const items = group.individualRecords.map((record) => ({
      type: "individual",
      record,
      priority: record.priority,
      createdAt: record.createdAt
    }));
    if (group.summaryRecords.length > 0) {
      items.push({
        type: "summary",
        records: group.summaryRecords,
        priority: getHighestPriority(group.summaryRecords),
        createdAt: getEarliestCreatedAt(group.summaryRecords)
      });
    }
    items.sort(compareDueDispatchItems);
    for (const item of items) {
      if (item.type === "summary") {
        try {
          const result = await sendNotificationSummary({ mastra, storage, records: item.records, now });
          delivered.push(...result.records);
          signals.push(result.signal);
        } catch (error) {
          for (const record of item.records) {
            await recordDeliveryFailure({ storage, record, now, error });
            failed.push({ record, error: errorMessage(error) });
          }
        }
        continue;
      }
      try {
        const result = await sendNotificationRecord({ mastra, storage, record: item.record, now, batchThreadState });
        if (!result) continue;
        delivered.push(result.record);
        signals.push(result.signal);
      } catch (error) {
        await recordDeliveryFailure({ storage, record: item.record, now, error });
        failed.push({ record: item.record, error: errorMessage(error) });
      }
    }
  }
  ungroupedIndividual.sort(
    (a, b) => compareDueDispatchItems(
      { priority: a.priority, createdAt: a.createdAt },
      { priority: b.priority, createdAt: b.createdAt }
    )
  );
  for (const record of ungroupedIndividual) {
    try {
      const result = await sendNotificationRecord({ mastra, storage, record, now });
      if (!result) continue;
      delivered.push(result.record);
      signals.push(result.signal);
    } catch (error) {
      await recordDeliveryFailure({ storage, record, now, error });
      failed.push({ record, error: errorMessage(error) });
    }
  }
  return { delivered, failed, signals };
}
var NotificationsStorage = class extends StorageDomain {
  constructor() {
    super({ component: "STORAGE", name: "NOTIFICATIONS" });
  }
};
var cloneDate = (value) => value ? new Date(value) : void 0;
var notificationKey = (threadId, id) => `${threadId}\0${id}`;
var cloneValue = (value) => value === void 0 ? void 0 : structuredClone(value);
var cloneRecord = (record) => ({
  ...record,
  createdAt: new Date(record.createdAt),
  updatedAt: new Date(record.updatedAt),
  deliveredAt: cloneDate(record.deliveredAt),
  seenAt: cloneDate(record.seenAt),
  dismissedAt: cloneDate(record.dismissedAt),
  archivedAt: cloneDate(record.archivedAt),
  discardedAt: cloneDate(record.discardedAt),
  deliverAt: cloneDate(record.deliverAt),
  summaryAt: cloneDate(record.summaryAt),
  lastDeliveryAttemptAt: cloneDate(record.lastDeliveryAttemptAt),
  payload: cloneValue(record.payload),
  attributes: cloneValue(record.attributes),
  metadata: cloneValue(record.metadata)
});
var statusTimestamp = (status, now) => {
  if (status === "delivered") return { deliveredAt: now };
  if (status === "seen") return { seenAt: now };
  if (status === "dismissed") return { dismissedAt: now };
  if (status === "archived") return { archivedAt: now };
  if (status === "discarded") return { discardedAt: now };
  return {};
};
var valueMatches = (value, filter) => {
  if (!filter) return true;
  return Array.isArray(filter) ? filter.includes(value) : value === filter;
};
var dueTime = (record) => {
  const deliverAt = record.deliverAt?.getTime();
  const summaryAt = record.summaryAt?.getTime();
  if (deliverAt !== void 0 && summaryAt !== void 0) return Math.min(deliverAt, summaryAt);
  return deliverAt ?? summaryAt ?? Number.POSITIVE_INFINITY;
};
var InMemoryNotificationsStorage = class extends NotificationsStorage {
  #notifications = /* @__PURE__ */ new Map();
  async createNotification(input) {
    const existing = this.findCoalescable(input);
    if (existing) {
      const now2 = /* @__PURE__ */ new Date();
      const next = {
        ...existing,
        summary: input.summary,
        payload: cloneValue(input.payload ?? existing.payload),
        priority: input.priority ?? existing.priority,
        attributes: input.attributes ? { ...cloneValue(existing.attributes), ...cloneValue(input.attributes) } : cloneValue(existing.attributes),
        updatedAt: now2,
        deliverAt: input.deliverAt ?? existing.deliverAt,
        summaryAt: input.summaryAt ?? existing.summaryAt,
        deliveryReason: input.deliveryReason ?? existing.deliveryReason,
        coalescedCount: (existing.coalescedCount ?? 1) + 1,
        metadata: input.metadata ? { ...cloneValue(existing.metadata), ...cloneValue(input.metadata) } : cloneValue(existing.metadata)
      };
      this.#notifications.set(notificationKey(next.threadId, next.id), next);
      return cloneRecord(next);
    }
    const now = input.createdAt ?? /* @__PURE__ */ new Date();
    const record = {
      id: input.id ?? randomUUID(),
      threadId: input.threadId,
      source: input.source,
      kind: input.kind,
      priority: input.priority ?? "medium",
      status: "pending",
      summary: input.summary,
      payload: cloneValue(input.payload),
      resourceId: input.resourceId,
      agentId: input.agentId,
      sourceId: input.sourceId,
      dedupeKey: input.dedupeKey,
      coalesceKey: input.coalesceKey,
      coalescedCount: 1,
      attributes: cloneValue(input.attributes),
      createdAt: now,
      updatedAt: now,
      deliverAt: input.deliverAt,
      summaryAt: input.summaryAt,
      deliveryReason: input.deliveryReason,
      deliveryAttempts: 0,
      metadata: cloneValue(input.metadata)
    };
    this.#notifications.set(notificationKey(record.threadId, record.id), record);
    return cloneRecord(record);
  }
  async listNotifications(input) {
    const search = input.search?.toLowerCase();
    const results = [...this.#notifications.values()].filter((record) => record.threadId === input.threadId).filter((record) => valueMatches(record.status, input.status)).filter((record) => valueMatches(record.priority, input.priority)).filter((record) => !input.source || record.source === input.source).filter((record) => !input.resourceId || record.resourceId === input.resourceId).filter((record) => !input.agentId || record.agentId === input.agentId).filter(
      (record) => !search || record.summary.toLowerCase().includes(search) || record.kind.toLowerCase().includes(search) || record.source.toLowerCase().includes(search)
    ).sort((a, b) => b.updatedAt.getTime() - a.updatedAt.getTime());
    return results.slice(0, input.limit ?? results.length).map(cloneRecord);
  }
  async listDueNotifications(input) {
    const now = input.now.getTime();
    const results = [...this.#notifications.values()].filter((record) => record.status === "pending").filter((record) => !input.agentId || record.agentId === input.agentId).filter((record) => !input.resourceId || record.resourceId === input.resourceId).filter((record) => dueTime(record) <= now).sort((a, b) => dueTime(a) - dueTime(b) || a.updatedAt.getTime() - b.updatedAt.getTime());
    return results.slice(0, input.limit ?? results.length).map(cloneRecord);
  }
  async getNotification(input) {
    const record = this.#notifications.get(notificationKey(input.threadId, input.id));
    if (!record) return null;
    return cloneRecord(record);
  }
  async updateNotification(input) {
    const existing = this.#notifications.get(notificationKey(input.threadId, input.id));
    if (!existing) {
      throw new Error(`Notification ${input.id} was not found for thread ${input.threadId}`);
    }
    const now = /* @__PURE__ */ new Date();
    const next = {
      ...existing,
      ...input.status ? { status: input.status, ...statusTimestamp(input.status, now) } : {},
      ...input.summary !== void 0 ? { summary: input.summary } : {},
      ...input.payload !== void 0 ? { payload: cloneValue(input.payload) } : {},
      ...input.attributes !== void 0 ? { attributes: cloneValue(input.attributes) } : {},
      ...input.metadata !== void 0 ? { metadata: cloneValue(input.metadata) } : {},
      ...input.deliverAt !== void 0 ? { deliverAt: input.deliverAt ?? void 0 } : {},
      ...input.summaryAt !== void 0 ? { summaryAt: input.summaryAt ?? void 0 } : {},
      ...input.deliveryReason !== void 0 ? { deliveryReason: input.deliveryReason } : {},
      ...input.deliveryAttempts !== void 0 ? { deliveryAttempts: input.deliveryAttempts } : {},
      ...input.lastDeliveryAttemptAt !== void 0 ? { lastDeliveryAttemptAt: input.lastDeliveryAttemptAt } : {},
      ...input.lastDeliveryError !== void 0 ? { lastDeliveryError: input.lastDeliveryError } : {},
      ...input.deliveredSignalId !== void 0 ? { deliveredSignalId: input.deliveredSignalId } : {},
      ...input.summarySignalId !== void 0 ? { summarySignalId: input.summarySignalId } : {},
      updatedAt: now
    };
    this.#notifications.set(notificationKey(next.threadId, next.id), next);
    return cloneRecord(next);
  }
  async dangerouslyClearAll() {
    this.#notifications.clear();
  }
  findCoalescable(input) {
    if (!input.dedupeKey && !input.coalesceKey) return void 0;
    return [...this.#notifications.values()].find((record) => {
      if (record.threadId !== input.threadId || record.source !== input.source || record.kind !== input.kind || record.status !== "pending")
        return false;
      if (record.agentId !== input.agentId || record.resourceId !== input.resourceId) return false;
      return Boolean(
        input.dedupeKey && record.dedupeKey === input.dedupeKey || input.coalesceKey && record.coalesceKey === input.coalesceKey
      );
    });
  }
};

export { InMemoryNotificationsStorage, NotificationsStorage, agentThreadStreamRuntime, createNotificationSignal, createNotificationSummarySignal, dispatchDueNotifications, notificationSignalAttributes, notificationSignalMetadata, notificationSummaryContents, notificationSummarySignalMetadata, summarizeNotifications };
