// @mastra/core/dist/chunk-RDZCRQ74.cjs
'use strict';

var chunk7DI6OU5T_cjs = require('./chunk-7DI6OU5T.cjs');

// src/tools/toolchecks.ts
function isMastraTool(tool) {
  return tool instanceof chunk7DI6OU5T_cjs.Tool || typeof tool === "object" && tool !== null && chunk7DI6OU5T_cjs.MASTRA_TOOL_MARKER in tool;
}
function isVercelTool(tool) {
  return !!(tool && !isMastraTool(tool) && ("parameters" in tool || "execute" in tool && typeof tool.execute === "function" && "inputSchema" in tool));
}
function isProviderDefinedTool(tool) {
  if (typeof tool !== "object" || tool === null) return false;
  const t = tool;
  const isProviderType = t.type === "provider-defined" || t.type === "provider";
  return isProviderType && typeof t.id === "string";
}
var isProviderTool = isProviderDefinedTool;
function getProviderToolName(providerId) {
  return providerId.split(".").slice(1).join(".");
}
function getNeedsApprovalFn(tool) {
  if (typeof tool !== "object" || tool === null) return void 0;
  const fn = tool.needsApprovalFn;
  return typeof fn === "function" ? fn : void 0;
}

exports.getNeedsApprovalFn = getNeedsApprovalFn;
exports.getProviderToolName = getProviderToolName;
exports.isMastraTool = isMastraTool;
exports.isProviderDefinedTool = isProviderDefinedTool;
exports.isProviderTool = isProviderTool;
exports.isVercelTool = isVercelTool;
