// @mastra/core/dist/fga-check-FMCKOSEK.cjs
'use strict';

var chunkETRMQ4H5_cjs = require('./chunk-ETRMQ4H5.cjs');



Object.defineProperty(exports, "FGADeniedError", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.FGADeniedError; }
});
Object.defineProperty(exports, "checkFGA", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.checkFGA; }
});
Object.defineProperty(exports, "getAgentFGAResourceId", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.getAgentFGAResourceId; }
});
Object.defineProperty(exports, "getAgentToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.getAgentToolFGAResourceId; }
});
Object.defineProperty(exports, "getMCPToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.getMCPToolFGAResourceId; }
});
Object.defineProperty(exports, "getStandaloneToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.getStandaloneToolFGAResourceId; }
});
Object.defineProperty(exports, "getWorkflowFGAResourceId", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.getWorkflowFGAResourceId; }
});
Object.defineProperty(exports, "requireFGA", {
  enumerable: true,
  get: function () { return chunkETRMQ4H5_cjs.requireFGA; }
});
