// @mastra/core/dist/provider-registry-2GTS6YNM.cjs
'use strict';

var chunkM37OYAPC_cjs = require('./chunk-M37OYAPC.cjs');



Object.defineProperty(exports, "GatewayRegistry", {
  enumerable: true,
  get: function () { return chunkM37OYAPC_cjs.GatewayRegistry; }
});
Object.defineProperty(exports, "PROVIDER_MODELS", {
  enumerable: true,
  get: function () { return chunkM37OYAPC_cjs.PROVIDER_MODELS; }
});
Object.defineProperty(exports, "PROVIDER_REGISTRY", {
  enumerable: true,
  get: function () { return chunkM37OYAPC_cjs.PROVIDER_REGISTRY; }
});
Object.defineProperty(exports, "getProviderConfig", {
  enumerable: true,
  get: function () { return chunkM37OYAPC_cjs.getProviderConfig; }
});
Object.defineProperty(exports, "getRegisteredProviders", {
  enumerable: true,
  get: function () { return chunkM37OYAPC_cjs.getRegisteredProviders; }
});
Object.defineProperty(exports, "isOfflineMode", {
  enumerable: true,
  get: function () { return chunkM37OYAPC_cjs.isOfflineMode; }
});
Object.defineProperty(exports, "isProviderRegistered", {
  enumerable: true,
  get: function () { return chunkM37OYAPC_cjs.isProviderRegistered; }
});
Object.defineProperty(exports, "isValidModelId", {
  enumerable: true,
  get: function () { return chunkM37OYAPC_cjs.isValidModelId; }
});
Object.defineProperty(exports, "modelSupportsAttachments", {
  enumerable: true,
  get: function () { return chunkM37OYAPC_cjs.modelSupportsAttachments; }
});
Object.defineProperty(exports, "parseModelString", {
  enumerable: true,
  get: function () { return chunkM37OYAPC_cjs.parseModelString; }
});
