// @mastra/memory/dist/processors/index.cjs
'use strict';

var chunk4TZ5SRFX_cjs = require('../chunk-4TZ5SRFX.cjs');
var chunkD4J4XPGM_cjs = require('../chunk-D4J4XPGM.cjs');



Object.defineProperty(exports, "ModelByInputTokens", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.ModelByInputTokens; }
});
Object.defineProperty(exports, "OBSERVER_SYSTEM_PROMPT", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.OBSERVER_SYSTEM_PROMPT; }
});
Object.defineProperty(exports, "ObservationalMemory", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.ObservationalMemory; }
});
Object.defineProperty(exports, "ObservationalMemoryProcessor", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.ObservationalMemoryProcessor; }
});
Object.defineProperty(exports, "TokenCounter", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.TokenCounter; }
});
Object.defineProperty(exports, "buildObserverPrompt", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.buildObserverPrompt; }
});
Object.defineProperty(exports, "buildObserverSystemPrompt", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.buildObserverSystemPrompt; }
});
Object.defineProperty(exports, "combineObservationGroupRanges", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.combineObservationGroupRanges; }
});
Object.defineProperty(exports, "deriveObservationGroupProvenance", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.deriveObservationGroupProvenance; }
});
Object.defineProperty(exports, "extractCurrentTask", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.extractCurrentTask; }
});
Object.defineProperty(exports, "formatMessagesForObserver", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.formatMessagesForObserver; }
});
Object.defineProperty(exports, "getObservationsAsOf", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.getObservationsAsOf; }
});
Object.defineProperty(exports, "hasCurrentTaskSection", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.hasCurrentTaskSection; }
});
Object.defineProperty(exports, "injectAnchorIds", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.injectAnchorIds; }
});
Object.defineProperty(exports, "optimizeObservationsForContext", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.optimizeObservationsForContext; }
});
Object.defineProperty(exports, "parseAnchorId", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.parseAnchorId; }
});
Object.defineProperty(exports, "parseObservationGroups", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.parseObservationGroups; }
});
Object.defineProperty(exports, "parseObserverOutput", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.parseObserverOutput; }
});
Object.defineProperty(exports, "reconcileObservationGroupsFromReflection", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.reconcileObservationGroupsFromReflection; }
});
Object.defineProperty(exports, "renderObservationGroupsForReflection", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.renderObservationGroupsForReflection; }
});
Object.defineProperty(exports, "stripEphemeralAnchorIds", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.stripEphemeralAnchorIds; }
});
Object.defineProperty(exports, "stripObservationGroups", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.stripObservationGroups; }
});
Object.defineProperty(exports, "wrapInObservationGroup", {
  enumerable: true,
  get: function () { return chunk4TZ5SRFX_cjs.wrapInObservationGroup; }
});
Object.defineProperty(exports, "OBSERVATIONAL_MEMORY_DEFAULTS", {
  enumerable: true,
  get: function () { return chunkD4J4XPGM_cjs.OBSERVATIONAL_MEMORY_DEFAULTS; }
});
Object.defineProperty(exports, "OBSERVATION_CONTEXT_INSTRUCTIONS", {
  enumerable: true,
  get: function () { return chunkD4J4XPGM_cjs.OBSERVATION_CONTEXT_INSTRUCTIONS; }
});
Object.defineProperty(exports, "OBSERVATION_CONTEXT_PROMPT", {
  enumerable: true,
  get: function () { return chunkD4J4XPGM_cjs.OBSERVATION_CONTEXT_PROMPT; }
});
Object.defineProperty(exports, "OBSERVATION_CONTINUATION_HINT", {
  enumerable: true,
  get: function () { return chunkD4J4XPGM_cjs.OBSERVATION_CONTINUATION_HINT; }
});
