// @mastra/server/dist/chunk-4BJSE3LM.js
import { providerSchema } from './chunk-4JILES3G.js';
import { z } from 'zod/v4';

var agentFeaturesSchema = z.object({
  tools: z.boolean().optional(),
  agents: z.boolean().optional(),
  workflows: z.boolean().optional(),
  scorers: z.boolean().optional(),
  skills: z.boolean().optional(),
  memory: z.boolean().optional(),
  variables: z.boolean().optional(),
  favorites: z.boolean().optional(),
  avatarUpload: z.boolean().optional(),
  browser: z.boolean().optional(),
  /**
   * Whether the model picker is visible in the Agent Builder.
   * Omitted ⇒ picker visible (default-on). Explicit `false` ⇒ picker hidden
   * (locked mode); `models.default` is required and applied.
   */
  model: z.boolean().optional()
});
var knownProviderEntrySchema = z.object({
  provider: z.string().min(1),
  modelId: z.string().min(1).optional()
}).strict();
var customProviderEntrySchema = z.object({
  kind: z.literal("custom"),
  provider: z.string().min(1),
  modelId: z.string().min(1).optional()
}).strict();
var knownDefaultModelEntrySchema = z.object({
  provider: z.string().min(1),
  modelId: z.string().min(1)
}).strict();
var customDefaultModelEntrySchema = z.object({
  kind: z.literal("custom"),
  provider: z.string().min(1),
  modelId: z.string().min(1)
}).strict();
var providerModelEntrySchema = z.union([customProviderEntrySchema, knownProviderEntrySchema]);
var defaultModelEntrySchema = z.union([customDefaultModelEntrySchema, knownDefaultModelEntrySchema]);
var agentModelsSchema = z.object({
  allowed: z.array(providerModelEntrySchema).optional(),
  default: defaultModelEntrySchema.optional()
});
var pickerAllowlistSchema = z.object({
  allowed: z.array(z.string()).optional()
}).strict();
var agentConfigurationSchema = z.object({
  models: agentModelsSchema.optional(),
  tools: pickerAllowlistSchema.optional(),
  agents: pickerAllowlistSchema.optional(),
  workflows: pickerAllowlistSchema.optional()
}).catchall(z.unknown());
var builderPickerSchema = z.object({
  visibleTools: z.array(z.string()).nullable(),
  visibleAgents: z.array(z.string()).nullable(),
  visibleWorkflows: z.array(z.string()).nullable()
});
var builderModelPolicySchema = z.object({
  active: z.boolean(),
  pickerVisible: z.boolean().optional(),
  allowed: z.array(providerModelEntrySchema).optional(),
  default: defaultModelEntrySchema.optional()
});
var builderSettingsResponseSchema = z.object({
  enabled: z.boolean(),
  features: z.object({
    agent: agentFeaturesSchema.optional()
  }).optional(),
  configuration: z.object({
    agent: agentConfigurationSchema.optional()
  }).optional(),
  modelPolicy: builderModelPolicySchema.optional(),
  /**
   * Resolved picker visibility for tools/agents/workflows. Always present when
   * the builder is enabled. Omitted when the builder is disabled.
   */
  picker: builderPickerSchema.optional(),
  /**
   * Non-fatal warnings produced by `EditorAgentBuilder`'s constructor-time
   * validation (e.g. allowlist entries with unknown provider strings, or
   * picker allowlist entries that don't match a registered ID). UI surfaces
   * these as a banner in the Builder admin view.
   */
  modelPolicyWarnings: z.array(z.string()).optional()
});
var infrastructureStatusResponseSchema = z.object({
  channels: z.object({
    providers: z.array(
      z.object({
        id: z.string(),
        name: z.string(),
        isConfigured: z.boolean(),
        routeCount: z.number()
      })
    )
  }),
  browser: z.object({
    type: z.string().nullable(),
    provider: z.string().nullable(),
    env: z.string().nullable(),
    registered: z.boolean(),
    availableProviders: z.array(z.string()),
    config: z.array(z.object({ key: z.string(), value: z.string() }))
  }),
  workspace: z.object({
    type: z.string().nullable(),
    workspaceId: z.string().nullable(),
    name: z.string().nullable(),
    source: z.string().nullable(),
    registered: z.boolean(),
    hasFilesystem: z.boolean(),
    hasSandbox: z.boolean(),
    filesystemProvider: z.string().nullable(),
    sandboxProvider: z.string().nullable(),
    config: z.array(z.object({ key: z.string(), value: z.string() }))
  }),
  registries: z.object({
    skillsSh: z.object({
      enabled: z.boolean()
    })
  })
});
var builderAvailableModelsResponseSchema = z.object({
  providers: z.array(providerSchema)
});

export { agentConfigurationSchema, agentFeaturesSchema, agentModelsSchema, builderAvailableModelsResponseSchema, builderModelPolicySchema, builderPickerSchema, builderSettingsResponseSchema, defaultModelEntrySchema, infrastructureStatusResponseSchema, pickerAllowlistSchema, providerModelEntrySchema };
