// @mastra/server/dist/chunk-WIOYCLJC.cjs
'use strict';

var chunkDIG2K5CV_cjs = require('./chunk-DIG2K5CV.cjs');
var v4 = require('zod/v4');

var storedMCPClientIdPathParams = v4.z.object({
  storedMCPClientId: v4.z.string().describe("Unique identifier for the stored MCP client")
});
var storageOrderBySchema = v4.z.object({
  field: v4.z.enum(["createdAt", "updatedAt"]).optional(),
  direction: v4.z.enum(["ASC", "DESC"]).optional()
});
var listStoredMCPClientsQuerySchema = chunkDIG2K5CV_cjs.createPagePaginationSchema(100).extend({
  orderBy: storageOrderBySchema.optional(),
  status: v4.z.enum(["draft", "published", "archived"]).optional().default("published").describe("Filter MCP clients by status (defaults to published)"),
  authorId: v4.z.string().optional().describe("Filter MCP clients by author identifier"),
  metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional().describe("Filter MCP clients by metadata key-value pairs")
});
var mcpServerConfigSchema = v4.z.object({
  type: v4.z.enum(["stdio", "http"]).describe("Transport type: stdio for local processes, http for remote servers"),
  command: v4.z.string().optional().describe("Command to run (stdio only)"),
  args: v4.z.array(v4.z.string()).optional().describe("Command arguments (stdio only)"),
  env: v4.z.record(v4.z.string(), v4.z.string()).optional().describe("Environment variables (stdio only)"),
  url: v4.z.string().optional().describe("Server URL (http only)"),
  timeout: v4.z.number().optional().describe("Connection timeout in milliseconds")
});
var snapshotConfigSchema = v4.z.object({
  name: v4.z.string().describe("Name of the MCP client"),
  description: v4.z.string().optional().describe("Description of the MCP client"),
  servers: v4.z.record(v4.z.string(), mcpServerConfigSchema).describe("Map of server name to server configuration")
});
var createStoredMCPClientBodySchema = v4.z.object({
  id: v4.z.string().optional().describe("Unique identifier. If not provided, derived from name."),
  authorId: v4.z.string().optional().describe("Author identifier for multi-tenant filtering"),
  metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional().describe("Additional metadata for the MCP client")
}).merge(snapshotConfigSchema);
var updateStoredMCPClientBodySchema = v4.z.object({
  authorId: v4.z.string().optional(),
  metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional()
}).partial().merge(snapshotConfigSchema.partial());
var storedMCPClientSchema = v4.z.object({
  id: v4.z.string(),
  status: v4.z.string().describe("MCP client status: draft, published, or archived"),
  activeVersionId: v4.z.string().optional(),
  authorId: v4.z.string().optional(),
  metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional(),
  createdAt: v4.z.coerce.date(),
  updatedAt: v4.z.coerce.date(),
  name: v4.z.string().describe("Name of the MCP client"),
  description: v4.z.string().optional().describe("Description of the MCP client"),
  servers: v4.z.record(v4.z.string(), mcpServerConfigSchema).describe("Map of server name to server configuration")
});
var listStoredMCPClientsResponseSchema = chunkDIG2K5CV_cjs.paginationInfoSchema.extend({
  mcpClients: v4.z.array(storedMCPClientSchema)
});
var getStoredMCPClientResponseSchema = storedMCPClientSchema;
var createStoredMCPClientResponseSchema = storedMCPClientSchema;
var updateStoredMCPClientResponseSchema = v4.z.union([
  v4.z.object({
    id: v4.z.string(),
    status: v4.z.string(),
    activeVersionId: v4.z.string().optional(),
    authorId: v4.z.string().optional(),
    metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional(),
    createdAt: v4.z.coerce.date(),
    updatedAt: v4.z.coerce.date()
  }),
  storedMCPClientSchema
]);
var deleteStoredMCPClientResponseSchema = v4.z.object({
  success: v4.z.boolean(),
  message: v4.z.string()
});

exports.createStoredMCPClientBodySchema = createStoredMCPClientBodySchema;
exports.createStoredMCPClientResponseSchema = createStoredMCPClientResponseSchema;
exports.deleteStoredMCPClientResponseSchema = deleteStoredMCPClientResponseSchema;
exports.getStoredMCPClientResponseSchema = getStoredMCPClientResponseSchema;
exports.listStoredMCPClientsQuerySchema = listStoredMCPClientsQuerySchema;
exports.listStoredMCPClientsResponseSchema = listStoredMCPClientsResponseSchema;
exports.storedMCPClientIdPathParams = storedMCPClientIdPathParams;
exports.storedMCPClientSchema = storedMCPClientSchema;
exports.updateStoredMCPClientBodySchema = updateStoredMCPClientBodySchema;
exports.updateStoredMCPClientResponseSchema = updateStoredMCPClientResponseSchema;
