// @mastra/server/dist/chunk-PMGFP7II.cjs
'use strict';

var chunk47PVLKIL_cjs = require('./chunk-47PVLKIL.cjs');
var chunkO7I5CWRX_cjs = require('./chunk-O7I5CWRX.cjs');

// src/server/handlers/voice.ts
var voice_exports = {};
chunkO7I5CWRX_cjs.__export(voice_exports, {
  GENERATE_SPEECH_DEPRECATED_ROUTE: () => GENERATE_SPEECH_DEPRECATED_ROUTE2,
  GENERATE_SPEECH_ROUTE: () => GENERATE_SPEECH_ROUTE2,
  GET_LISTENER_ROUTE: () => GET_LISTENER_ROUTE2,
  GET_SPEAKERS_DEPRECATED_ROUTE: () => GET_SPEAKERS_DEPRECATED_ROUTE2,
  GET_SPEAKERS_ROUTE: () => GET_SPEAKERS_ROUTE2,
  TRANSCRIBE_SPEECH_DEPRECATED_ROUTE: () => TRANSCRIBE_SPEECH_DEPRECATED_ROUTE2,
  TRANSCRIBE_SPEECH_ROUTE: () => TRANSCRIBE_SPEECH_ROUTE2
});
var GET_SPEAKERS_ROUTE2 = chunk47PVLKIL_cjs.GET_SPEAKERS_ROUTE;
var GET_SPEAKERS_DEPRECATED_ROUTE2 = chunk47PVLKIL_cjs.GET_SPEAKERS_DEPRECATED_ROUTE;
var GENERATE_SPEECH_ROUTE2 = chunk47PVLKIL_cjs.GENERATE_SPEECH_ROUTE;
var GENERATE_SPEECH_DEPRECATED_ROUTE2 = chunk47PVLKIL_cjs.GENERATE_SPEECH_DEPRECATED_ROUTE;
var TRANSCRIBE_SPEECH_ROUTE2 = chunk47PVLKIL_cjs.TRANSCRIBE_SPEECH_ROUTE;
var TRANSCRIBE_SPEECH_DEPRECATED_ROUTE2 = chunk47PVLKIL_cjs.TRANSCRIBE_SPEECH_DEPRECATED_ROUTE;
var GET_LISTENER_ROUTE2 = chunk47PVLKIL_cjs.GET_LISTENER_ROUTE;

exports.GENERATE_SPEECH_DEPRECATED_ROUTE = GENERATE_SPEECH_DEPRECATED_ROUTE2;
exports.GENERATE_SPEECH_ROUTE = GENERATE_SPEECH_ROUTE2;
exports.GET_LISTENER_ROUTE = GET_LISTENER_ROUTE2;
exports.GET_SPEAKERS_DEPRECATED_ROUTE = GET_SPEAKERS_DEPRECATED_ROUTE2;
exports.GET_SPEAKERS_ROUTE = GET_SPEAKERS_ROUTE2;
exports.TRANSCRIBE_SPEECH_DEPRECATED_ROUTE = TRANSCRIBE_SPEECH_DEPRECATED_ROUTE2;
exports.TRANSCRIBE_SPEECH_ROUTE = TRANSCRIBE_SPEECH_ROUTE2;
exports.voice_exports = voice_exports;
