// @mastra/server/dist/chunk-QQCQV7ZF.cjs
'use strict';

var chunkPFKYMKIH_cjs = require('./chunk-PFKYMKIH.cjs');
var chunkDIG2K5CV_cjs = require('./chunk-DIG2K5CV.cjs');
var v4 = require('zod/v4');

var jsonValueSchema = v4.z.lazy(
  () => v4.z.union([
    v4.z.string(),
    v4.z.number(),
    v4.z.boolean(),
    v4.z.null(),
    v4.z.array(jsonValueSchema),
    v4.z.record(v4.z.string(), jsonValueSchema)
  ])
);
var jsonRecordSchema = v4.z.record(v4.z.string(), jsonValueSchema);
var signalAttributesSchema = v4.z.record(
  v4.z.string(),
  v4.z.union([v4.z.string(), v4.z.number(), v4.z.boolean(), v4.z.null(), v4.z.undefined()])
);
var baseSignalSchema = v4.z.object({
  id: v4.z.string().optional(),
  createdAt: v4.z.union([v4.z.string(), v4.z.date()]).optional(),
  metadata: jsonRecordSchema.optional(),
  attributes: signalAttributesSchema.optional()
});
var partProviderOptionsSchema = v4.z.record(v4.z.string(), v4.z.record(v4.z.string(), jsonValueSchema)).optional();
var signalTextPartSchema = v4.z.object({
  type: v4.z.literal("text"),
  text: v4.z.string(),
  providerOptions: partProviderOptionsSchema
});
var signalFilePartSchema = v4.z.object({
  type: v4.z.literal("file"),
  data: v4.z.string(),
  mediaType: v4.z.string(),
  filename: v4.z.string().optional(),
  providerOptions: partProviderOptionsSchema
});
var userMessageSignalContentsSchema = v4.z.union([
  v4.z.string(),
  v4.z.array(v4.z.union([signalTextPartSchema, signalFilePartSchema]))
]);
var agentMessageInputObjectSchema = v4.z.object({
  contents: userMessageSignalContentsSchema,
  attributes: signalAttributesSchema.optional(),
  metadata: jsonRecordSchema.optional(),
  providerOptions: v4.z.record(v4.z.string(), v4.z.record(v4.z.string(), jsonValueSchema)).optional()
});
var agentMessageInputSchema = v4.z.union([userMessageSignalContentsSchema, agentMessageInputObjectSchema]);
var agentSignalSchema = baseSignalSchema.extend({
  type: v4.z.enum(["user", "state", "reactive", "notification", "user-message", "system-reminder"]),
  tagName: v4.z.string().optional(),
  contents: userMessageSignalContentsSchema,
  providerOptions: v4.z.record(v4.z.string(), v4.z.record(v4.z.string(), jsonValueSchema)).optional()
});
var agentIdPathParams = v4.z.object({
  agentId: v4.z.string().describe("Unique identifier for the agent")
});
var agentVersionQuerySchema = v4.z.object({
  status: v4.z.enum(["draft", "published"]).optional().describe(
    "Which stored config version to resolve: draft (latest, default) or published (active version). Mutually exclusive with versionId."
  ),
  versionId: v4.z.string().optional().describe(
    "Specific version ID to resolve. Mutually exclusive with status \u2014 if both are provided, versionId takes precedence."
  )
});
var toolIdPathParams = v4.z.object({
  toolId: v4.z.string().describe("Unique identifier for the tool")
});
var agentToolPathParams = agentIdPathParams.extend({
  toolId: v4.z.string().describe("Unique identifier for the tool")
});
var agentSkillPathParams = agentIdPathParams.extend({
  skillName: v4.z.string().describe("Name of the skill")
});
var modelConfigIdPathParams = agentIdPathParams.extend({
  modelConfigId: v4.z.string().describe("Unique identifier for the model configuration")
});
var serializedProcessorSchema = v4.z.object({
  id: v4.z.string(),
  name: v4.z.string().optional()
});
var serializedToolSchema = v4.z.object({
  id: v4.z.string(),
  description: v4.z.string().optional(),
  inputSchema: v4.z.string().optional(),
  outputSchema: v4.z.string().optional(),
  requireApproval: v4.z.boolean().optional()
});
var serializedWorkflowSchema = v4.z.object({
  name: v4.z.string(),
  steps: v4.z.record(
    v4.z.string(),
    v4.z.object({
      id: v4.z.string(),
      description: v4.z.string().optional()
    })
  ).optional()
});
var serializedAgentDefinitionSchema = v4.z.object({
  id: v4.z.string(),
  name: v4.z.string()
});
var systemMessageSchema = v4.z.union([
  v4.z.string(),
  v4.z.array(v4.z.string()),
  v4.z.any(),
  // CoreSystemMessage or SystemModelMessage
  v4.z.array(v4.z.any())
]);
var modelConfigSchema = v4.z.object({
  model: v4.z.object({
    modelId: v4.z.string(),
    provider: v4.z.string(),
    modelVersion: v4.z.string()
  })
  // Additional fields from AgentModelManagerConfig can be added here
});
var agentEditorConfigSchema = v4.z.union([
  v4.z.literal(false),
  v4.z.object({
    instructions: v4.z.boolean().optional(),
    tools: v4.z.union([v4.z.boolean(), v4.z.object({ description: v4.z.boolean().optional() })]).optional()
  })
]);
var serializedAgentSchema = v4.z.object({
  name: v4.z.string(),
  description: v4.z.string().optional(),
  metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional(),
  instructions: systemMessageSchema.optional(),
  tools: v4.z.record(v4.z.string(), serializedToolSchema),
  agents: v4.z.record(v4.z.string(), serializedAgentDefinitionSchema),
  workflows: v4.z.record(v4.z.string(), serializedWorkflowSchema),
  inputProcessors: v4.z.array(serializedProcessorSchema),
  outputProcessors: v4.z.array(serializedProcessorSchema),
  provider: v4.z.string().optional(),
  modelId: v4.z.string().optional(),
  modelVersion: v4.z.string().optional(),
  supportsMemory: v4.z.boolean().optional(),
  modelList: v4.z.array(modelConfigSchema).optional(),
  defaultOptions: chunkPFKYMKIH_cjs.defaultOptionsSchema.optional(),
  defaultGenerateOptionsLegacy: v4.z.record(v4.z.string(), v4.z.any()).optional(),
  defaultStreamOptionsLegacy: v4.z.record(v4.z.string(), v4.z.any()).optional(),
  source: v4.z.enum(["code", "stored"]).optional(),
  status: v4.z.enum(["draft", "published", "archived"]).optional(),
  activeVersionId: v4.z.string().optional(),
  hasDraft: v4.z.boolean().optional(),
  editor: agentEditorConfigSchema.optional()
});
var serializedAgentWithIdSchema = serializedAgentSchema.extend({
  id: v4.z.string()
});
var providerSchema = v4.z.object({
  id: v4.z.string(),
  name: v4.z.string(),
  label: v4.z.string().optional(),
  description: v4.z.string().optional(),
  envVar: v4.z.union([v4.z.string(), v4.z.array(v4.z.string())]),
  connected: v4.z.boolean(),
  docUrl: v4.z.string().optional(),
  models: v4.z.array(v4.z.string())
});
var providersResponseSchema = v4.z.object({
  providers: v4.z.array(providerSchema)
});
var listAgentsResponseSchema = v4.z.record(v4.z.string(), serializedAgentSchema);
var listToolsResponseSchema = v4.z.record(v4.z.string(), serializedToolSchema);
var agentMemoryOptionSchema = v4.z.object({
  thread: v4.z.union([v4.z.string(), v4.z.object({ id: v4.z.string() }).passthrough()]),
  resource: v4.z.string(),
  options: v4.z.record(v4.z.string(), v4.z.any()).optional(),
  readOnly: v4.z.boolean().optional()
});
var toolChoiceSchema = v4.z.union([
  v4.z.enum(["auto", "none", "required"]),
  v4.z.object({ type: v4.z.literal("tool"), toolName: v4.z.string() })
]);
var agentExecutionBodySchema = v4.z.object({
  // REQUIRED
  messages: v4.z.union([
    v4.z.array(chunkDIG2K5CV_cjs.coreMessageSchema),
    // Array of messages
    v4.z.string()
    // Single user message shorthand
  ]),
  // Message Configuration
  instructions: systemMessageSchema.optional(),
  system: systemMessageSchema.optional(),
  context: v4.z.array(chunkDIG2K5CV_cjs.coreMessageSchema).optional(),
  // Memory & Persistence
  memory: agentMemoryOptionSchema.optional(),
  runId: v4.z.string().optional(),
  savePerStep: v4.z.boolean().optional(),
  // Request Context (handler-specific field - merged with server's requestContext)
  requestContext: v4.z.record(v4.z.string(), v4.z.any()).optional(),
  // Version overrides for sub-agents (and future primitives)
  versions: v4.z.object({
    agents: v4.z.record(
      v4.z.string(),
      v4.z.union([v4.z.object({ versionId: v4.z.string() }), v4.z.object({ status: v4.z.enum(["draft", "published"]) })])
    ).optional(),
    defaultStatus: v4.z.enum(["draft", "published"]).optional()
  }).optional(),
  // Execution Control
  maxSteps: v4.z.number().optional(),
  stopWhen: v4.z.any().optional(),
  // Model Configuration
  providerOptions: v4.z.object({
    anthropic: v4.z.record(v4.z.string(), v4.z.any()).optional(),
    google: v4.z.record(v4.z.string(), v4.z.any()).optional(),
    openai: v4.z.record(v4.z.string(), v4.z.any()).optional(),
    xai: v4.z.record(v4.z.string(), v4.z.any()).optional()
  }).optional(),
  modelSettings: v4.z.any().optional(),
  // Tool Configuration
  activeTools: v4.z.array(v4.z.string()).optional(),
  toolsets: v4.z.record(v4.z.string(), v4.z.any()).optional(),
  clientTools: v4.z.record(v4.z.string(), v4.z.any()).optional(),
  toolChoice: toolChoiceSchema.optional(),
  requireToolApproval: v4.z.boolean().optional(),
  // Evaluation
  scorers: v4.z.union([
    v4.z.record(v4.z.string(), v4.z.any()),
    v4.z.record(
      v4.z.string(),
      v4.z.object({
        scorer: v4.z.string(),
        sampling: v4.z.any().optional()
      })
    )
  ]).optional(),
  returnScorerData: v4.z.boolean().optional(),
  // Observability
  tracingOptions: chunkDIG2K5CV_cjs.tracingOptionsSchema.optional(),
  // Structured Output
  output: v4.z.any().optional(),
  // Zod schema, JSON schema, or structured output object
  structuredOutput: v4.z.object({
    schema: v4.z.object({}).passthrough(),
    model: v4.z.union([v4.z.string(), v4.z.any()]).optional(),
    instructions: v4.z.string().optional(),
    jsonPromptInjection: v4.z.boolean().optional(),
    errorStrategy: v4.z.enum(["strict", "warn", "fallback"]).optional(),
    fallbackValue: v4.z.any().optional()
  }).optional(),
  // Idle-loop streaming (collapses streamUntilIdle into stream)
  untilIdle: v4.z.union([v4.z.boolean(), v4.z.object({ maxIdleMs: v4.z.number().int().positive().optional() })]).optional()
}).passthrough();
var agentExecutionLegacyBodySchema = agentExecutionBodySchema.extend({
  resourceId: v4.z.string().optional(),
  resourceid: v4.z.string().optional(),
  // lowercase variant
  threadId: v4.z.string().optional()
});
var streamUntilIdleBodySchema = agentExecutionBodySchema.extend({
  maxIdleMs: v4.z.number().int().positive().optional(),
  untilIdle: v4.z.union([v4.z.boolean(), v4.z.object({ maxIdleMs: v4.z.number().int().positive().optional() })]).optional()
});
var resumeStreamUntilIdleBodySchema = agentExecutionBodySchema.omit({ messages: true }).extend({
  runId: v4.z.string(),
  resumeData: v4.z.unknown().refine((x) => x !== void 0, { message: "resumeData is required" }),
  toolCallId: v4.z.string().optional(),
  maxIdleMs: v4.z.number().int().positive().optional()
});
var executeToolDataBodySchema = v4.z.object({
  data: v4.z.unknown().refine((x) => x !== void 0, { message: "data is required" })
});
var executeToolBodySchema = executeToolDataBodySchema.extend({
  requestContext: v4.z.record(v4.z.string(), v4.z.any()).optional()
});
var executeToolContextBodySchema = executeToolDataBodySchema.extend({
  requestContext: v4.z.record(v4.z.string(), v4.z.any()).optional()
});
var toolCallActionBodySchema = v4.z.object({
  runId: v4.z.string(),
  requestContext: v4.z.record(v4.z.string(), v4.z.any()).optional(),
  toolCallId: v4.z.string(),
  format: v4.z.string().optional()
});
var networkToolCallActionBodySchema = v4.z.object({
  runId: v4.z.string(),
  requestContext: v4.z.record(v4.z.string(), v4.z.any()).optional(),
  format: v4.z.string().optional()
});
var approveToolCallBodySchema = toolCallActionBodySchema;
var declineToolCallBodySchema = toolCallActionBodySchema;
var approveNetworkToolCallBodySchema = networkToolCallActionBodySchema;
var declineNetworkToolCallBodySchema = networkToolCallActionBodySchema;
var toolCallResponseSchema = v4.z.object({
  fullStream: v4.z.any()
  // ReadableStream
});
var sendToolApprovalResponseSchema = v4.z.object({
  accepted: v4.z.literal(true),
  runId: v4.z.string(),
  toolCallId: v4.z.string().optional()
});
var resumeStreamBodySchema = agentExecutionBodySchema.omit({ messages: true }).extend({
  runId: v4.z.string(),
  resumeData: v4.z.unknown().refine((x) => x !== void 0, { message: "resumeData is required" }),
  toolCallId: v4.z.string().optional()
});
var updateAgentModelBodySchema = v4.z.object({
  modelId: v4.z.string(),
  provider: v4.z.string()
});
var reorderAgentModelListBodySchema = v4.z.object({
  reorderedModelIds: v4.z.array(v4.z.string())
});
var updateAgentModelInModelListBodySchema = v4.z.object({
  model: v4.z.object({
    modelId: v4.z.string(),
    provider: v4.z.string()
  }).optional(),
  maxRetries: v4.z.number().optional(),
  enabled: v4.z.boolean().optional()
});
var modelManagementResponseSchema = chunkDIG2K5CV_cjs.messageResponseSchema;
var generateResponseSchema = v4.z.any();
var streamResponseSchema = v4.z.any();
var executeToolResponseSchema = v4.z.any();
var enhanceInstructionsBodySchema = v4.z.object({
  instructions: v4.z.string().describe("The current agent instructions to enhance"),
  comment: v4.z.string().describe("User comment describing how to enhance the instructions")
});
var enhanceInstructionsResponseSchema = v4.z.object({
  explanation: v4.z.string().describe("Explanation of the changes made"),
  new_prompt: v4.z.string().describe("The enhanced instructions")
});
var observeAgentBodySchema = v4.z.object({
  runId: v4.z.string().describe("The run ID to observe/reconnect to"),
  offset: v4.z.number().optional().describe("Resume from this event index (0-based). If omitted, replays all events.")
});
var signalActiveBehaviorSchema = v4.z.enum(["deliver", "persist", "discard"]);
var signalIdleBehaviorSchema = v4.z.enum(["wake", "persist", "discard"]);
var signalTargetBaseBodySchema = v4.z.object({
  ifActive: v4.z.object({
    behavior: signalActiveBehaviorSchema.optional(),
    attributes: signalAttributesSchema.optional()
  }).optional()
});
var signalTargetBodySchema = v4.z.union([
  signalTargetBaseBodySchema.extend({
    runId: v4.z.string(),
    resourceId: v4.z.string().optional(),
    threadId: v4.z.string().optional(),
    ifIdle: v4.z.undefined().optional()
  }),
  signalTargetBaseBodySchema.extend({
    runId: v4.z.undefined().optional(),
    resourceId: v4.z.string(),
    threadId: v4.z.string(),
    ifIdle: v4.z.object({
      behavior: signalIdleBehaviorSchema.optional(),
      streamOptions: agentExecutionBodySchema.omit({ messages: true }).optional(),
      attributes: signalAttributesSchema.optional()
    }).optional()
  })
]);
var sendAgentSignalBodySchema = v4.z.union([
  signalTargetBodySchema.options[0].extend({ signal: agentSignalSchema }),
  signalTargetBodySchema.options[1].extend({ signal: agentSignalSchema })
]);
var sendAgentMessageBodySchema = v4.z.union([
  signalTargetBodySchema.options[0].extend({ message: agentMessageInputSchema }),
  signalTargetBodySchema.options[1].extend({ message: agentMessageInputSchema })
]);
var queueAgentMessageBodySchema = sendAgentMessageBodySchema;
var subscribeAgentThreadBodySchema = v4.z.object({
  resourceId: v4.z.string().optional(),
  threadId: v4.z.string()
});
var abortAgentThreadBodySchema = subscribeAgentThreadBodySchema;
var sendToolApprovalBodySchema = v4.z.object({
  resourceId: v4.z.string(),
  threadId: v4.z.string(),
  requestContext: v4.z.record(v4.z.string(), v4.z.any()).optional(),
  toolCallId: v4.z.string(),
  approved: v4.z.boolean(),
  format: v4.z.string().optional(),
  messages: v4.z.array(chunkDIG2K5CV_cjs.coreMessageSchema).optional(),
  streamOptions: v4.z.any().optional()
});
var abortAgentThreadResponseSchema = v4.z.object({
  aborted: v4.z.boolean()
});
var observeAgentResponseSchema = v4.z.any();

exports.abortAgentThreadBodySchema = abortAgentThreadBodySchema;
exports.abortAgentThreadResponseSchema = abortAgentThreadResponseSchema;
exports.agentExecutionBodySchema = agentExecutionBodySchema;
exports.agentExecutionLegacyBodySchema = agentExecutionLegacyBodySchema;
exports.agentIdPathParams = agentIdPathParams;
exports.agentSkillPathParams = agentSkillPathParams;
exports.agentToolPathParams = agentToolPathParams;
exports.agentVersionQuerySchema = agentVersionQuerySchema;
exports.approveNetworkToolCallBodySchema = approveNetworkToolCallBodySchema;
exports.approveToolCallBodySchema = approveToolCallBodySchema;
exports.declineNetworkToolCallBodySchema = declineNetworkToolCallBodySchema;
exports.declineToolCallBodySchema = declineToolCallBodySchema;
exports.enhanceInstructionsBodySchema = enhanceInstructionsBodySchema;
exports.enhanceInstructionsResponseSchema = enhanceInstructionsResponseSchema;
exports.executeToolBodySchema = executeToolBodySchema;
exports.executeToolContextBodySchema = executeToolContextBodySchema;
exports.executeToolResponseSchema = executeToolResponseSchema;
exports.generateResponseSchema = generateResponseSchema;
exports.listAgentsResponseSchema = listAgentsResponseSchema;
exports.listToolsResponseSchema = listToolsResponseSchema;
exports.modelConfigIdPathParams = modelConfigIdPathParams;
exports.modelManagementResponseSchema = modelManagementResponseSchema;
exports.observeAgentBodySchema = observeAgentBodySchema;
exports.observeAgentResponseSchema = observeAgentResponseSchema;
exports.providerSchema = providerSchema;
exports.providersResponseSchema = providersResponseSchema;
exports.queueAgentMessageBodySchema = queueAgentMessageBodySchema;
exports.reorderAgentModelListBodySchema = reorderAgentModelListBodySchema;
exports.resumeStreamBodySchema = resumeStreamBodySchema;
exports.resumeStreamUntilIdleBodySchema = resumeStreamUntilIdleBodySchema;
exports.sendAgentMessageBodySchema = sendAgentMessageBodySchema;
exports.sendAgentSignalBodySchema = sendAgentSignalBodySchema;
exports.sendToolApprovalBodySchema = sendToolApprovalBodySchema;
exports.sendToolApprovalResponseSchema = sendToolApprovalResponseSchema;
exports.serializedAgentDefinitionSchema = serializedAgentDefinitionSchema;
exports.serializedAgentSchema = serializedAgentSchema;
exports.serializedAgentWithIdSchema = serializedAgentWithIdSchema;
exports.serializedProcessorSchema = serializedProcessorSchema;
exports.serializedToolSchema = serializedToolSchema;
exports.serializedWorkflowSchema = serializedWorkflowSchema;
exports.streamResponseSchema = streamResponseSchema;
exports.streamUntilIdleBodySchema = streamUntilIdleBodySchema;
exports.subscribeAgentThreadBodySchema = subscribeAgentThreadBodySchema;
exports.toolCallResponseSchema = toolCallResponseSchema;
exports.toolIdPathParams = toolIdPathParams;
exports.updateAgentModelBodySchema = updateAgentModelBodySchema;
exports.updateAgentModelInModelListBodySchema = updateAgentModelInModelListBodySchema;
