// @mastra/server/dist/chunk-PWBN74RE.cjs
'use strict';

var chunkGFP7IMFR_cjs = require('./chunk-GFP7IMFR.cjs');
var chunkHVATI3FQ_cjs = require('./chunk-HVATI3FQ.cjs');
var v4 = require('zod/v4');

var listScorerVersionsQuerySchema = chunkGFP7IMFR_cjs.listVersionsQuerySchema;
var compareScorerVersionsQuerySchema = chunkGFP7IMFR_cjs.compareVersionsQuerySchema;
var createScorerVersionBodySchema = chunkGFP7IMFR_cjs.createVersionBodySchema;
var scorerVersionPathParams = v4.z.object({
  scorerId: v4.z.string().describe("Unique identifier for the stored scorer definition")
});
var scorerVersionIdPathParams = v4.z.object({
  scorerId: v4.z.string().describe("Unique identifier for the stored scorer definition"),
  versionId: v4.z.string().describe("Unique identifier for the version (UUID)")
});
var samplingConfigSchema = v4.z.union([
  v4.z.object({ type: v4.z.literal("none") }),
  v4.z.object({ type: v4.z.literal("ratio"), rate: v4.z.number().min(0).max(1) })
]);
var scorerTypeEnum = v4.z.enum([
  "llm-judge",
  "answer-relevancy",
  "answer-similarity",
  "bias",
  "context-precision",
  "context-relevance",
  "faithfulness",
  "hallucination",
  "noise-sensitivity",
  "prompt-alignment",
  "tool-call-accuracy",
  "toxicity"
]);
var scorerVersionSchema = v4.z.object({
  id: v4.z.string().describe("Unique identifier for the version (UUID)"),
  scorerDefinitionId: v4.z.string().describe("ID of the scorer this version belongs to"),
  versionNumber: v4.z.number().describe("Sequential version number (1, 2, 3, ...)"),
  // Snapshot config fields
  name: v4.z.string().describe("Name of the scorer"),
  description: v4.z.string().optional().describe("Description of the scorer"),
  type: scorerTypeEnum,
  model: chunkHVATI3FQ_cjs.modelConfigSchema.optional(),
  instructions: v4.z.string().optional(),
  scoreRange: v4.z.object({
    min: v4.z.number().optional(),
    max: v4.z.number().optional()
  }).optional(),
  presetConfig: v4.z.record(v4.z.string(), v4.z.unknown()).optional(),
  defaultSampling: samplingConfigSchema.optional(),
  // Version metadata
  changedFields: v4.z.array(v4.z.string()).optional().describe("Array of field names that changed from the previous version"),
  changeMessage: v4.z.string().optional().describe("Optional message describing the changes"),
  createdAt: v4.z.coerce.date().describe("When this version was created")
});
var listScorerVersionsResponseSchema = chunkGFP7IMFR_cjs.createListVersionsResponseSchema(scorerVersionSchema);
var getScorerVersionResponseSchema = scorerVersionSchema;
var createScorerVersionResponseSchema = scorerVersionSchema.partial().merge(
  v4.z.object({
    id: v4.z.string(),
    scorerDefinitionId: v4.z.string(),
    versionNumber: v4.z.number(),
    createdAt: v4.z.coerce.date()
  })
);
var activateScorerVersionResponseSchema = chunkGFP7IMFR_cjs.activateVersionResponseSchema;
var restoreScorerVersionResponseSchema = scorerVersionSchema;
var deleteScorerVersionResponseSchema = chunkGFP7IMFR_cjs.deleteVersionResponseSchema;
var compareScorerVersionsResponseSchema = chunkGFP7IMFR_cjs.createCompareVersionsResponseSchema(scorerVersionSchema);

exports.activateScorerVersionResponseSchema = activateScorerVersionResponseSchema;
exports.compareScorerVersionsQuerySchema = compareScorerVersionsQuerySchema;
exports.compareScorerVersionsResponseSchema = compareScorerVersionsResponseSchema;
exports.createScorerVersionBodySchema = createScorerVersionBodySchema;
exports.createScorerVersionResponseSchema = createScorerVersionResponseSchema;
exports.deleteScorerVersionResponseSchema = deleteScorerVersionResponseSchema;
exports.getScorerVersionResponseSchema = getScorerVersionResponseSchema;
exports.listScorerVersionsQuerySchema = listScorerVersionsQuerySchema;
exports.listScorerVersionsResponseSchema = listScorerVersionsResponseSchema;
exports.restoreScorerVersionResponseSchema = restoreScorerVersionResponseSchema;
exports.scorerVersionIdPathParams = scorerVersionIdPathParams;
exports.scorerVersionPathParams = scorerVersionPathParams;
exports.scorerVersionSchema = scorerVersionSchema;
