// @mastra/server/dist/chunk-PFKYMKIH.cjs
'use strict';

var v4 = require('zod/v4');

// src/server/schemas/default-options.ts
var defaultOptionsSchema = v4.z.object({
  /** Unique identifier for this execution run */
  runId: v4.z.string().optional(),
  /** Save messages incrementally after each stream step completes (default: false) */
  savePerStep: v4.z.boolean().optional(),
  /** Maximum number of steps to run */
  maxSteps: v4.z.number().optional(),
  /** Provider-specific options passed to the language model */
  /** Tools that are active for this execution (stored as tool IDs) */
  activeTools: v4.z.array(v4.z.string()).optional(),
  /** Maximum number of times processors can trigger a retry */
  maxProcessorRetries: v4.z.number().optional(),
  /** Tool selection strategy: 'auto', 'none', 'required', or specific tools */
  toolChoice: v4.z.union([
    v4.z.literal("auto"),
    v4.z.literal("none"),
    v4.z.literal("required"),
    v4.z.object({ type: v4.z.literal("tool"), toolName: v4.z.string() })
  ]).optional(),
  /** Model-specific settings like temperature, maxTokens, topP, etc. */
  modelSettings: v4.z.object({
    temperature: v4.z.number().optional(),
    maxTokens: v4.z.number().optional(),
    topP: v4.z.number().optional(),
    topK: v4.z.number().optional(),
    frequencyPenalty: v4.z.number().optional(),
    presencePenalty: v4.z.number().optional(),
    stopSequences: v4.z.array(v4.z.string()).optional(),
    seed: v4.z.number().optional(),
    maxRetries: v4.z.number().optional()
  }).optional(),
  /** Whether to return detailed scoring data in the response */
  returnScorerData: v4.z.boolean().optional(),
  /** Tracing options for starting new traces */
  tracingOptions: v4.z.object({
    traceName: v4.z.string().optional(),
    attributes: v4.z.record(v4.z.string(), v4.z.unknown()).optional(),
    spanId: v4.z.string().optional(),
    traceId: v4.z.string().optional()
  }).optional(),
  /** Require approval for all tool calls */
  requireToolApproval: v4.z.boolean().optional(),
  /** Automatically resume suspended tools */
  autoResumeSuspendedTools: v4.z.boolean().optional(),
  /** Maximum number of tool calls to execute concurrently */
  toolCallConcurrency: v4.z.number().optional(),
  /** Whether to include raw chunks in the stream output */
  includeRawChunks: v4.z.boolean().optional()
}).passthrough().describe("Default options for agent execution");

exports.defaultOptionsSchema = defaultOptionsSchema;
