// @mastra/server/dist/chunk-TXM2XTUT.cjs
'use strict';

var chunkGFP7IMFR_cjs = require('./chunk-GFP7IMFR.cjs');
var chunk24YYB3J7_cjs = require('./chunk-24YYB3J7.cjs');
var v4 = require('zod/v4');

var listPromptBlockVersionsQuerySchema = chunkGFP7IMFR_cjs.listVersionsQuerySchema;
var comparePromptBlockVersionsQuerySchema = chunkGFP7IMFR_cjs.compareVersionsQuerySchema;
var createPromptBlockVersionBodySchema = chunkGFP7IMFR_cjs.createVersionBodySchema;
var promptBlockVersionPathParams = v4.z.object({
  promptBlockId: v4.z.string().describe("Unique identifier for the stored prompt block")
});
var promptBlockVersionIdPathParams = v4.z.object({
  promptBlockId: v4.z.string().describe("Unique identifier for the stored prompt block"),
  versionId: v4.z.string().describe("Unique identifier for the version (UUID)")
});
var promptBlockVersionSchema = v4.z.object({
  id: v4.z.string().describe("Unique identifier for the version (UUID)"),
  blockId: v4.z.string().describe("ID of the prompt block this version belongs to"),
  versionNumber: v4.z.number().describe("Sequential version number (1, 2, 3, ...)"),
  // Snapshot config fields
  name: v4.z.string().describe("Display name of the prompt block"),
  description: v4.z.string().optional().describe("Purpose description"),
  content: v4.z.string().describe("Template content with {{variable}} interpolation"),
  rules: chunk24YYB3J7_cjs.ruleGroupSchema.optional().describe("Rules for conditional inclusion"),
  requestContextSchema: v4.z.record(v4.z.string(), v4.z.unknown()).optional().describe("JSON Schema defining available variables for {{variableName}} interpolation and conditions"),
  // Version metadata
  changedFields: v4.z.array(v4.z.string()).optional().describe("Array of field names that changed from the previous version"),
  changeMessage: v4.z.string().optional().describe("Optional message describing the changes"),
  createdAt: v4.z.coerce.date().describe("When this version was created")
});
var listPromptBlockVersionsResponseSchema = chunkGFP7IMFR_cjs.createListVersionsResponseSchema(promptBlockVersionSchema);
var getPromptBlockVersionResponseSchema = promptBlockVersionSchema;
var createPromptBlockVersionResponseSchema = promptBlockVersionSchema.partial().merge(
  v4.z.object({
    id: v4.z.string(),
    blockId: v4.z.string(),
    versionNumber: v4.z.number(),
    createdAt: v4.z.coerce.date()
  })
);
var activatePromptBlockVersionResponseSchema = chunkGFP7IMFR_cjs.activateVersionResponseSchema;
var restorePromptBlockVersionResponseSchema = promptBlockVersionSchema;
var deletePromptBlockVersionResponseSchema = chunkGFP7IMFR_cjs.deleteVersionResponseSchema;
var comparePromptBlockVersionsResponseSchema = chunkGFP7IMFR_cjs.createCompareVersionsResponseSchema(promptBlockVersionSchema);

exports.activatePromptBlockVersionResponseSchema = activatePromptBlockVersionResponseSchema;
exports.comparePromptBlockVersionsQuerySchema = comparePromptBlockVersionsQuerySchema;
exports.comparePromptBlockVersionsResponseSchema = comparePromptBlockVersionsResponseSchema;
exports.createPromptBlockVersionBodySchema = createPromptBlockVersionBodySchema;
exports.createPromptBlockVersionResponseSchema = createPromptBlockVersionResponseSchema;
exports.deletePromptBlockVersionResponseSchema = deletePromptBlockVersionResponseSchema;
exports.getPromptBlockVersionResponseSchema = getPromptBlockVersionResponseSchema;
exports.listPromptBlockVersionsQuerySchema = listPromptBlockVersionsQuerySchema;
exports.listPromptBlockVersionsResponseSchema = listPromptBlockVersionsResponseSchema;
exports.promptBlockVersionIdPathParams = promptBlockVersionIdPathParams;
exports.promptBlockVersionPathParams = promptBlockVersionPathParams;
exports.promptBlockVersionSchema = promptBlockVersionSchema;
exports.restorePromptBlockVersionResponseSchema = restorePromptBlockVersionResponseSchema;
