// @mastra/server/dist/chunk-IABGFWY2.cjs
'use strict';

var chunk24YYB3J7_cjs = require('./chunk-24YYB3J7.cjs');
var chunkDIG2K5CV_cjs = require('./chunk-DIG2K5CV.cjs');
var v4 = require('zod/v4');

var storedPromptBlockIdPathParams = v4.z.object({
  storedPromptBlockId: v4.z.string().describe("Unique identifier for the stored prompt block")
});
var storageOrderBySchema = v4.z.object({
  field: v4.z.enum(["createdAt", "updatedAt"]).optional(),
  direction: v4.z.enum(["ASC", "DESC"]).optional()
});
var listStoredPromptBlocksQuerySchema = chunkDIG2K5CV_cjs.createPagePaginationSchema(100).extend({
  orderBy: storageOrderBySchema.optional(),
  status: v4.z.enum(["draft", "published", "archived"]).optional().describe("Filter prompt blocks by status. When omitted, returns all prompt blocks regardless of status"),
  authorId: v4.z.string().optional().describe("Filter prompt blocks by author identifier"),
  metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional().describe("Filter prompt blocks by metadata key-value pairs")
});
var snapshotConfigSchema = v4.z.object({
  name: v4.z.string().describe("Display name of the prompt block"),
  description: v4.z.string().optional().describe("Purpose description"),
  content: v4.z.string().describe("Template content with {{variable}} interpolation"),
  rules: chunk24YYB3J7_cjs.ruleGroupSchema.optional().describe("Rules for conditional inclusion"),
  requestContextSchema: v4.z.record(v4.z.string(), v4.z.unknown()).optional().describe("JSON Schema defining available variables for {{variableName}} interpolation and conditions")
});
var createStoredPromptBlockBodySchema = v4.z.object({
  id: v4.z.string().optional().describe("Unique identifier. If not provided, derived from name."),
  authorId: v4.z.string().optional().describe("Author identifier for multi-tenant filtering"),
  metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional().describe("Additional metadata for the prompt block")
}).merge(snapshotConfigSchema);
var updateStoredPromptBlockBodySchema = v4.z.object({
  authorId: v4.z.string().optional(),
  metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional()
}).merge(snapshotConfigSchema.partial());
var storedPromptBlockSchema = v4.z.object({
  id: v4.z.string(),
  status: v4.z.string().describe("Prompt block status: draft, published, or archived"),
  activeVersionId: v4.z.string().optional(),
  hasDraft: v4.z.boolean().optional().describe("Whether the prompt block has unpublished draft changes"),
  authorId: v4.z.string().optional(),
  metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional(),
  createdAt: v4.z.coerce.date(),
  updatedAt: v4.z.coerce.date(),
  name: v4.z.string().describe("Display name of the prompt block"),
  description: v4.z.string().optional().describe("Purpose description"),
  content: v4.z.string().describe("Template content with {{variable}} interpolation"),
  rules: chunk24YYB3J7_cjs.ruleGroupSchema.optional().describe("Rules for conditional inclusion"),
  requestContextSchema: v4.z.record(v4.z.string(), v4.z.unknown()).optional().describe("JSON Schema defining available variables for {{variableName}} interpolation and conditions")
});
var listStoredPromptBlocksResponseSchema = chunkDIG2K5CV_cjs.paginationInfoSchema.extend({
  promptBlocks: v4.z.array(storedPromptBlockSchema)
});
var getStoredPromptBlockResponseSchema = storedPromptBlockSchema;
var createStoredPromptBlockResponseSchema = storedPromptBlockSchema;
var updateStoredPromptBlockResponseSchema = v4.z.union([
  v4.z.object({
    id: v4.z.string(),
    status: v4.z.string(),
    activeVersionId: v4.z.string().optional(),
    authorId: v4.z.string().optional(),
    metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional(),
    createdAt: v4.z.coerce.date(),
    updatedAt: v4.z.coerce.date()
  }),
  storedPromptBlockSchema
]);
var deleteStoredPromptBlockResponseSchema = v4.z.object({
  success: v4.z.boolean(),
  message: v4.z.string()
});

exports.createStoredPromptBlockBodySchema = createStoredPromptBlockBodySchema;
exports.createStoredPromptBlockResponseSchema = createStoredPromptBlockResponseSchema;
exports.deleteStoredPromptBlockResponseSchema = deleteStoredPromptBlockResponseSchema;
exports.getStoredPromptBlockResponseSchema = getStoredPromptBlockResponseSchema;
exports.listStoredPromptBlocksQuerySchema = listStoredPromptBlocksQuerySchema;
exports.listStoredPromptBlocksResponseSchema = listStoredPromptBlocksResponseSchema;
exports.storedPromptBlockIdPathParams = storedPromptBlockIdPathParams;
exports.storedPromptBlockSchema = storedPromptBlockSchema;
exports.updateStoredPromptBlockBodySchema = updateStoredPromptBlockBodySchema;
exports.updateStoredPromptBlockResponseSchema = updateStoredPromptBlockResponseSchema;
