// @mastra/server/dist/chunk-3BLT6ISZ.js
import { createListVersionsResponseSchema, createCompareVersionsResponseSchema, activateVersionResponseSchema, compareVersionsQuerySchema, createVersionBodySchema, deleteVersionResponseSchema, listVersionsQuerySchema } from './chunk-3OQMTFIV.js';
import { ruleGroupSchema } from './chunk-JA4UAHWP.js';
import { z } from 'zod/v4';

var listPromptBlockVersionsQuerySchema = listVersionsQuerySchema;
var comparePromptBlockVersionsQuerySchema = compareVersionsQuerySchema;
var createPromptBlockVersionBodySchema = createVersionBodySchema;
var promptBlockVersionPathParams = z.object({
  promptBlockId: z.string().describe("Unique identifier for the stored prompt block")
});
var promptBlockVersionIdPathParams = z.object({
  promptBlockId: z.string().describe("Unique identifier for the stored prompt block"),
  versionId: z.string().describe("Unique identifier for the version (UUID)")
});
var promptBlockVersionSchema = z.object({
  id: z.string().describe("Unique identifier for the version (UUID)"),
  blockId: z.string().describe("ID of the prompt block this version belongs to"),
  versionNumber: z.number().describe("Sequential version number (1, 2, 3, ...)"),
  // Snapshot config fields
  name: z.string().describe("Display name of the prompt block"),
  description: z.string().optional().describe("Purpose description"),
  content: z.string().describe("Template content with {{variable}} interpolation"),
  rules: ruleGroupSchema.optional().describe("Rules for conditional inclusion"),
  requestContextSchema: z.record(z.string(), z.unknown()).optional().describe("JSON Schema defining available variables for {{variableName}} interpolation and conditions"),
  // Version metadata
  changedFields: z.array(z.string()).optional().describe("Array of field names that changed from the previous version"),
  changeMessage: z.string().optional().describe("Optional message describing the changes"),
  createdAt: z.coerce.date().describe("When this version was created")
});
var listPromptBlockVersionsResponseSchema = createListVersionsResponseSchema(promptBlockVersionSchema);
var getPromptBlockVersionResponseSchema = promptBlockVersionSchema;
var createPromptBlockVersionResponseSchema = promptBlockVersionSchema.partial().merge(
  z.object({
    id: z.string(),
    blockId: z.string(),
    versionNumber: z.number(),
    createdAt: z.coerce.date()
  })
);
var activatePromptBlockVersionResponseSchema = activateVersionResponseSchema;
var restorePromptBlockVersionResponseSchema = promptBlockVersionSchema;
var deletePromptBlockVersionResponseSchema = deleteVersionResponseSchema;
var comparePromptBlockVersionsResponseSchema = createCompareVersionsResponseSchema(promptBlockVersionSchema);

export { activatePromptBlockVersionResponseSchema, comparePromptBlockVersionsQuerySchema, comparePromptBlockVersionsResponseSchema, createPromptBlockVersionBodySchema, createPromptBlockVersionResponseSchema, deletePromptBlockVersionResponseSchema, getPromptBlockVersionResponseSchema, listPromptBlockVersionsQuerySchema, listPromptBlockVersionsResponseSchema, promptBlockVersionIdPathParams, promptBlockVersionPathParams, promptBlockVersionSchema, restorePromptBlockVersionResponseSchema };
