// @mastra/server/dist/chunk-I7S765JQ.cjs
'use strict';

var chunk3QTG4V4W_cjs = require('./chunk-3QTG4V4W.cjs');
var chunkZ7LCIYK7_cjs = require('./chunk-Z7LCIYK7.cjs');
var chunkRZR7PLFW_cjs = require('./chunk-RZR7PLFW.cjs');
var fs = require('fs');

var GET_API_SCHEMA_ROUTE = chunkRZR7PLFW_cjs.createRoute({
  method: "GET",
  path: "/system/api-schema",
  responseType: "json",
  responseSchema: chunk3QTG4V4W_cjs.apiSchemaManifestResponseSchema,
  summary: "Get API schema manifest",
  description: "Returns the route-contract-derived API schema manifest for the machine-readable CLI",
  tags: ["System"],
  requiresAuth: true,
  handler: async () => {
    const { buildApiSchemaManifest } = await import('./api-schema-manifest-ZCDIT4ND.cjs');
    return buildApiSchemaManifest();
  }
});
var GET_SYSTEM_PACKAGES_ROUTE = chunkRZR7PLFW_cjs.createRoute({
  method: "GET",
  path: "/system/packages",
  responseType: "json",
  responseSchema: chunk3QTG4V4W_cjs.systemPackagesResponseSchema,
  summary: "Get installed Mastra packages",
  description: "Returns a list of all installed Mastra packages and their versions from the project",
  tags: ["System"],
  requiresAuth: true,
  handler: async ({ mastra }) => {
    try {
      const packagesFilePath = process.env.MASTRA_PACKAGES_FILE;
      let packages = [];
      if (packagesFilePath) {
        try {
          const fileContent = fs.readFileSync(packagesFilePath, "utf-8");
          packages = JSON.parse(fileContent);
        } catch {
          packages = [];
        }
      }
      const storage = mastra.getStorage();
      const storageType = storage?.name;
      const observabilityStorage = storage?.stores?.observability;
      const observabilityStorageType = observabilityStorage?.constructor.name;
      const observabilityRuntimeStrategy = observabilityStorage?.runtimeTracingStrategy;
      const observabilityEnabled = !!mastra.observability.getDefaultInstance();
      const editor = mastra.getEditor();
      const editorSource = editor?.getSource?.();
      return {
        packages,
        isDev: process.env.MASTRA_DEV === "true",
        cmsEnabled: !!editor,
        editorSource,
        observabilityEnabled,
        storageType,
        observabilityStorageType,
        observabilityRuntimeStrategy
      };
    } catch (error) {
      return chunkZ7LCIYK7_cjs.handleError(error, "Error getting system packages");
    }
  }
});

exports.GET_API_SCHEMA_ROUTE = GET_API_SCHEMA_ROUTE;
exports.GET_SYSTEM_PACKAGES_ROUTE = GET_SYSTEM_PACKAGES_ROUTE;
