// @mastra/server/dist/chunk-356JZOB2.js
import { createCombinedPaginationSchema } from './chunk-2YY3EMMS.js';
import { z } from 'zod/v4';

var mcpServerIdPathParams = z.object({
  serverId: z.string().describe("MCP server ID")
});
var mcpServerDetailPathParams = z.object({
  id: z.string().describe("MCP server ID")
});
var mcpServerToolPathParams = z.object({
  serverId: z.string().describe("MCP server ID"),
  toolId: z.string().describe("Tool ID")
});
var executeToolBodySchema = z.object({
  data: z.unknown().optional()
});
var listMcpServersQuerySchema = createCombinedPaginationSchema();
var getMcpServerDetailQuerySchema = z.object({
  version: z.string().optional()
});
var versionDetailSchema = z.object({
  version: z.string(),
  release_date: z.string(),
  is_latest: z.boolean()
});
var serverInfoSchema = z.object({
  id: z.string(),
  name: z.string(),
  version_detail: versionDetailSchema
});
var listMcpServersResponseSchema = z.object({
  servers: z.array(serverInfoSchema),
  total_count: z.number(),
  next: z.string().nullable()
});
var serverDetailSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().optional(),
  version_detail: versionDetailSchema,
  package_canonical: z.string().optional(),
  packages: z.array(z.unknown()).optional(),
  remotes: z.array(z.unknown()).optional()
});
var mcpToolInfoSchema = z.object({
  name: z.string(),
  description: z.string().optional(),
  inputSchema: z.unknown(),
  outputSchema: z.unknown().optional(),
  toolType: z.string().optional(),
  _meta: z.record(z.string(), z.unknown()).optional()
});
var listMcpServerToolsResponseSchema = z.object({
  tools: z.array(mcpToolInfoSchema)
});
var executeToolResponseSchema = z.object({
  result: z.unknown()
});
var mcpServerResourcePathParams = z.object({
  serverId: z.string().describe("MCP server ID")
});
var readResourceBodySchema = z.object({
  uri: z.string().describe("Resource URI to read")
});
var resourceContentSchema = z.object({
  uri: z.string(),
  text: z.string().optional(),
  blob: z.string().optional()
});
var readResourceResponseSchema = z.object({
  contents: z.array(resourceContentSchema)
});
var resourceInfoSchema = z.object({
  uri: z.string(),
  name: z.string(),
  description: z.string().optional(),
  mimeType: z.string().optional(),
  _meta: z.record(z.string(), z.unknown()).optional()
});
var listResourcesResponseSchema = z.object({
  resources: z.array(resourceInfoSchema)
});
var jsonRpcErrorSchema = z.object({
  jsonrpc: z.literal("2.0"),
  error: z.object({
    code: z.number(),
    message: z.string()
  }),
  id: z.null()
});

export { executeToolBodySchema, executeToolResponseSchema, getMcpServerDetailQuerySchema, jsonRpcErrorSchema, listMcpServerToolsResponseSchema, listMcpServersQuerySchema, listMcpServersResponseSchema, listResourcesResponseSchema, mcpServerDetailPathParams, mcpServerIdPathParams, mcpServerResourcePathParams, mcpServerToolPathParams, mcpToolInfoSchema, readResourceBodySchema, readResourceResponseSchema, serverDetailSchema, serverInfoSchema, versionDetailSchema };
