// @mastra/server/dist/chunk-3QTG4V4W.cjs
'use strict';

var v4 = require('zod/v4');

// src/server/schemas/system.ts
var mastraPackageSchema = v4.z.object({
  name: v4.z.string(),
  version: v4.z.string()
});
var observabilityRuntimeStrategySchema = v4.z.enum([
  "realtime",
  "batch-with-updates",
  "insert-only",
  "event-sourced"
]);
var editorSourceSchema = v4.z.enum(["code", "db"]);
var systemPackagesResponseSchema = v4.z.object({
  packages: v4.z.array(mastraPackageSchema),
  isDev: v4.z.boolean(),
  cmsEnabled: v4.z.boolean(),
  /**
   * The editor's configured source, when set. `'code'` swaps Studio's
   * Save/Publish UI for Download JSON + Open PR. `'db'` keeps the standard
   * Save/Publish flow. Omitted when the editor has no explicit source.
   */
  editorSource: editorSourceSchema.optional(),
  observabilityEnabled: v4.z.boolean(),
  storageType: v4.z.string().optional(),
  observabilityStorageType: v4.z.string().optional(),
  observabilityRuntimeStrategy: observabilityRuntimeStrategySchema.optional()
});
var jsonSchemaRecordSchema = v4.z.record(v4.z.string(), v4.z.unknown());
var apiSchemaResponseShapeSchema = v4.z.object({
  kind: v4.z.enum(["array", "record", "object-property", "single", "unknown"]),
  listProperty: v4.z.string().optional(),
  paginationProperty: v4.z.string().optional()
});
var apiSchemaManifestRouteSchema = v4.z.object({
  method: v4.z.string(),
  path: v4.z.string(),
  responseType: v4.z.string(),
  pathParamSchema: jsonSchemaRecordSchema.optional(),
  queryParamSchema: jsonSchemaRecordSchema.optional(),
  bodySchema: jsonSchemaRecordSchema.optional(),
  responseSchema: jsonSchemaRecordSchema.optional(),
  responseShape: apiSchemaResponseShapeSchema
});
var apiSchemaManifestResponseSchema = v4.z.object({
  version: v4.z.literal(1),
  routes: v4.z.array(apiSchemaManifestRouteSchema)
});

exports.apiSchemaManifestResponseSchema = apiSchemaManifestResponseSchema;
exports.apiSchemaManifestRouteSchema = apiSchemaManifestRouteSchema;
exports.apiSchemaResponseShapeSchema = apiSchemaResponseShapeSchema;
exports.editorSourceSchema = editorSourceSchema;
exports.mastraPackageSchema = mastraPackageSchema;
exports.observabilityRuntimeStrategySchema = observabilityRuntimeStrategySchema;
exports.systemPackagesResponseSchema = systemPackagesResponseSchema;
