// @mastra/server/dist/chunk-5DHLIFRG.js
import { apiSchemaManifestResponseSchema, systemPackagesResponseSchema } from './chunk-7ZW76TJN.js';
import { handleError } from './chunk-7ZWJX3AN.js';
import { createRoute } from './chunk-YQL3TXSD.js';
import { readFileSync } from 'fs';

var GET_API_SCHEMA_ROUTE = createRoute({
  method: "GET",
  path: "/system/api-schema",
  responseType: "json",
  responseSchema: apiSchemaManifestResponseSchema,
  summary: "Get API schema manifest",
  description: "Returns the route-contract-derived API schema manifest for the machine-readable CLI",
  tags: ["System"],
  requiresAuth: true,
  handler: async () => {
    const { buildApiSchemaManifest } = await import('./api-schema-manifest-AOVCLL52.js');
    return buildApiSchemaManifest();
  }
});
var GET_SYSTEM_PACKAGES_ROUTE = createRoute({
  method: "GET",
  path: "/system/packages",
  responseType: "json",
  responseSchema: systemPackagesResponseSchema,
  summary: "Get installed Mastra packages",
  description: "Returns a list of all installed Mastra packages and their versions from the project",
  tags: ["System"],
  requiresAuth: true,
  handler: async ({ mastra }) => {
    try {
      const packagesFilePath = process.env.MASTRA_PACKAGES_FILE;
      let packages = [];
      if (packagesFilePath) {
        try {
          const fileContent = readFileSync(packagesFilePath, "utf-8");
          packages = JSON.parse(fileContent);
        } catch {
          packages = [];
        }
      }
      const storage = mastra.getStorage();
      const storageType = storage?.name;
      const observabilityStorage = storage?.stores?.observability;
      const observabilityStorageType = observabilityStorage?.constructor.name;
      const observabilityRuntimeStrategy = observabilityStorage?.runtimeTracingStrategy;
      const observabilityEnabled = !!mastra.observability.getDefaultInstance();
      const editor = mastra.getEditor();
      const editorSource = editor?.getSource?.();
      return {
        packages,
        isDev: process.env.MASTRA_DEV === "true",
        cmsEnabled: !!editor,
        editorSource,
        observabilityEnabled,
        storageType,
        observabilityStorageType,
        observabilityRuntimeStrategy
      };
    } catch (error) {
      return handleError(error, "Error getting system packages");
    }
  }
});

export { GET_API_SCHEMA_ROUTE, GET_SYSTEM_PACKAGES_ROUTE };
