// @mastra/server/dist/chunk-42WZN6MM.cjs
'use strict';

var z = require('zod');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var z__default = /*#__PURE__*/_interopDefault(z);

// src/server/schemas/responses.ts
var responseIdPathParams = z__default.default.object({
  responseId: z__default.default.string().describe("Unique identifier for the stored response")
});
var responseInputTextPartSchema = z__default.default.object({
  type: z__default.default.enum(["input_text", "text", "output_text"]),
  text: z__default.default.string()
});
var responseInputMessageSchema = z__default.default.object({
  role: z__default.default.enum(["system", "developer", "user", "assistant"]),
  content: z__default.default.union([z__default.default.string(), z__default.default.array(responseInputTextPartSchema)])
});
var providerOptionValuesSchema = z__default.default.record(z__default.default.string(), z__default.default.unknown());
var providerOptionsSchema = z__default.default.object({
  openai: providerOptionValuesSchema.optional().describe("OpenAI provider options such as previousResponseId, conversation, or responseId"),
  azure: providerOptionValuesSchema.optional().describe("Azure OpenAI provider options such as previousResponseId, store, or itemId")
}).passthrough();
var responseTextFormatJsonObjectSchema = z__default.default.object({
  type: z__default.default.literal("json_object").describe("Requests JSON object output compatibility for the response")
});
var responseTextFormatJsonSchemaSchema = z__default.default.object({
  type: z__default.default.literal("json_schema"),
  name: z__default.default.string(),
  description: z__default.default.string().optional(),
  schema: z__default.default.record(z__default.default.string(), z__default.default.unknown()),
  strict: z__default.default.boolean().optional()
});
var responseTextFormatSchema = z__default.default.union([
  responseTextFormatJsonObjectSchema,
  responseTextFormatJsonSchemaSchema
]);
var responseTextSchema = z__default.default.object({
  format: responseTextFormatSchema
});
var createResponseBodySchema = z__default.default.object({
  model: z__default.default.string().optional().describe(
    "Optional model identifier override, such as openai/gpt-5. When omitted, the agent default model is used."
  ),
  agent_id: z__default.default.string().min(1).optional().describe("Mastra agent ID for the request. Required unless previous_response_id is provided."),
  input: z__default.default.union([z__default.default.string(), z__default.default.array(responseInputMessageSchema)]),
  instructions: z__default.default.string().optional(),
  text: responseTextSchema.optional().describe(
    "Optional text output format. Supports `json_object` for JSON mode and `json_schema` for schema-constrained structured output."
  ),
  conversation_id: z__default.default.string().optional().describe("Optional conversation ID. In Mastra this is the raw threadId."),
  providerOptions: providerOptionsSchema.optional().describe("Optional provider-specific options passed through to the underlying model call"),
  stream: z__default.default.boolean().optional().default(false),
  store: z__default.default.boolean().optional().default(false),
  previous_response_id: z__default.default.string().min(1).optional()
}).passthrough().refine((data) => data.agent_id || data.previous_response_id, {
  message: "agent_id is required unless previous_response_id is provided",
  path: []
});
var responseOutputTextSchema = z__default.default.object({
  type: z__default.default.literal("output_text"),
  text: z__default.default.string(),
  annotations: z__default.default.array(z__default.default.unknown()).optional(),
  logprobs: z__default.default.array(z__default.default.unknown()).optional()
});
var responseInputTextContentSchema = z__default.default.object({
  type: z__default.default.literal("input_text"),
  text: z__default.default.string()
});
var conversationMessageContentSchema = z__default.default.union([responseInputTextContentSchema, responseOutputTextSchema]);
var responseOutputMessageSchema = z__default.default.object({
  id: z__default.default.string(),
  type: z__default.default.literal("message"),
  role: z__default.default.literal("assistant"),
  status: z__default.default.enum(["in_progress", "completed", "incomplete"]),
  content: z__default.default.array(responseOutputTextSchema)
});
var responseOutputFunctionCallSchema = z__default.default.object({
  id: z__default.default.string(),
  type: z__default.default.literal("function_call"),
  call_id: z__default.default.string(),
  name: z__default.default.string(),
  arguments: z__default.default.string(),
  status: z__default.default.enum(["in_progress", "completed", "incomplete"]).optional()
});
var responseOutputFunctionCallOutputSchema = z__default.default.object({
  id: z__default.default.string(),
  type: z__default.default.literal("function_call_output"),
  call_id: z__default.default.string(),
  output: z__default.default.string()
});
var responseUsageSchema = z__default.default.object({
  input_tokens: z__default.default.number(),
  output_tokens: z__default.default.number(),
  total_tokens: z__default.default.number(),
  input_tokens_details: z__default.default.object({
    cached_tokens: z__default.default.number()
  }).optional(),
  output_tokens_details: z__default.default.object({
    reasoning_tokens: z__default.default.number()
  }).optional()
});
var responseToolSchema = z__default.default.object({
  type: z__default.default.literal("function"),
  name: z__default.default.string(),
  description: z__default.default.string().optional(),
  parameters: z__default.default.unknown().optional()
});
var responseOutputItemSchema = z__default.default.union([
  responseOutputMessageSchema,
  responseOutputFunctionCallSchema,
  responseOutputFunctionCallOutputSchema
]);
var conversationItemMessageSchema = z__default.default.object({
  id: z__default.default.string(),
  type: z__default.default.literal("message"),
  role: z__default.default.enum(["system", "user", "assistant"]),
  status: z__default.default.literal("completed"),
  content: z__default.default.array(conversationMessageContentSchema)
});
var conversationItemSchema = z__default.default.union([
  conversationItemMessageSchema,
  responseOutputFunctionCallSchema,
  responseOutputFunctionCallOutputSchema
]);
var responseObjectSchema = z__default.default.object({
  id: z__default.default.string(),
  object: z__default.default.literal("response"),
  created_at: z__default.default.number(),
  completed_at: z__default.default.number().nullable(),
  model: z__default.default.string(),
  status: z__default.default.enum(["in_progress", "completed", "incomplete"]),
  output: z__default.default.array(responseOutputItemSchema),
  usage: responseUsageSchema.nullable(),
  error: z__default.default.null().optional(),
  incomplete_details: z__default.default.null().optional(),
  instructions: z__default.default.string().nullable().optional(),
  text: responseTextSchema.nullable().optional(),
  previous_response_id: z__default.default.string().nullable().optional(),
  conversation_id: z__default.default.string().nullable().optional(),
  providerOptions: providerOptionsSchema.optional(),
  tools: z__default.default.array(responseToolSchema).optional(),
  store: z__default.default.boolean().optional()
});
var deleteResponseSchema = z__default.default.object({
  id: z__default.default.string(),
  object: z__default.default.literal("response"),
  deleted: z__default.default.literal(true)
});

exports.conversationItemMessageSchema = conversationItemMessageSchema;
exports.conversationItemSchema = conversationItemSchema;
exports.conversationMessageContentSchema = conversationMessageContentSchema;
exports.createResponseBodySchema = createResponseBodySchema;
exports.deleteResponseSchema = deleteResponseSchema;
exports.responseIdPathParams = responseIdPathParams;
exports.responseInputMessageSchema = responseInputMessageSchema;
exports.responseInputTextContentSchema = responseInputTextContentSchema;
exports.responseInputTextPartSchema = responseInputTextPartSchema;
exports.responseObjectSchema = responseObjectSchema;
exports.responseOutputFunctionCallOutputSchema = responseOutputFunctionCallOutputSchema;
exports.responseOutputFunctionCallSchema = responseOutputFunctionCallSchema;
exports.responseOutputItemSchema = responseOutputItemSchema;
exports.responseOutputMessageSchema = responseOutputMessageSchema;
exports.responseOutputTextSchema = responseOutputTextSchema;
exports.responseTextFormatSchema = responseTextFormatSchema;
exports.responseToolSchema = responseToolSchema;
exports.responseUsageSchema = responseUsageSchema;
