// @mastra/server/dist/chunk-42NUNMRC.cjs
'use strict';

var coreStorage = require('@mastra/core/storage');
var v4 = require('zod/v4');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var coreStorage__namespace = /*#__PURE__*/_interopNamespace(coreStorage);

// src/server/handlers/observability-storage-schemas.ts
var ns = coreStorage__namespace;
var fallbackEmptyObject = v4.z.object({});
var fallbackBranchArgs = v4.z.object({
  traceId: v4.z.unknown(),
  spanId: v4.z.unknown(),
  depth: v4.z.unknown()
});
var fallbackSchema = v4.z.unknown();
var branchesFilterSchema = ns.branchesFilterSchema ?? fallbackEmptyObject;
var branchesOrderBySchema = ns.branchesOrderBySchema ?? fallbackEmptyObject;
var getBranchArgsSchema = ns.getBranchArgsSchema ?? fallbackBranchArgs;
var listBranchesResponseSchema = ns.listBranchesResponseSchema ?? fallbackSchema;
var getBranchResponseSchema = ns.getBranchResponseSchema ?? fallbackSchema;
var listTracesLightResponseSchema = ns.listTracesLightResponseSchema ?? fallbackSchema;

exports.branchesFilterSchema = branchesFilterSchema;
exports.branchesOrderBySchema = branchesOrderBySchema;
exports.getBranchArgsSchema = getBranchArgsSchema;
exports.getBranchResponseSchema = getBranchResponseSchema;
exports.listBranchesResponseSchema = listBranchesResponseSchema;
exports.listTracesLightResponseSchema = listTracesLightResponseSchema;
