// @mastra/server/dist/chunk-SFAWHLOX.js
import { z } from 'zod/v4';

// src/server/schemas/tool-providers.ts
var labelSchema = z.string().min(1, "Connection label is required").max(32, "Connection label must be \u2264 32 characters").regex(/^[A-Za-z0-9 _-]+$/, "Connection label may only contain letters, digits, spaces, _ and -");
var connectionScopeSchema = z.enum(["shared", "per-author", "caller-supplied"]);
var connectionSchema = z.object({
  kind: z.enum(["author", "invoker", "platform"]),
  toolkit: z.string().min(1),
  connectionId: z.string(),
  label: labelSchema.optional(),
  scope: connectionScopeSchema.optional()
});
var toolMetaSchema = z.object({
  toolkit: z.string().min(1).optional(),
  description: z.string().optional()
});
var toolProviderConfigSchema = z.object({
  tools: z.record(z.string(), toolMetaSchema),
  connections: z.record(z.string(), z.array(connectionSchema))
}).superRefine((value, ctx) => {
  for (const [toolkit, connections] of Object.entries(value.connections)) {
    if (connections.length < 2) continue;
    const seen = /* @__PURE__ */ new Map();
    for (let i = 0; i < connections.length; i++) {
      const conn = connections[i];
      const trimmed = conn.label?.trim() ?? "";
      if (trimmed.length === 0) {
        ctx.addIssue({
          code: "custom",
          path: ["connections", toolkit, i, "label"],
          message: `Connection label is required on toolkit "${toolkit}" once it has two or more connections`
        });
        continue;
      }
      const key = trimmed.toLocaleLowerCase();
      const prevIndex = seen.get(key);
      if (prevIndex !== void 0) {
        ctx.addIssue({
          code: "custom",
          path: ["connections", toolkit, i, "label"],
          message: `Duplicate connection label "${conn.label}" on toolkit "${toolkit}" (labels must be unique case-insensitively)`
        });
      } else {
        seen.set(key, i);
      }
    }
  }
});
var toolProvidersSchema = z.record(z.string(), toolProviderConfigSchema);
var toolProviderIdPathParams = z.object({
  providerId: z.string().describe("Unique identifier for the tool provider")
});
var toolSlugPathParams = toolProviderIdPathParams.extend({
  toolSlug: z.string().describe("Slug identifier for the tool")
});
var toolProviderAuthStatusPathParams = toolProviderIdPathParams.extend({
  authId: z.string().describe("Opaque auth handle returned by authorize")
});
var toolProviderConnectionPathParams = toolProviderIdPathParams.extend({
  connectionId: z.string().describe("Adapter-native connection id (e.g. Composio ca_...)")
});
var listToolProviderToolsQuerySchema = z.object({
  toolkit: z.string().optional().describe("Filter tools by toolkit slug"),
  search: z.string().optional().describe("Search tools by name or description"),
  page: z.coerce.number().optional().describe("Page number for pagination (1-indexed)"),
  perPage: z.coerce.number().optional().describe("Number of items per page")
});
var listConnectionFieldsQuerySchema = z.object({
  toolkit: z.string().describe("Toolkit slug whose connection field schema to list")
});
var listConnectionsQuerySchema = z.object({
  toolkit: z.string().describe("Toolkit slug whose connections to list"),
  authorId: z.string().optional().describe("Admin-only: restrict listing to a specific author. Silently ignored for non-admin callers."),
  scope: connectionScopeSchema.optional().describe("Filter results by scope. Omit to include shared + per-author pins for the caller."),
  page: z.coerce.number().int().positive().optional().describe("Page number for pagination (1-indexed)"),
  perPage: z.coerce.number().int().positive().max(200).optional().describe("Number of items per page (default 50, max 200)")
});
var disconnectConnectionQuerySchema = z.object({
  force: z.union([z.boolean(), z.enum(["true", "false"])]).optional().describe("When true, revoke at the provider and drop the row even if pinned by agents"),
  toolkit: z.string().optional().describe("Toolkit slug for the connection (used when the row was upserted with one)")
});
var connectionUsageQuerySchema = z.object({
  toolkit: z.string().optional().describe("Optional toolkit slug to scope the usage scan")
});
var authorizeToolProviderBodySchema = z.object({
  toolkit: z.string().describe("Toolkit slug being authorized"),
  connectionId: z.string().optional().describe("Existing connection bucket id when re-authorizing; omit for a brand-new connection"),
  toolName: z.string().optional().describe("Optional tool slug for tool-scoped authorization"),
  config: z.record(z.string(), z.unknown()).optional().describe("Provider-specific user-supplied connection fields (e.g. subdomain)"),
  label: z.string().min(1, "Connection label is required").max(32, "Connection label must be \u2264 32 characters").regex(/^[A-Za-z0-9 _-]+$/, "Connection label may only contain letters, digits, spaces, _ and -").nullish().describe(
    "Optional human label to persist on the resulting tool_provider_connections row. Must match the stored connection label rules (\u2264 32 chars, [A-Za-z0-9 _-]+)."
  ),
  scope: connectionScopeSchema.optional().describe(
    'Identity bucket. "shared" pins under SHARED_BUCKET_ID. "caller-supplied" pins under the request-context resourceId (returns 400 when missing). Defaults to "per-author".'
  )
});
var connectionStatusToolProviderBodySchema = z.object({
  items: z.array(
    z.object({
      connectionId: z.string(),
      toolkit: z.string()
    })
  ).describe("Connection tuples to batch-check")
});
var paginationSchema = z.object({
  total: z.number().optional(),
  page: z.number().optional(),
  perPage: z.number().optional(),
  hasMore: z.boolean()
}).optional();
var capabilitiesSchema = z.object({
  multipleConnectionsPerToolkit: z.boolean(),
  batchConnectionStatus: z.boolean(),
  reauthorizeReusesConnectionId: z.boolean(),
  supportsRevoke: z.boolean().optional()
});
var listToolProvidersResponseSchema = z.object({
  providers: z.array(
    z.object({
      id: z.string(),
      name: z.string(),
      description: z.string().optional(),
      displayName: z.string().optional(),
      capabilities: capabilitiesSchema.optional()
    })
  )
});
var listToolProviderToolkitsResponseSchema = z.object({
  data: z.array(
    z.object({
      slug: z.string(),
      name: z.string(),
      description: z.string().optional(),
      icon: z.string().optional()
    })
  ),
  pagination: paginationSchema
});
var listToolProviderToolsResponseSchema = z.object({
  data: z.array(
    z.object({
      slug: z.string(),
      name: z.string(),
      description: z.string().optional(),
      toolkit: z.string().optional()
    })
  ),
  pagination: paginationSchema
});
var getToolProviderToolSchemaResponseSchema = z.record(z.string(), z.unknown());
var authorizeToolProviderResponseSchema = z.object({
  url: z.string(),
  authId: z.string()
});
var authStatusToolProviderResponseSchema = z.object({
  status: z.enum(["pending", "completed", "failed"])
});
var connectionStatusToolProviderResponseSchema = z.object({
  items: z.record(z.string(), z.object({ connected: z.boolean() }))
});
var listConnectionsResponseSchema = z.object({
  items: z.array(
    z.object({
      connectionId: z.string(),
      status: z.enum(["active", "pending", "failed", "inactive"]),
      createdAt: z.string().optional(),
      label: z.string().nullish().describe("Persisted display label from tool_provider_connections, if any"),
      authorId: z.string().optional().describe("Owner of the connection (when known)"),
      scope: connectionScopeSchema.optional().describe("Persisted scope from tool_provider_connections. Missing for rows that predate the scope field.")
    })
  ),
  pagination: paginationSchema
});
var listConnectionFieldsResponseSchema = z.object({
  fields: z.array(
    z.object({
      name: z.string(),
      displayName: z.string().optional(),
      description: z.string().optional(),
      type: z.enum(["string", "number", "boolean"]),
      required: z.boolean(),
      default: z.unknown().optional()
    })
  )
});
var disconnectConnectionResponseSchema = z.object({
  ok: z.literal(true),
  revoked: z.boolean().describe("Whether the provider-side connection was revoked")
});
var updateConnectionBodySchema = z.object({
  label: z.union([labelSchema, z.literal(""), z.null()]).describe("New display label for the connection. Pass null (or empty string) to clear the existing label.")
});
var updateConnectionResponseSchema = z.object({
  ok: z.literal(true),
  label: z.string().nullable().describe("The persisted label after the update (null when cleared)")
});
var connectionUsageResponseSchema = z.object({
  agents: z.array(
    z.object({
      id: z.string(),
      name: z.string()
    })
  )
});
var toolProviderHealthResponseSchema = z.object({
  ok: z.boolean(),
  message: z.string().optional(),
  details: z.record(z.string(), z.unknown()).optional()
});

export { authStatusToolProviderResponseSchema, authorizeToolProviderBodySchema, authorizeToolProviderResponseSchema, connectionSchema, connectionScopeSchema, connectionStatusToolProviderBodySchema, connectionStatusToolProviderResponseSchema, connectionUsageQuerySchema, connectionUsageResponseSchema, disconnectConnectionQuerySchema, disconnectConnectionResponseSchema, getToolProviderToolSchemaResponseSchema, listConnectionFieldsQuerySchema, listConnectionFieldsResponseSchema, listConnectionsQuerySchema, listConnectionsResponseSchema, listToolProviderToolkitsResponseSchema, listToolProviderToolsQuerySchema, listToolProviderToolsResponseSchema, listToolProvidersResponseSchema, toolProviderAuthStatusPathParams, toolProviderConfigSchema, toolProviderConnectionPathParams, toolProviderHealthResponseSchema, toolProviderIdPathParams, toolProvidersSchema, toolSlugPathParams, updateConnectionBodySchema, updateConnectionResponseSchema };
