// @mastra/server/dist/chunk-HTBC6CVX.cjs
'use strict';

var chunk42WZN6MM_cjs = require('./chunk-42WZN6MM.cjs');
var chunkNOXRQV2G_cjs = require('./chunk-NOXRQV2G.cjs');
var z = require('zod');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var z__default = /*#__PURE__*/_interopDefault(z);

var conversationIdPathParams = z__default.default.object({
  conversationId: z__default.default.string().describe("Unique identifier for the conversation thread")
});
var createConversationBodySchema = z__default.default.object({
  agent_id: z__default.default.string().describe("Mastra agent ID used to create the conversation thread"),
  conversation_id: z__default.default.string().optional().describe("Optional conversation ID to use as the raw threadId"),
  resource_id: z__default.default.string().optional().describe("Optional resource ID to associate with the conversation"),
  title: z__default.default.string().optional(),
  metadata: z__default.default.record(z__default.default.string(), z__default.default.unknown()).optional()
});
var conversationObjectSchema = z__default.default.object({
  id: z__default.default.string(),
  object: z__default.default.literal("conversation"),
  thread: chunkNOXRQV2G_cjs.getThreadByIdResponseSchema
});
var conversationDeletedSchema = z__default.default.object({
  id: z__default.default.string(),
  object: z__default.default.literal("conversation.deleted"),
  deleted: z__default.default.literal(true)
});
var conversationItemsListSchema = z__default.default.object({
  object: z__default.default.literal("list"),
  data: z__default.default.array(chunk42WZN6MM_cjs.conversationItemSchema),
  first_id: z__default.default.string().nullable(),
  last_id: z__default.default.string().nullable(),
  has_more: z__default.default.boolean()
});

exports.conversationDeletedSchema = conversationDeletedSchema;
exports.conversationIdPathParams = conversationIdPathParams;
exports.conversationItemsListSchema = conversationItemsListSchema;
exports.conversationObjectSchema = conversationObjectSchema;
exports.createConversationBodySchema = createConversationBodySchema;
