// @mastra/server/dist/chunk-SGXBKIPB.cjs
'use strict';

var chunkGFP7IMFR_cjs = require('./chunk-GFP7IMFR.cjs');
var v4 = require('zod/v4');

var listMCPClientVersionsQuerySchema = chunkGFP7IMFR_cjs.listVersionsQuerySchema;
var compareMCPClientVersionsQuerySchema = chunkGFP7IMFR_cjs.compareVersionsQuerySchema;
var createMCPClientVersionBodySchema = chunkGFP7IMFR_cjs.createVersionBodySchema;
var mcpClientVersionPathParams = v4.z.object({
  mcpClientId: v4.z.string().describe("Unique identifier for the stored MCP client")
});
var mcpClientVersionIdPathParams = v4.z.object({
  mcpClientId: v4.z.string().describe("Unique identifier for the stored MCP client"),
  versionId: v4.z.string().describe("Unique identifier for the version (UUID)")
});
var mcpServerConfigSchema = v4.z.object({
  type: v4.z.enum(["stdio", "http"]),
  command: v4.z.string().optional(),
  args: v4.z.array(v4.z.string()).optional(),
  env: v4.z.record(v4.z.string(), v4.z.string()).optional(),
  url: v4.z.string().optional(),
  timeout: v4.z.number().optional()
});
var mcpClientVersionSchema = v4.z.object({
  id: v4.z.string().describe("Unique identifier for the version (UUID)"),
  mcpClientId: v4.z.string().describe("ID of the MCP client this version belongs to"),
  versionNumber: v4.z.number().describe("Sequential version number (1, 2, 3, ...)"),
  // Snapshot config fields
  name: v4.z.string().describe("Name of the MCP client"),
  description: v4.z.string().optional().describe("Description of the MCP client"),
  servers: v4.z.record(v4.z.string(), mcpServerConfigSchema),
  // Version metadata
  changedFields: v4.z.array(v4.z.string()).optional().describe("Array of field names that changed from the previous version"),
  changeMessage: v4.z.string().optional().describe("Optional message describing the changes"),
  createdAt: v4.z.coerce.date().describe("When this version was created")
});
var listMCPClientVersionsResponseSchema = chunkGFP7IMFR_cjs.createListVersionsResponseSchema(mcpClientVersionSchema);
var getMCPClientVersionResponseSchema = mcpClientVersionSchema;
var createMCPClientVersionResponseSchema = mcpClientVersionSchema.partial().merge(
  v4.z.object({
    id: v4.z.string(),
    mcpClientId: v4.z.string(),
    versionNumber: v4.z.number(),
    createdAt: v4.z.coerce.date()
  })
);
var activateMCPClientVersionResponseSchema = chunkGFP7IMFR_cjs.activateVersionResponseSchema;
var restoreMCPClientVersionResponseSchema = mcpClientVersionSchema;
var deleteMCPClientVersionResponseSchema = chunkGFP7IMFR_cjs.deleteVersionResponseSchema;
var compareMCPClientVersionsResponseSchema = chunkGFP7IMFR_cjs.createCompareVersionsResponseSchema(mcpClientVersionSchema);

exports.activateMCPClientVersionResponseSchema = activateMCPClientVersionResponseSchema;
exports.compareMCPClientVersionsQuerySchema = compareMCPClientVersionsQuerySchema;
exports.compareMCPClientVersionsResponseSchema = compareMCPClientVersionsResponseSchema;
exports.createMCPClientVersionBodySchema = createMCPClientVersionBodySchema;
exports.createMCPClientVersionResponseSchema = createMCPClientVersionResponseSchema;
exports.deleteMCPClientVersionResponseSchema = deleteMCPClientVersionResponseSchema;
exports.getMCPClientVersionResponseSchema = getMCPClientVersionResponseSchema;
exports.listMCPClientVersionsQuerySchema = listMCPClientVersionsQuerySchema;
exports.listMCPClientVersionsResponseSchema = listMCPClientVersionsResponseSchema;
exports.mcpClientVersionIdPathParams = mcpClientVersionIdPathParams;
exports.mcpClientVersionPathParams = mcpClientVersionPathParams;
exports.mcpClientVersionSchema = mcpClientVersionSchema;
exports.restoreMCPClientVersionResponseSchema = restoreMCPClientVersionResponseSchema;
