// @mastra/server/dist/chunk-V23ANIEC.cjs
'use strict';

var chunkQQCQV7ZF_cjs = require('./chunk-QQCQV7ZF.cjs');
var v4 = require('zod/v4');

var agentFeaturesSchema = v4.z.object({
  tools: v4.z.boolean().optional(),
  agents: v4.z.boolean().optional(),
  workflows: v4.z.boolean().optional(),
  scorers: v4.z.boolean().optional(),
  skills: v4.z.boolean().optional(),
  memory: v4.z.boolean().optional(),
  variables: v4.z.boolean().optional(),
  favorites: v4.z.boolean().optional(),
  avatarUpload: v4.z.boolean().optional(),
  browser: v4.z.boolean().optional(),
  /**
   * Whether the model picker is visible in the Agent Builder.
   * Omitted ⇒ picker visible (default-on). Explicit `false` ⇒ picker hidden
   * (locked mode); `models.default` is required and applied.
   */
  model: v4.z.boolean().optional()
});
var knownProviderEntrySchema = v4.z.object({
  provider: v4.z.string().min(1),
  modelId: v4.z.string().min(1).optional()
}).strict();
var customProviderEntrySchema = v4.z.object({
  kind: v4.z.literal("custom"),
  provider: v4.z.string().min(1),
  modelId: v4.z.string().min(1).optional()
}).strict();
var knownDefaultModelEntrySchema = v4.z.object({
  provider: v4.z.string().min(1),
  modelId: v4.z.string().min(1)
}).strict();
var customDefaultModelEntrySchema = v4.z.object({
  kind: v4.z.literal("custom"),
  provider: v4.z.string().min(1),
  modelId: v4.z.string().min(1)
}).strict();
var providerModelEntrySchema = v4.z.union([customProviderEntrySchema, knownProviderEntrySchema]);
var defaultModelEntrySchema = v4.z.union([customDefaultModelEntrySchema, knownDefaultModelEntrySchema]);
var agentModelsSchema = v4.z.object({
  allowed: v4.z.array(providerModelEntrySchema).optional(),
  default: defaultModelEntrySchema.optional()
});
var pickerAllowlistSchema = v4.z.object({
  allowed: v4.z.array(v4.z.string()).optional()
}).strict();
var agentConfigurationSchema = v4.z.object({
  models: agentModelsSchema.optional(),
  tools: pickerAllowlistSchema.optional(),
  agents: pickerAllowlistSchema.optional(),
  workflows: pickerAllowlistSchema.optional()
}).catchall(v4.z.unknown());
var builderPickerSchema = v4.z.object({
  visibleTools: v4.z.array(v4.z.string()).nullable(),
  visibleAgents: v4.z.array(v4.z.string()).nullable(),
  visibleWorkflows: v4.z.array(v4.z.string()).nullable()
});
var builderModelPolicySchema = v4.z.object({
  active: v4.z.boolean(),
  pickerVisible: v4.z.boolean().optional(),
  allowed: v4.z.array(providerModelEntrySchema).optional(),
  default: defaultModelEntrySchema.optional()
});
var builderSettingsResponseSchema = v4.z.object({
  enabled: v4.z.boolean(),
  features: v4.z.object({
    agent: agentFeaturesSchema.optional()
  }).optional(),
  configuration: v4.z.object({
    agent: agentConfigurationSchema.optional()
  }).optional(),
  modelPolicy: builderModelPolicySchema.optional(),
  /**
   * Resolved picker visibility for tools/agents/workflows. Always present when
   * the builder is enabled. Omitted when the builder is disabled.
   */
  picker: builderPickerSchema.optional(),
  /**
   * Non-fatal warnings produced by `EditorAgentBuilder`'s constructor-time
   * validation (e.g. allowlist entries with unknown provider strings, or
   * picker allowlist entries that don't match a registered ID). UI surfaces
   * these as a banner in the Builder admin view.
   */
  modelPolicyWarnings: v4.z.array(v4.z.string()).optional()
});
var infrastructureStatusResponseSchema = v4.z.object({
  channels: v4.z.object({
    providers: v4.z.array(
      v4.z.object({
        id: v4.z.string(),
        name: v4.z.string(),
        isConfigured: v4.z.boolean(),
        routeCount: v4.z.number()
      })
    )
  }),
  browser: v4.z.object({
    type: v4.z.string().nullable(),
    provider: v4.z.string().nullable(),
    env: v4.z.string().nullable(),
    registered: v4.z.boolean(),
    availableProviders: v4.z.array(v4.z.string()),
    config: v4.z.array(v4.z.object({ key: v4.z.string(), value: v4.z.string() }))
  }),
  workspace: v4.z.object({
    type: v4.z.string().nullable(),
    workspaceId: v4.z.string().nullable(),
    name: v4.z.string().nullable(),
    source: v4.z.string().nullable(),
    registered: v4.z.boolean(),
    hasFilesystem: v4.z.boolean(),
    hasSandbox: v4.z.boolean(),
    filesystemProvider: v4.z.string().nullable(),
    sandboxProvider: v4.z.string().nullable(),
    config: v4.z.array(v4.z.object({ key: v4.z.string(), value: v4.z.string() }))
  }),
  registries: v4.z.object({
    skillsSh: v4.z.object({
      enabled: v4.z.boolean()
    })
  })
});
var builderAvailableModelsResponseSchema = v4.z.object({
  providers: v4.z.array(chunkQQCQV7ZF_cjs.providerSchema)
});

exports.agentConfigurationSchema = agentConfigurationSchema;
exports.agentFeaturesSchema = agentFeaturesSchema;
exports.agentModelsSchema = agentModelsSchema;
exports.builderAvailableModelsResponseSchema = builderAvailableModelsResponseSchema;
exports.builderModelPolicySchema = builderModelPolicySchema;
exports.builderPickerSchema = builderPickerSchema;
exports.builderSettingsResponseSchema = builderSettingsResponseSchema;
exports.defaultModelEntrySchema = defaultModelEntrySchema;
exports.infrastructureStatusResponseSchema = infrastructureStatusResponseSchema;
exports.pickerAllowlistSchema = pickerAllowlistSchema;
exports.providerModelEntrySchema = providerModelEntrySchema;
