// @mastra/server/dist/chunk-45B57OCM.cjs
'use strict';

var chunkS5SAMHVP_cjs = require('./chunk-S5SAMHVP.cjs');
var chunkS4UGB72N_cjs = require('./chunk-S4UGB72N.cjs');
var chunkTIRYJGUG_cjs = require('./chunk-TIRYJGUG.cjs');
var chunkI7S765JQ_cjs = require('./chunk-I7S765JQ.cjs');
var chunkDBCEECWQ_cjs = require('./chunk-DBCEECWQ.cjs');
var chunk3ATWQ4MY_cjs = require('./chunk-3ATWQ4MY.cjs');
var chunk2DZPC7UL_cjs = require('./chunk-2DZPC7UL.cjs');
var chunkPIHZRT4S_cjs = require('./chunk-PIHZRT4S.cjs');
var chunkYIR53RXV_cjs = require('./chunk-YIR53RXV.cjs');
var chunkEBLD2M5C_cjs = require('./chunk-EBLD2M5C.cjs');
var chunkLLXZIBZA_cjs = require('./chunk-LLXZIBZA.cjs');
var chunkR2CBBOZ4_cjs = require('./chunk-R2CBBOZ4.cjs');
var chunkN2VPPQLD_cjs = require('./chunk-N2VPPQLD.cjs');
var chunkIQFKZCCF_cjs = require('./chunk-IQFKZCCF.cjs');
var chunkWDASEJZD_cjs = require('./chunk-WDASEJZD.cjs');
var chunkNV2MI4K7_cjs = require('./chunk-NV2MI4K7.cjs');
var chunkNQHBWI24_cjs = require('./chunk-NQHBWI24.cjs');
var chunkRBLDPJSS_cjs = require('./chunk-RBLDPJSS.cjs');
var chunkASSXEY4X_cjs = require('./chunk-ASSXEY4X.cjs');
var chunkEOYMMYEW_cjs = require('./chunk-EOYMMYEW.cjs');
var chunkUDTWZLLE_cjs = require('./chunk-UDTWZLLE.cjs');
var chunkXBXH5HQ7_cjs = require('./chunk-XBXH5HQ7.cjs');
var chunkURW7EH4T_cjs = require('./chunk-URW7EH4T.cjs');
var chunkPMGFP7II_cjs = require('./chunk-PMGFP7II.cjs');
var chunkAUFCEKYY_cjs = require('./chunk-AUFCEKYY.cjs');
var chunkHRRB4YVI_cjs = require('./chunk-HRRB4YVI.cjs');
var chunkMGC46CGD_cjs = require('./chunk-MGC46CGD.cjs');
var chunkZ6GOXY4U_cjs = require('./chunk-Z6GOXY4U.cjs');
var chunkMN4MQEDY_cjs = require('./chunk-MN4MQEDY.cjs');
var chunkFVUJRSNB_cjs = require('./chunk-FVUJRSNB.cjs');
var chunk6MMXLHED_cjs = require('./chunk-6MMXLHED.cjs');
var chunkNLIGVLFF_cjs = require('./chunk-NLIGVLFF.cjs');
var chunk47PV6AIJ_cjs = require('./chunk-47PV6AIJ.cjs');
var chunkT43AU375_cjs = require('./chunk-T43AU375.cjs');
var chunkN4H2ICZJ_cjs = require('./chunk-N4H2ICZJ.cjs');
var chunkS5DWQRTK_cjs = require('./chunk-S5DWQRTK.cjs');
var chunkQXWZ2DKR_cjs = require('./chunk-QXWZ2DKR.cjs');
var chunkHVJTEOVD_cjs = require('./chunk-HVJTEOVD.cjs');
var chunkBLR7PQHQ_cjs = require('./chunk-BLR7PQHQ.cjs');

// src/server/server-adapter/routes/permissions.ts
var METHOD_TO_ACTION = {
  GET: "read",
  POST: "write",
  // Default for POST, may be overridden to 'execute'
  PUT: "write",
  PATCH: "write",
  DELETE: "delete"
};
var EXECUTE_PATTERNS = [
  "/generate",
  "/stream",
  "/execute",
  "/start",
  "/resume",
  "/restart",
  "/cancel",
  "/approve",
  "/decline",
  "/speak",
  "/listen",
  "/query",
  "/search",
  "/observe",
  "/time-travel",
  "/enhance",
  "/clone"
];
var PUBLISH_PATTERNS = ["/publish", "/activate", "/restore"];
var STORED_RESOURCE_SEGMENTS = {
  agents: "stored-agents",
  "mcp-clients": "stored-mcp-clients",
  "prompt-blocks": "stored-prompt-blocks",
  scorers: "stored-scorers",
  skills: "stored-skills",
  workspaces: "stored-workspaces"
};
function extractResource(path) {
  const segments = path.replace(/^\//, "").split("/");
  if (segments.length === 0) {
    return null;
  }
  const firstSegment = segments[0];
  if (firstSegment === "stored" && segments[1]) {
    return STORED_RESOURCE_SEGMENTS[segments[1]] ?? null;
  }
  if (firstSegment === ".well-known") {
    return "a2a";
  }
  return firstSegment || null;
}
function deriveAction(method, path) {
  const upperMethod = method.toUpperCase();
  if (upperMethod === "POST") {
    if (path.startsWith("/stored/")) {
      const isPublishOperation = PUBLISH_PATTERNS.some((pattern) => path.endsWith(pattern));
      if (isPublishOperation) {
        return "publish";
      }
    }
    const isExecuteOperation = EXECUTE_PATTERNS.some((pattern) => path.includes(pattern));
    return isExecuteOperation ? "execute" : "write";
  }
  return METHOD_TO_ACTION[upperMethod] || "read";
}
function derivePermission(route) {
  if (route.method === "ALL") {
    return null;
  }
  const resource = extractResource(route.path);
  if (!resource) {
    return null;
  }
  const action = deriveAction(route.method, route.path);
  return `${resource}:${action}`;
}
function getEffectivePermission(route) {
  if (route.requiresAuth === false) {
    return null;
  }
  if (route.requiresPermission) {
    return route.requiresPermission;
  }
  return derivePermission(route);
}

// src/server/server-adapter/routes/a2a.ts
var A2A_ROUTES = [chunkQXWZ2DKR_cjs.GET_AGENT_CARD_ROUTE, chunkQXWZ2DKR_cjs.AGENT_EXECUTION_ROUTE];

// src/server/server-adapter/routes/agent-builder.ts
var AGENT_BUILDER_ROUTES = [
  chunkN4H2ICZJ_cjs.LIST_AGENT_BUILDER_ACTIONS_ROUTE,
  chunkN4H2ICZJ_cjs.GET_AGENT_BUILDER_ACTION_BY_ID_ROUTE,
  chunkN4H2ICZJ_cjs.LIST_AGENT_BUILDER_ACTION_RUNS_ROUTE,
  chunkN4H2ICZJ_cjs.GET_AGENT_BUILDER_ACTION_RUN_BY_ID_ROUTE,
  chunkN4H2ICZJ_cjs.CREATE_AGENT_BUILDER_ACTION_RUN_ROUTE,
  chunkN4H2ICZJ_cjs.STREAM_AGENT_BUILDER_ACTION_ROUTE,
  chunkN4H2ICZJ_cjs.STREAM_LEGACY_AGENT_BUILDER_ACTION_ROUTE,
  chunkN4H2ICZJ_cjs.START_ASYNC_AGENT_BUILDER_ACTION_ROUTE,
  chunkN4H2ICZJ_cjs.START_AGENT_BUILDER_ACTION_RUN_ROUTE,
  chunkN4H2ICZJ_cjs.OBSERVE_STREAM_AGENT_BUILDER_ACTION_ROUTE,
  chunkN4H2ICZJ_cjs.OBSERVE_STREAM_LEGACY_AGENT_BUILDER_ACTION_ROUTE,
  chunkN4H2ICZJ_cjs.RESUME_ASYNC_AGENT_BUILDER_ACTION_ROUTE,
  chunkN4H2ICZJ_cjs.RESUME_NO_WAIT_AGENT_BUILDER_ACTION_ROUTE,
  chunkN4H2ICZJ_cjs.RESUME_AGENT_BUILDER_ACTION_ROUTE,
  chunkN4H2ICZJ_cjs.RESUME_STREAM_AGENT_BUILDER_ACTION_ROUTE,
  chunkN4H2ICZJ_cjs.CANCEL_AGENT_BUILDER_ACTION_RUN_ROUTE
];

// src/server/server-adapter/routes/agents.ts
var AGENTS_ROUTES = [
  // ============================================================================
  // Agent Core Routes
  // ============================================================================
  chunkHVJTEOVD_cjs.LIST_AGENTS_ROUTE,
  chunkHVJTEOVD_cjs.GET_PROVIDERS_ROUTE,
  chunkHVJTEOVD_cjs.GET_AGENT_BY_ID_ROUTE,
  chunkHVJTEOVD_cjs.CLONE_AGENT_ROUTE,
  // ============================================================================
  // Voice Routes
  // ============================================================================
  chunkPMGFP7II_cjs.GET_SPEAKERS_ROUTE,
  chunkPMGFP7II_cjs.GET_SPEAKERS_DEPRECATED_ROUTE,
  // ============================================================================
  // Agent Execution Routes
  // ============================================================================
  chunkHVJTEOVD_cjs.GENERATE_AGENT_ROUTE,
  chunkHVJTEOVD_cjs.GENERATE_AGENT_VNEXT_ROUTE,
  chunkHVJTEOVD_cjs.STREAM_GENERATE_ROUTE,
  chunkHVJTEOVD_cjs.STREAM_UNTIL_IDLE_GENERATE_ROUTE,
  chunkHVJTEOVD_cjs.STREAM_GENERATE_VNEXT_DEPRECATED_ROUTE,
  // ============================================================================
  // Resumable Stream Routes
  // ============================================================================
  chunkHVJTEOVD_cjs.OBSERVE_AGENT_STREAM_ROUTE,
  chunkHVJTEOVD_cjs.SEND_AGENT_MESSAGE_ROUTE,
  chunkHVJTEOVD_cjs.QUEUE_AGENT_MESSAGE_ROUTE,
  chunkHVJTEOVD_cjs.SEND_AGENT_SIGNAL_ROUTE,
  chunkHVJTEOVD_cjs.ABORT_AGENT_THREAD_ROUTE,
  chunkHVJTEOVD_cjs.SUBSCRIBE_AGENT_THREAD_ROUTE,
  // ============================================================================
  // Tool Routes
  // ============================================================================
  chunkHRRB4YVI_cjs.EXECUTE_AGENT_TOOL_ROUTE,
  chunkHVJTEOVD_cjs.APPROVE_TOOL_CALL_ROUTE,
  chunkHVJTEOVD_cjs.SEND_TOOL_APPROVAL_ROUTE,
  chunkHVJTEOVD_cjs.DECLINE_TOOL_CALL_ROUTE,
  chunkHVJTEOVD_cjs.RESUME_STREAM_ROUTE,
  chunkHVJTEOVD_cjs.APPROVE_TOOL_CALL_GENERATE_ROUTE,
  chunkHVJTEOVD_cjs.DECLINE_TOOL_CALL_GENERATE_ROUTE,
  chunkHVJTEOVD_cjs.APPROVE_NETWORK_TOOL_CALL_ROUTE,
  chunkHVJTEOVD_cjs.DECLINE_NETWORK_TOOL_CALL_ROUTE,
  chunkHVJTEOVD_cjs.RESUME_STREAM_UNTIL_IDLE_ROUTE,
  // ============================================================================
  // Network Routes
  // ============================================================================
  chunkHVJTEOVD_cjs.STREAM_NETWORK_ROUTE,
  // ============================================================================
  // Model Management Routes
  // ============================================================================
  chunkHVJTEOVD_cjs.UPDATE_AGENT_MODEL_ROUTE,
  chunkHVJTEOVD_cjs.RESET_AGENT_MODEL_ROUTE,
  chunkHVJTEOVD_cjs.REORDER_AGENT_MODEL_LIST_ROUTE,
  chunkHVJTEOVD_cjs.UPDATE_AGENT_MODEL_IN_MODEL_LIST_ROUTE,
  // ============================================================================
  // Instruction Enhancement Routes
  // ============================================================================
  chunkHVJTEOVD_cjs.ENHANCE_INSTRUCTIONS_ROUTE,
  // ============================================================================
  // Agent Tool Routes
  // ============================================================================
  chunkHRRB4YVI_cjs.GET_AGENT_TOOL_ROUTE,
  // ============================================================================
  // Agent Skill Routes
  // ============================================================================
  chunkHVJTEOVD_cjs.GET_AGENT_SKILL_ROUTE,
  // ============================================================================
  // Voice/Speech Routes
  // ============================================================================
  chunkPMGFP7II_cjs.GENERATE_SPEECH_ROUTE,
  chunkPMGFP7II_cjs.GENERATE_SPEECH_DEPRECATED_ROUTE,
  chunkPMGFP7II_cjs.TRANSCRIBE_SPEECH_ROUTE,
  chunkPMGFP7II_cjs.TRANSCRIBE_SPEECH_DEPRECATED_ROUTE,
  chunkPMGFP7II_cjs.GET_LISTENER_ROUTE,
  // ============================================================================
  // Deprecated Routes
  // ============================================================================
  chunkHVJTEOVD_cjs.STREAM_VNEXT_DEPRECATED_ROUTE,
  chunkHVJTEOVD_cjs.STREAM_UI_MESSAGE_VNEXT_DEPRECATED_ROUTE,
  chunkHVJTEOVD_cjs.STREAM_UI_MESSAGE_DEPRECATED_ROUTE
];

// src/server/server-adapter/routes/background-tasks.ts
var BACKGROUND_TASK_ROUTES = [
  chunkUDTWZLLE_cjs.BACKGROUND_TASK_STREAM_ROUTE,
  chunkUDTWZLLE_cjs.LIST_BACKGROUND_TASKS_ROUTE,
  chunkUDTWZLLE_cjs.GET_BACKGROUND_TASK_ROUTE
];

// src/server/server-adapter/routes/channels.ts
var CHANNELS_ROUTES = [
  chunkNQHBWI24_cjs.LIST_CHANNEL_PLATFORMS_ROUTE,
  chunkNQHBWI24_cjs.LIST_CHANNEL_INSTALLATIONS_ROUTE,
  chunkNQHBWI24_cjs.CONNECT_CHANNEL_ROUTE,
  chunkNQHBWI24_cjs.DISCONNECT_CHANNEL_ROUTE
];

// src/server/server-adapter/routes/conversations.ts
var CONVERSATIONS_ROUTES = [
  chunk47PV6AIJ_cjs.CREATE_CONVERSATION_ROUTE,
  chunk47PV6AIJ_cjs.GET_CONVERSATION_ROUTE,
  chunk47PV6AIJ_cjs.GET_CONVERSATION_ITEMS_ROUTE,
  chunk47PV6AIJ_cjs.DELETE_CONVERSATION_ROUTE
];

// src/server/server-adapter/routes/datasets.ts
var DATASETS_ROUTES = [
  // Dataset CRUD
  chunkRBLDPJSS_cjs.LIST_DATASETS_ROUTE,
  chunkRBLDPJSS_cjs.CREATE_DATASET_ROUTE,
  chunkRBLDPJSS_cjs.GET_DATASET_ROUTE,
  chunkRBLDPJSS_cjs.UPDATE_DATASET_ROUTE,
  chunkRBLDPJSS_cjs.DELETE_DATASET_ROUTE,
  // Item list and add
  chunkRBLDPJSS_cjs.LIST_ITEMS_ROUTE,
  chunkRBLDPJSS_cjs.ADD_ITEM_ROUTE,
  // Batch operations - MUST come before item-specific routes to avoid /items/batch matching /items/:itemId
  chunkRBLDPJSS_cjs.BATCH_INSERT_ITEMS_ROUTE,
  chunkRBLDPJSS_cjs.BATCH_DELETE_ITEMS_ROUTE,
  // Item-specific CRUD (uses :itemId param)
  chunkRBLDPJSS_cjs.GET_ITEM_ROUTE,
  chunkRBLDPJSS_cjs.UPDATE_ITEM_ROUTE,
  chunkRBLDPJSS_cjs.DELETE_ITEM_ROUTE,
  // Version operations
  chunkRBLDPJSS_cjs.LIST_DATASET_VERSIONS_ROUTE,
  chunkRBLDPJSS_cjs.LIST_ITEM_VERSIONS_ROUTE,
  chunkRBLDPJSS_cjs.GET_ITEM_VERSION_ROUTE,
  // All experiments (cross-dataset) - MUST come before dataset-scoped experiments
  chunkRBLDPJSS_cjs.LIST_ALL_EXPERIMENTS_ROUTE,
  // Experiment review summary (cross-dataset) - MUST come before dataset-scoped experiments
  chunkRBLDPJSS_cjs.EXPERIMENT_REVIEW_SUMMARY_ROUTE,
  // Experiment operations (dataset-scoped)
  chunkRBLDPJSS_cjs.LIST_EXPERIMENTS_ROUTE,
  chunkRBLDPJSS_cjs.TRIGGER_EXPERIMENT_ROUTE,
  chunkRBLDPJSS_cjs.GET_EXPERIMENT_ROUTE,
  chunkRBLDPJSS_cjs.LIST_EXPERIMENT_RESULTS_ROUTE,
  chunkRBLDPJSS_cjs.UPDATE_EXPERIMENT_RESULT_ROUTE,
  // Analytics
  chunkRBLDPJSS_cjs.COMPARE_EXPERIMENTS_ROUTE,
  // AI generation
  chunkRBLDPJSS_cjs.GENERATE_ITEMS_ROUTE,
  // Failure analysis
  chunkRBLDPJSS_cjs.CLUSTER_FAILURES_ROUTE
];

// src/server/server-adapter/routes/editor-builder.ts
var EDITOR_BUILDER_ROUTES = [
  chunkASSXEY4X_cjs.GET_EDITOR_BUILDER_SETTINGS_ROUTE,
  chunkASSXEY4X_cjs.GET_EDITOR_BUILDER_AVAILABLE_MODELS_ROUTE,
  chunkASSXEY4X_cjs.GET_INFRASTRUCTURE_STATUS_ROUTE,
  chunkXBXH5HQ7_cjs.LIST_BUILDER_REGISTRIES_ROUTE,
  chunkXBXH5HQ7_cjs.BUILDER_REGISTRY_SEARCH_ROUTE,
  chunkXBXH5HQ7_cjs.BUILDER_REGISTRY_POPULAR_ROUTE,
  chunkXBXH5HQ7_cjs.BUILDER_REGISTRY_PREVIEW_ROUTE,
  chunkXBXH5HQ7_cjs.BUILDER_REGISTRY_INSTALL_ROUTE
];

// src/server/server-adapter/routes/legacy.ts
var LEGACY_ROUTES = [
  // ============================================================================
  // Legacy Agent Routes
  // ============================================================================
  chunkHVJTEOVD_cjs.GENERATE_LEGACY_ROUTE,
  chunkHVJTEOVD_cjs.STREAM_GENERATE_LEGACY_ROUTE,
  // ============================================================================
  // Legacy Workflow Routes
  // ============================================================================
  chunkS5DWQRTK_cjs.STREAM_LEGACY_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.OBSERVE_STREAM_LEGACY_WORKFLOW_ROUTE
];

// src/server/server-adapter/routes/logs.ts
var LOGS_ROUTES = [chunkT43AU375_cjs.LIST_LOG_TRANSPORTS_ROUTE, chunkT43AU375_cjs.LIST_LOGS_ROUTE, chunkT43AU375_cjs.LIST_LOGS_BY_RUN_ID_ROUTE];

// src/server/server-adapter/routes/mcp.ts
var MCP_ROUTES = [
  // ============================================================================
  // MCP Server Registry Routes
  // ============================================================================
  chunkMN4MQEDY_cjs.LIST_MCP_SERVERS_ROUTE,
  chunkMN4MQEDY_cjs.GET_MCP_SERVER_DETAIL_ROUTE,
  // ============================================================================
  // MCP Server Tool Routes
  // ============================================================================
  chunkMN4MQEDY_cjs.LIST_MCP_SERVER_TOOLS_ROUTE,
  chunkMN4MQEDY_cjs.GET_MCP_SERVER_TOOL_DETAIL_ROUTE,
  chunkMN4MQEDY_cjs.EXECUTE_MCP_SERVER_TOOL_ROUTE,
  // ============================================================================
  // MCP Server Resource Routes
  // ============================================================================
  chunkMN4MQEDY_cjs.LIST_MCP_SERVER_RESOURCES_ROUTE,
  chunkMN4MQEDY_cjs.READ_MCP_SERVER_RESOURCE_ROUTE,
  // ============================================================================
  // MCP Transport Routes (handled by adapters)
  // ============================================================================
  chunkMN4MQEDY_cjs.MCP_HTTP_TRANSPORT_ROUTE,
  chunkMN4MQEDY_cjs.MCP_SSE_TRANSPORT_ROUTE,
  chunkMN4MQEDY_cjs.MCP_SSE_MESSAGES_ROUTE
];

// src/server/server-adapter/routes/memory.ts
var MEMORY_ROUTES = [
  chunkFVUJRSNB_cjs.GET_MEMORY_STATUS_ROUTE,
  chunkFVUJRSNB_cjs.GET_MEMORY_CONFIG_ROUTE,
  chunkFVUJRSNB_cjs.GET_OBSERVATIONAL_MEMORY_ROUTE,
  chunkFVUJRSNB_cjs.AWAIT_BUFFER_STATUS_ROUTE,
  chunkFVUJRSNB_cjs.LIST_THREADS_ROUTE,
  chunkFVUJRSNB_cjs.GET_THREAD_BY_ID_ROUTE,
  chunkFVUJRSNB_cjs.LIST_MESSAGES_ROUTE,
  chunkFVUJRSNB_cjs.GET_WORKING_MEMORY_ROUTE,
  chunkFVUJRSNB_cjs.SAVE_MESSAGES_ROUTE,
  chunkFVUJRSNB_cjs.CREATE_THREAD_ROUTE,
  chunkFVUJRSNB_cjs.UPDATE_THREAD_ROUTE,
  chunkFVUJRSNB_cjs.DELETE_THREAD_ROUTE,
  chunkFVUJRSNB_cjs.CLONE_THREAD_ROUTE,
  chunkFVUJRSNB_cjs.UPDATE_WORKING_MEMORY_ROUTE,
  chunkFVUJRSNB_cjs.DELETE_MESSAGES_ROUTE,
  chunkFVUJRSNB_cjs.SEARCH_MEMORY_ROUTE,
  chunkFVUJRSNB_cjs.GET_MEMORY_STATUS_NETWORK_ROUTE,
  chunkFVUJRSNB_cjs.LIST_THREADS_NETWORK_ROUTE,
  chunkFVUJRSNB_cjs.GET_THREAD_BY_ID_NETWORK_ROUTE,
  chunkFVUJRSNB_cjs.LIST_MESSAGES_NETWORK_ROUTE,
  chunkFVUJRSNB_cjs.SAVE_MESSAGES_NETWORK_ROUTE,
  chunkFVUJRSNB_cjs.CREATE_THREAD_NETWORK_ROUTE,
  chunkFVUJRSNB_cjs.UPDATE_THREAD_NETWORK_ROUTE,
  chunkFVUJRSNB_cjs.DELETE_THREAD_NETWORK_ROUTE,
  chunkFVUJRSNB_cjs.DELETE_MESSAGES_NETWORK_ROUTE
];

// src/server/server-adapter/routes/observability.ts
var OBSERVABILITY_ROUTES = [
  // Legacy
  chunk6MMXLHED_cjs.LIST_TRACES_ROUTE,
  chunk6MMXLHED_cjs.LIST_TRACES_LIGHT_ROUTE,
  chunk6MMXLHED_cjs.LIST_BRANCHES_ROUTE,
  chunk6MMXLHED_cjs.GET_BRANCH_ROUTE,
  chunk6MMXLHED_cjs.GET_TRACE_ROUTE,
  chunk6MMXLHED_cjs.GET_TRACE_LIGHT_ROUTE,
  chunk6MMXLHED_cjs.GET_SPAN_ROUTE,
  chunk6MMXLHED_cjs.GET_TRACE_TRAJECTORY_ROUTE,
  chunk6MMXLHED_cjs.SCORE_TRACES_ROUTE,
  chunk6MMXLHED_cjs.LIST_SCORES_BY_SPAN_ROUTE,
  // New observability routes
  chunkNLIGVLFF_cjs.LIST_METRICS,
  chunkNLIGVLFF_cjs.LIST_LOGS,
  chunkNLIGVLFF_cjs.LIST_SCORES,
  chunkNLIGVLFF_cjs.CREATE_SCORE,
  chunkNLIGVLFF_cjs.GET_SCORE,
  chunkNLIGVLFF_cjs.GET_SCORE_AGGREGATE,
  chunkNLIGVLFF_cjs.GET_SCORE_BREAKDOWN,
  chunkNLIGVLFF_cjs.GET_SCORE_TIME_SERIES,
  chunkNLIGVLFF_cjs.GET_SCORE_PERCENTILES,
  chunkNLIGVLFF_cjs.LIST_FEEDBACK,
  chunkNLIGVLFF_cjs.CREATE_FEEDBACK,
  chunkNLIGVLFF_cjs.GET_FEEDBACK_AGGREGATE,
  chunkNLIGVLFF_cjs.GET_FEEDBACK_BREAKDOWN,
  chunkNLIGVLFF_cjs.GET_FEEDBACK_TIME_SERIES,
  chunkNLIGVLFF_cjs.GET_FEEDBACK_PERCENTILES,
  chunkNLIGVLFF_cjs.GET_METRIC_AGGREGATE,
  chunkNLIGVLFF_cjs.GET_METRIC_BREAKDOWN,
  chunkNLIGVLFF_cjs.GET_METRIC_TIME_SERIES,
  chunkNLIGVLFF_cjs.GET_METRIC_PERCENTILES,
  chunkNLIGVLFF_cjs.GET_METRIC_NAMES,
  chunkNLIGVLFF_cjs.GET_METRIC_LABEL_KEYS,
  chunkNLIGVLFF_cjs.GET_METRIC_LABEL_VALUES,
  chunkNLIGVLFF_cjs.GET_ENTITY_TYPES,
  chunkNLIGVLFF_cjs.GET_ENTITY_NAMES,
  chunkNLIGVLFF_cjs.GET_SERVICE_NAMES,
  chunkNLIGVLFF_cjs.GET_ENVIRONMENTS,
  chunkNLIGVLFF_cjs.GET_TAGS
];

// src/server/server-adapter/routes/processor-providers.ts
var PROCESSOR_PROVIDER_ROUTES = [chunkR2CBBOZ4_cjs.LIST_PROCESSOR_PROVIDERS_ROUTE, chunkR2CBBOZ4_cjs.GET_PROCESSOR_PROVIDER_ROUTE];

// src/server/server-adapter/routes/processors.ts
var PROCESSORS_ROUTES = [chunkN2VPPQLD_cjs.LIST_PROCESSORS_ROUTE, chunkN2VPPQLD_cjs.GET_PROCESSOR_BY_ID_ROUTE, chunkN2VPPQLD_cjs.EXECUTE_PROCESSOR_ROUTE];

// src/server/server-adapter/routes/responses.ts
var RESPONSES_ROUTES = [chunkZ6GOXY4U_cjs.CREATE_RESPONSE_ROUTE, chunkZ6GOXY4U_cjs.GET_RESPONSE_ROUTE, chunkZ6GOXY4U_cjs.DELETE_RESPONSE_ROUTE];

// src/server/server-adapter/routes/schedules.ts
var SCHEDULES_ROUTES = [
  chunkWDASEJZD_cjs.LIST_SCHEDULES_ROUTE,
  chunkWDASEJZD_cjs.GET_SCHEDULE_ROUTE,
  chunkWDASEJZD_cjs.LIST_SCHEDULE_TRIGGERS_ROUTE,
  chunkWDASEJZD_cjs.PAUSE_SCHEDULE_ROUTE,
  chunkWDASEJZD_cjs.RESUME_SCHEDULE_ROUTE
];

// src/server/server-adapter/routes/scorers.ts
var SCORES_ROUTES = [
  chunkMGC46CGD_cjs.LIST_SCORERS_ROUTE,
  chunkMGC46CGD_cjs.GET_SCORER_ROUTE,
  chunkMGC46CGD_cjs.LIST_SCORES_BY_RUN_ID_ROUTE,
  chunkMGC46CGD_cjs.LIST_SCORES_BY_SCORER_ID_ROUTE,
  chunkMGC46CGD_cjs.LIST_SCORES_BY_ENTITY_ID_ROUTE,
  chunkMGC46CGD_cjs.SAVE_SCORE_ROUTE
];

// src/server/server-adapter/routes/stored-agents.ts
var STORED_AGENTS_ROUTES = [
  // ============================================================================
  // Stored Agents CRUD Routes
  // IMPORTANT: Routes with literal paths (e.g., /preview-instructions) must come
  // BEFORE routes with path parameters (e.g., /:storedAgentId) to ensure correct matching.
  // ============================================================================
  chunkPIHZRT4S_cjs.LIST_STORED_AGENTS_ROUTE,
  chunkPIHZRT4S_cjs.PREVIEW_INSTRUCTIONS_ROUTE,
  // Must be before GET_STORED_AGENT_ROUTE
  chunkPIHZRT4S_cjs.GET_STORED_AGENT_DEPENDENTS_ROUTE,
  // Must be before GET_STORED_AGENT_ROUTE (longer literal)
  chunkPIHZRT4S_cjs.EXPORT_STORED_AGENT_ROUTE,
  // Must be before GET_STORED_AGENT_ROUTE
  chunkPIHZRT4S_cjs.GET_STORED_AGENT_ROUTE,
  chunkPIHZRT4S_cjs.CREATE_STORED_AGENT_ROUTE,
  chunkPIHZRT4S_cjs.UPDATE_STORED_AGENT_ROUTE,
  chunkPIHZRT4S_cjs.DELETE_STORED_AGENT_ROUTE,
  // ============================================================================
  // Agent Versions Routes
  // IMPORTANT: Routes with literal paths (e.g., /compare) must come BEFORE
  // routes with path parameters (e.g., /:versionId) to ensure correct matching.
  // ============================================================================
  chunkEOYMMYEW_cjs.LIST_AGENT_VERSIONS_ROUTE,
  chunkEOYMMYEW_cjs.CREATE_AGENT_VERSION_ROUTE,
  chunkEOYMMYEW_cjs.COMPARE_AGENT_VERSIONS_ROUTE,
  // Must be before GET_AGENT_VERSION_ROUTE
  chunkEOYMMYEW_cjs.GET_AGENT_VERSION_ROUTE,
  chunkEOYMMYEW_cjs.ACTIVATE_AGENT_VERSION_ROUTE,
  chunkEOYMMYEW_cjs.RESTORE_AGENT_VERSION_ROUTE,
  chunkEOYMMYEW_cjs.DELETE_AGENT_VERSION_ROUTE,
  // ============================================================================
  // Favorites (EE)
  // ============================================================================
  chunk2DZPC7UL_cjs.FAVORITE_STORED_AGENT_ROUTE,
  chunk2DZPC7UL_cjs.UNFAVORITE_STORED_AGENT_ROUTE
];

// src/server/server-adapter/routes/stored-mcp-clients.ts
var STORED_MCP_CLIENTS_ROUTES = [
  // Stored MCP Clients CRUD Routes
  chunkYIR53RXV_cjs.LIST_STORED_MCP_CLIENTS_ROUTE,
  chunkYIR53RXV_cjs.GET_STORED_MCP_CLIENT_ROUTE,
  chunkYIR53RXV_cjs.CREATE_STORED_MCP_CLIENT_ROUTE,
  chunkYIR53RXV_cjs.UPDATE_STORED_MCP_CLIENT_ROUTE,
  chunkYIR53RXV_cjs.DELETE_STORED_MCP_CLIENT_ROUTE,
  // MCP Client Versions Routes
  // IMPORTANT: Routes with literal paths (e.g., /compare) must come BEFORE
  // routes with path parameters (e.g., /:versionId) to ensure correct matching.
  chunkNV2MI4K7_cjs.LIST_MCP_CLIENT_VERSIONS_ROUTE,
  chunkNV2MI4K7_cjs.CREATE_MCP_CLIENT_VERSION_ROUTE,
  chunkNV2MI4K7_cjs.COMPARE_MCP_CLIENT_VERSIONS_ROUTE,
  chunkNV2MI4K7_cjs.GET_MCP_CLIENT_VERSION_ROUTE,
  chunkNV2MI4K7_cjs.ACTIVATE_MCP_CLIENT_VERSION_ROUTE,
  chunkNV2MI4K7_cjs.RESTORE_MCP_CLIENT_VERSION_ROUTE,
  chunkNV2MI4K7_cjs.DELETE_MCP_CLIENT_VERSION_ROUTE
];

// src/server/server-adapter/routes/stored-prompt-blocks.ts
var STORED_PROMPT_BLOCKS_ROUTES = [
  // Stored Prompt Blocks CRUD Routes
  chunkEBLD2M5C_cjs.LIST_STORED_PROMPT_BLOCKS_ROUTE,
  chunkEBLD2M5C_cjs.GET_STORED_PROMPT_BLOCK_ROUTE,
  chunkEBLD2M5C_cjs.CREATE_STORED_PROMPT_BLOCK_ROUTE,
  chunkEBLD2M5C_cjs.UPDATE_STORED_PROMPT_BLOCK_ROUTE,
  chunkEBLD2M5C_cjs.DELETE_STORED_PROMPT_BLOCK_ROUTE,
  // Prompt Block Versions Routes
  // IMPORTANT: Routes with literal paths (e.g., /compare) must come BEFORE
  // routes with path parameters (e.g., /:versionId) to ensure correct matching.
  chunkIQFKZCCF_cjs.LIST_PROMPT_BLOCK_VERSIONS_ROUTE,
  chunkIQFKZCCF_cjs.CREATE_PROMPT_BLOCK_VERSION_ROUTE,
  chunkIQFKZCCF_cjs.COMPARE_PROMPT_BLOCK_VERSIONS_ROUTE,
  chunkIQFKZCCF_cjs.GET_PROMPT_BLOCK_VERSION_ROUTE,
  chunkIQFKZCCF_cjs.ACTIVATE_PROMPT_BLOCK_VERSION_ROUTE,
  chunkIQFKZCCF_cjs.RESTORE_PROMPT_BLOCK_VERSION_ROUTE,
  chunkIQFKZCCF_cjs.DELETE_PROMPT_BLOCK_VERSION_ROUTE
];

// src/server/server-adapter/routes/stored-scorers.ts
var STORED_SCORERS_ROUTES = [
  // Stored Scorers CRUD Routes
  chunkLLXZIBZA_cjs.LIST_STORED_SCORERS_ROUTE,
  chunkLLXZIBZA_cjs.GET_STORED_SCORER_ROUTE,
  chunkLLXZIBZA_cjs.CREATE_STORED_SCORER_ROUTE,
  chunkLLXZIBZA_cjs.UPDATE_STORED_SCORER_ROUTE,
  chunkLLXZIBZA_cjs.DELETE_STORED_SCORER_ROUTE,
  // Scorer Versions Routes
  // IMPORTANT: Routes with literal paths (e.g., /compare) must come BEFORE
  // routes with path parameters (e.g., /:versionId) to ensure correct matching.
  chunk3ATWQ4MY_cjs.LIST_SCORER_VERSIONS_ROUTE,
  chunk3ATWQ4MY_cjs.CREATE_SCORER_VERSION_ROUTE,
  chunk3ATWQ4MY_cjs.COMPARE_SCORER_VERSIONS_ROUTE,
  chunk3ATWQ4MY_cjs.GET_SCORER_VERSION_ROUTE,
  chunk3ATWQ4MY_cjs.ACTIVATE_SCORER_VERSION_ROUTE,
  chunk3ATWQ4MY_cjs.RESTORE_SCORER_VERSION_ROUTE,
  chunk3ATWQ4MY_cjs.DELETE_SCORER_VERSION_ROUTE
];

// src/server/server-adapter/routes/stored-skills.ts
var STORED_SKILLS_ROUTES = [
  // Stored Skills CRUD Routes
  chunkS4UGB72N_cjs.LIST_STORED_SKILLS_ROUTE,
  chunkS4UGB72N_cjs.GET_STORED_SKILL_ROUTE,
  chunkS4UGB72N_cjs.CREATE_STORED_SKILL_ROUTE,
  chunkS4UGB72N_cjs.UPDATE_STORED_SKILL_ROUTE,
  chunkS4UGB72N_cjs.DELETE_STORED_SKILL_ROUTE,
  // Publish
  chunkS4UGB72N_cjs.PUBLISH_STORED_SKILL_ROUTE,
  // Favorites (EE)
  chunkS5SAMHVP_cjs.FAVORITE_STORED_SKILL_ROUTE,
  chunkS5SAMHVP_cjs.UNFAVORITE_STORED_SKILL_ROUTE
];

// src/server/server-adapter/routes/stored-workspaces.ts
var STORED_WORKSPACES_ROUTES = [
  // Stored Workspaces CRUD Routes
  chunkTIRYJGUG_cjs.LIST_STORED_WORKSPACES_ROUTE,
  chunkTIRYJGUG_cjs.GET_STORED_WORKSPACE_ROUTE,
  chunkTIRYJGUG_cjs.CREATE_STORED_WORKSPACE_ROUTE,
  chunkTIRYJGUG_cjs.UPDATE_STORED_WORKSPACE_ROUTE,
  chunkTIRYJGUG_cjs.DELETE_STORED_WORKSPACE_ROUTE
];

// src/server/server-adapter/routes/system.ts
var SYSTEM_ROUTES = [chunkI7S765JQ_cjs.GET_SYSTEM_PACKAGES_ROUTE, chunkI7S765JQ_cjs.GET_API_SCHEMA_ROUTE];

// src/server/server-adapter/routes/tool-providers.ts
var TOOL_PROVIDER_ROUTES = [
  chunkDBCEECWQ_cjs.LIST_TOOL_PROVIDERS_ROUTE,
  chunkDBCEECWQ_cjs.LIST_TOOL_PROVIDER_TOOLKITS_ROUTE,
  chunkDBCEECWQ_cjs.LIST_TOOL_PROVIDER_TOOLS_ROUTE,
  chunkDBCEECWQ_cjs.GET_TOOL_PROVIDER_TOOL_SCHEMA_ROUTE,
  chunkDBCEECWQ_cjs.AUTHORIZE_TOOL_PROVIDER_ROUTE,
  chunkDBCEECWQ_cjs.GET_TOOL_PROVIDER_AUTH_STATUS_ROUTE,
  chunkDBCEECWQ_cjs.TOOL_PROVIDER_CONNECTION_STATUS_ROUTE,
  chunkDBCEECWQ_cjs.LIST_TOOL_PROVIDER_CONNECTIONS_ROUTE,
  chunkDBCEECWQ_cjs.LIST_TOOL_PROVIDER_CONNECTION_FIELDS_ROUTE,
  chunkDBCEECWQ_cjs.DISCONNECT_TOOL_PROVIDER_CONNECTION_ROUTE,
  chunkDBCEECWQ_cjs.UPDATE_TOOL_PROVIDER_CONNECTION_ROUTE,
  chunkDBCEECWQ_cjs.GET_TOOL_PROVIDER_CONNECTION_USAGE_ROUTE,
  chunkDBCEECWQ_cjs.GET_TOOL_PROVIDER_HEALTH_ROUTE
];

// src/server/server-adapter/routes/tools.ts
var TOOLS_ROUTES = [chunkHRRB4YVI_cjs.LIST_TOOLS_ROUTE, chunkHRRB4YVI_cjs.GET_TOOL_BY_ID_ROUTE, chunkHRRB4YVI_cjs.EXECUTE_TOOL_ROUTE];

// src/server/server-adapter/routes/vectors.ts
var VECTORS_ROUTES = [
  chunkURW7EH4T_cjs.UPSERT_VECTORS_ROUTE,
  chunkURW7EH4T_cjs.CREATE_INDEX_ROUTE,
  chunkURW7EH4T_cjs.QUERY_VECTORS_ROUTE,
  chunkURW7EH4T_cjs.LIST_INDEXES_ROUTE,
  chunkURW7EH4T_cjs.DESCRIBE_INDEX_ROUTE,
  chunkURW7EH4T_cjs.DELETE_INDEX_ROUTE,
  chunkURW7EH4T_cjs.LIST_VECTORS_ROUTE,
  chunkURW7EH4T_cjs.LIST_EMBEDDERS_ROUTE
];

// src/server/server-adapter/routes/workflows.ts
var WORKFLOWS_ROUTES = [
  chunkS5DWQRTK_cjs.LIST_WORKFLOWS_ROUTE,
  chunkS5DWQRTK_cjs.GET_WORKFLOW_BY_ID_ROUTE,
  chunkS5DWQRTK_cjs.LIST_WORKFLOW_RUNS_ROUTE,
  chunkS5DWQRTK_cjs.GET_WORKFLOW_RUN_BY_ID_ROUTE,
  chunkS5DWQRTK_cjs.DELETE_WORKFLOW_RUN_BY_ID_ROUTE,
  chunkS5DWQRTK_cjs.CREATE_WORKFLOW_RUN_ROUTE,
  chunkS5DWQRTK_cjs.STREAM_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.RESUME_STREAM_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.START_ASYNC_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.START_WORKFLOW_RUN_ROUTE,
  chunkS5DWQRTK_cjs.OBSERVE_STREAM_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.RESUME_ASYNC_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.RESUME_NO_WAIT_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.RESUME_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.CANCEL_WORKFLOW_RUN_ROUTE,
  chunkS5DWQRTK_cjs.TIME_TRAVEL_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.TIME_TRAVEL_ASYNC_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.TIME_TRAVEL_STREAM_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.RESTART_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.RESTART_ASYNC_WORKFLOW_ROUTE,
  chunkS5DWQRTK_cjs.RESTART_ALL_ACTIVE_WORKFLOW_RUNS_ROUTE,
  chunkS5DWQRTK_cjs.RESTART_ALL_ACTIVE_WORKFLOW_RUNS_ASYNC_ROUTE,
  chunkS5DWQRTK_cjs.EXECUTE_WORKFLOW_STEP_ROUTE,
  chunkS5DWQRTK_cjs.RECEIVE_WORKFLOW_EVENT_ROUTE
];

// src/server/server-adapter/routes/workspace.ts
var WORKSPACE_ROUTES = [
  // List all workspaces route (at /api/workspaces)
  chunkAUFCEKYY_cjs.LIST_WORKSPACES_ROUTE,
  // Get workspace route (at /api/workspaces/:workspaceId)
  chunkAUFCEKYY_cjs.GET_WORKSPACE_ROUTE,
  // Filesystem routes (at /api/workspaces/:workspaceId/fs/*)
  ...chunkAUFCEKYY_cjs.WORKSPACE_FS_ROUTES,
  // Search routes (at /api/workspaces/:workspaceId/search, /api/workspaces/:workspaceId/index)
  ...chunkAUFCEKYY_cjs.WORKSPACE_SEARCH_ROUTES,
  // Skills routes (search must come before parameterized routes)
  ...chunkAUFCEKYY_cjs.WORKSPACE_SKILLS_ROUTES,
  // skills.sh proxy routes (at /api/workspaces/:workspaceId/skills-sh/*)
  ...chunkAUFCEKYY_cjs.WORKSPACE_SKILLS_SH_ROUTES
];

// src/server/server-adapter/routes/index.ts
var SERVER_ROUTES = [
  ...AGENTS_ROUTES,
  ...chunkBLR7PQHQ_cjs.AUTH_ROUTES,
  ...WORKFLOWS_ROUTES,
  ...TOOLS_ROUTES,
  ...PROCESSORS_ROUTES,
  ...RESPONSES_ROUTES,
  ...CONVERSATIONS_ROUTES,
  ...MEMORY_ROUTES,
  ...SCORES_ROUTES,
  ...OBSERVABILITY_ROUTES,
  ...LOGS_ROUTES,
  ...VECTORS_ROUTES,
  ...A2A_ROUTES,
  ...WORKSPACE_ROUTES,
  ...LEGACY_ROUTES,
  ...MCP_ROUTES,
  ...STORED_AGENTS_ROUTES,
  ...STORED_MCP_CLIENTS_ROUTES,
  ...STORED_PROMPT_BLOCKS_ROUTES,
  ...STORED_SCORERS_ROUTES,
  ...STORED_WORKSPACES_ROUTES,
  ...STORED_SKILLS_ROUTES,
  ...TOOL_PROVIDER_ROUTES,
  ...PROCESSOR_PROVIDER_ROUTES,
  ...SYSTEM_ROUTES,
  ...DATASETS_ROUTES,
  ...BACKGROUND_TASK_ROUTES,
  ...EDITOR_BUILDER_ROUTES,
  ...AGENT_BUILDER_ROUTES,
  ...SCHEDULES_ROUTES,
  ...CHANNELS_ROUTES
];

exports.SERVER_ROUTES = SERVER_ROUTES;
exports.deriveAction = deriveAction;
exports.derivePermission = derivePermission;
exports.extractResource = extractResource;
exports.getEffectivePermission = getEffectivePermission;
