// @mastra/server/dist/chunk-DCDQNIYY.cjs
'use strict';

var chunkDIG2K5CV_cjs = require('./chunk-DIG2K5CV.cjs');
var v4 = require('zod/v4');

var scoringSamplingConfigSchema = v4.z.object({});
var mastraScorerConfigSchema = v4.z.object({
  id: v4.z.string(),
  name: v4.z.string().optional(),
  description: v4.z.string(),
  type: v4.z.unknown().optional(),
  judge: v4.z.unknown().optional()
});
var mastraScorerSchema = v4.z.object({
  config: mastraScorerConfigSchema
});
var scorerEntrySchema = v4.z.object({
  scorer: mastraScorerSchema,
  sampling: scoringSamplingConfigSchema.optional(),
  agentIds: v4.z.array(v4.z.string()),
  agentNames: v4.z.array(v4.z.string()),
  workflowIds: v4.z.array(v4.z.string()),
  isRegistered: v4.z.boolean(),
  source: v4.z.enum(["code", "stored"])
});
var listScorersResponseSchema = v4.z.record(v4.z.string(), scorerEntrySchema);
var scorerIdPathParams = v4.z.object({
  scorerId: v4.z.string().describe("Unique identifier for the scorer")
});
var entityPathParams = v4.z.object({
  entityType: v4.z.string().describe("Type of the entity (AGENT or WORKFLOW)"),
  entityId: v4.z.string().describe("Unique identifier for the entity")
});
var listScoresByRunIdQuerySchema = v4.z.object({
  page: v4.z.coerce.number().optional().default(0),
  perPage: v4.z.coerce.number().optional().default(10)
});
var listScoresByScorerIdQuerySchema = v4.z.object({
  page: v4.z.coerce.number().optional().default(0),
  perPage: v4.z.coerce.number().optional().default(10),
  entityId: v4.z.string().optional(),
  entityType: v4.z.string().optional()
});
var listScoresByEntityIdQuerySchema = v4.z.object({
  page: v4.z.coerce.number().optional().default(0),
  perPage: v4.z.coerce.number().optional().default(10)
});
var saveScoreBodySchema = v4.z.object({
  score: v4.z.unknown()
  // ScoreRowData - complex type
});
var scoresWithPaginationResponseSchema = v4.z.object({
  pagination: chunkDIG2K5CV_cjs.paginationInfoSchema,
  scores: v4.z.array(v4.z.unknown())
  // Array of score records
});
var saveScoreResponseSchema = v4.z.object({
  score: v4.z.unknown()
  // ScoreRowData
});

exports.entityPathParams = entityPathParams;
exports.listScorersResponseSchema = listScorersResponseSchema;
exports.listScoresByEntityIdQuerySchema = listScoresByEntityIdQuerySchema;
exports.listScoresByRunIdQuerySchema = listScoresByRunIdQuerySchema;
exports.listScoresByScorerIdQuerySchema = listScoresByScorerIdQuerySchema;
exports.saveScoreBodySchema = saveScoreBodySchema;
exports.saveScoreResponseSchema = saveScoreResponseSchema;
exports.scorerEntrySchema = scorerEntrySchema;
exports.scorerIdPathParams = scorerIdPathParams;
exports.scoresWithPaginationResponseSchema = scoresWithPaginationResponseSchema;
