// @mastra/server/dist/chunk-UIRQU4GW.js
import { createPagePaginationSchema, paginationInfoSchema } from './chunk-2YY3EMMS.js';
import { z } from 'zod/v4';

var storedSkillIdPathParams = z.object({
  storedSkillId: z.string().describe("Unique identifier for the stored skill")
});
var storageOrderBySchema = z.object({
  field: z.enum(["createdAt", "updatedAt"]).optional(),
  direction: z.enum(["ASC", "DESC"]).optional()
});
var listStoredSkillsQuerySchema = createPagePaginationSchema(100).extend({
  orderBy: storageOrderBySchema.optional(),
  status: z.enum(["draft", "published", "archived"]).optional().describe("Filter skills by status"),
  authorId: z.string().optional().describe("Filter skills by author identifier"),
  visibility: z.enum(["public"]).optional().describe("Filter to only public skills"),
  metadata: z.record(z.string(), z.unknown()).optional().describe("Filter skills by metadata key-value pairs"),
  favoritedOnly: z.stringbool().optional().describe("When true, return only skills favorited by the caller (requires the `favorites` EE feature)"),
  pinFavoritedFor: z.string().optional().describe(
    "When set, treat the given subject (user/role) as the favoriting principal for `favoritedOnly` instead of the caller"
  )
});
var sourceSchema = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("external"),
    packagePath: z.string().describe("Package path for external source")
  }),
  z.object({
    type: z.literal("local"),
    projectPath: z.string().describe("Project path for local source")
  }),
  z.object({
    type: z.literal("managed"),
    mastraPath: z.string().describe("Mastra path for managed source")
  })
]);
var fileNodeSchema = z.object({
  id: z.string().optional(),
  name: z.string(),
  type: z.enum(["file", "folder"]),
  content: z.string().optional(),
  children: z.lazy(() => z.array(fileNodeSchema)).optional()
});
var skillOriginSchema = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("skills-sh"),
    owner: z.string().describe("Repository owner on skills.sh"),
    repo: z.string().describe("Repository name on skills.sh"),
    skillName: z.string().describe("Original skill name on skills.sh"),
    installedAt: z.string().describe("ISO-8601 timestamp of the install")
  }),
  z.object({
    type: z.literal("library-copy"),
    sourceSkillId: z.string().describe("ID of the public Library skill this was copied from"),
    sourceSkillName: z.string().describe("Name of the source skill at copy time"),
    sourceAuthorId: z.string().optional().describe("Author of the source skill at copy time, when known"),
    copiedAt: z.string().describe("ISO-8601 timestamp of the copy")
  })
]);
var SKILL_ORIGIN_METADATA_KEY = "origin";
function readSkillOrigin(metadata) {
  if (!metadata) return null;
  const raw = metadata[SKILL_ORIGIN_METADATA_KEY];
  if (!raw) return null;
  const parsed = skillOriginSchema.safeParse(raw);
  return parsed.success ? parsed.data : null;
}
function buildOriginMetadata(origin) {
  return { [SKILL_ORIGIN_METADATA_KEY]: origin };
}
var snapshotConfigSchema = z.object({
  name: z.string().describe("Name of the skill"),
  description: z.string().describe("Description of what the skill does and when to use it"),
  instructions: z.string().describe("Markdown instructions for the skill"),
  license: z.string().optional().describe("License identifier for the skill"),
  compatibility: z.unknown().optional().describe("Compatibility requirements"),
  source: sourceSchema.optional().describe("Source location of the skill"),
  references: z.array(z.string()).optional().describe("List of reference file paths"),
  scripts: z.array(z.string()).optional().describe("List of script file paths"),
  assets: z.array(z.string()).optional().describe("List of asset file paths"),
  files: z.array(fileNodeSchema).optional().describe("Full file tree structure for the skill"),
  metadata: z.record(z.string(), z.unknown()).optional().describe("Additional metadata for the skill")
});
var createStoredSkillBodySchema = z.object({
  id: z.string().optional().describe("Unique identifier. If not provided, derived from name."),
  authorId: z.string().optional().describe("Author identifier for multi-tenant filtering"),
  visibility: z.enum(["private", "public"]).optional().describe("Skill visibility: private (owner/admin only) or public (any reader)")
}).merge(snapshotConfigSchema);
var updateStoredSkillBodySchema = z.object({
  authorId: z.string().optional(),
  visibility: z.enum(["private", "public"]).optional().describe("Skill visibility: private (owner/admin only) or public (any reader)")
}).partial().merge(snapshotConfigSchema.partial());
var storedSkillSchema = z.object({
  id: z.string(),
  status: z.string().describe("Skill status: draft, published, or archived"),
  activeVersionId: z.string().optional(),
  authorId: z.string().optional(),
  visibility: z.enum(["private", "public"]).optional(),
  favoriteCount: z.number().int().nonnegative().optional().describe("Number of users who have favorited this skill"),
  isFavorited: z.boolean().optional().describe("Whether the requesting user has favorited this skill"),
  createdAt: z.coerce.date(),
  updatedAt: z.coerce.date(),
  name: z.string().describe("Name of the skill"),
  description: z.string().describe("Description of what the skill does and when to use it"),
  instructions: z.string().describe("Markdown instructions for the skill"),
  license: z.string().optional().describe("License identifier for the skill"),
  compatibility: z.unknown().optional().describe("Compatibility requirements"),
  source: sourceSchema.optional().describe("Source location of the skill"),
  references: z.array(z.string()).optional().describe("List of reference file paths"),
  scripts: z.array(z.string()).optional().describe("List of script file paths"),
  assets: z.array(z.string()).optional().describe("List of asset file paths"),
  files: z.array(fileNodeSchema).optional().describe("Full file tree structure for the skill"),
  metadata: z.record(z.string(), z.unknown()).optional().describe("Additional metadata for the skill")
});
var listStoredSkillsResponseSchema = paginationInfoSchema.extend({
  skills: z.array(storedSkillSchema)
});
var getStoredSkillResponseSchema = storedSkillSchema;
var createStoredSkillResponseSchema = storedSkillSchema;
var updateStoredSkillResponseSchema = z.union([
  z.object({
    id: z.string(),
    status: z.string(),
    activeVersionId: z.string().optional(),
    authorId: z.string().optional(),
    visibility: z.enum(["private", "public"]).optional(),
    createdAt: z.coerce.date(),
    updatedAt: z.coerce.date()
  }),
  storedSkillSchema
]);
var deleteStoredSkillResponseSchema = z.object({
  success: z.boolean(),
  message: z.string()
});
var publishStoredSkillBodySchema = z.object({
  skillPath: z.string().describe("Path to the skill directory on the server filesystem (containing SKILL.md)")
});
var publishStoredSkillResponseSchema = storedSkillSchema;

export { SKILL_ORIGIN_METADATA_KEY, buildOriginMetadata, createStoredSkillBodySchema, createStoredSkillResponseSchema, deleteStoredSkillResponseSchema, getStoredSkillResponseSchema, listStoredSkillsQuerySchema, listStoredSkillsResponseSchema, publishStoredSkillBodySchema, publishStoredSkillResponseSchema, readSkillOrigin, skillOriginSchema, storedSkillIdPathParams, storedSkillSchema, updateStoredSkillBodySchema, updateStoredSkillResponseSchema };
