// @mastra/server/dist/chunk-LZ32IIH2.cjs
'use strict';

var chunkDIG2K5CV_cjs = require('./chunk-DIG2K5CV.cjs');
var v4 = require('zod/v4');

var vectorNamePathParams = v4.z.object({
  vectorName: v4.z.string().describe("Name of the vector store")
});
var vectorIndexPathParams = vectorNamePathParams.extend({
  indexName: v4.z.string().describe("Name of the index")
});
var indexBodyBaseSchema = v4.z.object({
  indexName: v4.z.string()
});
var upsertVectorsBodySchema = indexBodyBaseSchema.extend({
  vectors: v4.z.array(v4.z.array(v4.z.number())),
  metadata: v4.z.array(v4.z.record(v4.z.string(), v4.z.any())).optional(),
  ids: v4.z.array(v4.z.string()).optional()
});
var createIndexBodySchema = indexBodyBaseSchema.extend({
  dimension: v4.z.number(),
  metric: v4.z.enum(["cosine", "euclidean", "dotproduct"]).optional()
});
var queryVectorsBodySchema = indexBodyBaseSchema.extend({
  queryVector: v4.z.array(v4.z.number()),
  topK: v4.z.number().optional(),
  filter: v4.z.record(v4.z.string(), v4.z.any()).optional(),
  includeVector: v4.z.boolean().optional()
});
var upsertVectorsResponseSchema = v4.z.object({
  ids: v4.z.array(v4.z.string())
});
var createIndexResponseSchema = chunkDIG2K5CV_cjs.successResponseSchema;
var queryVectorsResponseSchema = v4.z.array(v4.z.unknown());
var listIndexesResponseSchema = v4.z.array(v4.z.string());
var describeIndexResponseSchema = v4.z.object({
  dimension: v4.z.number(),
  count: v4.z.number(),
  metric: v4.z.string().optional()
});
var deleteIndexResponseSchema = chunkDIG2K5CV_cjs.successResponseSchema;
var listVectorsResponseSchema = v4.z.object({
  vectors: v4.z.array(
    v4.z.object({
      id: v4.z.string(),
      name: v4.z.string(),
      type: v4.z.string(),
      description: v4.z.string().optional()
    })
  )
});
var listEmbeddersResponseSchema = v4.z.object({
  embedders: v4.z.array(
    v4.z.object({
      id: v4.z.string(),
      provider: v4.z.string(),
      name: v4.z.string(),
      description: v4.z.string(),
      dimensions: v4.z.number(),
      maxInputTokens: v4.z.number()
    })
  )
});

exports.createIndexBodySchema = createIndexBodySchema;
exports.createIndexResponseSchema = createIndexResponseSchema;
exports.deleteIndexResponseSchema = deleteIndexResponseSchema;
exports.describeIndexResponseSchema = describeIndexResponseSchema;
exports.listEmbeddersResponseSchema = listEmbeddersResponseSchema;
exports.listIndexesResponseSchema = listIndexesResponseSchema;
exports.listVectorsResponseSchema = listVectorsResponseSchema;
exports.queryVectorsBodySchema = queryVectorsBodySchema;
exports.queryVectorsResponseSchema = queryVectorsResponseSchema;
exports.upsertVectorsBodySchema = upsertVectorsBodySchema;
exports.upsertVectorsResponseSchema = upsertVectorsResponseSchema;
exports.vectorIndexPathParams = vectorIndexPathParams;
exports.vectorNamePathParams = vectorNamePathParams;
