// @mastra/server/dist/chunk-K6WBL3KZ.js
import { createPagePaginationSchema, paginationInfoSchema } from './chunk-2YY3EMMS.js';
import { z } from 'zod/v4';

var storedMCPClientIdPathParams = z.object({
  storedMCPClientId: z.string().describe("Unique identifier for the stored MCP client")
});
var storageOrderBySchema = z.object({
  field: z.enum(["createdAt", "updatedAt"]).optional(),
  direction: z.enum(["ASC", "DESC"]).optional()
});
var listStoredMCPClientsQuerySchema = createPagePaginationSchema(100).extend({
  orderBy: storageOrderBySchema.optional(),
  status: z.enum(["draft", "published", "archived"]).optional().default("published").describe("Filter MCP clients by status (defaults to published)"),
  authorId: z.string().optional().describe("Filter MCP clients by author identifier"),
  metadata: z.record(z.string(), z.unknown()).optional().describe("Filter MCP clients by metadata key-value pairs")
});
var mcpServerConfigSchema = z.object({
  type: z.enum(["stdio", "http"]).describe("Transport type: stdio for local processes, http for remote servers"),
  command: z.string().optional().describe("Command to run (stdio only)"),
  args: z.array(z.string()).optional().describe("Command arguments (stdio only)"),
  env: z.record(z.string(), z.string()).optional().describe("Environment variables (stdio only)"),
  url: z.string().optional().describe("Server URL (http only)"),
  timeout: z.number().optional().describe("Connection timeout in milliseconds")
});
var snapshotConfigSchema = z.object({
  name: z.string().describe("Name of the MCP client"),
  description: z.string().optional().describe("Description of the MCP client"),
  servers: z.record(z.string(), mcpServerConfigSchema).describe("Map of server name to server configuration")
});
var createStoredMCPClientBodySchema = z.object({
  id: z.string().optional().describe("Unique identifier. If not provided, derived from name."),
  authorId: z.string().optional().describe("Author identifier for multi-tenant filtering"),
  metadata: z.record(z.string(), z.unknown()).optional().describe("Additional metadata for the MCP client")
}).merge(snapshotConfigSchema);
var updateStoredMCPClientBodySchema = z.object({
  authorId: z.string().optional(),
  metadata: z.record(z.string(), z.unknown()).optional()
}).partial().merge(snapshotConfigSchema.partial());
var storedMCPClientSchema = z.object({
  id: z.string(),
  status: z.string().describe("MCP client status: draft, published, or archived"),
  activeVersionId: z.string().optional(),
  authorId: z.string().optional(),
  metadata: z.record(z.string(), z.unknown()).optional(),
  createdAt: z.coerce.date(),
  updatedAt: z.coerce.date(),
  name: z.string().describe("Name of the MCP client"),
  description: z.string().optional().describe("Description of the MCP client"),
  servers: z.record(z.string(), mcpServerConfigSchema).describe("Map of server name to server configuration")
});
var listStoredMCPClientsResponseSchema = paginationInfoSchema.extend({
  mcpClients: z.array(storedMCPClientSchema)
});
var getStoredMCPClientResponseSchema = storedMCPClientSchema;
var createStoredMCPClientResponseSchema = storedMCPClientSchema;
var updateStoredMCPClientResponseSchema = z.union([
  z.object({
    id: z.string(),
    status: z.string(),
    activeVersionId: z.string().optional(),
    authorId: z.string().optional(),
    metadata: z.record(z.string(), z.unknown()).optional(),
    createdAt: z.coerce.date(),
    updatedAt: z.coerce.date()
  }),
  storedMCPClientSchema
]);
var deleteStoredMCPClientResponseSchema = z.object({
  success: z.boolean(),
  message: z.string()
});

export { createStoredMCPClientBodySchema, createStoredMCPClientResponseSchema, deleteStoredMCPClientResponseSchema, getStoredMCPClientResponseSchema, listStoredMCPClientsQuerySchema, listStoredMCPClientsResponseSchema, storedMCPClientIdPathParams, storedMCPClientSchema, updateStoredMCPClientBodySchema, updateStoredMCPClientResponseSchema };
