// @mastra/server/dist/chunk-Z7LCIYK7.cjs
'use strict';

var chunk64ITUOXI_cjs = require('./chunk-64ITUOXI.cjs');

// src/server/handlers/error.ts
var MODEL_NOT_ALLOWED_CODE = "MODEL_NOT_ALLOWED";
function isModelNotAllowedError(error) {
  return error instanceof Error && error.code === MODEL_NOT_ALLOWED_CODE;
}
function isZodError(error) {
  return !!error && typeof error === "object" && error.name === "ZodError" && Array.isArray(error.issues);
}
function formatZodError(error, context) {
  const issues = error.issues.map((e) => ({
    field: e.path.length > 0 ? e.path.map((p) => String(p)).join(".") : "root",
    message: e.message
  }));
  return {
    error: `Invalid ${context}`,
    issues
  };
}
function handleError(error, defaultMessage) {
  if (isModelNotAllowedError(error)) {
    const body = {
      error: {
        code: error.code,
        message: error.message,
        allowed: error.allowed,
        attempted: error.attempted,
        offendingLabel: error.offendingLabel
      }
    };
    const res = new Response(JSON.stringify(body), {
      status: 422,
      headers: { "content-type": "application/json" }
    });
    throw new chunk64ITUOXI_cjs.HTTPException(422, {
      res,
      message: error.message,
      cause: error
    });
  }
  const apiError = error;
  const apiErrorStatus = apiError.status || apiError.details?.status || 500;
  throw new chunk64ITUOXI_cjs.HTTPException(apiErrorStatus, {
    message: apiError.message || defaultMessage,
    stack: apiError.stack,
    cause: apiError.cause
  });
}

exports.formatZodError = formatZodError;
exports.handleError = handleError;
exports.isZodError = isZodError;
