// @mastra/server/dist/chunk-L56ROTL6.cjs
'use strict';

var chunkXOGNYKAF_cjs = require('./chunk-XOGNYKAF.cjs');
var v4 = require('zod/v4');

var actionIdPathParams = v4.z.object({
  actionId: v4.z.string().describe("Unique identifier for the agent-builder action")
});
var actionRunPathParams = v4.z.object({
  actionId: v4.z.string().describe("Unique identifier for the agent-builder action"),
  runId: v4.z.string().describe("Unique identifier for the action run")
});
var streamAgentBuilderBodySchema = chunkXOGNYKAF_cjs.streamWorkflowBodySchema;
var streamLegacyAgentBuilderBodySchema = chunkXOGNYKAF_cjs.streamLegacyWorkflowBodySchema;
var resumeAgentBuilderBodySchema = chunkXOGNYKAF_cjs.resumeBodySchema;
var startAsyncAgentBuilderBodySchema = chunkXOGNYKAF_cjs.startAsyncWorkflowBodySchema;

exports.actionIdPathParams = actionIdPathParams;
exports.actionRunPathParams = actionRunPathParams;
exports.resumeAgentBuilderBodySchema = resumeAgentBuilderBodySchema;
exports.startAsyncAgentBuilderBodySchema = startAsyncAgentBuilderBodySchema;
exports.streamAgentBuilderBodySchema = streamAgentBuilderBodySchema;
exports.streamLegacyAgentBuilderBodySchema = streamLegacyAgentBuilderBodySchema;
