// @mastra/server/dist/chunk-CHP6ZIVF.js
import { createListVersionsResponseSchema, createCompareVersionsResponseSchema, activateVersionResponseSchema, compareVersionsQuerySchema, createVersionBodySchema, deleteVersionResponseSchema, listVersionsQuerySchema } from './chunk-3OQMTFIV.js';
import { modelConfigSchema } from './chunk-GBVJ4OYO.js';
import { z } from 'zod/v4';

var listScorerVersionsQuerySchema = listVersionsQuerySchema;
var compareScorerVersionsQuerySchema = compareVersionsQuerySchema;
var createScorerVersionBodySchema = createVersionBodySchema;
var scorerVersionPathParams = z.object({
  scorerId: z.string().describe("Unique identifier for the stored scorer definition")
});
var scorerVersionIdPathParams = z.object({
  scorerId: z.string().describe("Unique identifier for the stored scorer definition"),
  versionId: z.string().describe("Unique identifier for the version (UUID)")
});
var samplingConfigSchema = z.union([
  z.object({ type: z.literal("none") }),
  z.object({ type: z.literal("ratio"), rate: z.number().min(0).max(1) })
]);
var scorerTypeEnum = z.enum([
  "llm-judge",
  "answer-relevancy",
  "answer-similarity",
  "bias",
  "context-precision",
  "context-relevance",
  "faithfulness",
  "hallucination",
  "noise-sensitivity",
  "prompt-alignment",
  "tool-call-accuracy",
  "toxicity"
]);
var scorerVersionSchema = z.object({
  id: z.string().describe("Unique identifier for the version (UUID)"),
  scorerDefinitionId: z.string().describe("ID of the scorer this version belongs to"),
  versionNumber: z.number().describe("Sequential version number (1, 2, 3, ...)"),
  // Snapshot config fields
  name: z.string().describe("Name of the scorer"),
  description: z.string().optional().describe("Description of the scorer"),
  type: scorerTypeEnum,
  model: modelConfigSchema.optional(),
  instructions: z.string().optional(),
  scoreRange: z.object({
    min: z.number().optional(),
    max: z.number().optional()
  }).optional(),
  presetConfig: z.record(z.string(), z.unknown()).optional(),
  defaultSampling: samplingConfigSchema.optional(),
  // Version metadata
  changedFields: z.array(z.string()).optional().describe("Array of field names that changed from the previous version"),
  changeMessage: z.string().optional().describe("Optional message describing the changes"),
  createdAt: z.coerce.date().describe("When this version was created")
});
var listScorerVersionsResponseSchema = createListVersionsResponseSchema(scorerVersionSchema);
var getScorerVersionResponseSchema = scorerVersionSchema;
var createScorerVersionResponseSchema = scorerVersionSchema.partial().merge(
  z.object({
    id: z.string(),
    scorerDefinitionId: z.string(),
    versionNumber: z.number(),
    createdAt: z.coerce.date()
  })
);
var activateScorerVersionResponseSchema = activateVersionResponseSchema;
var restoreScorerVersionResponseSchema = scorerVersionSchema;
var deleteScorerVersionResponseSchema = deleteVersionResponseSchema;
var compareScorerVersionsResponseSchema = createCompareVersionsResponseSchema(scorerVersionSchema);

export { activateScorerVersionResponseSchema, compareScorerVersionsQuerySchema, compareScorerVersionsResponseSchema, createScorerVersionBodySchema, createScorerVersionResponseSchema, deleteScorerVersionResponseSchema, getScorerVersionResponseSchema, listScorerVersionsQuerySchema, listScorerVersionsResponseSchema, restoreScorerVersionResponseSchema, scorerVersionIdPathParams, scorerVersionPathParams, scorerVersionSchema };
