// @mastra/server/dist/chunk-7ZW76TJN.js
import { z } from 'zod/v4';

// src/server/schemas/system.ts
var mastraPackageSchema = z.object({
  name: z.string(),
  version: z.string()
});
var observabilityRuntimeStrategySchema = z.enum([
  "realtime",
  "batch-with-updates",
  "insert-only",
  "event-sourced"
]);
var editorSourceSchema = z.enum(["code", "db"]);
var systemPackagesResponseSchema = z.object({
  packages: z.array(mastraPackageSchema),
  isDev: z.boolean(),
  cmsEnabled: z.boolean(),
  /**
   * The editor's configured source, when set. `'code'` swaps Studio's
   * Save/Publish UI for Download JSON + Open PR. `'db'` keeps the standard
   * Save/Publish flow. Omitted when the editor has no explicit source.
   */
  editorSource: editorSourceSchema.optional(),
  observabilityEnabled: z.boolean(),
  storageType: z.string().optional(),
  observabilityStorageType: z.string().optional(),
  observabilityRuntimeStrategy: observabilityRuntimeStrategySchema.optional()
});
var jsonSchemaRecordSchema = z.record(z.string(), z.unknown());
var apiSchemaResponseShapeSchema = z.object({
  kind: z.enum(["array", "record", "object-property", "single", "unknown"]),
  listProperty: z.string().optional(),
  paginationProperty: z.string().optional()
});
var apiSchemaManifestRouteSchema = z.object({
  method: z.string(),
  path: z.string(),
  responseType: z.string(),
  pathParamSchema: jsonSchemaRecordSchema.optional(),
  queryParamSchema: jsonSchemaRecordSchema.optional(),
  bodySchema: jsonSchemaRecordSchema.optional(),
  responseSchema: jsonSchemaRecordSchema.optional(),
  responseShape: apiSchemaResponseShapeSchema
});
var apiSchemaManifestResponseSchema = z.object({
  version: z.literal(1),
  routes: z.array(apiSchemaManifestRouteSchema)
});

export { apiSchemaManifestResponseSchema, apiSchemaManifestRouteSchema, apiSchemaResponseShapeSchema, editorSourceSchema, mastraPackageSchema, observabilityRuntimeStrategySchema, systemPackagesResponseSchema };
