// @mastra/server/dist/chunk-OC4S3EDO.js
import { paginationInfoSchema } from './chunk-2YY3EMMS.js';
import { z } from 'zod/v4';

var scoringSamplingConfigSchema = z.object({});
var mastraScorerConfigSchema = z.object({
  id: z.string(),
  name: z.string().optional(),
  description: z.string(),
  type: z.unknown().optional(),
  judge: z.unknown().optional()
});
var mastraScorerSchema = z.object({
  config: mastraScorerConfigSchema
});
var scorerEntrySchema = z.object({
  scorer: mastraScorerSchema,
  sampling: scoringSamplingConfigSchema.optional(),
  agentIds: z.array(z.string()),
  agentNames: z.array(z.string()),
  workflowIds: z.array(z.string()),
  isRegistered: z.boolean(),
  source: z.enum(["code", "stored"])
});
var listScorersResponseSchema = z.record(z.string(), scorerEntrySchema);
var scorerIdPathParams = z.object({
  scorerId: z.string().describe("Unique identifier for the scorer")
});
var entityPathParams = z.object({
  entityType: z.string().describe("Type of the entity (AGENT or WORKFLOW)"),
  entityId: z.string().describe("Unique identifier for the entity")
});
var listScoresByRunIdQuerySchema = z.object({
  page: z.coerce.number().optional().default(0),
  perPage: z.coerce.number().optional().default(10)
});
var listScoresByScorerIdQuerySchema = z.object({
  page: z.coerce.number().optional().default(0),
  perPage: z.coerce.number().optional().default(10),
  entityId: z.string().optional(),
  entityType: z.string().optional()
});
var listScoresByEntityIdQuerySchema = z.object({
  page: z.coerce.number().optional().default(0),
  perPage: z.coerce.number().optional().default(10)
});
var saveScoreBodySchema = z.object({
  score: z.unknown()
  // ScoreRowData - complex type
});
var scoresWithPaginationResponseSchema = z.object({
  pagination: paginationInfoSchema,
  scores: z.array(z.unknown())
  // Array of score records
});
var saveScoreResponseSchema = z.object({
  score: z.unknown()
  // ScoreRowData
});

export { entityPathParams, listScorersResponseSchema, listScoresByEntityIdQuerySchema, listScoresByRunIdQuerySchema, listScoresByScorerIdQuerySchema, saveScoreBodySchema, saveScoreResponseSchema, scorerEntrySchema, scorerIdPathParams, scoresWithPaginationResponseSchema };
