// @mastra/server/dist/chunk-TMGV2NRM.cjs
'use strict';

var v4 = require('zod/v4');

// src/server/schemas/processors.ts
var processorIdPathParams = v4.z.object({
  processorId: v4.z.string().describe("Unique identifier for the processor")
});
var processorConfigurationSchema = v4.z.object({
  agentId: v4.z.string(),
  agentName: v4.z.string(),
  type: v4.z.enum(["input", "output"])
});
var processorListConfigurationSchema = v4.z.object({
  agentId: v4.z.string(),
  type: v4.z.enum(["input", "output"])
});
var serializedProcessorSchema = v4.z.object({
  id: v4.z.string(),
  name: v4.z.string().optional(),
  description: v4.z.string().optional(),
  phases: v4.z.array(v4.z.enum(["input", "inputStep", "outputStream", "outputResult", "outputStep"])),
  agentIds: v4.z.array(v4.z.string()),
  configurations: v4.z.array(processorListConfigurationSchema),
  isWorkflow: v4.z.boolean()
});
var serializedProcessorDetailSchema = v4.z.object({
  id: v4.z.string(),
  name: v4.z.string().optional(),
  description: v4.z.string().optional(),
  phases: v4.z.array(v4.z.enum(["input", "inputStep", "outputStream", "outputResult", "outputStep"])),
  configurations: v4.z.array(processorConfigurationSchema),
  isWorkflow: v4.z.boolean()
});
var listProcessorsResponseSchema = v4.z.record(v4.z.string(), serializedProcessorSchema);
var messageContentSchema = v4.z.object({
  format: v4.z.literal(2).optional(),
  parts: v4.z.array(v4.z.any()).optional(),
  content: v4.z.string().optional()
}).passthrough();
var processorMessageSchema = v4.z.object({
  id: v4.z.string(),
  role: v4.z.enum(["user", "assistant", "system", "tool", "signal"]),
  createdAt: v4.z.coerce.date().optional(),
  content: v4.z.union([messageContentSchema, v4.z.string()])
}).passthrough();
var executeProcessorBodySchema = v4.z.object({
  phase: v4.z.enum(["input", "inputStep", "outputStream", "outputResult", "outputStep"]),
  messages: v4.z.array(processorMessageSchema),
  agentId: v4.z.string().optional(),
  requestContext: v4.z.record(v4.z.string(), v4.z.any()).optional()
});
var tripwireSchema = v4.z.object({
  triggered: v4.z.boolean(),
  reason: v4.z.string().optional(),
  metadata: v4.z.any().optional()
});
var executeProcessorResponseSchema = v4.z.object({
  success: v4.z.boolean(),
  phase: v4.z.string(),
  messages: v4.z.array(processorMessageSchema).optional(),
  messageList: v4.z.object({
    messages: v4.z.array(processorMessageSchema)
  }).optional(),
  tripwire: tripwireSchema.optional(),
  error: v4.z.string().optional()
});

exports.executeProcessorBodySchema = executeProcessorBodySchema;
exports.executeProcessorResponseSchema = executeProcessorResponseSchema;
exports.listProcessorsResponseSchema = listProcessorsResponseSchema;
exports.processorConfigurationSchema = processorConfigurationSchema;
exports.processorIdPathParams = processorIdPathParams;
exports.serializedProcessorDetailSchema = serializedProcessorDetailSchema;
exports.serializedProcessorSchema = serializedProcessorSchema;
