// @mastra/server/dist/chunk-LQZHT5U7.js
import { TRANSCRIBE_SPEECH_ROUTE, TRANSCRIBE_SPEECH_DEPRECATED_ROUTE, GET_SPEAKERS_ROUTE, GET_SPEAKERS_DEPRECATED_ROUTE, GET_LISTENER_ROUTE, GENERATE_SPEECH_ROUTE, GENERATE_SPEECH_DEPRECATED_ROUTE } from './chunk-LPT77UR6.js';
import { __export } from './chunk-PR4QN5HX.js';

// src/server/handlers/voice.ts
var voice_exports = {};
__export(voice_exports, {
  GENERATE_SPEECH_DEPRECATED_ROUTE: () => GENERATE_SPEECH_DEPRECATED_ROUTE2,
  GENERATE_SPEECH_ROUTE: () => GENERATE_SPEECH_ROUTE2,
  GET_LISTENER_ROUTE: () => GET_LISTENER_ROUTE2,
  GET_SPEAKERS_DEPRECATED_ROUTE: () => GET_SPEAKERS_DEPRECATED_ROUTE2,
  GET_SPEAKERS_ROUTE: () => GET_SPEAKERS_ROUTE2,
  TRANSCRIBE_SPEECH_DEPRECATED_ROUTE: () => TRANSCRIBE_SPEECH_DEPRECATED_ROUTE2,
  TRANSCRIBE_SPEECH_ROUTE: () => TRANSCRIBE_SPEECH_ROUTE2
});
var GET_SPEAKERS_ROUTE2 = GET_SPEAKERS_ROUTE;
var GET_SPEAKERS_DEPRECATED_ROUTE2 = GET_SPEAKERS_DEPRECATED_ROUTE;
var GENERATE_SPEECH_ROUTE2 = GENERATE_SPEECH_ROUTE;
var GENERATE_SPEECH_DEPRECATED_ROUTE2 = GENERATE_SPEECH_DEPRECATED_ROUTE;
var TRANSCRIBE_SPEECH_ROUTE2 = TRANSCRIBE_SPEECH_ROUTE;
var TRANSCRIBE_SPEECH_DEPRECATED_ROUTE2 = TRANSCRIBE_SPEECH_DEPRECATED_ROUTE;
var GET_LISTENER_ROUTE2 = GET_LISTENER_ROUTE;

export { GENERATE_SPEECH_DEPRECATED_ROUTE2 as GENERATE_SPEECH_DEPRECATED_ROUTE, GENERATE_SPEECH_ROUTE2 as GENERATE_SPEECH_ROUTE, GET_LISTENER_ROUTE2 as GET_LISTENER_ROUTE, GET_SPEAKERS_DEPRECATED_ROUTE2 as GET_SPEAKERS_DEPRECATED_ROUTE, GET_SPEAKERS_ROUTE2 as GET_SPEAKERS_ROUTE, TRANSCRIBE_SPEECH_DEPRECATED_ROUTE2 as TRANSCRIBE_SPEECH_DEPRECATED_ROUTE, TRANSCRIBE_SPEECH_ROUTE2 as TRANSCRIBE_SPEECH_ROUTE, voice_exports };
