// @mastra/server/dist/chunk-RVICRXZF.js
import { successResponseSchema } from './chunk-2YY3EMMS.js';
import { z } from 'zod/v4';

var vectorNamePathParams = z.object({
  vectorName: z.string().describe("Name of the vector store")
});
var vectorIndexPathParams = vectorNamePathParams.extend({
  indexName: z.string().describe("Name of the index")
});
var indexBodyBaseSchema = z.object({
  indexName: z.string()
});
var upsertVectorsBodySchema = indexBodyBaseSchema.extend({
  vectors: z.array(z.array(z.number())),
  metadata: z.array(z.record(z.string(), z.any())).optional(),
  ids: z.array(z.string()).optional()
});
var createIndexBodySchema = indexBodyBaseSchema.extend({
  dimension: z.number(),
  metric: z.enum(["cosine", "euclidean", "dotproduct"]).optional()
});
var queryVectorsBodySchema = indexBodyBaseSchema.extend({
  queryVector: z.array(z.number()),
  topK: z.number().optional(),
  filter: z.record(z.string(), z.any()).optional(),
  includeVector: z.boolean().optional()
});
var upsertVectorsResponseSchema = z.object({
  ids: z.array(z.string())
});
var createIndexResponseSchema = successResponseSchema;
var queryVectorsResponseSchema = z.array(z.unknown());
var listIndexesResponseSchema = z.array(z.string());
var describeIndexResponseSchema = z.object({
  dimension: z.number(),
  count: z.number(),
  metric: z.string().optional()
});
var deleteIndexResponseSchema = successResponseSchema;
var listVectorsResponseSchema = z.object({
  vectors: z.array(
    z.object({
      id: z.string(),
      name: z.string(),
      type: z.string(),
      description: z.string().optional()
    })
  )
});
var listEmbeddersResponseSchema = z.object({
  embedders: z.array(
    z.object({
      id: z.string(),
      provider: z.string(),
      name: z.string(),
      description: z.string(),
      dimensions: z.number(),
      maxInputTokens: z.number()
    })
  )
});

export { createIndexBodySchema, createIndexResponseSchema, deleteIndexResponseSchema, describeIndexResponseSchema, listEmbeddersResponseSchema, listIndexesResponseSchema, listVectorsResponseSchema, queryVectorsBodySchema, queryVectorsResponseSchema, upsertVectorsBodySchema, upsertVectorsResponseSchema, vectorIndexPathParams, vectorNamePathParams };
