// @mastra/server/dist/chunk-GFP7IMFR.cjs
'use strict';

var chunkDIG2K5CV_cjs = require('./chunk-DIG2K5CV.cjs');
var v4 = require('zod/v4');

var versionOrderBySchema = v4.z.object({
  field: v4.z.enum(["versionNumber", "createdAt"]).optional(),
  direction: v4.z.enum(["ASC", "DESC"]).optional()
});
var listVersionsQuerySchema = chunkDIG2K5CV_cjs.createPagePaginationSchema(20).extend({
  orderBy: versionOrderBySchema.optional()
});
var compareVersionsQuerySchema = v4.z.object({
  from: v4.z.string().describe("Version ID (UUID) to compare from"),
  to: v4.z.string().describe("Version ID (UUID) to compare to")
});
var createVersionBodySchema = v4.z.object({
  changeMessage: v4.z.string().max(500).optional().describe("Optional message describing the changes")
});
var activateVersionResponseSchema = v4.z.object({
  success: v4.z.boolean(),
  message: v4.z.string(),
  activeVersionId: v4.z.string()
});
var deleteVersionResponseSchema = v4.z.object({
  success: v4.z.boolean(),
  message: v4.z.string()
});
var versionDiffEntrySchema = v4.z.object({
  field: v4.z.string().describe("The field path that changed"),
  previousValue: v4.z.unknown().describe('The value in the "from" version'),
  currentValue: v4.z.unknown().describe('The value in the "to" version')
});
function createListVersionsResponseSchema(versionSchema) {
  return chunkDIG2K5CV_cjs.paginationInfoSchema.extend({
    versions: v4.z.array(versionSchema)
  });
}
function createCompareVersionsResponseSchema(versionSchema) {
  return v4.z.object({
    diffs: v4.z.array(versionDiffEntrySchema).describe("List of differences between versions"),
    fromVersion: versionSchema.describe("The source version"),
    toVersion: versionSchema.describe("The target version")
  });
}

exports.activateVersionResponseSchema = activateVersionResponseSchema;
exports.compareVersionsQuerySchema = compareVersionsQuerySchema;
exports.createCompareVersionsResponseSchema = createCompareVersionsResponseSchema;
exports.createListVersionsResponseSchema = createListVersionsResponseSchema;
exports.createVersionBodySchema = createVersionBodySchema;
exports.deleteVersionResponseSchema = deleteVersionResponseSchema;
exports.listVersionsQuerySchema = listVersionsQuerySchema;
exports.versionDiffEntrySchema = versionDiffEntrySchema;
exports.versionOrderBySchema = versionOrderBySchema;
