// @mastra/server/dist/chunk-67F5GKI5.cjs
'use strict';

var v4 = require('zod/v4');

// src/server/schemas/processor-providers.ts
var processorProviderIdPathParams = v4.z.object({
  providerId: v4.z.string().describe("Unique identifier for the processor provider")
});
var processorPhaseSchema = v4.z.enum([
  "processInput",
  "processInputStep",
  "processOutputStream",
  "processOutputResult",
  "processOutputStep"
]);
var getProcessorProvidersResponseSchema = v4.z.object({
  providers: v4.z.array(
    v4.z.object({
      id: v4.z.string(),
      name: v4.z.string(),
      description: v4.z.string().optional(),
      availablePhases: v4.z.array(processorPhaseSchema)
    })
  )
});
var getProcessorProviderResponseSchema = v4.z.object({
  id: v4.z.string(),
  name: v4.z.string(),
  description: v4.z.string().optional(),
  availablePhases: v4.z.array(processorPhaseSchema),
  configSchema: v4.z.record(v4.z.string(), v4.z.unknown())
});

exports.getProcessorProviderResponseSchema = getProcessorProviderResponseSchema;
exports.getProcessorProvidersResponseSchema = getProcessorProvidersResponseSchema;
exports.processorProviderIdPathParams = processorProviderIdPathParams;
