// @mastra/server/dist/chunk-4JILES3G.js
import { defaultOptionsSchema } from './chunk-TKP6FJQ3.js';
import { tracingOptionsSchema, coreMessageSchema, messageResponseSchema } from './chunk-2YY3EMMS.js';
import { z } from 'zod/v4';

var jsonValueSchema = z.lazy(
  () => z.union([
    z.string(),
    z.number(),
    z.boolean(),
    z.null(),
    z.array(jsonValueSchema),
    z.record(z.string(), jsonValueSchema)
  ])
);
var jsonRecordSchema = z.record(z.string(), jsonValueSchema);
var signalAttributesSchema = z.record(
  z.string(),
  z.union([z.string(), z.number(), z.boolean(), z.null(), z.undefined()])
);
var baseSignalSchema = z.object({
  id: z.string().optional(),
  createdAt: z.union([z.string(), z.date()]).optional(),
  metadata: jsonRecordSchema.optional(),
  attributes: signalAttributesSchema.optional()
});
var partProviderOptionsSchema = z.record(z.string(), z.record(z.string(), jsonValueSchema)).optional();
var signalTextPartSchema = z.object({
  type: z.literal("text"),
  text: z.string(),
  providerOptions: partProviderOptionsSchema
});
var signalFilePartSchema = z.object({
  type: z.literal("file"),
  data: z.string(),
  mediaType: z.string(),
  filename: z.string().optional(),
  providerOptions: partProviderOptionsSchema
});
var userMessageSignalContentsSchema = z.union([
  z.string(),
  z.array(z.union([signalTextPartSchema, signalFilePartSchema]))
]);
var agentMessageInputObjectSchema = z.object({
  contents: userMessageSignalContentsSchema,
  attributes: signalAttributesSchema.optional(),
  metadata: jsonRecordSchema.optional(),
  providerOptions: z.record(z.string(), z.record(z.string(), jsonValueSchema)).optional()
});
var agentMessageInputSchema = z.union([userMessageSignalContentsSchema, agentMessageInputObjectSchema]);
var agentSignalSchema = baseSignalSchema.extend({
  type: z.enum(["user", "state", "reactive", "notification", "user-message", "system-reminder"]),
  tagName: z.string().optional(),
  contents: userMessageSignalContentsSchema,
  providerOptions: z.record(z.string(), z.record(z.string(), jsonValueSchema)).optional()
});
var agentIdPathParams = z.object({
  agentId: z.string().describe("Unique identifier for the agent")
});
var agentVersionQuerySchema = z.object({
  status: z.enum(["draft", "published"]).optional().describe(
    "Which stored config version to resolve: draft (latest, default) or published (active version). Mutually exclusive with versionId."
  ),
  versionId: z.string().optional().describe(
    "Specific version ID to resolve. Mutually exclusive with status \u2014 if both are provided, versionId takes precedence."
  )
});
var toolIdPathParams = z.object({
  toolId: z.string().describe("Unique identifier for the tool")
});
var agentToolPathParams = agentIdPathParams.extend({
  toolId: z.string().describe("Unique identifier for the tool")
});
var agentSkillPathParams = agentIdPathParams.extend({
  skillName: z.string().describe("Name of the skill")
});
var modelConfigIdPathParams = agentIdPathParams.extend({
  modelConfigId: z.string().describe("Unique identifier for the model configuration")
});
var serializedProcessorSchema = z.object({
  id: z.string(),
  name: z.string().optional()
});
var serializedToolSchema = z.object({
  id: z.string(),
  description: z.string().optional(),
  inputSchema: z.string().optional(),
  outputSchema: z.string().optional(),
  requireApproval: z.boolean().optional()
});
var serializedWorkflowSchema = z.object({
  name: z.string(),
  steps: z.record(
    z.string(),
    z.object({
      id: z.string(),
      description: z.string().optional()
    })
  ).optional()
});
var serializedAgentDefinitionSchema = z.object({
  id: z.string(),
  name: z.string()
});
var systemMessageSchema = z.union([
  z.string(),
  z.array(z.string()),
  z.any(),
  // CoreSystemMessage or SystemModelMessage
  z.array(z.any())
]);
var modelConfigSchema = z.object({
  model: z.object({
    modelId: z.string(),
    provider: z.string(),
    modelVersion: z.string()
  })
  // Additional fields from AgentModelManagerConfig can be added here
});
var agentEditorConfigSchema = z.union([
  z.literal(false),
  z.object({
    instructions: z.boolean().optional(),
    tools: z.union([z.boolean(), z.object({ description: z.boolean().optional() })]).optional()
  })
]);
var serializedAgentSchema = z.object({
  name: z.string(),
  description: z.string().optional(),
  metadata: z.record(z.string(), z.unknown()).optional(),
  instructions: systemMessageSchema.optional(),
  tools: z.record(z.string(), serializedToolSchema),
  agents: z.record(z.string(), serializedAgentDefinitionSchema),
  workflows: z.record(z.string(), serializedWorkflowSchema),
  inputProcessors: z.array(serializedProcessorSchema),
  outputProcessors: z.array(serializedProcessorSchema),
  provider: z.string().optional(),
  modelId: z.string().optional(),
  modelVersion: z.string().optional(),
  supportsMemory: z.boolean().optional(),
  modelList: z.array(modelConfigSchema).optional(),
  defaultOptions: defaultOptionsSchema.optional(),
  defaultGenerateOptionsLegacy: z.record(z.string(), z.any()).optional(),
  defaultStreamOptionsLegacy: z.record(z.string(), z.any()).optional(),
  source: z.enum(["code", "stored"]).optional(),
  status: z.enum(["draft", "published", "archived"]).optional(),
  activeVersionId: z.string().optional(),
  hasDraft: z.boolean().optional(),
  editor: agentEditorConfigSchema.optional()
});
var serializedAgentWithIdSchema = serializedAgentSchema.extend({
  id: z.string()
});
var providerSchema = z.object({
  id: z.string(),
  name: z.string(),
  label: z.string().optional(),
  description: z.string().optional(),
  envVar: z.union([z.string(), z.array(z.string())]),
  connected: z.boolean(),
  docUrl: z.string().optional(),
  models: z.array(z.string())
});
var providersResponseSchema = z.object({
  providers: z.array(providerSchema)
});
var listAgentsResponseSchema = z.record(z.string(), serializedAgentSchema);
var listToolsResponseSchema = z.record(z.string(), serializedToolSchema);
var agentMemoryOptionSchema = z.object({
  thread: z.union([z.string(), z.object({ id: z.string() }).passthrough()]),
  resource: z.string(),
  options: z.record(z.string(), z.any()).optional(),
  readOnly: z.boolean().optional()
});
var toolChoiceSchema = z.union([
  z.enum(["auto", "none", "required"]),
  z.object({ type: z.literal("tool"), toolName: z.string() })
]);
var agentExecutionBodySchema = z.object({
  // REQUIRED
  messages: z.union([
    z.array(coreMessageSchema),
    // Array of messages
    z.string()
    // Single user message shorthand
  ]),
  // Message Configuration
  instructions: systemMessageSchema.optional(),
  system: systemMessageSchema.optional(),
  context: z.array(coreMessageSchema).optional(),
  // Memory & Persistence
  memory: agentMemoryOptionSchema.optional(),
  runId: z.string().optional(),
  savePerStep: z.boolean().optional(),
  // Request Context (handler-specific field - merged with server's requestContext)
  requestContext: z.record(z.string(), z.any()).optional(),
  // Version overrides for sub-agents (and future primitives)
  versions: z.object({
    agents: z.record(
      z.string(),
      z.union([z.object({ versionId: z.string() }), z.object({ status: z.enum(["draft", "published"]) })])
    ).optional(),
    defaultStatus: z.enum(["draft", "published"]).optional()
  }).optional(),
  // Execution Control
  maxSteps: z.number().optional(),
  stopWhen: z.any().optional(),
  // Model Configuration
  providerOptions: z.object({
    anthropic: z.record(z.string(), z.any()).optional(),
    google: z.record(z.string(), z.any()).optional(),
    openai: z.record(z.string(), z.any()).optional(),
    xai: z.record(z.string(), z.any()).optional()
  }).optional(),
  modelSettings: z.any().optional(),
  // Tool Configuration
  activeTools: z.array(z.string()).optional(),
  toolsets: z.record(z.string(), z.any()).optional(),
  clientTools: z.record(z.string(), z.any()).optional(),
  toolChoice: toolChoiceSchema.optional(),
  requireToolApproval: z.boolean().optional(),
  // Evaluation
  scorers: z.union([
    z.record(z.string(), z.any()),
    z.record(
      z.string(),
      z.object({
        scorer: z.string(),
        sampling: z.any().optional()
      })
    )
  ]).optional(),
  returnScorerData: z.boolean().optional(),
  // Observability
  tracingOptions: tracingOptionsSchema.optional(),
  // Structured Output
  output: z.any().optional(),
  // Zod schema, JSON schema, or structured output object
  structuredOutput: z.object({
    schema: z.object({}).passthrough(),
    model: z.union([z.string(), z.any()]).optional(),
    instructions: z.string().optional(),
    jsonPromptInjection: z.boolean().optional(),
    errorStrategy: z.enum(["strict", "warn", "fallback"]).optional(),
    fallbackValue: z.any().optional()
  }).optional(),
  // Idle-loop streaming (collapses streamUntilIdle into stream)
  untilIdle: z.union([z.boolean(), z.object({ maxIdleMs: z.number().int().positive().optional() })]).optional()
}).passthrough();
var agentExecutionLegacyBodySchema = agentExecutionBodySchema.extend({
  resourceId: z.string().optional(),
  resourceid: z.string().optional(),
  // lowercase variant
  threadId: z.string().optional()
});
var streamUntilIdleBodySchema = agentExecutionBodySchema.extend({
  maxIdleMs: z.number().int().positive().optional(),
  untilIdle: z.union([z.boolean(), z.object({ maxIdleMs: z.number().int().positive().optional() })]).optional()
});
var resumeStreamUntilIdleBodySchema = agentExecutionBodySchema.omit({ messages: true }).extend({
  runId: z.string(),
  resumeData: z.unknown().refine((x) => x !== void 0, { message: "resumeData is required" }),
  toolCallId: z.string().optional(),
  maxIdleMs: z.number().int().positive().optional()
});
var executeToolDataBodySchema = z.object({
  data: z.unknown().refine((x) => x !== void 0, { message: "data is required" })
});
var executeToolBodySchema = executeToolDataBodySchema.extend({
  requestContext: z.record(z.string(), z.any()).optional()
});
var executeToolContextBodySchema = executeToolDataBodySchema.extend({
  requestContext: z.record(z.string(), z.any()).optional()
});
var toolCallActionBodySchema = z.object({
  runId: z.string(),
  requestContext: z.record(z.string(), z.any()).optional(),
  toolCallId: z.string(),
  format: z.string().optional()
});
var networkToolCallActionBodySchema = z.object({
  runId: z.string(),
  requestContext: z.record(z.string(), z.any()).optional(),
  format: z.string().optional()
});
var approveToolCallBodySchema = toolCallActionBodySchema;
var declineToolCallBodySchema = toolCallActionBodySchema;
var approveNetworkToolCallBodySchema = networkToolCallActionBodySchema;
var declineNetworkToolCallBodySchema = networkToolCallActionBodySchema;
var toolCallResponseSchema = z.object({
  fullStream: z.any()
  // ReadableStream
});
var sendToolApprovalResponseSchema = z.object({
  accepted: z.literal(true),
  runId: z.string(),
  toolCallId: z.string().optional()
});
var resumeStreamBodySchema = agentExecutionBodySchema.omit({ messages: true }).extend({
  runId: z.string(),
  resumeData: z.unknown().refine((x) => x !== void 0, { message: "resumeData is required" }),
  toolCallId: z.string().optional()
});
var updateAgentModelBodySchema = z.object({
  modelId: z.string(),
  provider: z.string()
});
var reorderAgentModelListBodySchema = z.object({
  reorderedModelIds: z.array(z.string())
});
var updateAgentModelInModelListBodySchema = z.object({
  model: z.object({
    modelId: z.string(),
    provider: z.string()
  }).optional(),
  maxRetries: z.number().optional(),
  enabled: z.boolean().optional()
});
var modelManagementResponseSchema = messageResponseSchema;
var generateResponseSchema = z.any();
var streamResponseSchema = z.any();
var executeToolResponseSchema = z.any();
var enhanceInstructionsBodySchema = z.object({
  instructions: z.string().describe("The current agent instructions to enhance"),
  comment: z.string().describe("User comment describing how to enhance the instructions")
});
var enhanceInstructionsResponseSchema = z.object({
  explanation: z.string().describe("Explanation of the changes made"),
  new_prompt: z.string().describe("The enhanced instructions")
});
var observeAgentBodySchema = z.object({
  runId: z.string().describe("The run ID to observe/reconnect to"),
  offset: z.number().optional().describe("Resume from this event index (0-based). If omitted, replays all events.")
});
var signalActiveBehaviorSchema = z.enum(["deliver", "persist", "discard"]);
var signalIdleBehaviorSchema = z.enum(["wake", "persist", "discard"]);
var signalTargetBaseBodySchema = z.object({
  ifActive: z.object({
    behavior: signalActiveBehaviorSchema.optional(),
    attributes: signalAttributesSchema.optional()
  }).optional()
});
var signalTargetBodySchema = z.union([
  signalTargetBaseBodySchema.extend({
    runId: z.string(),
    resourceId: z.string().optional(),
    threadId: z.string().optional(),
    ifIdle: z.undefined().optional()
  }),
  signalTargetBaseBodySchema.extend({
    runId: z.undefined().optional(),
    resourceId: z.string(),
    threadId: z.string(),
    ifIdle: z.object({
      behavior: signalIdleBehaviorSchema.optional(),
      streamOptions: agentExecutionBodySchema.omit({ messages: true }).optional(),
      attributes: signalAttributesSchema.optional()
    }).optional()
  })
]);
var sendAgentSignalBodySchema = z.union([
  signalTargetBodySchema.options[0].extend({ signal: agentSignalSchema }),
  signalTargetBodySchema.options[1].extend({ signal: agentSignalSchema })
]);
var sendAgentMessageBodySchema = z.union([
  signalTargetBodySchema.options[0].extend({ message: agentMessageInputSchema }),
  signalTargetBodySchema.options[1].extend({ message: agentMessageInputSchema })
]);
var queueAgentMessageBodySchema = sendAgentMessageBodySchema;
var subscribeAgentThreadBodySchema = z.object({
  resourceId: z.string().optional(),
  threadId: z.string()
});
var abortAgentThreadBodySchema = subscribeAgentThreadBodySchema;
var sendToolApprovalBodySchema = z.object({
  resourceId: z.string(),
  threadId: z.string(),
  requestContext: z.record(z.string(), z.any()).optional(),
  toolCallId: z.string(),
  approved: z.boolean(),
  format: z.string().optional(),
  messages: z.array(coreMessageSchema).optional(),
  streamOptions: z.any().optional()
});
var abortAgentThreadResponseSchema = z.object({
  aborted: z.boolean()
});
var observeAgentResponseSchema = z.any();

export { abortAgentThreadBodySchema, abortAgentThreadResponseSchema, agentExecutionBodySchema, agentExecutionLegacyBodySchema, agentIdPathParams, agentSkillPathParams, agentToolPathParams, agentVersionQuerySchema, approveNetworkToolCallBodySchema, approveToolCallBodySchema, declineNetworkToolCallBodySchema, declineToolCallBodySchema, enhanceInstructionsBodySchema, enhanceInstructionsResponseSchema, executeToolBodySchema, executeToolContextBodySchema, executeToolResponseSchema, generateResponseSchema, listAgentsResponseSchema, listToolsResponseSchema, modelConfigIdPathParams, modelManagementResponseSchema, observeAgentBodySchema, observeAgentResponseSchema, providerSchema, providersResponseSchema, queueAgentMessageBodySchema, reorderAgentModelListBodySchema, resumeStreamBodySchema, resumeStreamUntilIdleBodySchema, sendAgentMessageBodySchema, sendAgentSignalBodySchema, sendToolApprovalBodySchema, sendToolApprovalResponseSchema, serializedAgentDefinitionSchema, serializedAgentSchema, serializedAgentWithIdSchema, serializedProcessorSchema, serializedToolSchema, serializedWorkflowSchema, streamResponseSchema, streamUntilIdleBodySchema, subscribeAgentThreadBodySchema, toolCallResponseSchema, toolIdPathParams, updateAgentModelBodySchema, updateAgentModelInModelListBodySchema };
