// @mastra/server/dist/chunk-DIG2K5CV.cjs
'use strict';

var v4 = require('zod/v4');

// src/server/schemas/common.ts
var runIdSchema = v4.z.object({
  runId: v4.z.string().describe("Unique identifier for the run")
});
var optionalRunIdSchema = v4.z.object({
  runId: v4.z.string().optional()
});
var paginationInfoSchema = v4.z.object({
  total: v4.z.number(),
  page: v4.z.number(),
  perPage: v4.z.union([v4.z.number(), v4.z.literal(false)]),
  hasMore: v4.z.boolean()
});
var createPagePaginationSchema = (defaultPerPage) => {
  const baseSchema = {
    page: v4.z.coerce.number().optional().default(0)
  };
  if (defaultPerPage !== void 0) {
    return v4.z.object({
      ...baseSchema,
      perPage: v4.z.coerce.number().optional().default(defaultPerPage)
    });
  } else {
    return v4.z.object({
      ...baseSchema,
      perPage: v4.z.coerce.number().optional()
    });
  }
};
var createCombinedPaginationSchema = () => {
  return v4.z.object({
    page: v4.z.coerce.number().optional(),
    perPage: v4.z.coerce.number().optional(),
    /**
     * @deprecated Use page and perPage instead
     */
    offset: v4.z.coerce.number().optional(),
    /**
     * @deprecated Use page and perPage instead
     */
    limit: v4.z.coerce.number().optional()
  });
};
var tracingOptionsSchema = v4.z.object({
  metadata: v4.z.record(v4.z.string(), v4.z.unknown()).optional(),
  requestContextKeys: v4.z.array(v4.z.string()).optional(),
  traceId: v4.z.string().optional(),
  parentSpanId: v4.z.string().optional(),
  tags: v4.z.array(v4.z.string()).optional(),
  hideInput: v4.z.boolean().optional(),
  hideOutput: v4.z.boolean().optional()
});
var coreMessageSchema = v4.z.any();
var successResponseSchema = v4.z.object({
  success: v4.z.boolean()
});
var messageResponseSchema = v4.z.object({
  message: v4.z.string()
});
var partialQuerySchema = v4.z.object({
  partial: v4.z.string().optional()
});
var statusQuerySchema = v4.z.object({
  status: v4.z.enum(["draft", "published", "archived"]).optional().default("published").describe("Which version to resolve: published (active version) or draft (latest version)")
});
var baseLogMessageSchema = v4.z.object({
  level: v4.z.enum(["debug", "info", "warn", "error", "silent"]),
  msg: v4.z.string(),
  time: v4.z.date(),
  context: v4.z.record(v4.z.string(), v4.z.unknown()).optional(),
  runId: v4.z.string().optional(),
  pid: v4.z.number(),
  hostname: v4.z.string(),
  name: v4.z.string()
});

exports.baseLogMessageSchema = baseLogMessageSchema;
exports.coreMessageSchema = coreMessageSchema;
exports.createCombinedPaginationSchema = createCombinedPaginationSchema;
exports.createPagePaginationSchema = createPagePaginationSchema;
exports.messageResponseSchema = messageResponseSchema;
exports.optionalRunIdSchema = optionalRunIdSchema;
exports.paginationInfoSchema = paginationInfoSchema;
exports.partialQuerySchema = partialQuerySchema;
exports.runIdSchema = runIdSchema;
exports.statusQuerySchema = statusQuerySchema;
exports.successResponseSchema = successResponseSchema;
exports.tracingOptionsSchema = tracingOptionsSchema;
