// @mastra/server/dist/chunk-BIUL4Y6W.js
import z from 'zod';

// src/server/schemas/responses.ts
var responseIdPathParams = z.object({
  responseId: z.string().describe("Unique identifier for the stored response")
});
var responseInputTextPartSchema = z.object({
  type: z.enum(["input_text", "text", "output_text"]),
  text: z.string()
});
var responseInputMessageSchema = z.object({
  role: z.enum(["system", "developer", "user", "assistant"]),
  content: z.union([z.string(), z.array(responseInputTextPartSchema)])
});
var providerOptionValuesSchema = z.record(z.string(), z.unknown());
var providerOptionsSchema = z.object({
  openai: providerOptionValuesSchema.optional().describe("OpenAI provider options such as previousResponseId, conversation, or responseId"),
  azure: providerOptionValuesSchema.optional().describe("Azure OpenAI provider options such as previousResponseId, store, or itemId")
}).passthrough();
var responseTextFormatJsonObjectSchema = z.object({
  type: z.literal("json_object").describe("Requests JSON object output compatibility for the response")
});
var responseTextFormatJsonSchemaSchema = z.object({
  type: z.literal("json_schema"),
  name: z.string(),
  description: z.string().optional(),
  schema: z.record(z.string(), z.unknown()),
  strict: z.boolean().optional()
});
var responseTextFormatSchema = z.union([
  responseTextFormatJsonObjectSchema,
  responseTextFormatJsonSchemaSchema
]);
var responseTextSchema = z.object({
  format: responseTextFormatSchema
});
var createResponseBodySchema = z.object({
  model: z.string().optional().describe(
    "Optional model identifier override, such as openai/gpt-5. When omitted, the agent default model is used."
  ),
  agent_id: z.string().min(1).optional().describe("Mastra agent ID for the request. Required unless previous_response_id is provided."),
  input: z.union([z.string(), z.array(responseInputMessageSchema)]),
  instructions: z.string().optional(),
  text: responseTextSchema.optional().describe(
    "Optional text output format. Supports `json_object` for JSON mode and `json_schema` for schema-constrained structured output."
  ),
  conversation_id: z.string().optional().describe("Optional conversation ID. In Mastra this is the raw threadId."),
  providerOptions: providerOptionsSchema.optional().describe("Optional provider-specific options passed through to the underlying model call"),
  stream: z.boolean().optional().default(false),
  store: z.boolean().optional().default(false),
  previous_response_id: z.string().min(1).optional()
}).passthrough().refine((data) => data.agent_id || data.previous_response_id, {
  message: "agent_id is required unless previous_response_id is provided",
  path: []
});
var responseOutputTextSchema = z.object({
  type: z.literal("output_text"),
  text: z.string(),
  annotations: z.array(z.unknown()).optional(),
  logprobs: z.array(z.unknown()).optional()
});
var responseInputTextContentSchema = z.object({
  type: z.literal("input_text"),
  text: z.string()
});
var conversationMessageContentSchema = z.union([responseInputTextContentSchema, responseOutputTextSchema]);
var responseOutputMessageSchema = z.object({
  id: z.string(),
  type: z.literal("message"),
  role: z.literal("assistant"),
  status: z.enum(["in_progress", "completed", "incomplete"]),
  content: z.array(responseOutputTextSchema)
});
var responseOutputFunctionCallSchema = z.object({
  id: z.string(),
  type: z.literal("function_call"),
  call_id: z.string(),
  name: z.string(),
  arguments: z.string(),
  status: z.enum(["in_progress", "completed", "incomplete"]).optional()
});
var responseOutputFunctionCallOutputSchema = z.object({
  id: z.string(),
  type: z.literal("function_call_output"),
  call_id: z.string(),
  output: z.string()
});
var responseUsageSchema = z.object({
  input_tokens: z.number(),
  output_tokens: z.number(),
  total_tokens: z.number(),
  input_tokens_details: z.object({
    cached_tokens: z.number()
  }).optional(),
  output_tokens_details: z.object({
    reasoning_tokens: z.number()
  }).optional()
});
var responseToolSchema = z.object({
  type: z.literal("function"),
  name: z.string(),
  description: z.string().optional(),
  parameters: z.unknown().optional()
});
var responseOutputItemSchema = z.union([
  responseOutputMessageSchema,
  responseOutputFunctionCallSchema,
  responseOutputFunctionCallOutputSchema
]);
var conversationItemMessageSchema = z.object({
  id: z.string(),
  type: z.literal("message"),
  role: z.enum(["system", "user", "assistant"]),
  status: z.literal("completed"),
  content: z.array(conversationMessageContentSchema)
});
var conversationItemSchema = z.union([
  conversationItemMessageSchema,
  responseOutputFunctionCallSchema,
  responseOutputFunctionCallOutputSchema
]);
var responseObjectSchema = z.object({
  id: z.string(),
  object: z.literal("response"),
  created_at: z.number(),
  completed_at: z.number().nullable(),
  model: z.string(),
  status: z.enum(["in_progress", "completed", "incomplete"]),
  output: z.array(responseOutputItemSchema),
  usage: responseUsageSchema.nullable(),
  error: z.null().optional(),
  incomplete_details: z.null().optional(),
  instructions: z.string().nullable().optional(),
  text: responseTextSchema.nullable().optional(),
  previous_response_id: z.string().nullable().optional(),
  conversation_id: z.string().nullable().optional(),
  providerOptions: providerOptionsSchema.optional(),
  tools: z.array(responseToolSchema).optional(),
  store: z.boolean().optional()
});
var deleteResponseSchema = z.object({
  id: z.string(),
  object: z.literal("response"),
  deleted: z.literal(true)
});

export { conversationItemMessageSchema, conversationItemSchema, conversationMessageContentSchema, createResponseBodySchema, deleteResponseSchema, responseIdPathParams, responseInputMessageSchema, responseInputTextContentSchema, responseInputTextPartSchema, responseObjectSchema, responseOutputFunctionCallOutputSchema, responseOutputFunctionCallSchema, responseOutputItemSchema, responseOutputMessageSchema, responseOutputTextSchema, responseTextFormatSchema, responseToolSchema, responseUsageSchema };
