// @mastra/server/dist/chunk-EFJIK2PW.js
import { createListVersionsResponseSchema, createCompareVersionsResponseSchema, activateVersionResponseSchema, compareVersionsQuerySchema, createVersionBodySchema, deleteVersionResponseSchema, listVersionsQuerySchema } from './chunk-3OQMTFIV.js';
import { z } from 'zod/v4';

var listMCPClientVersionsQuerySchema = listVersionsQuerySchema;
var compareMCPClientVersionsQuerySchema = compareVersionsQuerySchema;
var createMCPClientVersionBodySchema = createVersionBodySchema;
var mcpClientVersionPathParams = z.object({
  mcpClientId: z.string().describe("Unique identifier for the stored MCP client")
});
var mcpClientVersionIdPathParams = z.object({
  mcpClientId: z.string().describe("Unique identifier for the stored MCP client"),
  versionId: z.string().describe("Unique identifier for the version (UUID)")
});
var mcpServerConfigSchema = z.object({
  type: z.enum(["stdio", "http"]),
  command: z.string().optional(),
  args: z.array(z.string()).optional(),
  env: z.record(z.string(), z.string()).optional(),
  url: z.string().optional(),
  timeout: z.number().optional()
});
var mcpClientVersionSchema = z.object({
  id: z.string().describe("Unique identifier for the version (UUID)"),
  mcpClientId: z.string().describe("ID of the MCP client this version belongs to"),
  versionNumber: z.number().describe("Sequential version number (1, 2, 3, ...)"),
  // Snapshot config fields
  name: z.string().describe("Name of the MCP client"),
  description: z.string().optional().describe("Description of the MCP client"),
  servers: z.record(z.string(), mcpServerConfigSchema),
  // Version metadata
  changedFields: z.array(z.string()).optional().describe("Array of field names that changed from the previous version"),
  changeMessage: z.string().optional().describe("Optional message describing the changes"),
  createdAt: z.coerce.date().describe("When this version was created")
});
var listMCPClientVersionsResponseSchema = createListVersionsResponseSchema(mcpClientVersionSchema);
var getMCPClientVersionResponseSchema = mcpClientVersionSchema;
var createMCPClientVersionResponseSchema = mcpClientVersionSchema.partial().merge(
  z.object({
    id: z.string(),
    mcpClientId: z.string(),
    versionNumber: z.number(),
    createdAt: z.coerce.date()
  })
);
var activateMCPClientVersionResponseSchema = activateVersionResponseSchema;
var restoreMCPClientVersionResponseSchema = mcpClientVersionSchema;
var deleteMCPClientVersionResponseSchema = deleteVersionResponseSchema;
var compareMCPClientVersionsResponseSchema = createCompareVersionsResponseSchema(mcpClientVersionSchema);

export { activateMCPClientVersionResponseSchema, compareMCPClientVersionsQuerySchema, compareMCPClientVersionsResponseSchema, createMCPClientVersionBodySchema, createMCPClientVersionResponseSchema, deleteMCPClientVersionResponseSchema, getMCPClientVersionResponseSchema, listMCPClientVersionsQuerySchema, listMCPClientVersionsResponseSchema, mcpClientVersionIdPathParams, mcpClientVersionPathParams, mcpClientVersionSchema, restoreMCPClientVersionResponseSchema };
