// @mastra/server/dist/server/auth/index.cjs
'use strict';

var chunk4PQR3D5Y_cjs = require('../../chunk-4PQR3D5Y.cjs');
var chunkSXKQ7CSX_cjs = require('../../chunk-SXKQ7CSX.cjs');



Object.defineProperty(exports, "canAccessPublicly", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.canAccessPublicly; }
});
Object.defineProperty(exports, "checkRules", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.checkRules; }
});
Object.defineProperty(exports, "coreAuthMiddleware", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.coreAuthMiddleware; }
});
Object.defineProperty(exports, "defaultAuthConfig", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.defaultAuthConfig; }
});
Object.defineProperty(exports, "findMatchingCustomRoute", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.findMatchingCustomRoute; }
});
Object.defineProperty(exports, "getAuthenticatedUser", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.getAuthenticatedUser; }
});
Object.defineProperty(exports, "isCustomRoutePublic", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.isCustomRoutePublic; }
});
Object.defineProperty(exports, "isDevPlaygroundRequest", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.isDevPlaygroundRequest; }
});
Object.defineProperty(exports, "isProtectedCustomRoute", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.isProtectedCustomRoute; }
});
Object.defineProperty(exports, "isProtectedPath", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.isProtectedPath; }
});
Object.defineProperty(exports, "matchesOrIncludes", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.matchesOrIncludes; }
});
Object.defineProperty(exports, "pathMatchesPattern", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.pathMatchesPattern; }
});
Object.defineProperty(exports, "pathMatchesRule", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.pathMatchesRule; }
});
Object.defineProperty(exports, "supportsSessionRefresh", {
  enumerable: true,
  get: function () { return chunk4PQR3D5Y_cjs.supportsSessionRefresh; }
});
Object.defineProperty(exports, "MASTRA_USER_KEY", {
  enumerable: true,
  get: function () { return chunkSXKQ7CSX_cjs.MASTRA_USER_KEY; }
});
Object.defineProperty(exports, "MASTRA_USER_PERMISSIONS_KEY", {
  enumerable: true,
  get: function () { return chunkSXKQ7CSX_cjs.MASTRA_USER_PERMISSIONS_KEY; }
});
Object.defineProperty(exports, "MASTRA_USER_ROLES_KEY", {
  enumerable: true,
  get: function () { return chunkSXKQ7CSX_cjs.MASTRA_USER_ROLES_KEY; }
});
