// @mastra/upstash/dist/index.cjs
'use strict';

var redis = require('@mastra/redis');
var redis$1 = require('@upstash/redis');
var storage = require('@mastra/core/storage');
var error = require('@mastra/core/error');
var agent = require('@mastra/core/agent');
var evals = require('@mastra/core/evals');
var crypto$1 = require('crypto');
var vector = require('@mastra/core/vector');
var vector$1 = require('@upstash/vector');
var filter = require('@mastra/core/vector/filter');

// src/cache/index.ts
var UpstashServerCache = class extends redis.RedisServerCache {
  constructor(config, options = {}) {
    let client;
    if ("client" in config) {
      client = config.client;
    } else {
      client = new redis$1.Redis({ url: config.url, token: config.token });
    }
    const redisOptions = {
      ...redis.upstashPreset,
      keyPrefix: options.keyPrefix,
      ttlSeconds: options.ttlSeconds
    };
    super({ client }, redisOptions);
  }
};
function getKey(tableName, keys) {
  const keyParts = Object.entries(keys).filter(([_, value]) => value !== void 0).map(([key, value]) => `${key}:${value}`);
  return `${tableName}:${keyParts.join(":")}`;
}
function processRecord(tableName, record) {
  let key;
  if (tableName === storage.TABLE_MESSAGES) {
    key = getKey(tableName, { threadId: record.threadId, id: record.id });
  } else if (tableName === storage.TABLE_WORKFLOW_SNAPSHOT) {
    key = getKey(tableName, {
      namespace: record.namespace || "workflows",
      workflow_name: record.workflow_name,
      run_id: record.run_id,
      ...record.resourceId ? { resourceId: record.resourceId } : {}
    });
  } else if (tableName === storage.TABLE_SCORERS) {
    key = getKey(tableName, { id: record.id });
  } else {
    key = getKey(tableName, { id: record.id });
  }
  const processedRecord = {
    ...record,
    createdAt: storage.serializeDate(record.createdAt),
    updatedAt: storage.serializeDate(record.updatedAt)
  };
  return { key, processedRecord };
}

// src/storage/db/index.ts
function resolveUpstashConfig(config) {
  if ("client" in config) {
    return config.client;
  }
  return new redis$1.Redis({
    url: config.url,
    token: config.token
  });
}
var UpstashDB = class {
  client;
  constructor({ client }) {
    this.client = client;
  }
  async insert({ tableName, record }) {
    const { key, processedRecord } = processRecord(tableName, record);
    try {
      await this.client.set(key, processedRecord);
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "INSERT", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            tableName
          }
        },
        error$1
      );
    }
  }
  async get({ tableName, keys }) {
    const key = getKey(tableName, keys);
    try {
      const data = await this.client.get(key);
      return data || null;
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "LOAD", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            tableName
          }
        },
        error$1
      );
    }
  }
  async scanAndDelete(pattern, batchSize = 1e4) {
    let cursor = "0";
    let totalDeleted = 0;
    do {
      const [nextCursor, keys] = await this.client.scan(cursor, {
        match: pattern,
        count: batchSize
      });
      if (keys.length > 0) {
        await this.client.del(...keys);
        totalDeleted += keys.length;
      }
      cursor = nextCursor;
    } while (cursor !== "0");
    return totalDeleted;
  }
  async scanKeys(pattern, batchSize = 1e4) {
    let cursor = "0";
    let keys = [];
    do {
      const [nextCursor, batch] = await this.client.scan(cursor, {
        match: pattern,
        count: batchSize
      });
      keys.push(...batch);
      cursor = nextCursor;
    } while (cursor !== "0");
    return keys;
  }
  async deleteData({ tableName }) {
    const pattern = `${tableName}:*`;
    try {
      await this.scanAndDelete(pattern);
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "CLEAR_TABLE", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            tableName
          }
        },
        error$1
      );
    }
  }
};

// src/storage/domains/background-tasks/index.ts
function toStorageRecord(task) {
  return {
    id: task.id,
    tool_call_id: task.toolCallId,
    tool_name: task.toolName,
    agent_id: task.agentId,
    thread_id: task.threadId ?? null,
    resource_id: task.resourceId ?? null,
    run_id: task.runId,
    status: task.status,
    args: task.args,
    result: task.result ?? null,
    error: task.error ?? null,
    suspend_payload: task.suspendPayload ?? null,
    retry_count: task.retryCount,
    max_retries: task.maxRetries,
    timeout_ms: task.timeoutMs,
    createdAt: task.createdAt.toISOString(),
    startedAt: task.startedAt?.toISOString() ?? null,
    suspendedAt: task.suspendedAt?.toISOString() ?? null,
    completedAt: task.completedAt?.toISOString() ?? null
  };
}
function fromStorageRecord(record) {
  return {
    id: record.id,
    status: record.status,
    toolName: record.tool_name,
    toolCallId: record.tool_call_id,
    args: record.args ?? {},
    agentId: record.agent_id,
    threadId: record.thread_id ?? void 0,
    resourceId: record.resource_id ?? void 0,
    runId: record.run_id ?? "",
    result: record.result ?? void 0,
    error: record.error ?? void 0,
    suspendPayload: record.suspend_payload ?? void 0,
    retryCount: Number(record.retry_count ?? 0),
    maxRetries: Number(record.max_retries ?? 0),
    timeoutMs: Number(record.timeout_ms ?? 3e5),
    createdAt: new Date(record.createdAt),
    startedAt: record.startedAt ? new Date(record.startedAt) : void 0,
    suspendedAt: record.suspendedAt ? new Date(record.suspendedAt) : void 0,
    completedAt: record.completedAt ? new Date(record.completedAt) : void 0
  };
}
var BackgroundTasksUpstash = class extends storage.BackgroundTasksStorage {
  client;
  #db;
  constructor(config) {
    super();
    const client = resolveUpstashConfig(config);
    this.client = client;
    this.#db = new UpstashDB({ client });
  }
  async dangerouslyClearAll() {
    await this.#db.deleteData({ tableName: storage.TABLE_BACKGROUND_TASKS });
  }
  async createTask(task) {
    const record = toStorageRecord(task);
    const { key, processedRecord } = processRecord(storage.TABLE_BACKGROUND_TASKS, record);
    await this.client.set(key, processedRecord);
  }
  async updateTask(taskId, update) {
    const existing = await this.getTask(taskId);
    if (!existing) return;
    const merged = { ...existing };
    if ("status" in update) merged.status = update.status;
    if ("result" in update) merged.result = update.result;
    if ("error" in update) merged.error = update.error;
    if ("suspendPayload" in update) merged.suspendPayload = update.suspendPayload;
    if ("retryCount" in update) merged.retryCount = update.retryCount;
    if ("startedAt" in update) merged.startedAt = update.startedAt;
    if ("suspendedAt" in update) merged.suspendedAt = update.suspendedAt;
    if ("completedAt" in update) merged.completedAt = update.completedAt;
    const record = toStorageRecord(merged);
    const { key, processedRecord } = processRecord(storage.TABLE_BACKGROUND_TASKS, record);
    await this.client.set(key, processedRecord);
  }
  async getTask(taskId) {
    const key = getKey(storage.TABLE_BACKGROUND_TASKS, { id: taskId });
    const data = await this.client.get(key);
    if (!data) return null;
    return fromStorageRecord(data);
  }
  async listTasks(filter) {
    const keys = await this.#db.scanKeys(`${storage.TABLE_BACKGROUND_TASKS}:*`);
    if (keys.length === 0) return { tasks: [], total: 0 };
    const pipeline = this.client.pipeline();
    for (const key of keys) {
      pipeline.get(key);
    }
    const results = await pipeline.exec();
    let tasks = results.filter(Boolean).map((r) => fromStorageRecord(r));
    if (filter.status) {
      const statuses = Array.isArray(filter.status) ? filter.status : [filter.status];
      tasks = tasks.filter((t) => statuses.includes(t.status));
    }
    if (filter.agentId) {
      tasks = tasks.filter((t) => t.agentId === filter.agentId);
    }
    if (filter.threadId) {
      tasks = tasks.filter((t) => t.threadId === filter.threadId);
    }
    if (filter.resourceId) {
      tasks = tasks.filter((t) => t.resourceId === filter.resourceId);
    }
    if (filter.runId) {
      tasks = tasks.filter((t) => t.runId === filter.runId);
    }
    if (filter.toolName) {
      tasks = tasks.filter((t) => t.toolName === filter.toolName);
    }
    if (filter.toolCallId) {
      tasks = tasks.filter((t) => t.toolCallId === filter.toolCallId);
    }
    const dateCol = filter.dateFilterBy ?? "createdAt";
    if (filter.fromDate) {
      tasks = tasks.filter((t) => {
        const val = t[dateCol];
        return val != null && val >= filter.fromDate;
      });
    }
    if (filter.toDate) {
      tasks = tasks.filter((t) => {
        const val = t[dateCol];
        return val != null && val < filter.toDate;
      });
    }
    const orderBy = filter.orderBy ?? "createdAt";
    const direction = filter.orderDirection ?? "asc";
    tasks.sort((a, b) => {
      const aVal = a[orderBy]?.getTime() ?? 0;
      const bVal = b[orderBy]?.getTime() ?? 0;
      return direction === "asc" ? aVal - bVal : bVal - aVal;
    });
    const total = tasks.length;
    if (filter.page != null && filter.perPage != null) {
      const start = filter.page * filter.perPage;
      tasks = tasks.slice(start, start + filter.perPage);
    } else if (filter.perPage != null) {
      tasks = tasks.slice(0, filter.perPage);
    }
    return { tasks, total };
  }
  async deleteTask(taskId) {
    const key = getKey(storage.TABLE_BACKGROUND_TASKS, { id: taskId });
    await this.client.del(key);
  }
  async deleteTasks(filter) {
    const { tasks } = await this.listTasks(filter);
    if (tasks.length === 0) return;
    const keys = tasks.map((t) => getKey(storage.TABLE_BACKGROUND_TASKS, { id: t.id }));
    if (keys.length > 0) {
      await this.client.del(...keys);
    }
  }
  async getRunningCount() {
    const { total } = await this.listTasks({ status: "running" });
    return total;
  }
  async getRunningCountByAgent(agentId) {
    const { total } = await this.listTasks({ status: "running", agentId });
    return total;
  }
};
function getThreadMessagesKey(threadId) {
  return `thread:${threadId}:messages`;
}
function getMessageKey(threadId, messageId) {
  const key = getKey(storage.TABLE_MESSAGES, { threadId, id: messageId });
  return key;
}
function getMessageIndexKey(messageId) {
  return `msg-idx:${messageId}`;
}
var StoreMemoryUpstash = class extends storage.MemoryStorage {
  client;
  #db;
  constructor(config) {
    super();
    const client = resolveUpstashConfig(config);
    this.client = client;
    this.#db = new UpstashDB({ client });
  }
  async dangerouslyClearAll() {
    await this.#db.deleteData({ tableName: storage.TABLE_THREADS });
    await this.#db.deleteData({ tableName: storage.TABLE_MESSAGES });
    await this.#db.deleteData({ tableName: storage.TABLE_RESOURCES });
  }
  async getThreadById({
    threadId,
    resourceId
  }) {
    try {
      const thread = await this.#db.get({
        tableName: storage.TABLE_THREADS,
        keys: { id: threadId }
      });
      if (!thread || resourceId !== void 0 && thread.resourceId !== resourceId) return null;
      return {
        ...thread,
        createdAt: storage.ensureDate(thread.createdAt),
        updatedAt: storage.ensureDate(thread.updatedAt),
        metadata: typeof thread.metadata === "string" ? JSON.parse(thread.metadata) : thread.metadata
      };
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "GET_THREAD_BY_ID", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            threadId
          }
        },
        error$1
      );
    }
  }
  async listThreads(args) {
    const { page = 0, perPage: perPageInput, orderBy, filter } = args;
    const { field, direction } = this.parseOrderBy(orderBy);
    try {
      this.validatePaginationInput(page, perPageInput ?? 100);
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "LIST_THREADS", "INVALID_PAGE"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.USER,
          details: { page, ...perPageInput !== void 0 && { perPage: perPageInput } }
        },
        error$1 instanceof Error ? error$1 : new Error("Invalid pagination parameters")
      );
    }
    const perPage = storage.normalizePerPage(perPageInput, 100);
    try {
      this.validateMetadataKeys(filter?.metadata);
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "LIST_THREADS", "INVALID_METADATA_KEY"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.USER,
          details: { metadataKeys: filter?.metadata ? Object.keys(filter.metadata).join(", ") : "" }
        },
        error$1 instanceof Error ? error$1 : new Error("Invalid metadata key")
      );
    }
    const { offset, perPage: perPageForResponse } = storage.calculatePagination(page, perPageInput, perPage);
    try {
      let allThreads = [];
      const pattern = `${storage.TABLE_THREADS}:*`;
      const keys = await this.#db.scanKeys(pattern);
      if (keys.length === 0) {
        return {
          threads: [],
          total: 0,
          page,
          perPage: perPageForResponse,
          hasMore: false
        };
      }
      const pipeline = this.client.pipeline();
      keys.forEach((key) => pipeline.get(key));
      const results = await pipeline.exec();
      for (let i = 0; i < results.length; i++) {
        const thread = results[i];
        if (!thread) continue;
        if (filter?.resourceId && thread.resourceId !== filter.resourceId) {
          continue;
        }
        if (filter?.metadata && Object.keys(filter.metadata).length > 0) {
          const threadMetadata = typeof thread.metadata === "string" ? JSON.parse(thread.metadata) : thread.metadata;
          const matches = Object.entries(filter.metadata).every(([key, value]) => threadMetadata?.[key] === value);
          if (!matches) continue;
        }
        allThreads.push({
          ...thread,
          createdAt: storage.ensureDate(thread.createdAt),
          updatedAt: storage.ensureDate(thread.updatedAt),
          metadata: typeof thread.metadata === "string" ? JSON.parse(thread.metadata) : thread.metadata
        });
      }
      const sortedThreads = this.sortThreads(allThreads, field, direction);
      const total = sortedThreads.length;
      const end = perPageInput === false ? total : offset + perPage;
      const paginatedThreads = sortedThreads.slice(offset, end);
      const hasMore = perPageInput === false ? false : end < total;
      return {
        threads: paginatedThreads,
        total,
        page,
        perPage: perPageForResponse,
        hasMore
      };
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "LIST_THREADS", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            ...filter?.resourceId && { resourceId: filter.resourceId },
            hasMetadataFilter: !!filter?.metadata,
            page,
            perPage
          }
        },
        error$1
      );
      this.logger?.trackException(mastraError);
      this.logger.error(mastraError.toString());
      return {
        threads: [],
        total: 0,
        page,
        perPage: perPageForResponse,
        hasMore: false
      };
    }
  }
  async saveThread({ thread }) {
    try {
      await this.#db.insert({
        tableName: storage.TABLE_THREADS,
        record: thread
      });
      return thread;
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "SAVE_THREAD", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            threadId: thread.id
          }
        },
        error$1
      );
      this.logger?.trackException(mastraError);
      this.logger.error(mastraError.toString());
      throw mastraError;
    }
  }
  async updateThread({
    id,
    title,
    metadata
  }) {
    const thread = await this.getThreadById({ threadId: id });
    if (!thread) {
      throw new error.MastraError({
        id: storage.createStorageErrorId("UPSTASH", "UPDATE_THREAD", "FAILED"),
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        text: `Thread ${id} not found`,
        details: {
          threadId: id
        }
      });
    }
    const updatedThread = {
      ...thread,
      title,
      metadata: {
        ...thread.metadata,
        ...metadata
      }
    };
    try {
      await this.saveThread({ thread: updatedThread });
      return updatedThread;
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "UPDATE_THREAD", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            threadId: id
          }
        },
        error$1
      );
    }
  }
  async deleteThread({ threadId }) {
    const threadKey = getKey(storage.TABLE_THREADS, { id: threadId });
    const threadMessagesKey = getThreadMessagesKey(threadId);
    try {
      const messageIds = await this.client.zrange(threadMessagesKey, 0, -1);
      const pipeline = this.client.pipeline();
      pipeline.del(threadKey);
      pipeline.del(threadMessagesKey);
      for (let i = 0; i < messageIds.length; i++) {
        const messageId = messageIds[i];
        const messageKey = getMessageKey(threadId, messageId);
        pipeline.del(messageKey);
      }
      await pipeline.exec();
      await this.#db.scanAndDelete(getMessageKey(threadId, "*"));
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "DELETE_THREAD", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            threadId
          }
        },
        error$1
      );
    }
  }
  async saveMessages(args) {
    const { messages } = args;
    if (messages.length === 0) return { messages: [] };
    try {
      for (const message of messages) {
        if (!message.threadId) {
          throw new Error("Thread ID is required");
        }
        if (!message.resourceId) {
          throw new Error(
            `Expected to find a resourceId for message, but couldn't find one. An unexpected error has occurred.`
          );
        }
      }
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "SAVE_MESSAGES", "INVALID_ARGS"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.USER
        },
        error$1
      );
    }
    const messagesWithIndex = messages.map((message, index) => {
      return {
        ...message,
        _index: index
      };
    });
    try {
      const batchSize = 1e3;
      const targetThreadIds = new Set(messagesWithIndex.map((message) => message.threadId));
      const existingThreadIds = [];
      const touchedThreadIds = new Set(targetThreadIds);
      for (let i = 0; i < messagesWithIndex.length; i += batchSize) {
        const batch = messagesWithIndex.slice(i, i + batchSize);
        const indexLookupPipeline = this.client.pipeline();
        batch.forEach((message) => {
          indexLookupPipeline.get(getMessageIndexKey(message.id));
        });
        const batchExistingThreadIds = await indexLookupPipeline.exec();
        existingThreadIds.push(...batchExistingThreadIds);
        batchExistingThreadIds.forEach((existingThreadId) => {
          if (existingThreadId) {
            touchedThreadIds.add(existingThreadId);
          }
        });
      }
      const touchedThreadIdList = Array.from(touchedThreadIds);
      const threadLookupPipeline = this.client.pipeline();
      touchedThreadIdList.forEach((touchedThreadId) => {
        threadLookupPipeline.get(getKey(storage.TABLE_THREADS, { id: touchedThreadId }));
      });
      const threadLookupResults = await threadLookupPipeline.exec();
      const threadRecordsById = /* @__PURE__ */ new Map();
      touchedThreadIdList.forEach((touchedThreadId, index) => {
        const threadRecord = threadLookupResults[index];
        if (threadRecord) {
          threadRecordsById.set(touchedThreadId, threadRecord);
        }
      });
      for (const targetThreadId of targetThreadIds) {
        if (!threadRecordsById.has(targetThreadId)) {
          throw new error.MastraError(
            {
              id: storage.createStorageErrorId("UPSTASH", "SAVE_MESSAGES", "INVALID_ARGS"),
              domain: error.ErrorDomain.STORAGE,
              category: error.ErrorCategory.USER
            },
            new Error(`Thread ${targetThreadId} not found`)
          );
        }
      }
      for (let i = 0; i < messagesWithIndex.length; i += batchSize) {
        const batch = messagesWithIndex.slice(i, i + batchSize);
        const pipeline = this.client.pipeline();
        const batchTouchedThreadIds = /* @__PURE__ */ new Set();
        const batchExistingThreadIds = existingThreadIds.slice(i, i + batch.length);
        for (const [batchIndex, message] of batch.entries()) {
          const key = getMessageKey(message.threadId, message.id);
          const createdAtScore = new Date(message.createdAt).getTime();
          const score = message._index !== void 0 ? message._index : createdAtScore;
          batchTouchedThreadIds.add(message.threadId);
          const existingThreadId = batchExistingThreadIds[batchIndex];
          if (existingThreadId && existingThreadId !== message.threadId) {
            pipeline.del(getMessageKey(existingThreadId, message.id));
            pipeline.zrem(getThreadMessagesKey(existingThreadId), message.id);
            batchTouchedThreadIds.add(existingThreadId);
          }
          pipeline.set(key, message);
          pipeline.set(getMessageIndexKey(message.id), message.threadId);
          pipeline.zadd(getThreadMessagesKey(message.threadId), {
            score,
            member: message.id
          });
        }
        const now = /* @__PURE__ */ new Date();
        for (const touchedThreadId of batchTouchedThreadIds) {
          const existingThread = threadRecordsById.get(touchedThreadId);
          if (!existingThread) {
            continue;
          }
          const updatedThread = {
            ...existingThread,
            updatedAt: now
          };
          const threadKey = getKey(storage.TABLE_THREADS, { id: touchedThreadId });
          pipeline.set(threadKey, processRecord(storage.TABLE_THREADS, updatedThread).processedRecord);
          threadRecordsById.set(touchedThreadId, updatedThread);
        }
        await pipeline.exec();
      }
      const list = new agent.MessageList().add(messages, "memory");
      return { messages: list.get.all.db() };
    } catch (error$1) {
      if (error$1 instanceof error.MastraError) {
        throw error$1;
      }
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "SAVE_MESSAGES", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            threadIds: Array.from(new Set(messages.map((message) => message.threadId).filter(Boolean))).join(",")
          }
        },
        error$1
      );
    }
  }
  /**
   * Lookup threadId for a message - tries index first (O(1)), falls back to scan (backwards compatible)
   */
  async _getThreadIdForMessage(messageId) {
    const indexedThreadId = await this.client.get(getMessageIndexKey(messageId));
    if (indexedThreadId) {
      return indexedThreadId;
    }
    const existingKeyPattern = getMessageKey("*", messageId);
    const keys = await this.#db.scanKeys(existingKeyPattern);
    if (keys.length === 0) return null;
    const messageData = await this.client.get(keys[0]);
    if (!messageData) return null;
    if (messageData.threadId) {
      await this.client.set(getMessageIndexKey(messageId), messageData.threadId);
    }
    return messageData.threadId || null;
  }
  _sortMessages(messages, field, direction) {
    return messages.sort((a, b) => {
      const getVal = (msg) => {
        if (field === "createdAt") {
          return new Date(msg.createdAt).getTime();
        }
        const value = msg[field];
        if (typeof value === "number") return value;
        if (value instanceof Date) return value.getTime();
        return 0;
      };
      const aValue = getVal(a);
      const bValue = getVal(b);
      return direction === "ASC" ? aValue - bValue : bValue - aValue;
    });
  }
  async _getIncludedMessages(include) {
    if (!include?.length) return [];
    const messageIds = /* @__PURE__ */ new Set();
    const messageIdToThreadIds = {};
    for (const item of include) {
      const itemThreadId = await this._getThreadIdForMessage(item.id);
      if (!itemThreadId) continue;
      messageIds.add(item.id);
      messageIdToThreadIds[item.id] = itemThreadId;
      const itemThreadMessagesKey = getThreadMessagesKey(itemThreadId);
      const rank = await this.client.zrank(itemThreadMessagesKey, item.id);
      if (rank === null) continue;
      if (item.withPreviousMessages) {
        const start = Math.max(0, rank - item.withPreviousMessages);
        const prevIds = rank === 0 ? [] : await this.client.zrange(itemThreadMessagesKey, start, rank - 1);
        prevIds.forEach((id) => {
          messageIds.add(id);
          messageIdToThreadIds[id] = itemThreadId;
        });
      }
      if (item.withNextMessages) {
        const nextIds = await this.client.zrange(itemThreadMessagesKey, rank + 1, rank + item.withNextMessages);
        nextIds.forEach((id) => {
          messageIds.add(id);
          messageIdToThreadIds[id] = itemThreadId;
        });
      }
    }
    if (messageIds.size === 0) return [];
    const pipeline = this.client.pipeline();
    Array.from(messageIds).forEach((id) => {
      const tId = messageIdToThreadIds[id];
      pipeline.get(getMessageKey(tId, id));
    });
    const results = await pipeline.exec();
    return results.filter((result) => result !== null);
  }
  parseStoredMessage(storedMessage) {
    const defaultMessageContent = { format: 2, parts: [{ type: "text", text: "" }] };
    const { _index, ...rest } = storedMessage;
    return {
      ...rest,
      createdAt: new Date(rest.createdAt),
      content: rest.content || defaultMessageContent
    };
  }
  async listMessagesById({ messageIds }) {
    if (messageIds.length === 0) return { messages: [] };
    try {
      const rawMessages = [];
      const indexPipeline = this.client.pipeline();
      messageIds.forEach((id) => indexPipeline.get(getMessageIndexKey(id)));
      const indexResults = await indexPipeline.exec();
      const indexedIds = [];
      const unindexedIds = [];
      messageIds.forEach((id, i) => {
        const threadId = indexResults[i];
        if (threadId) {
          indexedIds.push({ messageId: id, threadId });
        } else {
          unindexedIds.push(id);
        }
      });
      if (indexedIds.length > 0) {
        const messagePipeline = this.client.pipeline();
        indexedIds.forEach(({ messageId, threadId }) => messagePipeline.get(getMessageKey(threadId, messageId)));
        const messageResults = await messagePipeline.exec();
        rawMessages.push(...messageResults.filter((msg) => msg !== null));
      }
      if (unindexedIds.length > 0) {
        const threadKeys = await this.client.keys("thread:*");
        const result = await Promise.all(
          threadKeys.map((threadKey) => {
            const threadId = threadKey.split(":")[1];
            if (!threadId) throw new Error(`Failed to parse thread ID from thread key "${threadKey}"`);
            return this.client.mget(
              unindexedIds.map((id) => getMessageKey(threadId, id))
            );
          })
        );
        const foundMessages = result.flat(1).filter((msg) => !!msg);
        rawMessages.push(...foundMessages);
        if (foundMessages.length > 0) {
          const backfillPipeline = this.client.pipeline();
          foundMessages.forEach((msg) => {
            if (msg.threadId) {
              backfillPipeline.set(getMessageIndexKey(msg.id), msg.threadId);
            }
          });
          await backfillPipeline.exec();
        }
      }
      const list = new agent.MessageList().add(rawMessages.map(this.parseStoredMessage), "memory");
      return { messages: list.get.all.db() };
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "LIST_MESSAGES_BY_ID", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            messageIds: JSON.stringify(messageIds)
          }
        },
        error$1
      );
    }
  }
  async listMessages(args) {
    const { threadId, resourceId, include, filter, perPage: perPageInput, page = 0, orderBy } = args;
    const threadIds = Array.isArray(threadId) ? threadId : [threadId];
    if (threadIds.length === 0 || threadIds.some((id) => !id.trim())) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "LIST_MESSAGES", "INVALID_THREAD_ID"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: { threadId: Array.isArray(threadId) ? threadId.join(",") : threadId }
        },
        new Error("threadId must be a non-empty string or array of non-empty strings")
      );
    }
    const perPage = storage.normalizePerPage(perPageInput, 40);
    const { offset, perPage: perPageForResponse } = storage.calculatePagination(page, perPageInput, perPage);
    try {
      if (page < 0) {
        throw new error.MastraError(
          {
            id: storage.createStorageErrorId("UPSTASH", "LIST_MESSAGES", "INVALID_PAGE"),
            domain: error.ErrorDomain.STORAGE,
            category: error.ErrorCategory.USER,
            details: { page }
          },
          new Error("page must be >= 0")
        );
      }
      const { field, direction } = this.parseOrderBy(orderBy, "ASC");
      if (perPage === 0 && (!include || include.length === 0)) {
        return { messages: [], total: 0, page, perPage: perPageForResponse, hasMore: false };
      }
      let includedMessages = [];
      if (include && include.length > 0) {
        const included = await this._getIncludedMessages(include);
        includedMessages = included.map(this.parseStoredMessage);
      }
      if (perPage === 0 && include && include.length > 0) {
        const list2 = new agent.MessageList().add(includedMessages, "memory");
        return {
          messages: this._sortMessages(list2.get.all.db(), field, direction),
          total: 0,
          page,
          perPage: perPageForResponse,
          hasMore: false
        };
      }
      const allMessageIdsWithThreads = [];
      for (const tid of threadIds) {
        const threadMessagesKey = getThreadMessagesKey(tid);
        const messageIds2 = await this.client.zrange(threadMessagesKey, 0, -1);
        for (const mid of messageIds2) {
          allMessageIdsWithThreads.push({ threadId: tid, messageId: mid });
        }
      }
      if (allMessageIdsWithThreads.length === 0) {
        return {
          messages: [],
          total: 0,
          page,
          perPage: perPageForResponse,
          hasMore: false
        };
      }
      const pipeline = this.client.pipeline();
      allMessageIdsWithThreads.forEach(({ threadId: tid, messageId }) => pipeline.get(getMessageKey(tid, messageId)));
      const results = await pipeline.exec();
      let messagesData = results.filter((msg) => msg !== null).map(this.parseStoredMessage);
      if (resourceId) {
        messagesData = messagesData.filter((msg) => msg.resourceId === resourceId);
      }
      messagesData = storage.filterByDateRange(
        messagesData,
        (msg) => new Date(msg.createdAt),
        filter?.dateRange
      );
      messagesData = this._sortMessages(messagesData, field, direction);
      const total = messagesData.length;
      const start = offset;
      const end = perPageInput === false ? total : start + perPage;
      const paginatedMessages = messagesData.slice(start, end);
      const messageIds = /* @__PURE__ */ new Set();
      const allMessages = [];
      for (const msg of paginatedMessages) {
        if (!messageIds.has(msg.id)) {
          allMessages.push(msg);
          messageIds.add(msg.id);
        }
      }
      for (const msg of includedMessages) {
        if (!messageIds.has(msg.id)) {
          allMessages.push(msg);
          messageIds.add(msg.id);
        }
      }
      const list = new agent.MessageList().add(allMessages, "memory");
      let finalMessages = list.get.all.db();
      finalMessages = this._sortMessages(finalMessages, field, direction);
      const returnedThreadMessageIds = new Set(finalMessages.filter((m) => m.threadId === threadId).map((m) => m.id));
      const allThreadMessagesReturned = returnedThreadMessageIds.size >= total;
      const hasMore = perPageInput !== false && !allThreadMessagesReturned && end < total;
      return {
        messages: finalMessages,
        total,
        page,
        perPage: perPageForResponse,
        hasMore
      };
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "LIST_MESSAGES", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            threadId: Array.isArray(threadId) ? threadId.join(",") : threadId,
            resourceId: resourceId ?? ""
          }
        },
        error$1
      );
      this.logger.error(mastraError.toString());
      this.logger?.trackException(mastraError);
      return {
        messages: [],
        total: 0,
        page,
        perPage: perPageForResponse,
        hasMore: false
      };
    }
  }
  async getResourceById({ resourceId }) {
    try {
      const key = `${storage.TABLE_RESOURCES}:${resourceId}`;
      const data = await this.client.get(key);
      if (!data) {
        return null;
      }
      return {
        ...data,
        createdAt: new Date(data.createdAt),
        updatedAt: new Date(data.updatedAt),
        // Ensure workingMemory is always returned as a string, regardless of automatic parsing
        workingMemory: typeof data.workingMemory === "object" ? JSON.stringify(data.workingMemory) : data.workingMemory,
        metadata: typeof data.metadata === "string" ? JSON.parse(data.metadata) : data.metadata
      };
    } catch (error) {
      this.logger.error("Error getting resource by ID:", error);
      throw error;
    }
  }
  async saveResource({ resource }) {
    try {
      const key = `${storage.TABLE_RESOURCES}:${resource.id}`;
      const serializedResource = {
        ...resource,
        metadata: JSON.stringify(resource.metadata),
        createdAt: resource.createdAt.toISOString(),
        updatedAt: resource.updatedAt.toISOString()
      };
      await this.client.set(key, serializedResource);
      return resource;
    } catch (error) {
      this.logger.error("Error saving resource:", error);
      throw error;
    }
  }
  async updateResource({
    resourceId,
    workingMemory,
    metadata
  }) {
    try {
      const existingResource = await this.getResourceById({ resourceId });
      if (!existingResource) {
        const newResource = {
          id: resourceId,
          workingMemory,
          metadata: metadata || {},
          createdAt: /* @__PURE__ */ new Date(),
          updatedAt: /* @__PURE__ */ new Date()
        };
        return this.saveResource({ resource: newResource });
      }
      const updatedResource = {
        ...existingResource,
        workingMemory: workingMemory !== void 0 ? workingMemory : existingResource.workingMemory,
        metadata: {
          ...existingResource.metadata,
          ...metadata
        },
        updatedAt: /* @__PURE__ */ new Date()
      };
      await this.saveResource({ resource: updatedResource });
      return updatedResource;
    } catch (error) {
      this.logger.error("Error updating resource:", error);
      throw error;
    }
  }
  async updateMessages(args) {
    const { messages } = args;
    if (messages.length === 0) {
      return [];
    }
    try {
      const messageIds = messages.map((m) => m.id);
      const updatesById = new Map(messages.map((message) => [message.id, message]));
      const existingMessages = [];
      const messageIdToKey = {};
      const backfillIndexValues = {};
      const indexPipeline = this.client.pipeline();
      messageIds.forEach((messageId) => indexPipeline.get(getMessageIndexKey(messageId)));
      const indexResults = await indexPipeline.exec();
      const indexedLookups = [];
      const fallbackMessageIds = [];
      messageIds.forEach((messageId, index) => {
        const indexedThreadId = indexResults[index];
        if (indexedThreadId) {
          indexedLookups.push({ messageId, threadId: indexedThreadId });
        } else {
          fallbackMessageIds.push(messageId);
        }
      });
      if (indexedLookups.length > 0) {
        const messagePipeline = this.client.pipeline();
        indexedLookups.forEach(({ messageId, threadId }) => {
          messagePipeline.get(getMessageKey(threadId, messageId));
        });
        const indexedMessages = await messagePipeline.exec();
        indexedLookups.forEach(({ messageId, threadId }, index) => {
          const key = getMessageKey(threadId, messageId);
          const message = indexedMessages[index];
          if (message && message.id === messageId) {
            existingMessages.push(message);
            messageIdToKey[messageId] = key;
          } else {
            fallbackMessageIds.push(messageId);
          }
        });
      }
      for (const messageId of fallbackMessageIds) {
        const pattern = getMessageKey("*", messageId);
        const keys = await this.#db.scanKeys(pattern);
        for (const key of keys) {
          const message = await this.client.get(key);
          if (message && message.id === messageId) {
            existingMessages.push(message);
            messageIdToKey[messageId] = key;
            if (message.threadId) {
              backfillIndexValues[messageId] = message.threadId;
            }
            break;
          }
        }
      }
      if (Object.keys(backfillIndexValues).length > 0) {
        const backfillPipeline = this.client.pipeline();
        for (const [messageId, threadId] of Object.entries(backfillIndexValues)) {
          backfillPipeline.set(getMessageIndexKey(messageId), threadId);
        }
        await backfillPipeline.exec();
      }
      if (existingMessages.length === 0) {
        return [];
      }
      const threadIdsToUpdate = /* @__PURE__ */ new Set();
      const destinationThreadIds = /* @__PURE__ */ new Set();
      for (const existingMessage of existingMessages) {
        const updatePayload = updatesById.get(existingMessage.id);
        if (!updatePayload) continue;
        const { id: _id, ...fieldsToUpdate } = updatePayload;
        if (Object.keys(fieldsToUpdate).length === 0) continue;
        threadIdsToUpdate.add(existingMessage.threadId);
        if (updatePayload.threadId && updatePayload.threadId !== existingMessage.threadId) {
          threadIdsToUpdate.add(updatePayload.threadId);
          destinationThreadIds.add(updatePayload.threadId);
        }
      }
      const threadRecordsById = /* @__PURE__ */ new Map();
      if (threadIdsToUpdate.size > 0) {
        const threadIdList = Array.from(threadIdsToUpdate);
        const threadLookupPipeline = this.client.pipeline();
        threadIdList.forEach((threadId) => {
          threadLookupPipeline.get(getKey(storage.TABLE_THREADS, { id: threadId }));
        });
        const threadLookupResults = await threadLookupPipeline.exec();
        threadIdList.forEach((threadId, index) => {
          const threadRecord = threadLookupResults[index];
          if (threadRecord) {
            threadRecordsById.set(threadId, threadRecord);
          }
        });
      }
      for (const destinationThreadId of destinationThreadIds) {
        if (!threadRecordsById.has(destinationThreadId)) {
          throw new error.MastraError(
            {
              id: storage.createStorageErrorId("UPSTASH", "UPDATE_MESSAGES", "INVALID_ARGS"),
              domain: error.ErrorDomain.STORAGE,
              category: error.ErrorCategory.USER
            },
            new Error(`Thread ${destinationThreadId} not found`)
          );
        }
      }
      const pipeline = this.client.pipeline();
      for (const existingMessage of existingMessages) {
        const updatePayload = updatesById.get(existingMessage.id);
        if (!updatePayload) continue;
        const { id, ...fieldsToUpdate } = updatePayload;
        if (Object.keys(fieldsToUpdate).length === 0) continue;
        const updatedMessage = { ...existingMessage };
        if (fieldsToUpdate.content) {
          const existingContent = existingMessage.content;
          const newContent = {
            ...existingContent,
            ...fieldsToUpdate.content,
            // Deep merge metadata if it exists on both
            ...existingContent?.metadata && fieldsToUpdate.content.metadata ? {
              metadata: {
                ...existingContent.metadata,
                ...fieldsToUpdate.content.metadata
              }
            } : {}
          };
          updatedMessage.content = newContent;
        }
        for (const key2 in fieldsToUpdate) {
          if (Object.prototype.hasOwnProperty.call(fieldsToUpdate, key2) && key2 !== "content") {
            updatedMessage[key2] = fieldsToUpdate[key2];
          }
        }
        const key = messageIdToKey[id];
        if (key) {
          if (updatePayload.threadId && updatePayload.threadId !== existingMessage.threadId) {
            const newThreadId = updatedMessage.threadId;
            const oldThreadMessagesKey = getThreadMessagesKey(existingMessage.threadId);
            pipeline.zrem(oldThreadMessagesKey, id);
            pipeline.del(key);
            const newKey = getMessageKey(newThreadId, id);
            pipeline.set(newKey, updatedMessage);
            pipeline.set(getMessageIndexKey(id), newThreadId);
            messageIdToKey[id] = newKey;
            const newThreadMessagesKey = getThreadMessagesKey(newThreadId);
            const score = updatedMessage._index !== void 0 ? updatedMessage._index : new Date(updatedMessage.createdAt).getTime();
            pipeline.zadd(newThreadMessagesKey, { score, member: id });
          } else {
            pipeline.set(key, updatedMessage);
          }
        }
      }
      const now = /* @__PURE__ */ new Date();
      for (const threadId of threadIdsToUpdate) {
        if (threadId) {
          const existingThread = threadRecordsById.get(threadId);
          if (existingThread) {
            const updatedThread = {
              ...existingThread,
              updatedAt: now
            };
            const threadKey = getKey(storage.TABLE_THREADS, { id: threadId });
            pipeline.set(threadKey, processRecord(storage.TABLE_THREADS, updatedThread).processedRecord);
            threadRecordsById.set(threadId, updatedThread);
          }
        }
      }
      await pipeline.exec();
      const updatedMessages = [];
      for (const messageId of messageIds) {
        const key = messageIdToKey[messageId];
        if (key) {
          const updatedMessage = await this.client.get(key);
          if (updatedMessage) {
            updatedMessages.push(updatedMessage);
          }
        }
      }
      return updatedMessages;
    } catch (error$1) {
      if (error$1 instanceof error.MastraError) {
        throw error$1;
      }
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "UPDATE_MESSAGES", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            messageIds: messages.map((m) => m.id).join(",")
          }
        },
        error$1
      );
    }
  }
  async deleteMessages(messageIds) {
    if (!messageIds || messageIds.length === 0) {
      return;
    }
    try {
      const threadIds = /* @__PURE__ */ new Set();
      const messageKeys = [];
      const foundMessageIds = [];
      const indexPipeline = this.client.pipeline();
      messageIds.forEach((id) => indexPipeline.get(getMessageIndexKey(id)));
      const indexResults = await indexPipeline.exec();
      const indexedMessages = [];
      const unindexedMessageIds = [];
      messageIds.forEach((id, i) => {
        const threadId = indexResults[i];
        if (threadId) {
          indexedMessages.push({ messageId: id, threadId });
        } else {
          unindexedMessageIds.push(id);
        }
      });
      for (const { messageId, threadId } of indexedMessages) {
        messageKeys.push(getMessageKey(threadId, messageId));
        foundMessageIds.push(messageId);
        threadIds.add(threadId);
      }
      for (const messageId of unindexedMessageIds) {
        const pattern = getMessageKey("*", messageId);
        const keys = await this.#db.scanKeys(pattern);
        for (const key of keys) {
          const message = await this.client.get(key);
          if (message && message.id === messageId) {
            messageKeys.push(key);
            foundMessageIds.push(messageId);
            if (message.threadId) {
              threadIds.add(message.threadId);
            }
            break;
          }
        }
      }
      if (messageKeys.length === 0) {
        return;
      }
      const pipeline = this.client.pipeline();
      for (const key of messageKeys) {
        pipeline.del(key);
      }
      for (const messageId of foundMessageIds) {
        pipeline.del(getMessageIndexKey(messageId));
      }
      if (threadIds.size > 0) {
        for (const threadId of threadIds) {
          const threadKey = getKey(storage.TABLE_THREADS, { id: threadId });
          const thread = await this.client.get(threadKey);
          if (thread) {
            const updatedThread = {
              ...thread,
              updatedAt: /* @__PURE__ */ new Date()
            };
            pipeline.set(threadKey, processRecord(storage.TABLE_THREADS, updatedThread).processedRecord);
          }
        }
      }
      await pipeline.exec();
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "DELETE_MESSAGES", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: { messageIds: messageIds.join(", ") }
        },
        error$1
      );
    }
  }
  sortThreads(threads, field, direction) {
    return threads.sort((a, b) => {
      const aValue = new Date(a[field]).getTime();
      const bValue = new Date(b[field]).getTime();
      if (direction === "ASC") {
        return aValue - bValue;
      } else {
        return bValue - aValue;
      }
    });
  }
  async cloneThread(args) {
    const { sourceThreadId, newThreadId: providedThreadId, resourceId, title, metadata, options } = args;
    const sourceThread = await this.getThreadById({ threadId: sourceThreadId });
    if (!sourceThread) {
      throw new error.MastraError({
        id: storage.createStorageErrorId("UPSTASH", "CLONE_THREAD", "SOURCE_NOT_FOUND"),
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        text: `Source thread with id ${sourceThreadId} not found`,
        details: { sourceThreadId }
      });
    }
    const newThreadId = providedThreadId || crypto.randomUUID();
    const existingThread = await this.getThreadById({ threadId: newThreadId });
    if (existingThread) {
      throw new error.MastraError({
        id: storage.createStorageErrorId("UPSTASH", "CLONE_THREAD", "THREAD_EXISTS"),
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        text: `Thread with id ${newThreadId} already exists`,
        details: { newThreadId }
      });
    }
    try {
      const threadMessagesKey = getThreadMessagesKey(sourceThreadId);
      const messageIds = await this.client.zrange(threadMessagesKey, 0, -1);
      const pipeline = this.client.pipeline();
      for (const mid of messageIds) {
        pipeline.get(getMessageKey(sourceThreadId, mid));
      }
      const results = await pipeline.exec();
      let sourceMessages = results.filter((msg) => msg !== null).map((msg) => ({
        ...msg,
        createdAt: new Date(msg.createdAt)
      }));
      if (options?.messageFilter?.startDate || options?.messageFilter?.endDate) {
        sourceMessages = storage.filterByDateRange(sourceMessages, (msg) => new Date(msg.createdAt), {
          start: options.messageFilter?.startDate,
          end: options.messageFilter?.endDate
        });
      }
      if (options?.messageFilter?.messageIds && options.messageFilter.messageIds.length > 0) {
        const messageIdSet = new Set(options.messageFilter.messageIds);
        sourceMessages = sourceMessages.filter((msg) => messageIdSet.has(msg.id));
      }
      sourceMessages.sort((a, b) => new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime());
      if (options?.messageLimit && options.messageLimit > 0 && sourceMessages.length > options.messageLimit) {
        sourceMessages = sourceMessages.slice(-options.messageLimit);
      }
      const now = /* @__PURE__ */ new Date();
      const lastMessageId = sourceMessages.length > 0 ? sourceMessages[sourceMessages.length - 1].id : void 0;
      const cloneMetadata = {
        sourceThreadId,
        clonedAt: now,
        ...lastMessageId && { lastMessageId }
      };
      const newThread = {
        id: newThreadId,
        resourceId: resourceId || sourceThread.resourceId,
        title: title || (sourceThread.title ? `Clone of ${sourceThread.title}` : ""),
        metadata: {
          ...metadata,
          clone: cloneMetadata
        },
        createdAt: now,
        updatedAt: now
      };
      const writePipeline = this.client.pipeline();
      const threadKey = getKey(storage.TABLE_THREADS, { id: newThreadId });
      writePipeline.set(threadKey, processRecord(storage.TABLE_THREADS, newThread).processedRecord);
      const clonedMessages = [];
      const messageIdMap = {};
      const targetResourceId = resourceId || sourceThread.resourceId;
      const newThreadMessagesKey = getThreadMessagesKey(newThreadId);
      for (let i = 0; i < sourceMessages.length; i++) {
        const sourceMsg = sourceMessages[i];
        const newMessageId = crypto.randomUUID();
        messageIdMap[sourceMsg.id] = newMessageId;
        const { _index, ...restMsg } = sourceMsg;
        const newMessage = {
          ...restMsg,
          id: newMessageId,
          threadId: newThreadId,
          resourceId: targetResourceId
        };
        const messageKey = getMessageKey(newThreadId, newMessageId);
        writePipeline.set(messageKey, newMessage);
        writePipeline.set(getMessageIndexKey(newMessageId), newThreadId);
        writePipeline.zadd(newThreadMessagesKey, {
          score: i,
          member: newMessageId
        });
        clonedMessages.push(newMessage);
      }
      await writePipeline.exec();
      return {
        thread: newThread,
        clonedMessages,
        messageIdMap
      };
    } catch (error$1) {
      if (error$1 instanceof error.MastraError) {
        throw error$1;
      }
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "CLONE_THREAD", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: { sourceThreadId, newThreadId }
        },
        error$1
      );
    }
  }
};
function transformScoreRow(row) {
  return storage.transformScoreRow(row);
}
var ScoresUpstash = class extends storage.ScoresStorage {
  client;
  #db;
  constructor(config) {
    super();
    const client = resolveUpstashConfig(config);
    this.client = client;
    this.#db = new UpstashDB({ client });
  }
  async dangerouslyClearAll() {
    await this.#db.deleteData({ tableName: storage.TABLE_SCORERS });
  }
  async getScoreById({ id }) {
    try {
      const data = await this.#db.get({
        tableName: storage.TABLE_SCORERS,
        keys: { id }
      });
      if (!data) return null;
      return transformScoreRow(data);
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "GET_SCORE_BY_ID", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            ...id && { id }
          }
        },
        error$1
      );
    }
  }
  async listScoresByScorerId({
    scorerId,
    entityId,
    entityType,
    source,
    pagination = { page: 0, perPage: 20 }
  }) {
    const pattern = `${storage.TABLE_SCORERS}:*`;
    const keys = await this.#db.scanKeys(pattern);
    const { page, perPage: perPageInput } = pagination;
    if (keys.length === 0) {
      return {
        scores: [],
        pagination: { total: 0, page, perPage: perPageInput, hasMore: false }
      };
    }
    const pipeline = this.client.pipeline();
    keys.forEach((key) => pipeline.get(key));
    const results = await pipeline.exec();
    const filtered = results.map((raw) => {
      if (!raw) return null;
      if (typeof raw === "string") {
        try {
          return JSON.parse(raw);
        } catch {
          return null;
        }
      }
      return raw;
    }).filter((row) => {
      if (!row || typeof row !== "object") return false;
      if (row.scorerId !== scorerId) return false;
      if (entityId && row.entityId !== entityId) return false;
      if (entityType && row.entityType !== entityType) return false;
      if (source && row.source !== source) return false;
      return true;
    });
    const perPage = storage.normalizePerPage(perPageInput, 100);
    const { offset: start, perPage: perPageForResponse } = storage.calculatePagination(page, perPageInput, perPage);
    const end = perPageInput === false ? filtered.length : start + perPage;
    const total = filtered.length;
    const paged = filtered.slice(start, end);
    const scores = paged.map((row) => transformScoreRow(row));
    return {
      scores,
      pagination: {
        total,
        page,
        perPage: perPageForResponse,
        hasMore: end < total
      }
    };
  }
  async saveScore(score) {
    let validatedScore;
    try {
      validatedScore = evals.saveScorePayloadSchema.parse(score);
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "SAVE_SCORE", "VALIDATION_FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.USER,
          details: {
            scorer: typeof score.scorer?.id === "string" ? score.scorer.id : String(score.scorer?.id ?? "unknown"),
            entityId: score.entityId ?? "unknown",
            entityType: score.entityType ?? "unknown",
            traceId: score.traceId ?? "",
            spanId: score.spanId ?? ""
          }
        },
        error$1
      );
    }
    const now = /* @__PURE__ */ new Date();
    const id = crypto.randomUUID();
    const createdAt = now;
    const updatedAt = now;
    const scoreWithId = {
      ...validatedScore,
      id,
      createdAt,
      updatedAt
    };
    const { key, processedRecord } = processRecord(storage.TABLE_SCORERS, scoreWithId);
    try {
      await this.client.set(key, processedRecord);
      return { score: { ...validatedScore, id, createdAt, updatedAt } };
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "SAVE_SCORE", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: { id }
        },
        error$1
      );
    }
  }
  async listScoresByRunId({
    runId,
    pagination = { page: 0, perPage: 20 }
  }) {
    const pattern = `${storage.TABLE_SCORERS}:*`;
    const keys = await this.#db.scanKeys(pattern);
    const { page, perPage: perPageInput } = pagination;
    if (keys.length === 0) {
      return {
        scores: [],
        pagination: { total: 0, page, perPage: perPageInput, hasMore: false }
      };
    }
    const pipeline = this.client.pipeline();
    keys.forEach((key) => pipeline.get(key));
    const results = await pipeline.exec();
    const filtered = results.map((raw) => {
      if (!raw) return null;
      if (typeof raw === "string") {
        try {
          return JSON.parse(raw);
        } catch {
          return null;
        }
      }
      return raw;
    }).filter((row) => !!row && typeof row === "object" && row.runId === runId);
    const total = filtered.length;
    const perPage = storage.normalizePerPage(perPageInput, 100);
    const { offset: start, perPage: perPageForResponse } = storage.calculatePagination(page, perPageInput, perPage);
    const end = perPageInput === false ? filtered.length : start + perPage;
    const paged = filtered.slice(start, end);
    const scores = paged.map((row) => transformScoreRow(row));
    return {
      scores,
      pagination: {
        total,
        page,
        perPage: perPageForResponse,
        hasMore: end < total
      }
    };
  }
  async listScoresByEntityId({
    entityId,
    entityType,
    pagination = { page: 0, perPage: 20 }
  }) {
    const pattern = `${storage.TABLE_SCORERS}:*`;
    const keys = await this.#db.scanKeys(pattern);
    const { page, perPage: perPageInput } = pagination;
    if (keys.length === 0) {
      return {
        scores: [],
        pagination: { total: 0, page, perPage: perPageInput, hasMore: false }
      };
    }
    const pipeline = this.client.pipeline();
    keys.forEach((key) => pipeline.get(key));
    const results = await pipeline.exec();
    const filtered = results.map((raw) => {
      if (!raw) return null;
      if (typeof raw === "string") {
        try {
          return JSON.parse(raw);
        } catch {
          return null;
        }
      }
      return raw;
    }).filter((row) => {
      if (!row || typeof row !== "object") return false;
      if (row.entityId !== entityId) return false;
      if (entityType && row.entityType !== entityType) return false;
      return true;
    });
    const total = filtered.length;
    const perPage = storage.normalizePerPage(perPageInput, 100);
    const { offset: start, perPage: perPageForResponse } = storage.calculatePagination(page, perPageInput, perPage);
    const end = perPageInput === false ? filtered.length : start + perPage;
    const paged = filtered.slice(start, end);
    const scores = paged.map((row) => transformScoreRow(row));
    return {
      scores,
      pagination: {
        total,
        page,
        perPage: perPageForResponse,
        hasMore: end < total
      }
    };
  }
  async listScoresBySpan({
    traceId,
    spanId,
    pagination = { page: 0, perPage: 20 }
  }) {
    const pattern = `${storage.TABLE_SCORERS}:*`;
    const keys = await this.#db.scanKeys(pattern);
    const { page, perPage: perPageInput } = pagination;
    if (keys.length === 0) {
      return {
        scores: [],
        pagination: { total: 0, page, perPage: perPageInput, hasMore: false }
      };
    }
    const pipeline = this.client.pipeline();
    keys.forEach((key) => pipeline.get(key));
    const results = await pipeline.exec();
    const filtered = results.map((raw) => {
      if (!raw) return null;
      if (typeof raw === "string") {
        try {
          return JSON.parse(raw);
        } catch {
          return null;
        }
      }
      return raw;
    }).filter((row) => {
      if (!row || typeof row !== "object") return false;
      if (row.traceId !== traceId) return false;
      if (row.spanId !== spanId) return false;
      return true;
    });
    const total = filtered.length;
    const perPage = storage.normalizePerPage(perPageInput, 100);
    const { offset: start, perPage: perPageForResponse } = storage.calculatePagination(page, perPageInput, perPage);
    const end = perPageInput === false ? filtered.length : start + perPage;
    const paged = filtered.slice(start, end);
    const scores = paged.map((row) => transformScoreRow(row));
    return {
      scores,
      pagination: {
        total,
        page,
        perPage: perPageForResponse,
        hasMore: end < total
      }
    };
  }
};
var WorkflowsUpstash = class extends storage.WorkflowsStorage {
  client;
  #db;
  constructor(config) {
    super();
    const client = resolveUpstashConfig(config);
    this.client = client;
    this.#db = new UpstashDB({ client });
  }
  supportsConcurrentUpdates() {
    return true;
  }
  parseWorkflowRun(row) {
    let parsedSnapshot = row.snapshot;
    if (typeof parsedSnapshot === "string") {
      try {
        parsedSnapshot = JSON.parse(row.snapshot);
      } catch (e) {
        this.logger.warn(`Failed to parse snapshot for workflow ${row.workflow_name}: ${e}`);
      }
    }
    return {
      workflowName: row.workflow_name,
      runId: row.run_id,
      snapshot: parsedSnapshot,
      createdAt: storage.ensureDate(row.createdAt),
      updatedAt: storage.ensureDate(row.updatedAt),
      resourceId: row.resourceId
    };
  }
  async dangerouslyClearAll() {
    await this.#db.deleteData({ tableName: storage.TABLE_WORKFLOW_SNAPSHOT });
  }
  async updateWorkflowResults({
    workflowName,
    runId,
    stepId,
    result,
    requestContext
  }) {
    try {
      const key = getKey(storage.TABLE_WORKFLOW_SNAPSHOT, {
        namespace: "workflows",
        workflow_name: workflowName,
        run_id: runId
      });
      const now = (/* @__PURE__ */ new Date()).toISOString();
      const luaScript = `
        local key = KEYS[1]
        local stepId = ARGV[1]
        local resultJson = ARGV[2]
        local requestContextJson = ARGV[3]
        local now = ARGV[4]
        local namespace = ARGV[5]
        local workflowName = ARGV[6]
        local runId = ARGV[7]
        local timestamp = tonumber(ARGV[8])

        -- Get existing data
        local existing = redis.call('GET', key)
        local data
        local snapshot

        if existing then
          data = cjson.decode(existing)
          snapshot = data.snapshot
          if type(snapshot) == 'string' then
            snapshot = cjson.decode(snapshot)
          end
        else
          -- Create new record with default snapshot
          snapshot = {
            context = {},
            activePaths = {},
            timestamp = timestamp,
            suspendedPaths = {},
            activeStepsPath = {},
            resumeLabels = {},
            serializedStepGraph = {},
            status = 'pending',
            value = {},
            waitingPaths = {},
            runId = runId,
            requestContext = {}
          }
          data = {
            namespace = namespace,
            workflow_name = workflowName,
            run_id = runId,
            createdAt = now,
            updatedAt = now
          }
        end

        -- Initialize context if nil
        if snapshot.context == nil then
          snapshot.context = {}
        end

        -- Merge the new step result
        local stepResult = cjson.decode(resultJson)
        snapshot.context[stepId] = stepResult

        -- Merge request context
        local newRequestContext = cjson.decode(requestContextJson)
        if snapshot.requestContext == nil then
          snapshot.requestContext = {}
        end
        for k, v in pairs(newRequestContext) do
          snapshot.requestContext[k] = v
        end

        -- Update the record
        data.snapshot = snapshot
        data.updatedAt = now

        -- Save back
        redis.call('SET', key, cjson.encode(data))

        -- Return the full updated data
        return cjson.encode(data)
      `;
      const resultJson = await this.client.eval(
        luaScript,
        [key],
        [
          stepId,
          JSON.stringify(result),
          JSON.stringify(requestContext),
          now,
          "workflows",
          workflowName,
          runId,
          String(Date.now())
        ]
      );
      let data;
      if (typeof resultJson === "string") {
        data = JSON.parse(resultJson);
      } else {
        data = resultJson;
      }
      const snapshot = typeof data.snapshot === "string" ? JSON.parse(data.snapshot) : data.snapshot;
      return snapshot.context;
    } catch (error$1) {
      if (error$1 instanceof error.MastraError) throw error$1;
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "UPDATE_WORKFLOW_RESULTS", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: { workflowName, runId, stepId }
        },
        error$1
      );
    }
  }
  async updateWorkflowState({
    workflowName,
    runId,
    opts
  }) {
    try {
      const key = getKey(storage.TABLE_WORKFLOW_SNAPSHOT, {
        namespace: "workflows",
        workflow_name: workflowName,
        run_id: runId
      });
      const now = (/* @__PURE__ */ new Date()).toISOString();
      const luaScript = `
        local key = KEYS[1]
        local optsJson = ARGV[1]
        local now = ARGV[2]

        -- Get existing data
        local existing = redis.call('GET', key)

        if not existing then
          return nil
        end

        local data = cjson.decode(existing)
        local snapshot = data.snapshot

        if type(snapshot) == 'string' then
          snapshot = cjson.decode(snapshot)
        end

        if not snapshot or not snapshot.context then
          return nil
        end

        -- Merge the new options with the existing snapshot
        local opts = cjson.decode(optsJson)
        for k, v in pairs(opts) do
          snapshot[k] = v
        end

        -- Update the record
        data.snapshot = snapshot
        data.updatedAt = now

        -- Save back
        redis.call('SET', key, cjson.encode(data))

        -- Return the full updated data
        return cjson.encode(data)
      `;
      const resultJson = await this.client.eval(luaScript, [key], [JSON.stringify(opts), now]);
      if (!resultJson) {
        return void 0;
      }
      let data;
      if (typeof resultJson === "string") {
        data = JSON.parse(resultJson);
      } else {
        data = resultJson;
      }
      const snapshot = typeof data.snapshot === "string" ? JSON.parse(data.snapshot) : data.snapshot;
      return snapshot;
    } catch (error$1) {
      if (error$1 instanceof error.MastraError) throw error$1;
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "UPDATE_WORKFLOW_STATE", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: { workflowName, runId }
        },
        error$1
      );
    }
  }
  async persistWorkflowSnapshot(params) {
    const { namespace = "workflows", workflowName, runId, resourceId, snapshot, createdAt, updatedAt } = params;
    try {
      await this.#db.insert({
        tableName: storage.TABLE_WORKFLOW_SNAPSHOT,
        record: {
          namespace,
          workflow_name: workflowName,
          run_id: runId,
          resourceId,
          snapshot,
          createdAt: createdAt ?? /* @__PURE__ */ new Date(),
          updatedAt: updatedAt ?? /* @__PURE__ */ new Date()
        }
      });
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "PERSIST_WORKFLOW_SNAPSHOT", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            namespace,
            workflowName,
            runId
          }
        },
        error$1
      );
    }
  }
  async loadWorkflowSnapshot(params) {
    const { namespace = "workflows", workflowName, runId } = params;
    const key = getKey(storage.TABLE_WORKFLOW_SNAPSHOT, {
      namespace,
      workflow_name: workflowName,
      run_id: runId
    });
    try {
      const data = await this.client.get(key);
      if (!data) return null;
      return data.snapshot;
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "LOAD_WORKFLOW_SNAPSHOT", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            namespace,
            workflowName,
            runId
          }
        },
        error$1
      );
    }
  }
  async getWorkflowRunById({
    runId,
    workflowName
  }) {
    try {
      const key = getKey(storage.TABLE_WORKFLOW_SNAPSHOT, { namespace: "workflows", workflow_name: workflowName, run_id: runId }) + "*";
      const keys = await this.#db.scanKeys(key);
      const workflows = await Promise.all(
        keys.map(async (key2) => {
          const data2 = await this.client.get(key2);
          return data2;
        })
      );
      const data = workflows.find((w) => w?.run_id === runId && w?.workflow_name === workflowName);
      if (!data) return null;
      return this.parseWorkflowRun(data);
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "GET_WORKFLOW_RUN_BY_ID", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            namespace: "workflows",
            runId,
            workflowName: workflowName || ""
          }
        },
        error$1
      );
    }
  }
  async deleteWorkflowRunById({ runId, workflowName }) {
    const key = getKey(storage.TABLE_WORKFLOW_SNAPSHOT, { namespace: "workflows", workflow_name: workflowName, run_id: runId });
    try {
      await this.client.del(key);
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "DELETE_WORKFLOW_RUN_BY_ID", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            namespace: "workflows",
            runId,
            workflowName
          }
        },
        error$1
      );
    }
  }
  async listWorkflowRuns({
    workflowName,
    fromDate,
    toDate,
    perPage,
    page,
    resourceId,
    status
  } = {}) {
    try {
      if (page !== void 0 && page < 0) {
        throw new error.MastraError(
          {
            id: storage.createStorageErrorId("UPSTASH", "LIST_WORKFLOW_RUNS", "INVALID_PAGE"),
            domain: error.ErrorDomain.STORAGE,
            category: error.ErrorCategory.USER,
            details: { page }
          },
          new Error("page must be >= 0")
        );
      }
      let pattern = getKey(storage.TABLE_WORKFLOW_SNAPSHOT, { namespace: "workflows" }) + ":*";
      if (workflowName && resourceId) {
        pattern = getKey(storage.TABLE_WORKFLOW_SNAPSHOT, {
          namespace: "workflows",
          workflow_name: workflowName,
          run_id: "*",
          resourceId
        });
      } else if (workflowName) {
        pattern = getKey(storage.TABLE_WORKFLOW_SNAPSHOT, { namespace: "workflows", workflow_name: workflowName }) + ":*";
      } else if (resourceId) {
        pattern = getKey(storage.TABLE_WORKFLOW_SNAPSHOT, {
          namespace: "workflows",
          workflow_name: "*",
          run_id: "*",
          resourceId
        });
      }
      const keys = await this.#db.scanKeys(pattern);
      if (keys.length === 0) {
        return { runs: [], total: 0 };
      }
      const pipeline = this.client.pipeline();
      keys.forEach((key) => pipeline.get(key));
      const results = await pipeline.exec();
      let runs = results.map((result) => result).filter(
        (record) => record !== null && record !== void 0 && typeof record === "object" && "workflow_name" in record
      ).filter((record) => !workflowName || record.workflow_name === workflowName).map((w) => this.parseWorkflowRun(w)).filter((w) => {
        if (fromDate && w.createdAt < fromDate) return false;
        if (toDate && w.createdAt > toDate) return false;
        if (status) {
          let snapshot = w.snapshot;
          if (typeof snapshot === "string") {
            try {
              snapshot = JSON.parse(snapshot);
            } catch (e) {
              this.logger.warn(`Failed to parse snapshot for workflow ${w.workflowName}: ${e}`);
              return false;
            }
          }
          return snapshot.status === status;
        }
        return true;
      }).sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
      const total = runs.length;
      if (typeof perPage === "number" && typeof page === "number") {
        const normalizedPerPage = storage.normalizePerPage(perPage, Number.MAX_SAFE_INTEGER);
        const offset = page * normalizedPerPage;
        runs = runs.slice(offset, offset + normalizedPerPage);
      }
      return { runs, total };
    } catch (error$1) {
      if (error$1 instanceof error.MastraError) throw error$1;
      throw new error.MastraError(
        {
          id: storage.createStorageErrorId("UPSTASH", "LIST_WORKFLOW_RUNS", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            namespace: "workflows",
            workflowName: workflowName || "",
            resourceId: resourceId || ""
          }
        },
        error$1
      );
    }
  }
};

// src/storage/index.ts
var isClientConfig = (config) => {
  return "client" in config;
};
var UpstashStore = class extends storage.MastraCompositeStore {
  redis;
  stores;
  constructor(config) {
    super({ id: config.id, name: "Upstash", disableInit: config.disableInit });
    if (isClientConfig(config)) {
      this.redis = config.client;
    } else {
      if (!config.url || typeof config.url !== "string" || config.url.trim() === "") {
        throw new Error("UpstashStore: url is required and cannot be empty.");
      }
      if (!config.token || typeof config.token !== "string" || config.token.trim() === "") {
        throw new Error("UpstashStore: token is required and cannot be empty.");
      }
      this.redis = new redis$1.Redis({
        url: config.url,
        token: config.token
      });
    }
    const scores = new ScoresUpstash({ client: this.redis });
    const workflows = new WorkflowsUpstash({ client: this.redis });
    const memory = new StoreMemoryUpstash({ client: this.redis });
    const backgroundTasks = new BackgroundTasksUpstash({ client: this.redis });
    this.stores = {
      scores,
      workflows,
      memory,
      backgroundTasks
    };
  }
  async close() {
  }
};
var UpstashFilterTranslator = class extends filter.BaseFilterTranslator {
  getSupportedOperators() {
    return {
      ...filter.BaseFilterTranslator.DEFAULT_OPERATORS,
      array: ["$in", "$nin", "$all"],
      regex: ["$regex"],
      custom: ["$contains"]
    };
  }
  translate(filter) {
    if (this.isEmpty(filter)) return void 0;
    this.validateFilter(filter);
    return this.translateNode(filter);
  }
  translateNode(node, path = "") {
    if (this.isRegex(node)) {
      throw new Error("Direct regex pattern format is not supported in Upstash");
    }
    if (node === null || node === void 0) {
      throw new Error("Filtering for null/undefined values is not supported by Upstash Vector");
    }
    if (this.isPrimitive(node)) {
      if (node === null || node === void 0) {
        throw new Error("Filtering for null/undefined values is not supported by Upstash Vector");
      }
      return this.formatComparison(path, "=", node);
    }
    if (Array.isArray(node)) {
      if (node.length === 0) {
        return "(HAS FIELD empty AND HAS NOT FIELD empty)";
      }
      return `${path} IN (${this.formatArray(node)})`;
    }
    const entries = Object.entries(node);
    const conditions = [];
    for (const [key, value] of entries) {
      const newPath = path ? `${path}.${key}` : key;
      if (this.isOperator(key)) {
        conditions.push(this.translateOperator(key, value, path));
      } else if (typeof value === "object" && value !== null) {
        conditions.push(this.translateNode(value, newPath));
      } else if (value === null || value === void 0) {
        throw new Error("Filtering for null/undefined values is not supported by Upstash Vector");
      } else {
        conditions.push(this.formatComparison(newPath, "=", value));
      }
    }
    return conditions.length > 1 ? `(${conditions.join(" AND ")})` : conditions[0] ?? "";
  }
  COMPARISON_OPS = {
    $eq: "=",
    $ne: "!=",
    $gt: ">",
    $gte: ">=",
    $lt: "<",
    $lte: "<="
  };
  translateOperator(operator, value, path) {
    if (this.isBasicOperator(operator) || this.isNumericOperator(operator)) {
      return this.formatComparison(path, this.COMPARISON_OPS[operator], value);
    }
    switch (operator) {
      case "$in":
        if (!Array.isArray(value) || value.length === 0) {
          return "(HAS FIELD empty AND HAS NOT FIELD empty)";
        }
        return `${path} IN (${this.formatArray(value)})`;
      case "$nin":
        return `${path} NOT IN (${this.formatArray(value)})`;
      case "$contains":
        return `${path} CONTAINS ${this.formatValue(value)}`;
      case "$regex":
        return `${path} GLOB ${this.formatValue(value)}`;
      case "$exists":
        return value ? `HAS FIELD ${path}` : `HAS NOT FIELD ${path}`;
      case "$and":
        if (!Array.isArray(value) || value.length === 0) {
          return "(HAS FIELD empty OR HAS NOT FIELD empty)";
        }
        return this.joinConditions(value, "AND");
      case "$or":
        if (!Array.isArray(value) || value.length === 0) {
          return "(HAS FIELD empty AND HAS NOT FIELD empty)";
        }
        return this.joinConditions(value, "OR");
      case "$not":
        return this.formatNot(path, value);
      case "$nor":
        return this.formatNot("", { $or: value });
      case "$all":
        return this.translateOperator(
          "$and",
          value.map((item) => ({ [path]: { $contains: item } })),
          ""
        );
      default:
        throw new Error(`Unsupported operator: ${operator}`);
    }
  }
  NEGATED_OPERATORS = {
    $eq: "$ne",
    $ne: "$eq",
    $gt: "$lte",
    $gte: "$lt",
    $lt: "$gte",
    $lte: "$gt",
    $in: "$nin",
    $nin: "$in",
    $exists: "$exists"
    // Special case - we'll flip the value
  };
  formatNot(path, value) {
    if (typeof value !== "object") {
      return `${path} != ${this.formatValue(value)}`;
    }
    if (!Object.keys(value).some((k) => k.startsWith("$"))) {
      const [fieldName, fieldValue] = Object.entries(value)[0] ?? [];
      if (typeof fieldValue === "object" && fieldValue !== null && Object.keys(fieldValue)[0]?.startsWith("$")) {
        const [op2, val2] = Object.entries(fieldValue)[0] ?? [];
        const negatedOp = this.NEGATED_OPERATORS[op2];
        if (!negatedOp) throw new Error(`Unsupported operator in NOT: ${op2}`);
        if (op2 === "$exists") {
          return this.translateOperator(op2, !val2, fieldName ?? "");
        }
        return this.translateOperator(negatedOp, val2, fieldName ?? "");
      }
      return `${fieldName} != ${this.formatValue(fieldValue)}`;
    }
    const [op, val] = Object.entries(value)[0] ?? [];
    if (op === "$lt") return `${path} >= ${this.formatValue(val)}`;
    if (op === "$lte") return `${path} > ${this.formatValue(val)}`;
    if (op === "$gt") return `${path} <= ${this.formatValue(val)}`;
    if (op === "$gte") return `${path} < ${this.formatValue(val)}`;
    if (op === "$ne") return `${path} = ${this.formatValue(val)}`;
    if (op === "$eq") return `${path} != ${this.formatValue(val)}`;
    if (op === "$contains") return `${path} NOT CONTAINS ${this.formatValue(val)}`;
    if (op === "$regex") return `${path} NOT GLOB ${this.formatValue(val)}`;
    if (op === "$in") return `${path} NOT IN (${this.formatArray(val)})`;
    if (op === "$exists") return val ? `HAS NOT FIELD ${path}` : `HAS FIELD ${path}`;
    if (op === "$and" || op === "$or") {
      const newOp = op === "$and" ? "$or" : "$and";
      const conditions = val.map((condition) => {
        const [fieldName, fieldValue] = Object.entries(condition)[0] ?? [];
        return { [fieldName]: { $not: fieldValue } };
      });
      return this.translateOperator(newOp, conditions, "");
    }
    if (op === "$nor") {
      return this.translateOperator("$or", val, "");
    }
    return `${path} != ${this.formatValue(val)}`;
  }
  formatValue(value) {
    if (value === null || value === void 0) {
      throw new Error("Filtering for null/undefined values is not supported by Upstash Vector");
    }
    if (typeof value === "string") {
      const hasSingleQuote = /'/g.test(value);
      const hasDoubleQuote = /"/g.test(value);
      if (hasSingleQuote && hasDoubleQuote) {
        return `'${value.replace(/\\/g, "\\\\").replace(/'/g, "\\'")}'`;
      }
      if (hasSingleQuote) {
        return `"${value}"`;
      }
      return `'${value}'`;
    }
    if (typeof value === "number") {
      if (Math.abs(value) < 1e-6 || Math.abs(value) > 1e6) {
        return value.toFixed(20).replace(/\.?0+$/, "");
      }
      return value.toString();
    }
    return String(value);
  }
  formatArray(values) {
    return values.map((value) => {
      if (value === null || value === void 0) {
        throw new Error("Filtering for null/undefined values is not supported by Upstash Vector");
      }
      return this.formatValue(value);
    }).join(", ");
  }
  formatComparison(path, op, value) {
    return `${path} ${op} ${this.formatValue(value)}`;
  }
  joinConditions(conditions, operator) {
    const translated = Array.isArray(conditions) ? conditions.map((c) => this.translateNode(c)) : [this.translateNode(conditions)];
    return `(${translated.join(` ${operator} `)})`;
  }
};

// src/vector/index.ts
var UpstashVector = class extends vector.MastraVector {
  client;
  /**
   * Creates a new UpstashVector instance.
   * @param {object} params - The parameters for the UpstashVector.
   * @param {string} params.id - The unique identifier for this vector store instance.
   * @param {string} params.url - The URL of the Upstash vector index.
   * @param {string} params.token - The token for the Upstash vector index.
   */
  constructor({ url, token, id }) {
    super({ id });
    this.client = new vector$1.Index({
      url,
      token
    });
  }
  /**
   * Upserts vectors into the index.
   * @param {UpsertVectorParams} params - The parameters for the upsert operation.
   * @returns {Promise<string[]>} A promise that resolves to the IDs of the upserted vectors.
   */
  async upsert({
    indexName: namespace,
    vectors,
    metadata,
    ids,
    sparseVectors
  }) {
    const generatedIds = ids || vectors.map(() => crypto$1.randomUUID());
    const points = vectors.map((vector, index) => ({
      id: generatedIds[index],
      vector,
      ...sparseVectors?.[index] && { sparseVector: sparseVectors[index] },
      metadata: metadata?.[index]
    }));
    try {
      await this.client.upsert(points, {
        namespace
      });
      return generatedIds;
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createVectorErrorId("UPSTASH", "UPSERT", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: { namespace, vectorCount: vectors.length }
        },
        error$1
      );
    }
  }
  /**
   * Transforms a Mastra vector filter into an Upstash-compatible filter string.
   * @param {UpstashVectorFilter} [filter] - The filter to transform.
   * @returns {string | undefined} The transformed filter string, or undefined if no filter is provided.
   */
  transformFilter(filter) {
    const translator = new UpstashFilterTranslator();
    return translator.translate(filter);
  }
  /**
   * Creates a new index. For Upstash, this is a no-op as indexes (known as namespaces in Upstash) are created on-the-fly.
   * @param {CreateIndexParams} _params - The parameters for creating the index (ignored).
   * @returns {Promise<void>} A promise that resolves when the operation is complete.
   */
  async createIndex(_params) {
    this.logger.debug("No need to call createIndex for Upstash");
  }
  /**
   * Queries the vector index.
   * @param {QueryVectorParams} params - The parameters for the query operation. indexName is the namespace in Upstash.
   * @returns {Promise<QueryResult[]>} A promise that resolves to the query results.
   */
  async query({
    indexName: namespace,
    queryVector,
    topK = 10,
    filter,
    includeVector = false,
    sparseVector,
    fusionAlgorithm,
    queryMode
  }) {
    if (!queryVector) {
      throw new error.MastraError({
        id: storage.createVectorErrorId("UPSTASH", "QUERY", "MISSING_VECTOR"),
        text: "queryVector is required for Upstash queries. Metadata-only queries are not supported by this vector store.",
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        details: { indexName: namespace }
      });
    }
    try {
      const ns = this.client.namespace(namespace);
      const filterString = this.transformFilter(filter);
      const results = await ns.query({
        topK,
        vector: queryVector,
        ...sparseVector && { sparseVector },
        includeVectors: includeVector,
        includeMetadata: true,
        ...filterString ? { filter: filterString } : {},
        ...fusionAlgorithm && { fusionAlgorithm },
        ...queryMode && { queryMode }
      });
      return (results || []).map((result) => ({
        id: `${result.id}`,
        score: result.score,
        metadata: result.metadata,
        ...includeVector && { vector: result.vector || [] }
      }));
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createVectorErrorId("UPSTASH", "QUERY", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: { namespace, topK }
        },
        error$1
      );
    }
  }
  /**
   * Lists all namespaces in the Upstash vector index, which correspond to indexes.
   * @returns {Promise<string[]>} A promise that resolves to a list of index names.
   */
  async listIndexes() {
    try {
      const indexes = await this.client.listNamespaces();
      return indexes.filter(Boolean);
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createVectorErrorId("UPSTASH", "LIST_INDEXES", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY
        },
        error$1
      );
    }
  }
  /**
   * Retrieves statistics about a vector index.
   *
   * @param {string} indexName - The name of the namespace to describe
   * @returns A promise that resolves to the index statistics including dimension, count and metric
   */
  async describeIndex({ indexName: namespace }) {
    try {
      const info = await this.client.info();
      return {
        dimension: info.dimension,
        count: info.namespaces?.[namespace]?.vectorCount || 0,
        metric: info?.similarityFunction?.toLowerCase()
      };
    } catch (error$1) {
      throw new error.MastraError(
        {
          id: storage.createVectorErrorId("UPSTASH", "DESCRIBE_INDEX", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: { namespace }
        },
        error$1
      );
    }
  }
  /**
   * Deletes an index (namespace).
   * @param {DeleteIndexParams} params - The parameters for the delete operation.
   * @returns {Promise<void>} A promise that resolves when the deletion is complete.
   */
  async deleteIndex({ indexName: namespace }) {
    try {
      await this.client.deleteNamespace(namespace);
    } catch (error$1) {
      const errorMessage = error$1?.message || "";
      if (errorMessage.includes("does not exist") || errorMessage.includes("not found")) {
        this.logger.info(`Namespace ${namespace} does not exist, treating as already deleted`);
        return;
      }
      throw new error.MastraError(
        {
          id: storage.createVectorErrorId("UPSTASH", "DELETE_INDEX", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: { namespace }
        },
        error$1
      );
    }
  }
  /**
   * Updates a vector by its ID or multiple vectors matching a filter.
   * @param params - Parameters containing the id or filter for targeting the vector(s) to update
   * @param params.indexName - The name of the namespace containing the vector.
   * @param params.id - The ID of the vector to update (mutually exclusive with filter).
   * @param params.filter - Filter to match multiple vectors to update (mutually exclusive with id).
   * @param params.update - An object containing the vector and/or metadata to update.
   * @returns A promise that resolves when the update is complete.
   * @throws Will throw an error if no updates are provided or if the update operation fails.
   */
  async updateVector(params) {
    const { indexName: namespace, update } = params;
    const upstashUpdate = update;
    const sparseVector = upstashUpdate.sparseVector;
    if ("id" in params && params.id && "filter" in params && params.filter) {
      throw new error.MastraError({
        id: storage.createVectorErrorId("UPSTASH", "UPDATE_VECTOR", "MUTUALLY_EXCLUSIVE"),
        text: "Cannot specify both id and filter - they are mutually exclusive",
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        details: { namespace }
      });
    }
    if (!("id" in params && params.id) && !("filter" in params && params.filter)) {
      throw new error.MastraError({
        id: storage.createVectorErrorId("UPSTASH", "UPDATE_VECTOR", "NO_TARGET"),
        text: "Either id or filter must be provided",
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        details: { namespace }
      });
    }
    if (!update.vector && !update.metadata && !sparseVector) {
      throw new error.MastraError({
        id: storage.createVectorErrorId("UPSTASH", "UPDATE_VECTOR", "NO_PAYLOAD"),
        text: "No update data provided",
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        details: { namespace }
      });
    }
    if ("filter" in params && params.filter && Object.keys(params.filter).length === 0) {
      throw new error.MastraError({
        id: storage.createVectorErrorId("UPSTASH", "UPDATE_VECTOR", "EMPTY_FILTER"),
        text: "Filter cannot be an empty filter object",
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        details: { namespace }
      });
    }
    try {
      const ns = this.client.namespace(namespace);
      if ("id" in params && params.id) {
        const points = { id: params.id };
        if (!update.vector || !update.metadata) {
          try {
            const existing = await ns.fetch([params.id], {
              includeVectors: true,
              includeMetadata: true
            });
            if (existing && existing.length > 0 && existing[0]) {
              if (!update.vector && existing[0]?.vector) {
                points.vector = existing[0].vector;
              }
              if (!update.metadata && existing[0]?.metadata) {
                points.metadata = existing[0].metadata;
              }
            }
          } catch (fetchError) {
            this.logger.warn(`Failed to fetch existing vector ${params.id} for partial update: ${fetchError}`);
          }
        }
        if (update.vector) points.vector = update.vector;
        if (update.metadata) points.metadata = update.metadata;
        if (sparseVector) points.sparseVector = sparseVector;
        await ns.upsert(points);
      } else if ("filter" in params && params.filter) {
        const filterString = this.transformFilter(params.filter);
        if (filterString) {
          const stats = await this.describeIndex({ indexName: namespace });
          const dummyVector = new Array(stats.dimension).fill(1 / Math.sqrt(stats.dimension));
          const needsVectors = !update.vector;
          const results = await ns.query({
            vector: dummyVector,
            topK: 1e3,
            // Upstash's max query limit
            filter: filterString,
            includeVectors: needsVectors,
            includeMetadata: needsVectors
          });
          for (const result of results) {
            const points = { id: `${result.id}` };
            if (update.vector) {
              points.vector = update.vector;
            } else if (result.vector) {
              points.vector = result.vector;
            }
            if (update.metadata) {
              points.metadata = update.metadata;
            } else if (result.metadata) {
              points.metadata = result.metadata;
            }
            if (sparseVector) points.sparseVector = sparseVector;
            await ns.upsert(points);
          }
        }
      }
    } catch (error$1) {
      if (error$1 instanceof error.MastraError) throw error$1;
      throw new error.MastraError(
        {
          id: storage.createVectorErrorId("UPSTASH", "UPDATE_VECTOR", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            namespace,
            ..."id" in params && params.id && { id: params.id },
            ..."filter" in params && params.filter && { filter: JSON.stringify(params.filter) }
          }
        },
        error$1
      );
    }
  }
  /**
   * Deletes a vector by its ID.
   * @param indexName - The name of the namespace containing the vector.
   * @param id - The ID of the vector to delete.
   * @returns A promise that resolves when the deletion is complete.
   * @throws Will throw an error if the deletion operation fails.
   */
  async deleteVector({ indexName: namespace, id }) {
    try {
      const ns = this.client.namespace(namespace);
      await ns.delete(id);
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: storage.createVectorErrorId("UPSTASH", "DELETE_VECTOR", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            namespace,
            ...id && { id }
          }
        },
        error$1
      );
      this.logger?.error(mastraError.toString());
    }
  }
  /**
   * Deletes multiple vectors by IDs or filter.
   * @param indexName - The name of the namespace containing the vectors.
   * @param ids - Array of vector IDs to delete (mutually exclusive with filter).
   * @param filter - Filter to match vectors to delete (mutually exclusive with ids).
   * @returns A promise that resolves when the deletion is complete.
   * @throws Will throw an error if both ids and filter are provided, or if neither is provided.
   */
  async deleteVectors({ indexName: namespace, filter, ids }) {
    if (ids && filter) {
      throw new error.MastraError({
        id: storage.createVectorErrorId("UPSTASH", "DELETE_VECTORS", "MUTUALLY_EXCLUSIVE"),
        text: "Cannot specify both ids and filter - they are mutually exclusive",
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        details: { namespace }
      });
    }
    if (!ids && !filter) {
      throw new error.MastraError({
        id: storage.createVectorErrorId("UPSTASH", "DELETE_VECTORS", "NO_TARGET"),
        text: "Either filter or ids must be provided",
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        details: { namespace }
      });
    }
    if (ids && ids.length === 0) {
      throw new error.MastraError({
        id: storage.createVectorErrorId("UPSTASH", "DELETE_VECTORS", "EMPTY_IDS"),
        text: "Cannot delete with empty ids array",
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        details: { namespace }
      });
    }
    if (filter && Object.keys(filter).length === 0) {
      throw new error.MastraError({
        id: storage.createVectorErrorId("UPSTASH", "DELETE_VECTORS", "EMPTY_FILTER"),
        text: "Cannot delete with empty filter object",
        domain: error.ErrorDomain.STORAGE,
        category: error.ErrorCategory.USER,
        details: { namespace }
      });
    }
    try {
      const ns = this.client.namespace(namespace);
      if (ids) {
        await ns.delete(ids);
      } else if (filter) {
        const filterString = this.transformFilter(filter);
        if (filterString) {
          const stats = await this.describeIndex({ indexName: namespace });
          const dummyVector = new Array(stats.dimension).fill(1 / Math.sqrt(stats.dimension));
          const results = await ns.query({
            vector: dummyVector,
            topK: 1e3,
            // Upstash's max query limit
            filter: filterString,
            includeVectors: false,
            includeMetadata: false
          });
          const idsToDelete = results.map((r) => `${r.id}`);
          if (idsToDelete.length > 0) {
            await ns.delete(idsToDelete);
          }
        }
      }
    } catch (error$1) {
      if (error$1 instanceof error.MastraError) throw error$1;
      throw new error.MastraError(
        {
          id: storage.createVectorErrorId("UPSTASH", "DELETE_VECTORS", "FAILED"),
          domain: error.ErrorDomain.STORAGE,
          category: error.ErrorCategory.THIRD_PARTY,
          details: {
            namespace,
            ...filter && { filter: JSON.stringify(filter) },
            ...ids && { idsCount: ids.length }
          }
        },
        error$1
      );
    }
  }
};

// src/vector/prompt.ts
var UPSTASH_PROMPT = `When querying Upstash Vector, you can ONLY use the operators listed below. Any other operators will be rejected.
Important: Don't explain how to construct the filter - use the specified operators and fields to search the content and return relevant results.
If a user tries to give an explicit operator that is not supported, reject the filter entirely and let them know that the operator is not supported.

Basic Comparison Operators:
- $eq: Exact match (default when using field: value)
  Example: { "category": "electronics" }
- $ne: Not equal
  Example: { "category": { "$ne": "electronics" } }
- $gt: Greater than
  Example: { "price": { "$gt": 100 } }
- $gte: Greater than or equal
  Example: { "price": { "$gte": 100 } }
- $lt: Less than
  Example: { "price": { "$lt": 100 } }
- $lte: Less than or equal
  Example: { "price": { "$lte": 100 } }

Array Operators:
- $in: Match any value in array
  Example: { "category": { "$in": ["electronics", "books"] } }
- $nin: Does not match any value in array
  Example: { "category": { "$nin": ["electronics", "books"] } }

Logical Operators:
- $and: Logical AND (can be implicit or explicit)
  Implicit Example: { "price": { "$gt": 100 }, "category": "electronics" }
  Explicit Example: { "$and": [{ "price": { "$gt": 100 } }, { "category": "electronics" }] }
- $or: Logical OR
  Example: { "$or": [{ "price": { "$lt": 50 } }, { "category": "books" }] }

Element Operators:
- $exists: Check if field exists
  Example: { "rating": { "$exists": true } }

Restrictions:
- Regex patterns are not supported
- Only $and and $or logical operators are supported at the top level
- Empty arrays in $in/$nin will return no results
- Nested fields are supported using dot notation
- Multiple conditions on the same field are supported with both implicit and explicit $and
- At least one key-value pair is required in filter object
- Empty objects and undefined values are treated as no filter
- Invalid types in comparison operators will throw errors
- All non-logical operators must be used within a field condition
  Valid: { "field": { "$gt": 100 } }
  Valid: { "$and": [...] }
  Invalid: { "$gt": 100 }
- Logical operators must contain field conditions, not direct operators
  Valid: { "$and": [{ "field": { "$gt": 100 } }] }
  Invalid: { "$and": [{ "$gt": 100 }] }
- Logical operators ($and, $or):
  - Can only be used at top level or nested within other logical operators
  - Can not be used on a field level, or be nested inside a field
  - Can not be used inside an operator
  - Valid: { "$and": [{ "field": { "$gt": 100 } }] }
  - Valid: { "$or": [{ "$and": [{ "field": { "$gt": 100 } }] }] }
  - Invalid: { "field": { "$and": [{ "$gt": 100 }] } }
  - Invalid: { "field": { "$or": [{ "$gt": 100 }] } }
  - Invalid: { "field": { "$gt": { "$and": [{...}] } } }

Example Complex Query:
{
  "$and": [
    { "category": { "$in": ["electronics", "computers"] } },
    { "price": { "$gte": 100, "$lte": 1000 } },
    { "rating": { "$exists": true, "$gt": 4 } },
    { "$or": [
      { "stock": { "$gt": 0 } },
      { "preorder": true }
    ]}
  ]
}`;

Object.defineProperty(exports, "RedisServerCache", {
  enumerable: true,
  get: function () { return redis.RedisServerCache; }
});
Object.defineProperty(exports, "upstashPreset", {
  enumerable: true,
  get: function () { return redis.upstashPreset; }
});
exports.BackgroundTasksUpstash = BackgroundTasksUpstash;
exports.ScoresUpstash = ScoresUpstash;
exports.StoreMemoryUpstash = StoreMemoryUpstash;
exports.UPSTASH_PROMPT = UPSTASH_PROMPT;
exports.UpstashServerCache = UpstashServerCache;
exports.UpstashStore = UpstashStore;
exports.UpstashVector = UpstashVector;
exports.WorkflowsUpstash = WorkflowsUpstash;
