// @mastra/core/dist/telemetry/index.cjs
'use strict';

var chunkOEW7GS6U_cjs = require('../chunk-OEW7GS6U.cjs');
var fs = require('fs');
var os = require('os');
var path = require('path');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var os__default = /*#__PURE__*/_interopDefault(os);
var path__default = /*#__PURE__*/_interopDefault(path);

var INPUT_TOKENS_METRIC = "mastra_model_total_input_tokens";
var OUTPUT_TOKENS_METRIC = "mastra_model_total_output_tokens";
var BREAKDOWN_LIMIT = 200;
var USAGE_TELEMETRY_EVENT = "mastra_model_token_usage";
function getDefaultCursorPath() {
  return path__default.default.join(os__default.default.homedir(), ".mastra", "usage-telemetry.json");
}
function readCursors(cursorPath) {
  try {
    const parsed = JSON.parse(fs.readFileSync(cursorPath, "utf-8"));
    if (parsed && typeof parsed === "object" && parsed.projects && typeof parsed.projects === "object") {
      return { projects: parsed.projects };
    }
  } catch {
  }
  return { projects: {} };
}
function readCursor(cursorPath, projectId) {
  const value = readCursors(cursorPath).projects[projectId];
  if (!value) {
    return void 0;
  }
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? void 0 : date;
}
function writeCursor(cursorPath, projectId, syncedAt) {
  const cursors = readCursors(cursorPath);
  cursors.projects[projectId] = syncedAt.toISOString();
  fs.mkdirSync(path__default.default.dirname(cursorPath), { recursive: true });
  fs.writeFileSync(cursorPath, JSON.stringify(cursors));
}
function applyBreakdown(rows, response, field) {
  for (const group of response.groups) {
    const provider = group.dimensions.provider ?? null;
    const model = group.dimensions.model ?? null;
    const key = `${provider}\0${model}`;
    let row = rows.get(key);
    if (!row) {
      row = { provider, model, inputTokens: 0, outputTokens: 0, totalInputTokens: 0, totalOutputTokens: 0 };
      rows.set(key, row);
    }
    row[field] += group.value ?? 0;
  }
}
async function syncUsageTelemetry(mastra, options = {}) {
  try {
    if (!chunkOEW7GS6U_cjs.isEETelemetryEnabled()) {
      return;
    }
    const observability = mastra.getStorage()?.stores?.observability;
    if (!observability || typeof observability.getMetricBreakdown !== "function") {
      return;
    }
    const projectRoot = process.env.MASTRA_PROJECT_ROOT || process.cwd();
    const projectId = chunkOEW7GS6U_cjs.hashTelemetryValue(projectRoot).slice(0, 16);
    const cursorPath = options.cursorPath ?? getDefaultCursorPath();
    const lastSyncedAt = readCursor(cursorPath, projectId);
    const now = options.now ?? /* @__PURE__ */ new Date();
    if (lastSyncedAt && lastSyncedAt.getTime() >= now.getTime()) {
      return;
    }
    const baseArgs = {
      groupBy: ["provider", "model"],
      aggregation: "sum",
      limit: BREAKDOWN_LIMIT
    };
    const deltaFilters = {
      timestamp: { ...lastSyncedAt ? { start: lastSyncedAt, startExclusive: true } : {}, end: now }
    };
    const totalFilters = { timestamp: { end: now } };
    const [inputDelta, outputDelta, inputTotal, outputTotal] = await Promise.all([
      observability.getMetricBreakdown({ ...baseArgs, name: [INPUT_TOKENS_METRIC], filters: deltaFilters }),
      observability.getMetricBreakdown({ ...baseArgs, name: [OUTPUT_TOKENS_METRIC], filters: deltaFilters }),
      observability.getMetricBreakdown({ ...baseArgs, name: [INPUT_TOKENS_METRIC], filters: totalFilters }),
      observability.getMetricBreakdown({ ...baseArgs, name: [OUTPUT_TOKENS_METRIC], filters: totalFilters })
    ]);
    const rows = /* @__PURE__ */ new Map();
    applyBreakdown(rows, inputDelta, "inputTokens");
    applyBreakdown(rows, outputDelta, "outputTokens");
    applyBreakdown(rows, inputTotal, "totalInputTokens");
    applyBreakdown(rows, outputTotal, "totalOutputTokens");
    const distinctId = process.env.MASTRA_CLI_DISTINCT_ID || void 0;
    const command = process.env.MASTRA_TELEMETRY_COMMAND || "server";
    const nodeEnv = process.env.NODE_ENV || "development";
    const isFirstSync = !lastSyncedAt;
    for (const row of rows.values()) {
      if (row.inputTokens <= 0 && row.outputTokens <= 0) {
        continue;
      }
      chunkOEW7GS6U_cjs.captureTelemetryEvent(USAGE_TELEMETRY_EVENT, distinctId, {
        provider: row.provider,
        model: row.model,
        input_tokens: row.inputTokens,
        output_tokens: row.outputTokens,
        total_input_tokens: row.totalInputTokens,
        total_output_tokens: row.totalOutputTokens,
        command,
        node_env: nodeEnv,
        project_id: projectId,
        is_first_sync: isFirstSync,
        window_start: lastSyncedAt?.toISOString() ?? null,
        window_end: now.toISOString()
      });
    }
    writeCursor(cursorPath, projectId, now);
  } catch {
  }
}

exports.USAGE_TELEMETRY_EVENT = USAGE_TELEMETRY_EVENT;
exports.syncUsageTelemetry = syncUsageTelemetry;
