// @mastra/server/dist/server/handlers/a2a.cjs
'use strict';

var chunkQXWZ2DKR_cjs = require('../../chunk-QXWZ2DKR.cjs');



Object.defineProperty(exports, "AGENT_EXECUTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.AGENT_EXECUTION_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_CARD_ROUTE", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.GET_AGENT_CARD_ROUTE; }
});
Object.defineProperty(exports, "getAgentCardByIdHandler", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.getAgentCardByIdHandler; }
});
Object.defineProperty(exports, "getAgentExecutionHandler", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.getAgentExecutionHandler; }
});
Object.defineProperty(exports, "handleDeleteTaskPushNotificationConfig", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.handleDeleteTaskPushNotificationConfig; }
});
Object.defineProperty(exports, "handleGetTaskPushNotificationConfig", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.handleGetTaskPushNotificationConfig; }
});
Object.defineProperty(exports, "handleListTaskPushNotificationConfig", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.handleListTaskPushNotificationConfig; }
});
Object.defineProperty(exports, "handleMessageSend", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.handleMessageSend; }
});
Object.defineProperty(exports, "handleMessageStream", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.handleMessageStream; }
});
Object.defineProperty(exports, "handleSetTaskPushNotificationConfig", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.handleSetTaskPushNotificationConfig; }
});
Object.defineProperty(exports, "handleTaskCancel", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.handleTaskCancel; }
});
Object.defineProperty(exports, "handleTaskGet", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.handleTaskGet; }
});
Object.defineProperty(exports, "handleTaskResubscribe", {
  enumerable: true,
  get: function () { return chunkQXWZ2DKR_cjs.handleTaskResubscribe; }
});
