// @mastra/server/dist/server/handlers/agents.cjs
'use strict';

var chunkHVJTEOVD_cjs = require('../../chunk-HVJTEOVD.cjs');



Object.defineProperty(exports, "ABORT_AGENT_THREAD_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.ABORT_AGENT_THREAD_ROUTE; }
});
Object.defineProperty(exports, "APPROVE_NETWORK_TOOL_CALL_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.APPROVE_NETWORK_TOOL_CALL_ROUTE; }
});
Object.defineProperty(exports, "APPROVE_TOOL_CALL_GENERATE_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.APPROVE_TOOL_CALL_GENERATE_ROUTE; }
});
Object.defineProperty(exports, "APPROVE_TOOL_CALL_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.APPROVE_TOOL_CALL_ROUTE; }
});
Object.defineProperty(exports, "CLONE_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.CLONE_AGENT_ROUTE; }
});
Object.defineProperty(exports, "DECLINE_NETWORK_TOOL_CALL_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.DECLINE_NETWORK_TOOL_CALL_ROUTE; }
});
Object.defineProperty(exports, "DECLINE_TOOL_CALL_GENERATE_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.DECLINE_TOOL_CALL_GENERATE_ROUTE; }
});
Object.defineProperty(exports, "DECLINE_TOOL_CALL_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.DECLINE_TOOL_CALL_ROUTE; }
});
Object.defineProperty(exports, "ENHANCE_INSTRUCTIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.ENHANCE_INSTRUCTIONS_ROUTE; }
});
Object.defineProperty(exports, "GENERATE_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.GENERATE_AGENT_ROUTE; }
});
Object.defineProperty(exports, "GENERATE_AGENT_VNEXT_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.GENERATE_AGENT_VNEXT_ROUTE; }
});
Object.defineProperty(exports, "GENERATE_LEGACY_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.GENERATE_LEGACY_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_BY_ID_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.GET_AGENT_BY_ID_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_SKILL_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.GET_AGENT_SKILL_ROUTE; }
});
Object.defineProperty(exports, "GET_PROVIDERS_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.GET_PROVIDERS_ROUTE; }
});
Object.defineProperty(exports, "LIST_AGENTS_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.LIST_AGENTS_ROUTE; }
});
Object.defineProperty(exports, "OBSERVE_AGENT_STREAM_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.OBSERVE_AGENT_STREAM_ROUTE; }
});
Object.defineProperty(exports, "QUEUE_AGENT_MESSAGE_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.QUEUE_AGENT_MESSAGE_ROUTE; }
});
Object.defineProperty(exports, "REORDER_AGENT_MODEL_LIST_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.REORDER_AGENT_MODEL_LIST_ROUTE; }
});
Object.defineProperty(exports, "RESET_AGENT_MODEL_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.RESET_AGENT_MODEL_ROUTE; }
});
Object.defineProperty(exports, "RESUME_STREAM_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.RESUME_STREAM_ROUTE; }
});
Object.defineProperty(exports, "RESUME_STREAM_UNTIL_IDLE_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.RESUME_STREAM_UNTIL_IDLE_ROUTE; }
});
Object.defineProperty(exports, "SEND_AGENT_MESSAGE_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.SEND_AGENT_MESSAGE_ROUTE; }
});
Object.defineProperty(exports, "SEND_AGENT_SIGNAL_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.SEND_AGENT_SIGNAL_ROUTE; }
});
Object.defineProperty(exports, "SEND_TOOL_APPROVAL_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.SEND_TOOL_APPROVAL_ROUTE; }
});
Object.defineProperty(exports, "STREAM_GENERATE_LEGACY_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.STREAM_GENERATE_LEGACY_ROUTE; }
});
Object.defineProperty(exports, "STREAM_GENERATE_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.STREAM_GENERATE_ROUTE; }
});
Object.defineProperty(exports, "STREAM_GENERATE_VNEXT_DEPRECATED_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.STREAM_GENERATE_VNEXT_DEPRECATED_ROUTE; }
});
Object.defineProperty(exports, "STREAM_NETWORK_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.STREAM_NETWORK_ROUTE; }
});
Object.defineProperty(exports, "STREAM_UI_MESSAGE_DEPRECATED_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.STREAM_UI_MESSAGE_DEPRECATED_ROUTE; }
});
Object.defineProperty(exports, "STREAM_UI_MESSAGE_VNEXT_DEPRECATED_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.STREAM_UI_MESSAGE_VNEXT_DEPRECATED_ROUTE; }
});
Object.defineProperty(exports, "STREAM_UNTIL_IDLE_GENERATE_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.STREAM_UNTIL_IDLE_GENERATE_ROUTE; }
});
Object.defineProperty(exports, "STREAM_VNEXT_DEPRECATED_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.STREAM_VNEXT_DEPRECATED_ROUTE; }
});
Object.defineProperty(exports, "SUBSCRIBE_AGENT_THREAD_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.SUBSCRIBE_AGENT_THREAD_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_AGENT_MODEL_IN_MODEL_LIST_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.UPDATE_AGENT_MODEL_IN_MODEL_LIST_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_AGENT_MODEL_ROUTE", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.UPDATE_AGENT_MODEL_ROUTE; }
});
Object.defineProperty(exports, "buildProvidersList", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.buildProvidersList; }
});
Object.defineProperty(exports, "extractVersionOptions", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.extractVersionOptions; }
});
Object.defineProperty(exports, "getAgentFromSystem", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.getAgentFromSystem; }
});
Object.defineProperty(exports, "getBrowserToolsFromAgent", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.getBrowserToolsFromAgent; }
});
Object.defineProperty(exports, "getSerializedAgentTools", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.getSerializedAgentTools; }
});
Object.defineProperty(exports, "getSerializedProcessors", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.getSerializedProcessors; }
});
Object.defineProperty(exports, "getSerializedSkillsFromAgent", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.getSerializedSkillsFromAgent; }
});
Object.defineProperty(exports, "getWorkspaceToolsFromAgent", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.getWorkspaceToolsFromAgent; }
});
Object.defineProperty(exports, "isProviderConnected", {
  enumerable: true,
  get: function () { return chunkHVJTEOVD_cjs.isProviderConnected; }
});
