// @mastra/server/dist/server/handlers/agent-builder.cjs
'use strict';

var chunkN4H2ICZJ_cjs = require('../../chunk-N4H2ICZJ.cjs');



Object.defineProperty(exports, "CANCEL_AGENT_BUILDER_ACTION_RUN_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.CANCEL_AGENT_BUILDER_ACTION_RUN_ROUTE; }
});
Object.defineProperty(exports, "CREATE_AGENT_BUILDER_ACTION_RUN_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.CREATE_AGENT_BUILDER_ACTION_RUN_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_BUILDER_ACTION_BY_ID_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.GET_AGENT_BUILDER_ACTION_BY_ID_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_BUILDER_ACTION_RUN_BY_ID_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.GET_AGENT_BUILDER_ACTION_RUN_BY_ID_ROUTE; }
});
Object.defineProperty(exports, "LIST_AGENT_BUILDER_ACTIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.LIST_AGENT_BUILDER_ACTIONS_ROUTE; }
});
Object.defineProperty(exports, "LIST_AGENT_BUILDER_ACTION_RUNS_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.LIST_AGENT_BUILDER_ACTION_RUNS_ROUTE; }
});
Object.defineProperty(exports, "OBSERVE_STREAM_AGENT_BUILDER_ACTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.OBSERVE_STREAM_AGENT_BUILDER_ACTION_ROUTE; }
});
Object.defineProperty(exports, "OBSERVE_STREAM_LEGACY_AGENT_BUILDER_ACTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.OBSERVE_STREAM_LEGACY_AGENT_BUILDER_ACTION_ROUTE; }
});
Object.defineProperty(exports, "RESUME_AGENT_BUILDER_ACTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.RESUME_AGENT_BUILDER_ACTION_ROUTE; }
});
Object.defineProperty(exports, "RESUME_ASYNC_AGENT_BUILDER_ACTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.RESUME_ASYNC_AGENT_BUILDER_ACTION_ROUTE; }
});
Object.defineProperty(exports, "RESUME_NO_WAIT_AGENT_BUILDER_ACTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.RESUME_NO_WAIT_AGENT_BUILDER_ACTION_ROUTE; }
});
Object.defineProperty(exports, "RESUME_STREAM_AGENT_BUILDER_ACTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.RESUME_STREAM_AGENT_BUILDER_ACTION_ROUTE; }
});
Object.defineProperty(exports, "START_AGENT_BUILDER_ACTION_RUN_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.START_AGENT_BUILDER_ACTION_RUN_ROUTE; }
});
Object.defineProperty(exports, "START_ASYNC_AGENT_BUILDER_ACTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.START_ASYNC_AGENT_BUILDER_ACTION_ROUTE; }
});
Object.defineProperty(exports, "STREAM_AGENT_BUILDER_ACTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.STREAM_AGENT_BUILDER_ACTION_ROUTE; }
});
Object.defineProperty(exports, "STREAM_LEGACY_AGENT_BUILDER_ACTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkN4H2ICZJ_cjs.STREAM_LEGACY_AGENT_BUILDER_ACTION_ROUTE; }
});
