// @mastra/server/dist/server/handlers/auth.cjs
'use strict';

var chunkBLR7PQHQ_cjs = require('../../chunk-BLR7PQHQ.cjs');



Object.defineProperty(exports, "AUTH_ROUTES", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.AUTH_ROUTES; }
});
Object.defineProperty(exports, "GET_AUTH_CAPABILITIES_ROUTE", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.GET_AUTH_CAPABILITIES_ROUTE; }
});
Object.defineProperty(exports, "GET_CURRENT_USER_ROUTE", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.GET_CURRENT_USER_ROUTE; }
});
Object.defineProperty(exports, "GET_PERMISSION_PATTERNS_ROUTE", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.GET_PERMISSION_PATTERNS_ROUTE; }
});
Object.defineProperty(exports, "GET_ROLE_PERMISSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.GET_ROLE_PERMISSIONS_ROUTE; }
});
Object.defineProperty(exports, "GET_SSO_CALLBACK_ROUTE", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.GET_SSO_CALLBACK_ROUTE; }
});
Object.defineProperty(exports, "GET_SSO_LOGIN_ROUTE", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.GET_SSO_LOGIN_ROUTE; }
});
Object.defineProperty(exports, "POST_CREDENTIALS_SIGN_IN_ROUTE", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.POST_CREDENTIALS_SIGN_IN_ROUTE; }
});
Object.defineProperty(exports, "POST_CREDENTIALS_SIGN_UP_ROUTE", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.POST_CREDENTIALS_SIGN_UP_ROUTE; }
});
Object.defineProperty(exports, "POST_LOGOUT_ROUTE", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.POST_LOGOUT_ROUTE; }
});
Object.defineProperty(exports, "POST_REFRESH_ROUTE", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.POST_REFRESH_ROUTE; }
});
Object.defineProperty(exports, "getPublicOrigin", {
  enumerable: true,
  get: function () { return chunkBLR7PQHQ_cjs.getPublicOrigin; }
});
