// @mastra/server/dist/server/handlers/datasets.cjs
'use strict';

var chunkRBLDPJSS_cjs = require('../../chunk-RBLDPJSS.cjs');



Object.defineProperty(exports, "ADD_ITEM_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.ADD_ITEM_ROUTE; }
});
Object.defineProperty(exports, "BATCH_DELETE_ITEMS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.BATCH_DELETE_ITEMS_ROUTE; }
});
Object.defineProperty(exports, "BATCH_INSERT_ITEMS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.BATCH_INSERT_ITEMS_ROUTE; }
});
Object.defineProperty(exports, "CLUSTER_FAILURES_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.CLUSTER_FAILURES_ROUTE; }
});
Object.defineProperty(exports, "COMPARE_EXPERIMENTS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.COMPARE_EXPERIMENTS_ROUTE; }
});
Object.defineProperty(exports, "CREATE_DATASET_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.CREATE_DATASET_ROUTE; }
});
Object.defineProperty(exports, "DELETE_DATASET_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.DELETE_DATASET_ROUTE; }
});
Object.defineProperty(exports, "DELETE_ITEM_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.DELETE_ITEM_ROUTE; }
});
Object.defineProperty(exports, "EXPERIMENT_REVIEW_SUMMARY_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.EXPERIMENT_REVIEW_SUMMARY_ROUTE; }
});
Object.defineProperty(exports, "GENERATE_ITEMS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.GENERATE_ITEMS_ROUTE; }
});
Object.defineProperty(exports, "GET_DATASET_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.GET_DATASET_ROUTE; }
});
Object.defineProperty(exports, "GET_EXPERIMENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.GET_EXPERIMENT_ROUTE; }
});
Object.defineProperty(exports, "GET_ITEM_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.GET_ITEM_ROUTE; }
});
Object.defineProperty(exports, "GET_ITEM_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.GET_ITEM_VERSION_ROUTE; }
});
Object.defineProperty(exports, "LIST_ALL_EXPERIMENTS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.LIST_ALL_EXPERIMENTS_ROUTE; }
});
Object.defineProperty(exports, "LIST_DATASETS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.LIST_DATASETS_ROUTE; }
});
Object.defineProperty(exports, "LIST_DATASET_VERSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.LIST_DATASET_VERSIONS_ROUTE; }
});
Object.defineProperty(exports, "LIST_EXPERIMENTS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.LIST_EXPERIMENTS_ROUTE; }
});
Object.defineProperty(exports, "LIST_EXPERIMENT_RESULTS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.LIST_EXPERIMENT_RESULTS_ROUTE; }
});
Object.defineProperty(exports, "LIST_ITEMS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.LIST_ITEMS_ROUTE; }
});
Object.defineProperty(exports, "LIST_ITEM_VERSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.LIST_ITEM_VERSIONS_ROUTE; }
});
Object.defineProperty(exports, "TRIGGER_EXPERIMENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.TRIGGER_EXPERIMENT_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_DATASET_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.UPDATE_DATASET_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_EXPERIMENT_RESULT_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.UPDATE_EXPERIMENT_RESULT_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_ITEM_ROUTE", {
  enumerable: true,
  get: function () { return chunkRBLDPJSS_cjs.UPDATE_ITEM_ROUTE; }
});
