// @mastra/server/dist/server/handlers/gateway-memory-client.cjs
'use strict';

var chunkLFZO2LHV_cjs = require('../../chunk-LFZO2LHV.cjs');



Object.defineProperty(exports, "GatewayMemoryClient", {
  enumerable: true,
  get: function () { return chunkLFZO2LHV_cjs.GatewayMemoryClient; }
});
Object.defineProperty(exports, "getGatewayClient", {
  enumerable: true,
  get: function () { return chunkLFZO2LHV_cjs.getGatewayClient; }
});
Object.defineProperty(exports, "isGatewayAgent", {
  enumerable: true,
  get: function () { return chunkLFZO2LHV_cjs.isGatewayAgent; }
});
Object.defineProperty(exports, "isGatewayAgentAsync", {
  enumerable: true,
  get: function () { return chunkLFZO2LHV_cjs.isGatewayAgentAsync; }
});
Object.defineProperty(exports, "toLocalMessage", {
  enumerable: true,
  get: function () { return chunkLFZO2LHV_cjs.toLocalMessage; }
});
Object.defineProperty(exports, "toLocalOMRecord", {
  enumerable: true,
  get: function () { return chunkLFZO2LHV_cjs.toLocalOMRecord; }
});
Object.defineProperty(exports, "toLocalThread", {
  enumerable: true,
  get: function () { return chunkLFZO2LHV_cjs.toLocalThread; }
});
