// @mastra/server/dist/server/handlers/mcp-client-versions.cjs
'use strict';

var chunkNV2MI4K7_cjs = require('../../chunk-NV2MI4K7.cjs');



Object.defineProperty(exports, "ACTIVATE_MCP_CLIENT_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkNV2MI4K7_cjs.ACTIVATE_MCP_CLIENT_VERSION_ROUTE; }
});
Object.defineProperty(exports, "COMPARE_MCP_CLIENT_VERSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkNV2MI4K7_cjs.COMPARE_MCP_CLIENT_VERSIONS_ROUTE; }
});
Object.defineProperty(exports, "CREATE_MCP_CLIENT_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkNV2MI4K7_cjs.CREATE_MCP_CLIENT_VERSION_ROUTE; }
});
Object.defineProperty(exports, "DELETE_MCP_CLIENT_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkNV2MI4K7_cjs.DELETE_MCP_CLIENT_VERSION_ROUTE; }
});
Object.defineProperty(exports, "GET_MCP_CLIENT_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkNV2MI4K7_cjs.GET_MCP_CLIENT_VERSION_ROUTE; }
});
Object.defineProperty(exports, "LIST_MCP_CLIENT_VERSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkNV2MI4K7_cjs.LIST_MCP_CLIENT_VERSIONS_ROUTE; }
});
Object.defineProperty(exports, "RESTORE_MCP_CLIENT_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkNV2MI4K7_cjs.RESTORE_MCP_CLIENT_VERSION_ROUTE; }
});
