// @mastra/server/dist/server/handlers/observability-new-endpoints.cjs
'use strict';

var chunkNLIGVLFF_cjs = require('../../chunk-NLIGVLFF.cjs');



Object.defineProperty(exports, "CREATE_FEEDBACK", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.CREATE_FEEDBACK; }
});
Object.defineProperty(exports, "CREATE_SCORE", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.CREATE_SCORE; }
});
Object.defineProperty(exports, "GET_ENTITY_NAMES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_ENTITY_NAMES; }
});
Object.defineProperty(exports, "GET_ENTITY_TYPES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_ENTITY_TYPES; }
});
Object.defineProperty(exports, "GET_ENVIRONMENTS", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_ENVIRONMENTS; }
});
Object.defineProperty(exports, "GET_FEEDBACK_AGGREGATE", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_FEEDBACK_AGGREGATE; }
});
Object.defineProperty(exports, "GET_FEEDBACK_BREAKDOWN", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_FEEDBACK_BREAKDOWN; }
});
Object.defineProperty(exports, "GET_FEEDBACK_PERCENTILES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_FEEDBACK_PERCENTILES; }
});
Object.defineProperty(exports, "GET_FEEDBACK_TIME_SERIES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_FEEDBACK_TIME_SERIES; }
});
Object.defineProperty(exports, "GET_METRIC_AGGREGATE", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_METRIC_AGGREGATE; }
});
Object.defineProperty(exports, "GET_METRIC_BREAKDOWN", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_METRIC_BREAKDOWN; }
});
Object.defineProperty(exports, "GET_METRIC_LABEL_KEYS", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_METRIC_LABEL_KEYS; }
});
Object.defineProperty(exports, "GET_METRIC_LABEL_VALUES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_METRIC_LABEL_VALUES; }
});
Object.defineProperty(exports, "GET_METRIC_NAMES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_METRIC_NAMES; }
});
Object.defineProperty(exports, "GET_METRIC_PERCENTILES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_METRIC_PERCENTILES; }
});
Object.defineProperty(exports, "GET_METRIC_TIME_SERIES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_METRIC_TIME_SERIES; }
});
Object.defineProperty(exports, "GET_SCORE", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_SCORE; }
});
Object.defineProperty(exports, "GET_SCORE_AGGREGATE", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_SCORE_AGGREGATE; }
});
Object.defineProperty(exports, "GET_SCORE_BREAKDOWN", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_SCORE_BREAKDOWN; }
});
Object.defineProperty(exports, "GET_SCORE_PERCENTILES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_SCORE_PERCENTILES; }
});
Object.defineProperty(exports, "GET_SCORE_TIME_SERIES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_SCORE_TIME_SERIES; }
});
Object.defineProperty(exports, "GET_SERVICE_NAMES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_SERVICE_NAMES; }
});
Object.defineProperty(exports, "GET_TAGS", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.GET_TAGS; }
});
Object.defineProperty(exports, "LIST_FEEDBACK", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.LIST_FEEDBACK; }
});
Object.defineProperty(exports, "LIST_LOGS", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.LIST_LOGS; }
});
Object.defineProperty(exports, "LIST_METRICS", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.LIST_METRICS; }
});
Object.defineProperty(exports, "LIST_SCORES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.LIST_SCORES; }
});
Object.defineProperty(exports, "NEW_ROUTES", {
  enumerable: true,
  get: function () { return chunkNLIGVLFF_cjs.NEW_ROUTES; }
});
