// @mastra/server/dist/server/handlers/observability-shared.cjs
'use strict';

var chunk6UHGOBM7_cjs = require('../../chunk-6UHGOBM7.cjs');



Object.defineProperty(exports, "NEW_ROUTE_DEFS", {
  enumerable: true,
  get: function () { return chunk6UHGOBM7_cjs.NEW_ROUTE_DEFS; }
});
Object.defineProperty(exports, "OBSERVABILITY_DELTA_POLLING_FEATURE", {
  enumerable: true,
  get: function () { return chunk6UHGOBM7_cjs.OBSERVABILITY_DELTA_POLLING_FEATURE; }
});
Object.defineProperty(exports, "OBSERVABILITY_DELTA_POLLING_UPGRADE_MESSAGE", {
  enumerable: true,
  get: function () { return chunk6UHGOBM7_cjs.OBSERVABILITY_DELTA_POLLING_UPGRADE_MESSAGE; }
});
Object.defineProperty(exports, "OBSERVABILITY_LIST_ENDPOINTS", {
  enumerable: true,
  get: function () { return chunk6UHGOBM7_cjs.OBSERVABILITY_LIST_ENDPOINTS; }
});
Object.defineProperty(exports, "assertObservabilityDeltaSupported", {
  enumerable: true,
  get: function () { return chunk6UHGOBM7_cjs.assertObservabilityDeltaSupported; }
});
Object.defineProperty(exports, "createObservabilityListQuerySchema", {
  enumerable: true,
  get: function () { return chunk6UHGOBM7_cjs.createObservabilityListQuerySchema; }
});
Object.defineProperty(exports, "getObservabilityStore", {
  enumerable: true,
  get: function () { return chunk6UHGOBM7_cjs.getObservabilityStore; }
});
Object.defineProperty(exports, "getStorage", {
  enumerable: true,
  get: function () { return chunk6UHGOBM7_cjs.getStorage; }
});
