// @mastra/server/dist/server/handlers/scorer-versions.cjs
'use strict';

var chunk3ATWQ4MY_cjs = require('../../chunk-3ATWQ4MY.cjs');



Object.defineProperty(exports, "ACTIVATE_SCORER_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunk3ATWQ4MY_cjs.ACTIVATE_SCORER_VERSION_ROUTE; }
});
Object.defineProperty(exports, "COMPARE_SCORER_VERSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunk3ATWQ4MY_cjs.COMPARE_SCORER_VERSIONS_ROUTE; }
});
Object.defineProperty(exports, "CREATE_SCORER_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunk3ATWQ4MY_cjs.CREATE_SCORER_VERSION_ROUTE; }
});
Object.defineProperty(exports, "DELETE_SCORER_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunk3ATWQ4MY_cjs.DELETE_SCORER_VERSION_ROUTE; }
});
Object.defineProperty(exports, "GET_SCORER_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunk3ATWQ4MY_cjs.GET_SCORER_VERSION_ROUTE; }
});
Object.defineProperty(exports, "LIST_SCORER_VERSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunk3ATWQ4MY_cjs.LIST_SCORER_VERSIONS_ROUTE; }
});
Object.defineProperty(exports, "RESTORE_SCORER_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunk3ATWQ4MY_cjs.RESTORE_SCORER_VERSION_ROUTE; }
});
