// @mastra/server/dist/server/handlers/stored-skills.cjs
'use strict';

var chunkS4UGB72N_cjs = require('../../chunk-S4UGB72N.cjs');



Object.defineProperty(exports, "CREATE_STORED_SKILL_ROUTE", {
  enumerable: true,
  get: function () { return chunkS4UGB72N_cjs.CREATE_STORED_SKILL_ROUTE; }
});
Object.defineProperty(exports, "DELETE_STORED_SKILL_ROUTE", {
  enumerable: true,
  get: function () { return chunkS4UGB72N_cjs.DELETE_STORED_SKILL_ROUTE; }
});
Object.defineProperty(exports, "GET_STORED_SKILL_ROUTE", {
  enumerable: true,
  get: function () { return chunkS4UGB72N_cjs.GET_STORED_SKILL_ROUTE; }
});
Object.defineProperty(exports, "LIST_STORED_SKILLS_ROUTE", {
  enumerable: true,
  get: function () { return chunkS4UGB72N_cjs.LIST_STORED_SKILLS_ROUTE; }
});
Object.defineProperty(exports, "PUBLISH_STORED_SKILL_ROUTE", {
  enumerable: true,
  get: function () { return chunkS4UGB72N_cjs.PUBLISH_STORED_SKILL_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_STORED_SKILL_ROUTE", {
  enumerable: true,
  get: function () { return chunkS4UGB72N_cjs.UPDATE_STORED_SKILL_ROUTE; }
});
