// @mastra/server/dist/server/handlers/stored-agents.cjs
'use strict';

var chunkPIHZRT4S_cjs = require('../../chunk-PIHZRT4S.cjs');



Object.defineProperty(exports, "CREATE_STORED_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkPIHZRT4S_cjs.CREATE_STORED_AGENT_ROUTE; }
});
Object.defineProperty(exports, "DELETE_STORED_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkPIHZRT4S_cjs.DELETE_STORED_AGENT_ROUTE; }
});
Object.defineProperty(exports, "EXPORT_STORED_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkPIHZRT4S_cjs.EXPORT_STORED_AGENT_ROUTE; }
});
Object.defineProperty(exports, "GET_STORED_AGENT_DEPENDENTS_ROUTE", {
  enumerable: true,
  get: function () { return chunkPIHZRT4S_cjs.GET_STORED_AGENT_DEPENDENTS_ROUTE; }
});
Object.defineProperty(exports, "GET_STORED_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkPIHZRT4S_cjs.GET_STORED_AGENT_ROUTE; }
});
Object.defineProperty(exports, "LIST_STORED_AGENTS_ROUTE", {
  enumerable: true,
  get: function () { return chunkPIHZRT4S_cjs.LIST_STORED_AGENTS_ROUTE; }
});
Object.defineProperty(exports, "PREVIEW_INSTRUCTIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkPIHZRT4S_cjs.PREVIEW_INSTRUCTIONS_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_STORED_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkPIHZRT4S_cjs.UPDATE_STORED_AGENT_ROUTE; }
});
