// @mastra/server/dist/server/handlers/tool-providers.cjs
'use strict';

var chunkDBCEECWQ_cjs = require('../../chunk-DBCEECWQ.cjs');



Object.defineProperty(exports, "AUTHORIZE_TOOL_PROVIDER_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.AUTHORIZE_TOOL_PROVIDER_ROUTE; }
});
Object.defineProperty(exports, "DISCONNECT_TOOL_PROVIDER_CONNECTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.DISCONNECT_TOOL_PROVIDER_CONNECTION_ROUTE; }
});
Object.defineProperty(exports, "GET_TOOL_PROVIDER_AUTH_STATUS_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.GET_TOOL_PROVIDER_AUTH_STATUS_ROUTE; }
});
Object.defineProperty(exports, "GET_TOOL_PROVIDER_CONNECTION_USAGE_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.GET_TOOL_PROVIDER_CONNECTION_USAGE_ROUTE; }
});
Object.defineProperty(exports, "GET_TOOL_PROVIDER_HEALTH_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.GET_TOOL_PROVIDER_HEALTH_ROUTE; }
});
Object.defineProperty(exports, "GET_TOOL_PROVIDER_TOOL_SCHEMA_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.GET_TOOL_PROVIDER_TOOL_SCHEMA_ROUTE; }
});
Object.defineProperty(exports, "LIST_TOOL_PROVIDERS_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.LIST_TOOL_PROVIDERS_ROUTE; }
});
Object.defineProperty(exports, "LIST_TOOL_PROVIDER_CONNECTIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.LIST_TOOL_PROVIDER_CONNECTIONS_ROUTE; }
});
Object.defineProperty(exports, "LIST_TOOL_PROVIDER_CONNECTION_FIELDS_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.LIST_TOOL_PROVIDER_CONNECTION_FIELDS_ROUTE; }
});
Object.defineProperty(exports, "LIST_TOOL_PROVIDER_TOOLKITS_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.LIST_TOOL_PROVIDER_TOOLKITS_ROUTE; }
});
Object.defineProperty(exports, "LIST_TOOL_PROVIDER_TOOLS_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.LIST_TOOL_PROVIDER_TOOLS_ROUTE; }
});
Object.defineProperty(exports, "TOOL_PROVIDER_CONNECTION_STATUS_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.TOOL_PROVIDER_CONNECTION_STATUS_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_TOOL_PROVIDER_CONNECTION_ROUTE", {
  enumerable: true,
  get: function () { return chunkDBCEECWQ_cjs.UPDATE_TOOL_PROVIDER_CONNECTION_ROUTE; }
});
