// @mastra/server/dist/server/handlers/vector.cjs
'use strict';

var chunkURW7EH4T_cjs = require('../../chunk-URW7EH4T.cjs');



Object.defineProperty(exports, "CREATE_INDEX_ROUTE", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.CREATE_INDEX_ROUTE; }
});
Object.defineProperty(exports, "DELETE_INDEX_ROUTE", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.DELETE_INDEX_ROUTE; }
});
Object.defineProperty(exports, "DESCRIBE_INDEX_ROUTE", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.DESCRIBE_INDEX_ROUTE; }
});
Object.defineProperty(exports, "LIST_EMBEDDERS_ROUTE", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.LIST_EMBEDDERS_ROUTE; }
});
Object.defineProperty(exports, "LIST_INDEXES_ROUTE", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.LIST_INDEXES_ROUTE; }
});
Object.defineProperty(exports, "LIST_VECTORS_ROUTE", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.LIST_VECTORS_ROUTE; }
});
Object.defineProperty(exports, "QUERY_VECTORS_ROUTE", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.QUERY_VECTORS_ROUTE; }
});
Object.defineProperty(exports, "UPSERT_VECTORS_ROUTE", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.UPSERT_VECTORS_ROUTE; }
});
Object.defineProperty(exports, "createIndex", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.createIndex; }
});
Object.defineProperty(exports, "deleteIndex", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.deleteIndex; }
});
Object.defineProperty(exports, "describeIndex", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.describeIndex; }
});
Object.defineProperty(exports, "listIndexes", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.listIndexes; }
});
Object.defineProperty(exports, "listVectorStores", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.listVectorStores; }
});
Object.defineProperty(exports, "queryVectors", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.queryVectors; }
});
Object.defineProperty(exports, "upsertVectors", {
  enumerable: true,
  get: function () { return chunkURW7EH4T_cjs.upsertVectors; }
});
