// @mastra/server/dist/server/handlers/workflows.cjs
'use strict';

var chunkS5DWQRTK_cjs = require('../../chunk-S5DWQRTK.cjs');



Object.defineProperty(exports, "CANCEL_WORKFLOW_RUN_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.CANCEL_WORKFLOW_RUN_ROUTE; }
});
Object.defineProperty(exports, "CREATE_WORKFLOW_RUN_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.CREATE_WORKFLOW_RUN_ROUTE; }
});
Object.defineProperty(exports, "DELETE_WORKFLOW_RUN_BY_ID_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.DELETE_WORKFLOW_RUN_BY_ID_ROUTE; }
});
Object.defineProperty(exports, "EXECUTE_WORKFLOW_STEP_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.EXECUTE_WORKFLOW_STEP_ROUTE; }
});
Object.defineProperty(exports, "GET_WORKFLOW_BY_ID_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.GET_WORKFLOW_BY_ID_ROUTE; }
});
Object.defineProperty(exports, "GET_WORKFLOW_RUN_BY_ID_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.GET_WORKFLOW_RUN_BY_ID_ROUTE; }
});
Object.defineProperty(exports, "LIST_WORKFLOWS_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.LIST_WORKFLOWS_ROUTE; }
});
Object.defineProperty(exports, "LIST_WORKFLOW_RUNS_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.LIST_WORKFLOW_RUNS_ROUTE; }
});
Object.defineProperty(exports, "OBSERVE_STREAM_LEGACY_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.OBSERVE_STREAM_LEGACY_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "OBSERVE_STREAM_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.OBSERVE_STREAM_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "RECEIVE_WORKFLOW_EVENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.RECEIVE_WORKFLOW_EVENT_ROUTE; }
});
Object.defineProperty(exports, "RESTART_ALL_ACTIVE_WORKFLOW_RUNS_ASYNC_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.RESTART_ALL_ACTIVE_WORKFLOW_RUNS_ASYNC_ROUTE; }
});
Object.defineProperty(exports, "RESTART_ALL_ACTIVE_WORKFLOW_RUNS_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.RESTART_ALL_ACTIVE_WORKFLOW_RUNS_ROUTE; }
});
Object.defineProperty(exports, "RESTART_ASYNC_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.RESTART_ASYNC_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "RESTART_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.RESTART_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "RESUME_ASYNC_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.RESUME_ASYNC_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "RESUME_NO_WAIT_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.RESUME_NO_WAIT_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "RESUME_STREAM_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.RESUME_STREAM_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "RESUME_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.RESUME_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "START_ASYNC_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.START_ASYNC_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "START_WORKFLOW_RUN_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.START_WORKFLOW_RUN_ROUTE; }
});
Object.defineProperty(exports, "STREAM_LEGACY_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.STREAM_LEGACY_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "STREAM_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.STREAM_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "TIME_TRAVEL_ASYNC_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.TIME_TRAVEL_ASYNC_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "TIME_TRAVEL_STREAM_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.TIME_TRAVEL_STREAM_WORKFLOW_ROUTE; }
});
Object.defineProperty(exports, "TIME_TRAVEL_WORKFLOW_ROUTE", {
  enumerable: true,
  get: function () { return chunkS5DWQRTK_cjs.TIME_TRAVEL_WORKFLOW_ROUTE; }
});
