// @mastra/server/dist/server/handlers/workspace.cjs
'use strict';

var chunkAUFCEKYY_cjs = require('../../chunk-AUFCEKYY.cjs');



Object.defineProperty(exports, "GET_WORKSPACE_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.GET_WORKSPACE_ROUTE; }
});
Object.defineProperty(exports, "LIST_WORKSPACES_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.LIST_WORKSPACES_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_FS_DELETE_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_FS_DELETE_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_FS_LIST_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_FS_LIST_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_FS_MKDIR_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_FS_MKDIR_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_FS_READ_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_FS_READ_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_FS_ROUTES", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_FS_ROUTES; }
});
Object.defineProperty(exports, "WORKSPACE_FS_STAT_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_FS_STAT_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_FS_WRITE_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_FS_WRITE_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_GET_SKILL_REFERENCE_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_GET_SKILL_REFERENCE_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_GET_SKILL_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_GET_SKILL_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_INDEX_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_INDEX_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_LIST_SKILLS_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_LIST_SKILLS_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_LIST_SKILL_REFERENCES_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_LIST_SKILL_REFERENCES_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_SEARCH_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SEARCH_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_SEARCH_ROUTES", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SEARCH_ROUTES; }
});
Object.defineProperty(exports, "WORKSPACE_SEARCH_SKILLS_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SEARCH_SKILLS_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_SKILLS_ROUTES", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SKILLS_ROUTES; }
});
Object.defineProperty(exports, "WORKSPACE_SKILLS_SH_INSTALL_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SKILLS_SH_INSTALL_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_SKILLS_SH_POPULAR_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SKILLS_SH_POPULAR_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_SKILLS_SH_PREVIEW_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SKILLS_SH_PREVIEW_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_SKILLS_SH_REMOVE_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SKILLS_SH_REMOVE_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_SKILLS_SH_ROUTES", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SKILLS_SH_ROUTES; }
});
Object.defineProperty(exports, "WORKSPACE_SKILLS_SH_SEARCH_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SKILLS_SH_SEARCH_ROUTE; }
});
Object.defineProperty(exports, "WORKSPACE_SKILLS_SH_UPDATE_ROUTE", {
  enumerable: true,
  get: function () { return chunkAUFCEKYY_cjs.WORKSPACE_SKILLS_SH_UPDATE_ROUTE; }
});
