// @mastra/core/dist/signals/index.cjs
'use strict';

var chunkY7R26VZI_cjs = require('../chunk-Y7R26VZI.cjs');
var chunkAOQKFEGW_cjs = require('../chunk-AOQKFEGW.cjs');

// src/signals/task-signal-provider.ts
var TaskSignalProvider = class extends chunkY7R26VZI_cjs.SignalProvider {
  id = "task-signals";
  #processor = new chunkAOQKFEGW_cjs.TaskStateProcessor();
  getInputProcessors() {
    return [this.#processor];
  }
  getTools() {
    return {
      task_write: chunkAOQKFEGW_cjs.taskWriteTool,
      task_update: chunkAOQKFEGW_cjs.taskUpdateTool,
      task_complete: chunkAOQKFEGW_cjs.taskCompleteTool,
      task_check: chunkAOQKFEGW_cjs.taskCheckTool
    };
  }
};

Object.defineProperty(exports, "SignalProvider", {
  enumerable: true,
  get: function () { return chunkY7R26VZI_cjs.SignalProvider; }
});
Object.defineProperty(exports, "WebhookSignalProvider", {
  enumerable: true,
  get: function () { return chunkY7R26VZI_cjs.WebhookSignalProvider; }
});
Object.defineProperty(exports, "isSignalProvider", {
  enumerable: true,
  get: function () { return chunkY7R26VZI_cjs.isSignalProvider; }
});
exports.TaskSignalProvider = TaskSignalProvider;
