// @mastra/deployer/dist/index.cjs
'use strict';

var chunkWE66TECO_cjs = require('./chunk-WE66TECO.cjs');
var chunkKBVOYAAL_cjs = require('./chunk-KBVOYAAL.cjs');
var chunkUEH6D24P_cjs = require('./chunk-UEH6D24P.cjs');

// src/deploy/base.ts
var Deployer = class extends chunkWE66TECO_cjs.Bundler {
  deps = new chunkUEH6D24P_cjs.DepsService();
  constructor(args) {
    super(args.name, "DEPLOYER");
    this.deps.__setLogger(this.logger);
  }
  getEnvFiles() {
    const possibleFiles = [".env.production", ".env.local", ".env"];
    try {
      const fileService = new chunkUEH6D24P_cjs.FileService();
      const envFile = fileService.getFirstExistingFile(possibleFiles);
      return Promise.resolve([envFile]);
    } catch {
    }
    return Promise.resolve([]);
  }
};

// src/build/deployer.ts
async function getDeployer(entryFile, outputDir, logger) {
  const result = await chunkKBVOYAAL_cjs.extractMastraOption("deployer", entryFile, outputDir, logger);
  if (!result) {
    return null;
  }
  return result.getConfig();
}

Object.defineProperty(exports, "Deps", {
  enumerable: true,
  get: function () { return chunkUEH6D24P_cjs.Deps; }
});
Object.defineProperty(exports, "FileService", {
  enumerable: true,
  get: function () { return chunkUEH6D24P_cjs.FileService; }
});
Object.defineProperty(exports, "createChildProcessLogger", {
  enumerable: true,
  get: function () { return chunkUEH6D24P_cjs.createChildProcessLogger; }
});
Object.defineProperty(exports, "createPinoStream", {
  enumerable: true,
  get: function () { return chunkUEH6D24P_cjs.createPinoStream; }
});
exports.Deployer = Deployer;
exports.getDeployer = getDeployer;
