// @mastra/core/dist/harness/v1/index.js
import { augmentWithInit } from '../../chunk-EIL3YZEF.js';
import { Workspace } from '../../chunk-V2MTR7E6.js';
import { createTool } from '../../chunk-3SZ22TE5.js';
import { toStandardSchema } from '../../chunk-6SRTDZ7S.js';
import { RequestContext } from '../../chunk-RI37F2KY.js';
import { randomUUID, createHash } from 'crypto';
import { z } from 'zod';

var HARNESS_EVENT_ID_PREFIX = "harness-v1";
var HarnessValidationError = class extends Error {
  constructor(path, message) {
    super(`${path}: ${message}`);
    this.path = path;
    this.name = "HarnessValidationError";
  }
  path;
};
var HarnessEventSerializationError = class extends HarnessValidationError {
  constructor(eventType, reason, sessionId) {
    super("event.payload", `custom event "${eventType}" payload is not JSON-serializable: ${reason}`);
    this.eventType = eventType;
    this.reason = reason;
    this.sessionId = sessionId;
    this.name = "HarnessEventSerializationError";
  }
  eventType;
  reason;
  sessionId;
};
function formatHarnessEventId(epoch, sequence) {
  if (epoch.length === 0 || epoch.includes(":")) {
    throw new HarnessValidationError("eventId.epoch", 'epoch must be non-empty and must not contain ":"');
  }
  if (!Number.isSafeInteger(sequence) || sequence < 0) {
    throw new HarnessValidationError("eventId.sequence", "sequence must be a non-negative safe integer");
  }
  return `${HARNESS_EVENT_ID_PREFIX}:${epoch}:${sequence}`;
}
function parseHarnessEventId(eventId) {
  const parts = eventId.split(":");
  if (parts.length !== 3 || parts[0] !== HARNESS_EVENT_ID_PREFIX || parts[1] === "" || parts[2] === "") {
    throw new HarnessValidationError("lastEventId", "expected event id grammar harness-v1:<epoch>:<seq>");
  }
  const sequenceText = parts[2];
  if (!/^(0|[1-9][0-9]*)$/.test(sequenceText)) {
    throw new HarnessValidationError("lastEventId", "event id sequence must be an unsigned decimal integer");
  }
  const sequence = Number(sequenceText);
  if (!Number.isSafeInteger(sequence)) {
    throw new HarnessValidationError("lastEventId", "event id sequence must be within JavaScript safe integer range");
  }
  return { epoch: parts[1], sequence };
}
function sessionCreatedPayload(record) {
  return {
    resourceId: record.resourceId,
    threadId: record.threadId,
    parentSessionId: record.parentSessionId,
    modeId: record.modeId,
    modelId: record.modelId
  };
}
function assertCustomEventType(eventType) {
  if (RESERVED_EVENT_TYPES.has(eventType) || RESERVED_EVENT_PREFIXES.some((prefix) => eventType.startsWith(prefix))) {
    throw new HarnessValidationError("event.type", `"${eventType}" is reserved by the harness`);
  }
  if (!eventType.includes(".")) {
    throw new HarnessValidationError("event.type", "custom event types must be namespaced with a dot");
  }
}
function assertJsonSerializable(eventType, sessionId, value) {
  visitJsonValue(eventType, sessionId, value, /* @__PURE__ */ new WeakSet());
}
var EventEmitter = class _EventEmitter {
  listeners = [];
  epoch;
  seq;
  scope;
  onEvent;
  constructor(scope = {}, opts = {}) {
    this.scope = scope;
    this.onEvent = opts.onEvent;
    this.epoch = opts.epoch;
    this.seq = opts.nextSequence ?? 0;
    formatHarnessEventId(this.epoch ?? "pending-epoch", this.seq);
  }
  scoped(scope) {
    return new _EventEmitter(
      { ...this.scope, ...scope },
      {
        onEvent: (event) => this.forward(event)
      }
    );
  }
  subscribe(listener) {
    this.listeners.push(listener);
    return () => {
      const index = this.listeners.indexOf(listener);
      if (index !== -1) this.listeners.splice(index, 1);
    };
  }
  emit(event, overrides) {
    const sessionId = overrides?.sessionId ?? this.scope.sessionId;
    this.validateCustomEvent(event, sessionId);
    const stamped = {
      ...event,
      id: formatHarnessEventId(this.epochId, this.seq++),
      timestamp: Date.now(),
      ...sessionId !== void 0 && { sessionId }
    };
    this.dispatch(stamped);
    return stamped;
  }
  forward(event) {
    this.dispatch(event);
  }
  get listenerCount() {
    return this.listeners.length;
  }
  get epochId() {
    this.epoch ??= randomUUID();
    return this.epoch;
  }
  validateCustomEvent(event, sessionId) {
    const eventType = event.type;
    if (typeof eventType !== "string" || RESERVED_EVENT_TYPES.has(eventType)) return;
    assertCustomEventType(eventType);
    if (Object.prototype.hasOwnProperty.call(event, "payload")) {
      assertJsonSerializable(eventType, sessionId, event.payload);
    }
  }
  dispatch(event) {
    if (this.onEvent) {
      try {
        const result = this.onEvent(event);
        if (result && typeof result.catch === "function") {
          result.catch((err) => console.error("[harness/v1] event persistence rejected:", err));
        }
      } catch (err) {
        console.error("[harness/v1] event persistence threw:", err);
      }
    }
    for (const listener of [...this.listeners]) {
      try {
        const result = listener(event);
        if (result && typeof result.catch === "function") {
          result.catch((err) => console.error("[harness/v1] event listener rejected:", err));
        }
      } catch (err) {
        console.error("[harness/v1] event listener threw:", err);
      }
    }
  }
};
var RESERVED_EVENT_TYPES = /* @__PURE__ */ new Set([
  "session_created",
  "mode_changed",
  "model_changed",
  "thread_cloned",
  "state_changed",
  "subagent_start"
]);
var RESERVED_EVENT_PREFIXES = ["session_", "thread_"];
function visitJsonValue(eventType, sessionId, value, seen) {
  if (value === void 0) throwSerialization(eventType, sessionId, "undefined");
  if (typeof value === "function") throwSerialization(eventType, sessionId, "function");
  if (typeof value === "symbol") throwSerialization(eventType, sessionId, "symbol");
  if (typeof value === "bigint") throwSerialization(eventType, sessionId, "bigint");
  if (value === null || typeof value === "string" || typeof value === "number" || typeof value === "boolean") return;
  if (value instanceof Date) throwSerialization(eventType, sessionId, "date");
  if (value instanceof Map) throwSerialization(eventType, sessionId, "map");
  if (value instanceof Set) throwSerialization(eventType, sessionId, "set");
  if (ArrayBuffer.isView(value)) throwSerialization(eventType, sessionId, "typed-array");
  if (typeof value !== "object") return;
  if (seen.has(value)) throwSerialization(eventType, sessionId, "cycle");
  seen.add(value);
  if (Array.isArray(value)) {
    for (const item of value) visitJsonValue(eventType, sessionId, item, seen);
    seen.delete(value);
    return;
  }
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) {
    throwSerialization(eventType, sessionId, "non-plain-object");
  }
  for (const item of Object.values(value)) {
    visitJsonValue(eventType, sessionId, item, seen);
  }
  seen.delete(value);
}
function throwSerialization(eventType, sessionId, reason) {
  throw new HarnessEventSerializationError(eventType, reason, sessionId);
}

// src/harness/v1/request-context.ts
function buildHarnessRequestContext({
  harnessContext
}) {
  const requestContext = new RequestContext();
  requestContext.set("harness", harnessContext);
  return requestContext;
}

// src/harness/v1/skills.types.ts
var HarnessSkillNotFoundError = class extends Error {
  name = "HarnessSkillNotFoundError";
  skillName;
  searchedSources;
  constructor(opts) {
    super(`Harness skill not found: ${opts.name} (searched: ${opts.searchedSources.join(", ") || "none"})`);
    this.skillName = opts.name;
    this.searchedSources = opts.searchedSources;
  }
};
var taskSchema = z.object({
  id: z.string().optional(),
  content: z.string(),
  status: z.enum(["pending", "in_progress", "completed"]),
  activeForm: z.string()
});
var taskUpdateSchema = z.object({
  id: z.string(),
  content: z.string().optional(),
  status: z.enum(["pending", "in_progress", "completed"]).optional(),
  activeForm: z.string().optional()
});
function getHarnessContext(context) {
  const harnessContext = context.requestContext?.get("harness");
  if (!harnessContext) {
    throw new Error('Harness tool execution requires context.requestContext.get("harness")');
  }
  return harnessContext;
}
function assertOwningSession(session, context) {
  const harnessContext = getHarnessContext(context);
  if (harnessContext.sessionId !== session.id) {
    throw new Error(
      `Harness tool execution context belongs to session "${harnessContext.sessionId}", not "${session.id}"`
    );
  }
  return harnessContext;
}
function recoverableError(error, code = "harness.tool_failed") {
  return {
    isError: true,
    code,
    message: error instanceof Error ? error.message : String(error)
  };
}
function taskSummary(tasks) {
  const completed = tasks.filter((task) => task.status === "completed").length;
  const inProgress = tasks.filter((task) => task.status === "in_progress").length;
  const pending = tasks.filter((task) => task.status === "pending").length;
  const incompleteTasks = tasks.filter((task) => task.status !== "completed");
  return {
    tasks,
    summary: {
      total: tasks.length,
      completed,
      inProgress,
      pending,
      incomplete: incompleteTasks.length,
      hasTasks: tasks.length > 0,
      allCompleted: tasks.length > 0 && incompleteTasks.length === 0
    },
    incompleteTasks
  };
}
function getTasks(session) {
  const state = session.getState();
  return (state.tasks ?? []).map((task) => ({ ...task }));
}
async function replaceTasks(session, tasks) {
  return session.updateState(() => ({
    updates: { tasks: tasks.map((task) => ({ ...task })) },
    result: tasks.map((task) => ({ ...task }))
  }));
}
async function updateTask(session, taskId, updates) {
  return session.updateState((state) => {
    const tasks = (state.tasks ?? []).map((task2) => ({ ...task2 }));
    const task = tasks.find((task2) => task2.id === taskId);
    if (!task) {
      throw new Error(`Harness task "${taskId}" was not found on session "${session.id}"`);
    }
    Object.assign(task, updates);
    return { updates: { tasks }, result: tasks };
  });
}
function buildHarnessBuiltInTools(session) {
  return {
    task_write: createTool({
      id: "task_write",
      description: "Replace the task list for the current harness session.",
      inputSchema: z.object({ tasks: z.array(taskSchema).min(1) }),
      execute: async ({ tasks }, context) => {
        try {
          assertOwningSession(session, context);
          const savedTasks = await replaceTasks(
            session,
            tasks.map((task) => ({ ...task, id: task.id ?? randomUUID() }))
          );
          return taskSummary(savedTasks);
        } catch (error) {
          return recoverableError(error);
        }
      }
    }),
    task_update: createTool({
      id: "task_update",
      description: "Update one task in the current harness session.",
      inputSchema: taskUpdateSchema,
      execute: async ({ id, ...updates }, context) => {
        try {
          assertOwningSession(session, context);
          return taskSummary(await updateTask(session, id, updates));
        } catch (error) {
          return recoverableError(error);
        }
      }
    }),
    task_complete: createTool({
      id: "task_complete",
      description: "Mark one task completed in the current harness session.",
      inputSchema: z.object({ id: z.string() }),
      execute: async ({ id }, context) => {
        try {
          assertOwningSession(session, context);
          return taskSummary(await updateTask(session, id, { status: "completed" }));
        } catch (error) {
          return recoverableError(error);
        }
      }
    }),
    task_check: createTool({
      id: "task_check",
      description: "Read the task completion state for the current harness session.",
      inputSchema: z.object({}).optional(),
      execute: async (_input, context) => {
        try {
          assertOwningSession(session, context);
          return taskSummary(getTasks(session));
        } catch (error) {
          return recoverableError(error);
        }
      }
    }),
    ask_user: createTool({
      id: "ask_user",
      description: "Register a pending question for the current harness session.",
      inputSchema: z.object({
        question: z.string(),
        options: z.array(z.object({ label: z.string(), description: z.string().nullable().optional() })).min(2).max(4).optional().nullable(),
        selectionMode: z.enum(["single_select", "multi_select"]).optional()
      }),
      execute: async (input, context) => {
        try {
          assertOwningSession(session, context);
          const pending = await session.registerPendingItem({
            id: randomUUID(),
            kind: "question",
            status: "pending",
            runId: context.agent?.runId ?? context.workflow?.runId ?? null,
            traceId: context.traceId ?? null,
            payload: input
          });
          return { isError: false, pendingItemId: pending.id, status: pending.status };
        } catch (error) {
          return recoverableError(error);
        }
      }
    }),
    submit_plan: createTool({
      id: "submit_plan",
      description: "Register a pending plan approval for the current harness session.",
      inputSchema: z.object({ title: z.string().nullable().optional(), plan: z.string() }),
      execute: async (input, context) => {
        try {
          assertOwningSession(session, context);
          const pending = await session.registerPendingItem({
            id: randomUUID(),
            kind: "plan-approval",
            status: "pending",
            runId: context.agent?.runId ?? context.workflow?.runId ?? null,
            traceId: context.traceId ?? null,
            payload: { ...input, transitionModeId: session.getMode().transitionsTo }
          });
          return { isError: false, pendingItemId: pending.id, status: pending.status };
        } catch (error) {
          return recoverableError(error);
        }
      }
    }),
    subagent: createTool({
      id: "subagent",
      description: "Create a durable child harness session for a configured subagent type.",
      inputSchema: z.object({
        agentType: z.string(),
        prompt: z.string(),
        modelId: z.string().optional().nullable(),
        forked: z.boolean().optional().nullable()
      }),
      execute: async (input, context) => {
        try {
          assertOwningSession(session, context);
          return await session.spawnSubagentSession({
            agentType: input.agentType,
            prompt: input.prompt,
            ...input.modelId ? { modelId: input.modelId } : {},
            ...input.forked !== void 0 && input.forked !== null ? { forked: input.forked } : {}
          });
        } catch (error) {
          return recoverableError(error);
        }
      }
    }),
    useSkill: createTool({
      id: "useSkill",
      description: "Activate a skill in the current harness session.",
      inputSchema: z.object({ ref: z.string() }),
      execute: async ({ ref }, context) => {
        try {
          assertOwningSession(session, context);
          return { instructions: await session.useSkill(ref) };
        } catch (error) {
          return recoverableError(error);
        }
      }
    })
  };
}
function buildSessionToolsets(opts = {}) {
  const { agentTools, modeOverrides, builtInTools, permissionRules, disabledTools } = opts;
  const base = modeOverrides?.tools ? { ...modeOverrides.tools } : { ...agentTools };
  if (modeOverrides?.additionalTools && !modeOverrides.tools) {
    Object.assign(base, modeOverrides.additionalTools);
  }
  if (builtInTools) {
    Object.assign(base, builtInTools);
  }
  const denySet = new Set(disabledTools ?? []);
  for (const [name, policy] of Object.entries(permissionRules?.tools ?? {})) {
    if (policy === "deny") denySet.add(name);
  }
  for (const name of denySet) {
    delete base[name];
  }
  return base;
}

// src/harness/v1/session.ts
var Session = class _Session {
  /** Stable identity. Frozen at construction. */
  #id;
  #ownerId;
  #resourceId;
  #threadId;
  #createdAt;
  #lastActivityAt;
  #agent;
  #storage;
  #runtimeCompatibilityGeneration;
  #parentSessionId;
  #subagentDepth;
  #source;
  #pending;
  #runStatus = "idle";
  #currentRunId = null;
  #currentTraceId = null;
  #memory;
  #events;
  #stateSchemaInput;
  #stateSchema;
  #state;
  #stateUpdateQueue = Promise.resolve();
  #workspace;
  #resolvedWorkspace;
  #workspaceResolved = false;
  #resolveAgent;
  #resolveMode;
  /**
   * Single-flight cache for workspace skill discovery (spec §4.6: concurrent
   * `listSkills`/`useSkill` calls must share the same in-flight promise so we
   * don't re-scan the workspace per call).
   */
  #workspaceSkillsPromise;
  #subagents;
  #resolveModel;
  #defaultPermissionPolicy;
  #toolCategoryResolver;
  // readonly parentSessionId?: string;
  // readonly subagentDepth: number;
  #modelId;
  #mode;
  constructor(config) {
    this.#id = config.id;
    this.#ownerId = config.ownerId;
    this.#resourceId = config.resourceId;
    this.#threadId = config.threadId;
    this.#mode = config.mode;
    this.#modelId = config.model;
    this.#createdAt = config.createdAt;
    this.#lastActivityAt = config.lastActivityAt;
    this.#storage = config.storage;
    this.#runtimeCompatibilityGeneration = config.runtimeCompatibilityGeneration;
    this.#parentSessionId = config.record?.parentSessionId;
    this.#subagentDepth = config.record?.subagentDepth ?? 0;
    this.#source = config.record?.source ? { type: config.record.source.type, parentSessionId: config.record.source.parentSessionId } : { type: config.record?.origin ?? "top-level", parentSessionId: config.record?.parentSessionId };
    this.#pending = (config.pending ?? config.record?.pending ?? []).map((item) => ({ ...item }));
    this.#memory = config.memory;
    this.#events = config.events;
    this.#stateSchemaInput = config.stateSchema;
    this.#stateSchema = config.stateSchema ? toStandardSchema(config.stateSchema) : void 0;
    this.#resolveAgent = config.resolveAgent;
    this.#resolveMode = config.resolveMode;
    this.#state = {
      ...this.#getSchemaDefaults(),
      ...config.initialState,
      ...config.record?.state
    };
    this.#workspace = config.workspace;
    this.#subagents = config.subagents;
    this.#resolveModel = config.resolveModel;
    this.#defaultPermissionPolicy = config.defaultPermissionPolicy ?? "ask";
    this.#toolCategoryResolver = config.toolCategoryResolver;
    this.#agent = config.agent;
  }
  get id() {
    return this.#id;
  }
  get ownerId() {
    return this.#ownerId;
  }
  get resourceId() {
    return this.#resourceId;
  }
  get threadId() {
    return this.#threadId;
  }
  get createdAt() {
    return this.#createdAt;
  }
  get lastActivityAt() {
    return this.#lastActivityAt;
  }
  get parentSessionId() {
    return this.#parentSessionId;
  }
  get subagentDepth() {
    return this.#subagentDepth;
  }
  isBusy() {
    return this.#isBusySnapshot();
  }
  async waitForIdle(opts = {}) {
    const timeout = opts.timeout ?? 3e4;
    const startedAt = Date.now();
    while (this.#isBusySnapshot()) {
      if (Date.now() - startedAt >= timeout) {
        throw new Error(`Harness session "${this.#id}" did not become idle within ${timeout}ms`);
      }
      await new Promise((resolve) => setTimeout(resolve, 25));
    }
  }
  getQueueDepth() {
    return this.#pending.filter((item) => item.status === "pending").length;
  }
  getCurrentRunId() {
    return this.#currentRunId;
  }
  getCurrentTraceId() {
    return this.#currentTraceId;
  }
  listPendingItems() {
    return this.#pending.map((item) => ({ ...item }));
  }
  async spawnSubagentSession(opts) {
    const maxDepth = this.#subagents?.maxDepth ?? 1;
    const attemptedDepth = this.#subagentDepth + 1;
    if (attemptedDepth > maxDepth) {
      return {
        isError: true,
        code: "harness.subagent_depth_exceeded",
        message: `Harness subagent depth ${attemptedDepth} exceeds the configured maximum of ${maxDepth}`,
        details: { maxDepth, attemptedDepth }
      };
    }
    const definition = this.#subagents?.types?.[opts.agentType];
    if (!definition) {
      throw new Error(`Harness subagent type "${opts.agentType}" was not found`);
    }
    if (!this.#resolveAgent) {
      throw new Error("Harness subagent spawn requires an agent resolver");
    }
    await this.#resolveAgent(definition.agentId);
    const modelId = opts.modelId ?? definition.defaultModelId ?? this.#modelId;
    if (!this.#resolveModel) {
      throw new Error("Harness subagent spawn requires a resolveModel function");
    }
    await this.#resolveModel(modelId);
    const now = /* @__PURE__ */ new Date();
    const record = {
      id: `sess-${randomUUID()}`,
      ownerId: this.#ownerId,
      resourceId: this.#resourceId,
      threadId: `thread-${randomUUID()}`,
      parentSessionId: this.#id,
      origin: "subagent-tool",
      source: { type: "subagent-tool", parentSessionId: this.#id },
      subagentDepth: attemptedDepth,
      runtimeCompatibilityGeneration: this.#runtimeCompatibilityGeneration,
      modeId: this.#mode.id,
      modelId,
      metadata: {
        agentType: opts.agentType,
        agentId: definition.agentId,
        prompt: opts.prompt,
        forked: opts.forked ?? definition.forked ?? false
      },
      state: this.getState(),
      pending: [],
      createdAt: now,
      lastActivityAt: now
    };
    await this.#storage.saveSession(record);
    this.#events.emit({ type: "session_created", ...sessionCreatedPayload(record) }, { sessionId: record.id });
    this.#events.emit({
      type: "subagent_start",
      subagentSessionId: record.id,
      payload: { agentType: opts.agentType, parentSessionId: this.#id, depth: attemptedDepth }
    });
    return {
      isError: false,
      subagentSessionId: record.id,
      threadId: record.threadId,
      resourceId: record.resourceId,
      agentType: opts.agentType,
      depth: attemptedDepth
    };
  }
  async registerPendingItem(item) {
    const now = /* @__PURE__ */ new Date();
    const record = {
      ...item,
      sessionId: this.#id,
      runtimeCompatibilityGeneration: item.runtimeCompatibilityGeneration ?? this.#runtimeCompatibilityGeneration,
      createdAt: item.createdAt ?? now,
      updatedAt: item.updatedAt ?? now
    };
    this.#pending = [...this.#pending, record];
    await this.#storage.appendPendingItem(this.#id, record);
    await this.#reloadRecordProjection();
    return { ...record };
  }
  async updatePendingItem(pendingItemId, updates) {
    await this.#storage.updatePendingItem(this.#id, pendingItemId, updates);
    await this.#reloadRecordProjection();
    const item = this.#pending.find((item2) => item2.id === pendingItemId);
    if (!item) {
      throw new Error(`Harness pending item "${pendingItemId}" was not found on session "${this.#id}"`);
    }
    return { ...item };
  }
  async removePendingItem(pendingItemId) {
    await this.#storage.removePendingItem(this.#id, pendingItemId);
    await this.#reloadRecordProjection();
  }
  async respondToToolApproval(pendingItemId, response) {
    return this.#respondToPendingItem(pendingItemId, "tool-approval", response);
  }
  async respondToToolSuspension(pendingItemId, response) {
    return this.#respondToPendingItem(pendingItemId, "tool-suspension", response);
  }
  async respondToQuestion(pendingItemId, response) {
    return this.#respondToPendingItem(pendingItemId, "question", response);
  }
  async respondToPlanApproval(pendingItemId, response) {
    return this.#respondToPendingItem(pendingItemId, "plan-approval", response);
  }
  async clone(opts = {}) {
    const result = await (await this.#resolveMemory()).cloneThread({
      sourceThreadId: this.#threadId,
      newThreadId: opts.threadId,
      resourceId: opts.resourceId ?? this.#resourceId,
      title: opts.title,
      metadata: opts.metadata,
      options: opts.messageLimit !== void 0 ? { messageLimit: opts.messageLimit } : void 0
    });
    const cloneId = opts.sessionId ?? randomUUID();
    const clone = new _Session({
      id: cloneId,
      ownerId: this.#ownerId,
      threadId: result.thread.id,
      resourceId: result.thread.resourceId,
      mode: opts.mode ?? this.#mode,
      model: opts.modelId ?? this.#modelId,
      createdAt: result.thread.createdAt,
      lastActivityAt: result.thread.updatedAt,
      agent: this.#agent,
      memory: this.#memory,
      storage: this.#storage,
      events: this.#events.scoped({ sessionId: cloneId }),
      stateSchema: this.#stateSchemaInput,
      initialState: this.getState(),
      workspace: this.#workspace,
      subagents: this.#subagents,
      resolveAgent: this.#resolveAgent,
      resolveMode: this.#resolveMode,
      resolveModel: this.#resolveModel,
      defaultPermissionPolicy: this.#defaultPermissionPolicy,
      toolCategoryResolver: this.#toolCategoryResolver
    });
    this.#events.emit({
      type: "thread_cloned",
      threadId: clone.threadId,
      resourceId: clone.resourceId,
      sourceThreadId: this.#threadId,
      title: opts.title
    });
    return clone;
  }
  async getThread() {
    return (await this.#resolveMemory()).getThreadById({ threadId: this.#threadId });
  }
  async getMessages() {
    const result = await (await this.#resolveMemory()).recall({ threadId: this.#threadId, resourceId: this.#resourceId });
    return result.messages;
  }
  async saveMessages(messages) {
    return (await this.#resolveMemory()).saveMessages({ messages });
  }
  getState() {
    return Object.freeze({ ...this.#state });
  }
  async setState(updates) {
    const run = this.#stateUpdateQueue.then(() => this.#applyStateUpdates(updates));
    this.#stateUpdateQueue = run.then(
      () => void 0,
      () => void 0
    );
    return run;
  }
  async updateState(updater) {
    const run = this.#stateUpdateQueue.then(async () => {
      const update = await updater(this.getState());
      if (update.updates && Object.keys(update.updates).length > 0) {
        await this.#applyStateUpdates(update.updates);
      }
      for (const event of update.events ?? []) {
        this.#events.emit(event);
      }
      return update.result;
    });
    this.#stateUpdateQueue = run.then(
      () => void 0,
      () => void 0
    );
    return run;
  }
  getModelId() {
    return this.#modelId;
  }
  setModelId(modelId) {
    const previousModelId = this.#modelId;
    this.#modelId = modelId;
    if (modelId !== previousModelId) {
      void this.#persistSession({ modelId });
      this.#events.emit({ type: "model_changed", modelId, previousModelId });
    }
  }
  getMode() {
    return this.#mode;
  }
  #getToolOverrides() {
    return { tools: this.#mode.tools, additionalTools: this.#mode.additionalTools };
  }
  async signal({ messages, ...options }) {
    if (!this.#resolveAgent) {
      throw new Error("Harness session cannot signal because no agent resolver is configured");
    }
    const agent = this.#agent;
    const runId = options.runId ?? randomUUID();
    this.#markRunning(runId);
    try {
      const requestContext = await this.#buildRequestContext();
      const agentTools = await agent.listTools({ requestContext });
      const tools = buildSessionToolsets({
        agentTools,
        modeOverrides: this.#getToolOverrides(),
        builtInTools: buildHarnessBuiltInTools(this)
      });
      const model = this.#resolveModel ? await this.#resolveModel(this.#modelId) : void 0;
      const result = await agent.generate(messages, {
        ...options,
        runId,
        requestContext,
        ...model ? { model } : {},
        toolsets: { harness: tools },
        activeTools: options.activeTools ? [.../* @__PURE__ */ new Set([...options.activeTools, ...Object.keys(tools)])] : Object.keys(tools)
      });
      this.#markIdle();
      return result;
    } catch (error) {
      this.#markIdle();
      throw error;
    }
  }
  setMode(mode) {
    const previousModeId = this.#mode.id;
    this.#mode = mode;
    if (mode.id !== previousModeId) {
      void this.#persistSession({ modeId: mode.id });
      this.#events.emit({ type: "mode_changed", modeId: mode.id, previousModeId });
    }
  }
  /**
   * Returns the workspace skill catalog. Workspace discovery is async on first
   * call and cached for the lifetime of the session (use `refreshSkills` to
   * invalidate).
   */
  async listSkills() {
    return this.#loadWorkspaceSkillMetadata();
  }
  /**
   * Look up a single skill by name. Returns `null` when no skill matches;
   * use `useSkill` when a missing skill should be a hard error.
   */
  async getSkill(name) {
    const workspace = await this.#getResolvedWorkspace();
    if (!workspace?.skills) return null;
    const workspaceSkills = await this.#loadWorkspaceSkillMetadata();
    if (!workspaceSkills.some((skill2) => skill2.name === name)) return null;
    const skill = await workspace.skills.get(name);
    return skill ? this.#toHarnessSkill(skill) : null;
  }
  /**
   * Activate a skill by name and return the canonical skill instructions string.
   *
   * Throws `HarnessSkillNotFoundError` when the skill cannot be resolved.
   */
  async useSkill(name) {
    const skill = await this.getSkill(name);
    if (!skill) {
      throw new HarnessSkillNotFoundError({
        name,
        searchedSources: this.#searchedSources()
      });
    }
    return skill.instructions;
  }
  /**
   * Invalidate the workspace skill discovery cache. The next `listSkills` or
   * `useSkill` call will re-query the workspace.
   */
  refreshSkills() {
    this.#workspaceSkillsPromise = void 0;
  }
  #toHarnessSkill(skill) {
    const metadata = this.#plainMetadata(skill.metadata);
    const category = typeof metadata?.category === "string" ? metadata.category : void 0;
    const filePath = skill.path;
    if (!filePath) {
      throw new Error(`Workspace skill "${skill.name}" is missing a file path`);
    }
    return {
      name: skill.name,
      description: skill.description,
      instructions: "instructions" in skill ? skill.instructions : "",
      filePath,
      ...category ? { category } : {},
      ...metadata ? { metadata } : {}
    };
  }
  #plainMetadata(metadata) {
    if (!metadata) return void 0;
    const copy = {};
    for (const [key, value] of Object.entries(metadata)) {
      if (key === "args") continue;
      if (this.#isJsonSerializable(value)) copy[key] = value;
    }
    return Object.keys(copy).length > 0 ? copy : void 0;
  }
  #isPlainObject(value) {
    return typeof value === "object" && value !== null && !Array.isArray(value) && Object.getPrototypeOf(value) === Object.prototype;
  }
  #isJsonSerializable(value) {
    if (value === null || typeof value === "string" || typeof value === "boolean") return true;
    if (typeof value === "number") return Number.isFinite(value);
    if (Array.isArray(value)) return value.every((item) => this.#isJsonSerializable(item));
    if (this.#isPlainObject(value))
      return Object.values(value).every((item) => item !== void 0 && this.#isJsonSerializable(item));
    return false;
  }
  #searchedSources() {
    return this.#workspace !== void 0 ? ["workspace"] : [];
  }
  async #loadWorkspaceSkillMetadata() {
    if (!this.#workspaceSkillsPromise) {
      this.#workspaceSkillsPromise = this.#discoverWorkspaceSkillMetadata().catch((err) => {
        this.#workspaceSkillsPromise = void 0;
        throw err;
      });
    }
    return this.#workspaceSkillsPromise;
  }
  async #discoverWorkspaceSkillMetadata() {
    const workspace = await this.#getResolvedWorkspace();
    const skillsApi = workspace?.skills;
    if (!skillsApi) return [];
    const skillMetadata = await skillsApi.list();
    const skills = await Promise.all(
      skillMetadata.map(async (metadata) => await skillsApi.get(metadata.name) ?? metadata)
    );
    return skills.map((skill) => this.#toHarnessSkill(skill));
  }
  async #getResolvedWorkspace(requestContext) {
    const workspace = this.#workspace;
    if (!workspace) return void 0;
    if (typeof workspace !== "function") return workspace;
    if (this.#workspaceResolved) return this.#resolvedWorkspace;
    const resolved = await workspace({ requestContext: requestContext ?? new RequestContext() });
    this.#resolvedWorkspace = resolved;
    this.#workspaceResolved = true;
    return resolved;
  }
  async #applyStateUpdates(updates) {
    const changedKeys = Object.keys(updates);
    const newState = { ...this.#state, ...updates };
    if (this.#stateSchema) {
      const result = await this.#stateSchema["~standard"].validate(newState);
      if (result.issues) {
        const messages = result.issues.map((issue) => issue.message).join("; ");
        throw new Error(`Invalid state update: ${messages}`);
      }
      this.#state = result.value;
    } else {
      this.#state = newState;
    }
    await this.#persistSession({ state: this.#state });
    this.#events.emit({
      type: "state_changed",
      state: this.#state,
      changedKeys
    });
  }
  #isBusySnapshot() {
    const hasActiveRun = this.#runStatus === "starting" || this.#runStatus === "running" || this.#runStatus === "waiting" || this.#runStatus === "resuming";
    return hasActiveRun || this.#pending.some((item) => item.status === "pending");
  }
  #markRunning(runId, traceId = null) {
    this.#runStatus = "running";
    this.#currentRunId = runId;
    this.#currentTraceId = traceId;
  }
  #markIdle() {
    this.#runStatus = "idle";
    this.#currentRunId = null;
    this.#currentTraceId = null;
  }
  async #persistSession(updates) {
    this.#lastActivityAt = updates.lastActivityAt ?? /* @__PURE__ */ new Date();
    await this.#storage.updateSession(this.#id, {
      ...updates,
      lastActivityAt: this.#lastActivityAt
    });
  }
  async #respondToPendingItem(pendingItemId, expectedKind, response) {
    const item = this.#pending.find((item2) => item2.id === pendingItemId);
    if (!item) {
      throw new Error(`Harness pending item "${pendingItemId}" was not found on session "${this.#id}"`);
    }
    if (item.kind !== expectedKind) {
      throw new Error(`Harness pending item "${pendingItemId}" is kind "${item.kind}", not "${expectedKind}"`);
    }
    if (item.status !== "pending") {
      throw new Error(`Harness pending item "${pendingItemId}" is already ${item.status}`);
    }
    if (item.runtimeCompatibilityGeneration !== this.#runtimeCompatibilityGeneration) {
      throw new Error("harness.runtime_dependency_drifted");
    }
    const resumeResult = await this.#resumePendingBoundary(item, response);
    const recordedResponse = resumeResult === void 0 ? response : { ...response, resumeResult };
    await this.#storage.updatePendingItem(this.#id, pendingItemId, { status: "responded", response: recordedResponse });
    await this.#reloadRecordProjection();
    const updated = this.#pending.find((item2) => item2.id === pendingItemId);
    if (!updated) {
      throw new Error(`Harness pending item "${pendingItemId}" was not found on session "${this.#id}"`);
    }
    return { ...updated };
  }
  async #resumePendingBoundary(item, response) {
    const payload = item.payload;
    if (!this.#isPlainObject(payload)) return void 0;
    if (item.kind === "tool-approval") {
      const approved = response.approved;
      const runId = typeof payload.runId === "string" ? payload.runId : item.runId;
      const toolCallId = typeof payload.toolCallId === "string" ? payload.toolCallId : void 0;
      const agent = typeof payload.agentId === "string" && this.#resolveAgent ? await this.#resolveAgent(payload.agentId) : this.#agent;
      if (typeof approved !== "boolean" || !runId || !this.#resolveAgent) return void 0;
      const result = approved ? await agent.approveToolCallGenerate({ runId, toolCallId, requestContext: await this.#buildRequestContext() }) : await agent.declineToolCallGenerate({ runId, toolCallId, requestContext: await this.#buildRequestContext() });
      return this.#isJsonSerializable(result) ? result : { resumed: true };
    }
    if (item.kind === "plan-approval" && response.approved === true) {
      const transitionModeId = typeof payload.transitionModeId === "string" ? payload.transitionModeId : void 0;
      if (transitionModeId && transitionModeId !== this.#mode.id && this.#resolveMode) {
        const mode = await this.#resolveMode(transitionModeId);
        this.setMode(mode);
        return { transitionModeId, modeChanged: true };
      }
    }
    return void 0;
  }
  async #reloadRecordProjection() {
    const record = await this.#storage.loadSession(this.#id);
    if (!record) {
      throw new Error(`Harness session "${this.#id}" was not found`);
    }
    this.#lastActivityAt = record.lastActivityAt;
    this.#pending = (record.pending ?? []).map((item) => ({ ...item }));
  }
  #getSchemaDefaults() {
    if (!this.#stateSchema) return {};
    const defaults = {};
    try {
      const jsonSchema = this.#stateSchema["~standard"].jsonSchema.output({ target: "draft-07" });
      for (const [key, prop] of Object.entries(jsonSchema.properties ?? {})) {
        if (prop.default !== void 0) {
          defaults[key] = prop.default;
        }
      }
    } catch {
    }
    return defaults;
  }
  async #buildRequestContext() {
    const overlay = buildHarnessRequestContext({ harnessContext: this.#createHarnessContext() });
    await this.#getResolvedWorkspace(overlay);
    return overlay;
  }
  #createHarnessContext() {
    return {
      harnessId: this.#ownerId,
      sessionId: this.#id,
      ownerId: this.#ownerId,
      resourceId: this.#resourceId,
      threadId: this.#threadId,
      modeId: this.#mode.id,
      modelId: this.#modelId,
      parentSessionId: this.#parentSessionId,
      subagentDepth: this.#subagentDepth,
      source: this.#source,
      getState: () => this.getState()
    };
  }
  async #resolveMemory() {
    const mem = this.#memory;
    if (!mem) {
      throw new Error("Memory is not configured on this Harness");
    }
    if (typeof mem !== "function") {
      return mem;
    }
    const requestContext = await this.#buildRequestContext();
    const resolved = await mem({ requestContext });
    if (!resolved) {
      throw new Error("Dynamic memory factory returned empty value");
    }
    return resolved;
  }
};

// src/harness/v1/harness.ts
function isHarnessStorage(storage) {
  return typeof storage.loadSession === "function";
}
var Harness = class {
  #ownerId;
  #defaultMode;
  #modesById = /* @__PURE__ */ new Map();
  #storage;
  #compositeStorage;
  #mastra;
  #memory;
  #events;
  #stateSchema;
  #initialState;
  #workspace;
  #agent;
  #subagents;
  #resolveModel;
  #runtimeCompatibilityGeneration;
  #defaultPermissionPolicy;
  #toolCategoryResolver;
  constructor(config) {
    if (!config.modes.length) {
      throw new Error("The harness needs modes to operate.");
    }
    this.#ownerId = config.ownerId ?? randomUUID();
    this.#defaultMode = config.defaultModeId ?? config.modes[0].id;
    this.#mastra = config.mastra;
    if (config.storage) {
      if (isHarnessStorage(config.storage)) {
        this.#storage = config.storage;
      } else {
        this.#compositeStorage = augmentWithInit(config.storage);
      }
    } else {
      this.#compositeStorage = config.mastra?.getStorage();
    }
    this.#memory = config.memory;
    this.#events = new EventEmitter();
    if (config.runtimeCompatibilityGeneration !== void 0 && config.runtimeCompatibilityGeneration.length === 0) {
      throw new Error("Harness runtimeCompatibilityGeneration must be a non-empty string when configured");
    }
    this.#runtimeCompatibilityGeneration = config.runtimeCompatibilityGeneration ?? null;
    this.#stateSchema = config.stateSchema;
    this.#initialState = config.initialState;
    if (config.workspace instanceof Workspace || typeof config.workspace === "function") {
      this.#workspace = config.workspace;
    } else if (config.workspace) {
      this.#workspace = new Workspace(config.workspace);
    }
    this.#agent = this.#mastra ? this.#mastra.getAgentById(config.agent) : config.agent;
    if (config.subagents) {
      const entries = Object.entries(config.subagents.types ?? {});
      if (entries.length > 0 && !config.resolveModel) {
        throw new Error('Harness "subagents" requires a "resolveModel" function to instantiate subagent models');
      }
      for (const [typeId, def] of entries) {
        if (!def?.agentId) {
          throw new Error(`Subagent "${typeId}" must declare an "agentId"`);
        }
      }
      const maxDepth = config.sessions?.maxSubagentDepth ?? config.subagents.maxDepth ?? 1;
      if (!Number.isInteger(maxDepth) || maxDepth < 0) {
        throw new Error("Harness sessions.maxSubagentDepth must be a non-negative integer");
      }
      this.#subagents = { ...config.subagents, maxDepth };
    }
    this.#resolveModel = config.resolveModel;
    this.#defaultPermissionPolicy = config.defaultPermissionPolicy ?? "ask";
    if (config.toolCategoryResolver) {
      this.#toolCategoryResolver = config.toolCategoryResolver;
    } else if (config.toolCategories) {
      const categories = config.toolCategories;
      this.#toolCategoryResolver = (toolName) => categories[toolName] ?? null;
    }
    const modes = config.modes ?? [];
    for (const mode of modes) {
      if (this.#modesById.has(mode.id)) {
        throw new Error(`Duplicate mode id "${mode.id}" found when creating the Harness`);
      }
      if (mode.tools && mode.additionalTools) {
        throw new Error(`Mode "${mode.id} cannot set both "tools" and "additionalTools" - choose replace OR augment`);
      }
      this.#modesById.set(mode.id, mode);
    }
  }
  get ownerId() {
    return this.#ownerId;
  }
  subscribe(listener) {
    return this.#events.subscribe(listener);
  }
  emit(event) {
    return this.#events.emit(event);
  }
  async init() {
    await this.#requireStorage();
  }
  async shutdown() {
  }
  listModes() {
    return [...this.#modesById.values()];
  }
  /**
   * Look up a single mode by id. Returns `undefined` if no mode with that id
   * is registered. For the throwing variant used during request resolution,
   * see the internal `_getMode` helper.
   */
  getMode(modeId) {
    return this.#modesById.get(modeId);
  }
  async listSessions() {
    const storage = await this.#requireStorage();
    return storage.listSessions();
  }
  async session(opts) {
    const storage = await this.#requireStorage();
    if ("threadId" in opts) {
      return this.#sessionByThread(storage, opts);
    }
    const record = await this.#loadSessionRecord(storage, opts.sessionId, opts.resourceId);
    return this.#sessionFromRecord(record, storage);
  }
  async cloneSession(session, opts = {}) {
    const storage = await this.#requireStorage();
    const source = await this.#loadSessionRecord(storage, session.id, session.resourceId);
    const modeId = opts.modeId ?? source.modeId;
    const mode = this.#modesById.get(modeId);
    if (!mode) {
      throw new Error(`Harness session "${source.id}" cannot clone into unknown mode "${modeId}"`);
    }
    const clone = await session.clone({
      ...opts,
      resourceId: opts.resourceId ?? source.resourceId,
      mode,
      modelId: opts.modelId ?? source.modelId
    });
    const parentSessionId = opts.parentSessionId ?? source.id;
    const cloneModeId = opts.modeId ?? source.modeId;
    const cloneModelId = opts.modelId ?? source.modelId;
    const record = {
      id: clone.id,
      ownerId: this.#ownerId,
      threadId: clone.threadId,
      resourceId: clone.resourceId,
      parentSessionId,
      origin: opts.origin ?? source.origin,
      source: {
        type: opts.origin ?? source.origin,
        parentSessionId
      },
      subagentDepth: (source.subagentDepth ?? 0) + 1,
      runtimeCompatibilityGeneration: this.#runtimeCompatibilityGeneration,
      modeId: cloneModeId,
      modelId: cloneModelId,
      title: opts.title ?? source.title,
      metadata: opts.metadata ?? source.metadata,
      state: clone.getState(),
      pending: [],
      createdAt: clone.createdAt,
      lastActivityAt: clone.lastActivityAt
    };
    await storage.saveSession(record);
    this.#events.emit({ type: "session_created", ...sessionCreatedPayload(record) });
    return this.#sessionFromRecord(record, storage);
  }
  async #sessionByThread(storage, opts) {
    const id = this.#sessionIdFor(opts.resourceId, opts.threadId);
    const existing = await storage.loadSession(id);
    if (existing) {
      return this.#sessionFromRecord(existing, storage);
    }
    const modeId = opts.modeId ?? this.#defaultMode;
    const mode = this.#modesById.get(modeId);
    if (!mode) {
      throw new Error(`Harness session for thread "${opts.threadId}" cannot use unknown mode "${modeId}"`);
    }
    const record = {
      id,
      ownerId: this.#ownerId,
      threadId: opts.threadId,
      resourceId: opts.resourceId,
      origin: "top-level",
      modeId,
      modelId: opts.modelId ?? mode.defaultModelId,
      source: { type: "top-level" },
      subagentDepth: 0,
      runtimeCompatibilityGeneration: this.#runtimeCompatibilityGeneration,
      ...this.#initialState ? { state: this.#initialState } : {},
      pending: [],
      createdAt: /* @__PURE__ */ new Date(),
      lastActivityAt: /* @__PURE__ */ new Date()
    };
    await storage.saveSession(record);
    this.#events.emit({ type: "session_created", ...sessionCreatedPayload(record) });
    return this.#sessionFromRecord(record, storage);
  }
  async #loadSessionRecord(storage, sessionId, resourceId) {
    const record = await storage.loadSession(sessionId);
    if (!record) {
      throw new Error(`Harness session "${sessionId}" was not found`);
    }
    if (resourceId && record.resourceId !== resourceId) {
      throw new Error(`Harness session "${sessionId}" does not belong to resource "${resourceId}"`);
    }
    return record;
  }
  async #requireStorage() {
    if (this.#storage) {
      return this.#storage;
    }
    if (!this.#compositeStorage) {
      throw new Error("Harness session storage is not configured");
    }
    const storage = await this.#compositeStorage.getStore("harness");
    if (!storage) {
      throw new Error("Harness session storage is not configured");
    }
    return storage;
  }
  #sessionFromRecord(record, storage) {
    const mode = record.modeId ? this.#modesById.get(record.modeId) : this.#modesById.values().next().value;
    if (!mode) {
      throw new Error(`Harness session "${record.id}" references unknown mode "${record.modeId}"`);
    }
    return new Session({
      id: record.id,
      ownerId: record.ownerId,
      threadId: record.threadId,
      resourceId: record.resourceId,
      mode,
      model: record.modelId,
      createdAt: record.createdAt,
      lastActivityAt: record.lastActivityAt,
      agent: this.#agent,
      memory: this.#memory,
      storage,
      record,
      runtimeCompatibilityGeneration: this.#runtimeCompatibilityGeneration,
      events: this.#events.scoped({ sessionId: record.id }),
      stateSchema: this.#stateSchema,
      initialState: this.#initialState,
      workspace: this.#workspace,
      subagents: this.#subagents,
      resolveAgent: (agentId) => this.#resolveAgent(agentId),
      resolveMode: (modeId) => this.#resolveMode(modeId),
      resolveModel: this.#resolveModel,
      defaultPermissionPolicy: this.#defaultPermissionPolicy,
      toolCategoryResolver: this.#toolCategoryResolver
    });
  }
  #resolveAgent(agentId) {
    if (this.#mastra) return this.#mastra.getAgentById(agentId);
    throw new Error(`Harness mode references unknown agent "${agentId}"`);
  }
  #resolveMode(modeId) {
    const mode = this.#modesById.get(modeId);
    if (!mode) {
      throw new Error(`Harness mode "${modeId}" was not found`);
    }
    return mode;
  }
  #sessionIdFor(resourceId, threadId) {
    const hash = createHash("sha256").update(`${resourceId}\0${threadId}`).digest("hex").slice(0, 32);
    return `sess-${hash}`;
  }
};

export { EventEmitter, Harness, HarnessSkillNotFoundError, Session, formatHarnessEventId, parseHarnessEventId };
