// @mastra/server/dist/server/a2a/store.cjs
'use strict';

// src/server/a2a/store.ts
function createAbortError() {
  return new DOMException("The operation was aborted.", "AbortError");
}
var InMemoryTaskStore = class {
  store = /* @__PURE__ */ new Map();
  versions = /* @__PURE__ */ new Map();
  listeners = /* @__PURE__ */ new Map();
  activeCancellations = /* @__PURE__ */ new Set();
  getKey(agentId, taskId) {
    return `${agentId}-${taskId}`;
  }
  async load({ agentId, taskId }) {
    const snapshot = this.loadWithVersion({ agentId, taskId });
    if (!snapshot) {
      return null;
    }
    return snapshot.task;
  }
  loadWithVersion({ agentId, taskId }) {
    const key = this.getKey(agentId, taskId);
    const entry = this.store.get(key);
    if (!entry) {
      return null;
    }
    return {
      task: { ...entry },
      version: this.versions.get(key) ?? 0
    };
  }
  async save({ agentId, data }) {
    const key = this.getKey(agentId, data.id);
    if (!data.id) {
      throw new Error("Task ID is required");
    }
    const storedTask = { ...data };
    const nextVersion = (this.versions.get(key) ?? 0) + 1;
    this.store.set(key, storedTask);
    this.versions.set(key, nextVersion);
    const listeners = this.listeners.get(key);
    if (listeners) {
      for (const listener of listeners) {
        listener({ task: { ...storedTask }, version: nextVersion });
      }
    }
  }
  getVersion({ agentId, taskId }) {
    return this.versions.get(this.getKey(agentId, taskId)) ?? 0;
  }
  async waitForNextUpdate({
    agentId,
    taskId,
    afterVersion,
    signal
  }) {
    const key = this.getKey(agentId, taskId);
    const snapshot = this.loadWithVersion({ agentId, taskId });
    if (snapshot && snapshot.version > afterVersion) {
      return snapshot;
    }
    if (signal?.aborted) {
      return Promise.reject(createAbortError());
    }
    return new Promise((resolve, reject) => {
      const listeners = this.listeners.get(key) ?? /* @__PURE__ */ new Set();
      let settled = false;
      const cleanup = () => {
        listeners.delete(listener);
        if (listeners.size === 0) {
          this.listeners.delete(key);
        }
        signal?.removeEventListener("abort", onAbort);
      };
      const onAbort = () => {
        if (settled) {
          return;
        }
        settled = true;
        cleanup();
        reject(createAbortError());
      };
      const listener = (update) => {
        if (settled || signal?.aborted) {
          return;
        }
        if (update.version <= afterVersion) {
          return;
        }
        settled = true;
        cleanup();
        resolve({ task: { ...update.task }, version: update.version });
      };
      listeners.add(listener);
      this.listeners.set(key, listeners);
      signal?.addEventListener("abort", onAbort, { once: true });
    });
  }
};

exports.InMemoryTaskStore = InMemoryTaskStore;
