// @mastra/core/dist/logger/index.cjs
'use strict';

var chunkBJ2BML66_cjs = require('../chunk-BJ2BML66.cjs');
var chunkYNLHR2WT_cjs = require('../chunk-YNLHR2WT.cjs');



Object.defineProperty(exports, "DualLogger", {
  enumerable: true,
  get: function () { return chunkBJ2BML66_cjs.DualLogger; }
});
Object.defineProperty(exports, "MultiLogger", {
  enumerable: true,
  get: function () { return chunkBJ2BML66_cjs.MultiLogger; }
});
Object.defineProperty(exports, "createLogger", {
  enumerable: true,
  get: function () { return chunkBJ2BML66_cjs.createLogger; }
});
Object.defineProperty(exports, "noopLogger", {
  enumerable: true,
  get: function () { return chunkBJ2BML66_cjs.noopLogger; }
});
Object.defineProperty(exports, "ConsoleLogger", {
  enumerable: true,
  get: function () { return chunkYNLHR2WT_cjs.ConsoleLogger; }
});
Object.defineProperty(exports, "LogLevel", {
  enumerable: true,
  get: function () { return chunkYNLHR2WT_cjs.LogLevel; }
});
Object.defineProperty(exports, "LoggerTransport", {
  enumerable: true,
  get: function () { return chunkYNLHR2WT_cjs.LoggerTransport; }
});
Object.defineProperty(exports, "MastraLogger", {
  enumerable: true,
  get: function () { return chunkYNLHR2WT_cjs.MastraLogger; }
});
Object.defineProperty(exports, "RegisteredLogger", {
  enumerable: true,
  get: function () { return chunkYNLHR2WT_cjs.RegisteredLogger; }
});
Object.defineProperty(exports, "createCustomTransport", {
  enumerable: true,
  get: function () { return chunkYNLHR2WT_cjs.createCustomTransport; }
});
